{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "import sys, os\n",
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "from importlib import reload"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "try:\n",
    "    sys.path.insert(0, \"/home/piotr/CodeProjects/mhcevolution/PyScripts/\")\n",
    "except Exception:\n",
    "    pass\n",
    "try:\n",
    "    sys.path.insert(0, \"/home/piotr/CodeProjects/MHCevo/PyScripts/\")\n",
    "except Exception:\n",
    "    pass    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "import packed_plots_of_MHC_alleles as ppma"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[Errno 2] No such file or directory: '/media/piotr/Elements/MHC_UAM/6000Antig/'\n",
      "/home/piotr/CodeProjects/MhcEvo2000/PyScripts\n"
     ]
    }
   ],
   "source": [
    "cd /media/piotr/Elements/MHC_UAM/6000Antig/"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "ename": "OSError",
     "evalue": "0.02_6000_OneChr_DataSlice.csv not found.",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mOSError\u001b[0m                                   Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-6-97be41e07a0f>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m      1\u001b[0m \u001b[0mx_Label\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m\"number of pathogen species\"\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      2\u001b[0m \u001b[0;31m#theData_002 = np.genfromtxt(\"0.02_DataSlice.csv\", dtype=ppma.outType)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 3\u001b[0;31m \u001b[0mtheData_002\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mnp\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mgenfromtxt\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m\"0.02_6000_OneChr_DataSlice.csv\"\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mdtype\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mppma\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0moutType\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      4\u001b[0m \u001b[0mmeanResult_002\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mppma\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mbuildStats95CI\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mtheData_002\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      5\u001b[0m \u001b[0;31m#theData_008 = np.genfromtxt(\"0.08_DataSlice.csv\", dtype=ppma.outType)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/anaconda3/envs/myDef/lib/python3.6/site-packages/numpy/lib/npyio.py\u001b[0m in \u001b[0;36mgenfromtxt\u001b[0;34m(fname, dtype, comments, delimiter, skip_header, skip_footer, converters, missing_values, filling_values, usecols, names, excludelist, deletechars, replace_space, autostrip, case_sensitive, defaultfmt, unpack, usemask, loose, invalid_raise, max_rows, encoding)\u001b[0m\n\u001b[1;32m   1698\u001b[0m             \u001b[0mfname\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mstr\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mfname\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1699\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0misinstance\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mfname\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mbasestring\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 1700\u001b[0;31m             \u001b[0mfhd\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0miter\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mnp\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mlib\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_datasource\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mopen\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mfname\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'rt'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mencoding\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mencoding\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   1701\u001b[0m             \u001b[0mown_fhd\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;32mTrue\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1702\u001b[0m         \u001b[0;32melse\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/anaconda3/envs/myDef/lib/python3.6/site-packages/numpy/lib/_datasource.py\u001b[0m in \u001b[0;36mopen\u001b[0;34m(path, mode, destpath, encoding, newline)\u001b[0m\n\u001b[1;32m    260\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    261\u001b[0m     \u001b[0mds\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mDataSource\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mdestpath\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 262\u001b[0;31m     \u001b[0;32mreturn\u001b[0m \u001b[0mds\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mopen\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mpath\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mmode\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mencoding\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mencoding\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mnewline\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mnewline\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    263\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    264\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/anaconda3/envs/myDef/lib/python3.6/site-packages/numpy/lib/_datasource.py\u001b[0m in \u001b[0;36mopen\u001b[0;34m(self, path, mode, encoding, newline)\u001b[0m\n\u001b[1;32m    616\u001b[0m                                       encoding=encoding, newline=newline)\n\u001b[1;32m    617\u001b[0m         \u001b[0;32melse\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 618\u001b[0;31m             \u001b[0;32mraise\u001b[0m \u001b[0mIOError\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m\"%s not found.\"\u001b[0m \u001b[0;34m%\u001b[0m \u001b[0mpath\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    619\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    620\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mOSError\u001b[0m: 0.02_6000_OneChr_DataSlice.csv not found."
     ]
    }
   ],
   "source": [
    "x_Label = \"number of pathogen species\"\n",
    "#theData_002 = np.genfromtxt(\"0.02_DataSlice.csv\", dtype=ppma.outType)\n",
    "theData_002 = np.genfromtxt(\"0.02_6000_OneChr_DataSlice.csv\", dtype=ppma.outType)\n",
    "meanResult_002 = ppma.buildStats95CI(theData_002)\n",
    "#theData_008 = np.genfromtxt(\"0.08_DataSlice.csv\", dtype=ppma.outType)\n",
    "theData_008 = np.genfromtxt(\"0.08_6000_OneChr_DataSlice.csv\", dtype=ppma.outType)\n",
    "meanResult_008 = ppma.buildStats95CI(theData_008)\n",
    "theData_002_mcromut = np.genfromtxt(\"/media/piotr/Elements/MHC_UAM/review/macroMut/0.02_6000_macro_chromoOne_DataSlice.csv\", dtype=ppma.outType)\n",
    "meanResult_002_macro = ppma.buildStats95CI(theData_002_mcromut)\n",
    "theData_008_mcromut = np.genfromtxt(\"/media/piotr/Elements/MHC_UAM/review/macroMut/0.08_6000_macro_chromoOne_DataSlice.csv\", dtype=ppma.outType)\n",
    "meanResult_008_macro = ppma.buildStats95CI(theData_008_mcromut)\n",
    "meanResult = [meanResult_002, meanResult_008, meanResult_002_macro, meanResult_008_macro]\n",
    "meanResult[0].dtype"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def plotAvgMHCsInIndiv(meanResult, x_label, y_label, indx=('meanAllel', 'stdAllel'), figNumb=1, \n",
    "                       yLim=25, yScale=1., legdLoc='upper left', logsc='linear'):\n",
    "    \"\"\"Uses the array generated by buildStats() function and plots a fancy plot\n",
    "    of averaged data.\"\"\"\n",
    "    FS = 24\n",
    "    titleFS = FS+4\n",
    "    figSize = (24, 9)\n",
    "    annoSize = int(1.1*FS)\n",
    "    xtix = (\"\", \"\", \"2\", \"\", \"\", \"4\", \"\", \"\", \"8\", \"\", \"\", \"\", \"\", \"\", \"\", \"\",\n",
    "            \"16\", \"\", \"\", \"\", \"\", \"\", \"\", \"\", \"\", \"\", \"\", \"\", \"\", \"\", \"\", \"\",\n",
    "            \"32\", \"\", \"\", \"\", \"\", \"\", \"\", \"\", \"\", \"\", \"\", \"\", \"\", \"\", \"\", \"\",\n",
    "            \"\", \"\", \"\", \"\", \"\", \"\", \"\", \"\", \"\", \"\", \"\", \"\", \"\", \"\", \"\", \"\",\n",
    "            \"64\", \"\", \"\", \"\", \"\", \"\", \"\")\n",
    "    k = (-0.35, 0.35)\n",
    "    legd = (r'$10^{-5}$', r'$5\\cdot10^{-5}$')\n",
    "    clr = ('C0', 'C1', 'C2', 'C3', 'C4', 'C5')\n",
    "    # One\n",
    "    ll = []\n",
    "    for itm in meanResult[0]:\n",
    "        if itm[0] in ll:\n",
    "            pass\n",
    "        else:\n",
    "            ll.append(itm[0])\n",
    "    # Second plot - unique MHC alleles in one chromosome\n",
    "    plt.figure(1, figsize=figSize)\n",
    "    plt.subplot(121)\n",
    "    for i, var in enumerate(ll):\n",
    "        ww = meanResult[0][meanResult[0]['VAR'] == var]\n",
    "        plt.errorbar(ww['VARX']+k[i], ww[indx[0]]*yScale, ww[indx[1]]*yScale, lw=2, marker='o', ms=10,\n",
    "                     color=clr[i], label=legd[i])\n",
    "    plt.title(r'           $\\alpha = 0.02$', fontsize=titleFS, loc='left')\n",
    "    plt.xlabel(str(x_label), fontsize=FS)\n",
    "    plt.ylabel(str(y_label), fontsize=FS)\n",
    "    plt.ylim((0, yLim*yScale))\n",
    "    plt.xscale(logsc)\n",
    "    plt.tick_params(axis='both', labelsize=annoSize, bottom=False)\n",
    "    plt.xticks(np.arange(70), xtix, fontsize=FS+2)\n",
    "    plt.grid(axis='y')\n",
    "    plt.legend(frameon=False, loc=legdLoc, fontsize=FS-2)\n",
    "    # Two\n",
    "    ll = []\n",
    "    for itm in meanResult[1]:\n",
    "        if itm[0] in ll:\n",
    "            pass\n",
    "        else:\n",
    "            ll.append(itm[0])\n",
    "    # Second plot - unique MHC alleles in one chromosome\n",
    "    plt.subplot(122)\n",
    "    for i, var in enumerate(ll):\n",
    "        ww = meanResult[1][meanResult[1]['VAR'] == var]\n",
    "        plt.errorbar(ww['VARX']+k[i], ww[indx[0]]*yScale, ww[indx[1]]*yScale, lw=2, marker=\"o\", \n",
    "                     ms=10, color=clr[i+2], label=legd[i])\n",
    "    plt.title(r'           $\\alpha = 0.08$', fontsize=titleFS, loc='left')\n",
    "    plt.xlabel(str(x_label), fontsize=FS)\n",
    "    #plt.ylabel(\"average number of MHC copies\\nin an individual\",\n",
    "    #           fontsize=FS)\n",
    "    plt.ylim((0, yLim*yScale))\n",
    "    plt.xscale(logsc)\n",
    "    plt.tick_params(axis='both', labelsize=annoSize, bottom=False)\n",
    "    plt.xticks(np.arange(70), xtix, fontsize=FS+2)\n",
    "    plt.grid(axis='y')\n",
    "    plt.legend(frameon=False, loc=legdLoc, fontsize=FS-2)\n",
    "    #plt.tight_layout(True)\n",
    "    \"\"\"\n",
    "    plt.subplot(223)\n",
    "    for i, var in enumerate(ll):\n",
    "        ww = meanResult[2][meanResult[2]['VAR'] == var]\n",
    "        plt.errorbar(ww['VARX']+k[i], ww[indx[0]]*yScale, ww[indx[1]]*yScale, lw=2, marker=\"x\", \n",
    "                     linestyle='--', ms=10, color=clr[i], label=legd[i])\n",
    "    plt.title(r'           $\\alpha = 0.02$ - recombinantions', fontsize=titleFS, loc='left')\n",
    "    plt.xlabel(str(x_label), fontsize=FS)\n",
    "    #plt.ylabel(\"average number of MHC copies\\nin an individual\",\n",
    "    #           fontsize=FS)\n",
    "    plt.ylim((0, yLim*yScale))\n",
    "    plt.xscale(logsc)\n",
    "    plt.tick_params(axis='both', labelsize=annoSize, bottom=False)\n",
    "    plt.xticks(np.arange(70), xtix, fontsize=FS+2)\n",
    "    plt.grid(axis='y')\n",
    "    plt.legend(frameon=False, loc=legdLoc, fontsize=FS-2)\n",
    "    plt.subplot(224)\n",
    "    for i, var in enumerate(ll):\n",
    "        ww = meanResult[3][meanResult[3]['VAR'] == var]\n",
    "        plt.errorbar(ww['VARX']+k[i], ww[indx[0]]*yScale, ww[indx[1]]*yScale, lw=2, marker=\"x\", \n",
    "                     linestyle='--', ms=10, color=clr[i+2], label=legd[i])\n",
    "    plt.title(r'           $\\alpha = 0.02$ - recombinantions', fontsize=titleFS, loc='left')\n",
    "    plt.xlabel(str(x_label), fontsize=FS)\n",
    "    #plt.ylabel(\"average number of MHC copies\\nin an individual\",\n",
    "    #           fontsize=FS)\n",
    "    plt.ylim((0, yLim*yScale))\n",
    "    plt.xscale(logsc)\n",
    "    plt.tick_params(axis='both', labelsize=annoSize, bottom=False)\n",
    "    plt.xticks(np.arange(70), xtix, fontsize=FS+2)\n",
    "    plt.grid(axis='y')\n",
    "    plt.legend(frameon=False, loc=legdLoc, fontsize=FS-2)\n",
    "    \"\"\"\n",
    "    figPDF =\"fig_\" + str(figNumb) + \".tiff\"\n",
    "    figPNG = \"fig_\" + str(figNumb) + \".png\"\n",
    "    plt.savefig(figPDF, bbox_inches='tight')\n",
    "    plt.savefig(figPNG, bbox_inches='tight')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "y_Label = \"average number of MHC variants\\nin individual (INV)\"\n",
    "plotAvgMHCsInIndiv(meanResult, x_Label, y_Label, ('indvMean', 'indvSTD'), 1, 25, 1., 'upper left')\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def plotAvgMHCsInIndiv(meanResult, x_label, y_label, indx=('meanAllel', 'stdAllel'), figNumb=1, \n",
    "                       yLim=25, yScale=1., legdLoc=('upper left', 'upper left'), logsc='linear'):\n",
    "    \"\"\"Uses the array generated by buildStats() function and plots a fancy plot\n",
    "    of averaged data.\"\"\"\n",
    "    FS = 24\n",
    "    titleFS = FS+4\n",
    "    figSize = (24, 9)\n",
    "    annoSize = int(1.1*FS)\n",
    "    lineForRecomb = '-'\n",
    "    xtix = (\"\", \"\", \"2\", \"\", \"\", \"4\", \"\", \"\", \"8\", \"\", \"\", \"\", \"\", \"\", \"\", \"\",\n",
    "            \"16\", \"\", \"\", \"\", \"\", \"\", \"\", \"\", \"\", \"\", \"\", \"\", \"\", \"\", \"\", \"\",\n",
    "            \"32\", \"\", \"\", \"\", \"\", \"\", \"\", \"\", \"\", \"\", \"\", \"\", \"\", \"\", \"\", \"\",\n",
    "            \"\", \"\", \"\", \"\", \"\", \"\", \"\", \"\", \"\", \"\", \"\", \"\", \"\", \"\", \"\", \"\",\n",
    "            \"64\", \"\", \"\", \"\", \"\", \"\", \"\")\n",
    "    k = (-0.35, 0.35, 0.7, -0.7)\n",
    "    legd = (r'   $10^{-5}$ point mut.', r'$5\\cdot10^{-5}$ point mut.', \n",
    "            r'   $10^{-5}$ recombin.', r'$5\\cdot10^{-5}$ recombin.')\n",
    "    clr = ('C0', 'C1', 'C2', 'C3', 'C4', 'C5')\n",
    "    one = 0.6\n",
    "    two = 0.7\n",
    "    clrPoint = ( (one, one, one, 0.75),  (two, two, two, 0.75))\n",
    "    markr = (\"o\", \"x\")\n",
    "    # One\n",
    "    ll = []\n",
    "    for itm in meanResult[0]:\n",
    "        if itm[0] in ll:\n",
    "            pass\n",
    "        else:\n",
    "            ll.append(itm[0])\n",
    "    # Second plot - unique MHC alleles in one chromosome\n",
    "    plt.figure(1, figsize=figSize)\n",
    "    plt.subplot(121)\n",
    "    for i, var in enumerate(ll):\n",
    "        ww = meanResult[0][meanResult[0]['VAR'] == var]\n",
    "        plt.errorbar(ww['VARX']+k[i], ww[indx[0]]*yScale, ww[indx[1]]*yScale, lw=2, marker=markr[i], ms=10,\n",
    "                     color=clrPoint[i], label=legd[i])\n",
    "    for i, var in enumerate(ll):\n",
    "        ww = meanResult[2][meanResult[2]['VAR'] == var]\n",
    "        plt.errorbar(ww['VARX']+k[i+2], ww[indx[0]]*yScale, ww[indx[1]]*yScale, lw=2, marker=markr[i], \n",
    "                     linestyle=lineForRecomb, ms=10, color=clr[i], label=legd[i+2])\n",
    "    plt.title(r'           $\\alpha = 0.02$', fontsize=titleFS, loc='left')\n",
    "    plt.xlabel(str(x_label), fontsize=FS)\n",
    "    plt.ylabel(str(y_label), fontsize=FS)\n",
    "    plt.ylim((0, yLim*yScale))\n",
    "    plt.xscale(logsc)\n",
    "    plt.tick_params(axis='both', labelsize=annoSize, bottom=False)\n",
    "    plt.xticks(np.arange(70), xtix, fontsize=FS+2)\n",
    "    plt.grid(axis='y')\n",
    "    plt.legend(frameon=False, loc=legdLoc[0], fontsize=FS-2)\n",
    "    # Two\n",
    "    ll = []\n",
    "    for itm in meanResult[1]:\n",
    "        if itm[0] in ll:\n",
    "            pass\n",
    "        else:\n",
    "            ll.append(itm[0])\n",
    "    # Second plot - unique MHC alleles in one chromosome\n",
    "    plt.subplot(122)\n",
    "    for i, var in enumerate(ll):\n",
    "        ww = meanResult[1][meanResult[1]['VAR'] == var]\n",
    "        plt.errorbar(ww['VARX']+k[i], ww[indx[0]]*yScale, ww[indx[1]]*yScale, lw=2, marker=markr[i], \n",
    "                     ms=10, color=clrPoint[i], label=legd[i])\n",
    "    for i, var in enumerate(ll):\n",
    "        ww = meanResult[3][meanResult[3]['VAR'] == var]\n",
    "        plt.errorbar(ww['VARX']+k[i+2], ww[indx[0]]*yScale, ww[indx[1]]*yScale, lw=2, marker=markr[i], \n",
    "                     linestyle=lineForRecomb, ms=10, color=clr[i+2], label=legd[i+2])\n",
    "    plt.title(r'           $\\alpha = 0.08$', fontsize=titleFS, loc='left')\n",
    "    plt.xlabel(str(x_label), fontsize=FS)\n",
    "    #plt.ylabel(\"average number of MHC copies\\nin an individual\",\n",
    "    #           fontsize=FS)\n",
    "    plt.ylim((0, yLim*yScale))\n",
    "    plt.xscale(logsc)\n",
    "    plt.tick_params(axis='both', labelsize=annoSize, bottom=False)\n",
    "    plt.xticks(np.arange(70), xtix, fontsize=FS+2)\n",
    "    plt.grid(axis='y')\n",
    "    plt.legend(frameon=False, loc=legdLoc[1], fontsize=FS-2)\n",
    "    #plt.tight_layout(True)\n",
    "    \n",
    "   \n",
    "    \n",
    "    figPDF =\"fig_\" + str(figNumb) + \"_B.tiff\"\n",
    "    figPNG = \"fig_\" + str(figNumb) + \"_B.png\"\n",
    "    plt.savefig(figPDF, bbox_inches='tight')\n",
    "    plt.savefig(figPNG, bbox_inches='tight')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "y_Label = \"average number of MHC copies\\nin an individual (INV)\"\n",
    "plotAvgMHCsInIndiv(meanResult, x_Label, y_Label, ('indvMean', 'indvSTD'), \"SX\", 25, 1., \n",
    "                   ('lower right', 'upper right'))\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "yTikzDef=np.arange(0, 200, 25)\n",
    "def plotAvgMHCsInIndiv(meanResult, x_label, y_label, indx=('meanAllel', 'stdAllel'), figNumb=1, \n",
    "                       yLim=25, yScale=1., yTikz=yTikzDef, legdLoc='upper left', logsc='linear'):\n",
    "    \"\"\"Uses the array generated by buildStats() function and plots a fancy plot\n",
    "    of averaged data.\"\"\"\n",
    "    FS = 24\n",
    "    titleFS = FS+4\n",
    "    figSize = (24, 9)\n",
    "    annoSize = int(1.1*FS)\n",
    "    xtix = (\"\", \"\", \"2\", \"\", \"\", \"4\", \"\", \"\", \"8\", \"\", \"\", \"\", \"\", \"\", \"\", \"\",\n",
    "            \"16\", \"\", \"\", \"\", \"\", \"\", \"\", \"\", \"\", \"\", \"\", \"\", \"\", \"\", \"\", \"\",\n",
    "            \"32\", \"\", \"\", \"\", \"\", \"\", \"\", \"\", \"\", \"\", \"\", \"\", \"\", \"\", \"\", \"\",\n",
    "            \"\", \"\", \"\", \"\", \"\", \"\", \"\", \"\", \"\", \"\", \"\", \"\", \"\", \"\", \"\", \"\",\n",
    "            \"64\", \"\", \"\", \"\", \"\", \"\", \"\")\n",
    "    k = (-0.35, 0.35)\n",
    "    legd = (r'$10^{-5}$', r'$5\\cdot10^{-5}$')\n",
    "    clr = ('C0', 'C1', 'C2', 'C3', 'C4', 'C5')\n",
    "    # One\n",
    "    ll = []\n",
    "    for itm in meanResult[0]:\n",
    "        if itm[0] in ll:\n",
    "            pass\n",
    "        else:\n",
    "            ll.append(itm[0])\n",
    "    # Second plot - unique MHC alleles in one chromosome\n",
    "    plt.figure(1, figsize=figSize)\n",
    "    plt.subplot(121)\n",
    "    for i, var in enumerate(ll):\n",
    "        ww = meanResult[0][meanResult[0]['VAR'] == var]\n",
    "        plt.errorbar(ww['VARX']+k[i], ww[indx[0]]*yScale, ww[indx[1]]*yScale, lw=2, marker='o', ms=10,\n",
    "                     color=clr[i], label=legd[i])\n",
    "    plt.title(r'           $\\alpha = 0.02$', fontsize=titleFS, loc='left')\n",
    "    plt.xlabel(str(x_label), fontsize=FS)\n",
    "    plt.ylabel(str(y_label), fontsize=FS)\n",
    "    plt.ylim((0, yLim*yScale))\n",
    "    plt.xscale(logsc)\n",
    "    plt.tick_params(axis='both', labelsize=annoSize, bottom=False)\n",
    "    plt.xticks(np.arange(70), xtix, fontsize=FS+2)\n",
    "    plt.yticks(yTikz)\n",
    "    plt.grid(axis='y')\n",
    "    plt.legend(frameon=False, loc=legdLoc, fontsize=FS-2)\n",
    "    # Two\n",
    "    ll = []\n",
    "    for itm in meanResult[1]:\n",
    "        if itm[0] in ll:\n",
    "            pass\n",
    "        else:\n",
    "            ll.append(itm[0])\n",
    "    # Second plot - unique MHC alleles in one chromosome\n",
    "    plt.subplot(122)\n",
    "    for i, var in enumerate(ll):\n",
    "        ww = meanResult[1][meanResult[1]['VAR'] == var]\n",
    "        plt.errorbar(ww['VARX']+k[i], ww[indx[0]]*yScale, ww[indx[1]]*yScale, lw=2, marker=\"o\", \n",
    "                     ms=10, color=clr[i+2], label=legd[i])\n",
    "    #plt.annotate(r'$\\alpha = 0.08$', xy=(2, 27), size=annoSize)\n",
    "    plt.title(r'           $\\alpha = 0.08$', fontsize=titleFS, loc='left')\n",
    "    plt.xlabel(str(x_label), fontsize=FS)\n",
    "    #plt.ylabel(\"average number of MHC copies\\nin an individual\",\n",
    "    #           fontsize=FS)\n",
    "    plt.ylim((0, yLim*yScale))\n",
    "    plt.xscale(logsc)\n",
    "    plt.tick_params(axis='both', labelsize=annoSize, bottom=False)\n",
    "    plt.xticks(np.arange(70), xtix, fontsize=FS+2)\n",
    "    plt.yticks(yTikz)\n",
    "    plt.grid(axis='y')\n",
    "    plt.legend(frameon=False, loc=legdLoc, fontsize=FS-2)\n",
    "    #plt.tight_layout(True)\n",
    "    figPDF =\"fig_\" + str(figNumb) + \".tiff\"\n",
    "    figPNG = \"fig_\" + str(figNumb) + \".png\"\n",
    "    plt.savefig(figPDF, bbox_inches='tight')\n",
    "    plt.savefig(figPNG, bbox_inches='tight')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "yTikz = np.arange(0, 1.01, 0.2)\n",
    "y_Label = \"average CV fitness of hosts x $10^{-2}$\"\n",
    "plotAvgMHCsInIndiv(meanResult, x_Label, y_Label, ('cvFitMean', 'cvFitSTD'), 2, 0.01, 100.,\n",
    "                   yTikz, 'upper right')\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def plotDotMeans(theData_002,theData_008):\n",
    "    \"\"\"Plots number of MHC alleles in population vs average number of MHC in\n",
    "    one chromosome.\"\"\"\n",
    "#    clrs = ['bo', 'go', 'ro', 'co', 'mo']  # , 'yo'  , 'ko', 'wo']\n",
    "    clrs = ['bo', 'go', 'ro', 'co', 'mo', 'yo']\n",
    "    clrs += ['bv', 'gv', 'rv', 'cv', 'mv', 'yv', 'kv', 'wv']\n",
    "    clrs += ['bo', 'go', 'ro', 'co', 'mo', 'yo', 'ko', 'wo']\n",
    "    muts = {1e-5: \"$10^{-5}$\", 5e-5: \"$5\\cdot10^{-5}$\"}\n",
    "    FS = 24\n",
    "    titleFS = FS+4\n",
    "    figSize = (24, 10)\n",
    "    annoSize = int(1*FS)\n",
    "    ll = []\n",
    "    comp = []\n",
    "    for itm in theData_002:\n",
    "        mts = (muts[itm['VAR']], int(itm['VARX']))\n",
    "        if (itm['VAR'], itm['VARX']) in ll:\n",
    "            pass\n",
    "        else:\n",
    "            ll.append((itm['VAR'], itm['VARX']))\n",
    "            comp.append(mts)\n",
    "    plt.figure(5, figsize=figSize)\n",
    "    plt.subplot(121)\n",
    "    k = 0\n",
    "    for ii, var in enumerate(ll):\n",
    "        ww = theData_002[theData_002['VAR'] == var[0]]\n",
    "        ww = ww[ww['VARX'] == var[1]]\n",
    "        lbl = str(comp[ii][0]) + \" ; \" + str(comp[ii][1])\n",
    "        plt.plot(ww['indvMean'], ww['meanAllel'], clrs[k], ms=8, label=lbl)\n",
    "        k += 1\n",
    "    plt.legend(loc='upper left',bbox_to_anchor=(0.11, -0.12), \n",
    "               numpoints=1, ncol=6, fontsize=20, frameon=False)\n",
    "    plt.ylabel(\"mean number of MHC alleles in population (PNV)\",fontsize=FS)\n",
    "    plt.xlabel(\"mean number of MHC alleles in one chromosome\", fontsize=FS)\n",
    "    plt.title(r'           $\\alpha = 0.02$', fontsize=titleFS, loc='left')\n",
    "    plt.xlim(left=0)\n",
    "    plt.xlim(right=30)\n",
    "    plt.ylim(bottom=0)\n",
    "    plt.tick_params(axis='both', labelsize=annoSize)\n",
    "    plt.grid(True)\n",
    "    # Second set of data\n",
    "    plt.subplot(122)\n",
    "    k = 0\n",
    "    for var in ll:\n",
    "        ww = theData_008[theData_008['VAR'] == var[0]]\n",
    "        ww = ww[ww['VARX'] == var[1]]\n",
    "        plt.plot(ww['indvMean'], ww['meanAllel'], clrs[k], ms=8)\n",
    "        k += 1\n",
    "    #plt.ylabel(\"mean number of MHC alleles in population\",\n",
    "    #           fontsize=FS)\n",
    "    plt.xlabel(\"mean number of MHC alleles in one chromosome\", fontsize=FS)\n",
    "    plt.title(r'           $\\alpha = 0.08$', fontsize=titleFS, loc='left')\n",
    "    plt.xlim(left=0)\n",
    "    plt.xlim(right=30)\n",
    "    plt.ylim(bottom=0)\n",
    "    plt.tick_params(axis='both', labelsize=annoSize)\n",
    "    plt.grid(True)\n",
    "    plt.savefig(\"fig_S5.png\", bbox_inches='tight')\n",
    "    plt.savefig(\"fig_S5.tiff\", bbox_inches='tight')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "plotDotMeans(theData_002, theData_008)\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "yTikz = np.arange(0, 211, 50)\n",
    "y_Label = \"mean number of MHCs in population\"\n",
    "plotAvgMHCsInIndiv(meanResult, x_Label, y_Label, ('meanAllel', 'stdAllel'), \n",
    "                   '4', 210, 1., yTikz, 'upper left')\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "yTikz = np.arange(0, 0.401, 0.1)\n",
    "y_Label = \"hosts average fitness (normalized)\"\n",
    "plotAvgMHCsInIndiv(meanResult, x_Label, y_Label, ('meanFitt', 'stdFitt'), 'SXX', 0.2,\n",
    "                   1., yTikz, 'upper right')\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "yTikz = np.arange(0, 0.401, 0.1)\n",
    "y_Label = \"presented pathogens (normalized)\"\n",
    "plotAvgMHCsInIndiv(meanResult, x_Label, y_Label, ('meanPato', 'stdPato'), 'SXXX', 0.2,\n",
    "                   1., yTikz, 'upper right')\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "plt.errorbar??"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
