<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml">
<head>
<meta http-equiv="Content-Type" content="text/xhtml;charset=UTF-8"/>
<meta http-equiv="X-UA-Compatible" content="IE=9"/>
<meta name="generator" content="Doxygen 1.8.11"/>
<title>Modelling MHC Evolutionary Dynamics: README.md Source File</title>
<link href="tabs.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="jquery.js"></script>
<script type="text/javascript" src="dynsections.js"></script>
<link href="navtree.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="resize.js"></script>
<script type="text/javascript" src="navtreedata.js"></script>
<script type="text/javascript" src="navtree.js"></script>
<script type="text/javascript">
  $(document).ready(initResizable);
  $(window).load(resizeHeight);
</script>
<link href="search/search.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="search/searchdata.js"></script>
<script type="text/javascript" src="search/search.js"></script>
<script type="text/javascript">
  $(document).ready(function() { init_search(); });
</script>
<link href="doxygen.css" rel="stylesheet" type="text/css" />
</head>
<body>
<div id="top"><!-- do not remove this div, it is closed by doxygen! -->
<div id="titlearea">
<table cellspacing="0" cellpadding="0">
 <tbody>
 <tr style="height: 56px;">
  <td id="projectalign" style="padding-left: 0.5em;">
   <div id="projectname">Modelling MHC Evolutionary Dynamics
   &#160;<span id="projectnumber">0.9.1</span>
   </div>
  </td>
 </tr>
 </tbody>
</table>
</div>
<!-- end header part -->
<!-- Generated by Doxygen 1.8.11 -->
<script type="text/javascript">
var searchBox = new SearchBox("searchBox", "search",false,'Search');
</script>
  <div id="navrow1" class="tabs">
    <ul class="tablist">
      <li><a href="index.html"><span>Main&#160;Page</span></a></li>
      <li><a href="pages.html"><span>Related&#160;Pages</span></a></li>
      <li><a href="annotated.html"><span>Classes</span></a></li>
      <li class="current"><a href="files.html"><span>Files</span></a></li>
      <li>
        <div id="MSearchBox" class="MSearchBoxInactive">
        <span class="left">
          <img id="MSearchSelect" src="search/mag_sel.png"
               onmouseover="return searchBox.OnSearchSelectShow()"
               onmouseout="return searchBox.OnSearchSelectHide()"
               alt=""/>
          <input type="text" id="MSearchField" value="Search" accesskey="S"
               onfocus="searchBox.OnSearchFieldFocus(true)" 
               onblur="searchBox.OnSearchFieldFocus(false)" 
               onkeyup="searchBox.OnSearchFieldChange(event)"/>
          </span><span class="right">
            <a id="MSearchClose" href="javascript:searchBox.CloseResultsWindow()"><img id="MSearchCloseImg" border="0" src="search/close.png" alt=""/></a>
          </span>
        </div>
      </li>
    </ul>
  </div>
  <div id="navrow2" class="tabs2">
    <ul class="tablist">
      <li><a href="files.html"><span>File&#160;List</span></a></li>
      <li><a href="globals.html"><span>File&#160;Members</span></a></li>
    </ul>
  </div>
</div><!-- top -->
<div id="side-nav" class="ui-resizable side-nav-resizable">
  <div id="nav-tree">
    <div id="nav-tree-contents">
      <div id="nav-sync" class="sync"></div>
    </div>
  </div>
  <div id="splitbar" style="-moz-user-select:none;" 
       class="ui-resizable-handle">
  </div>
</div>
<script type="text/javascript">
$(document).ready(function(){initNavTree('_r_e_a_d_m_e_8md.html','');});
</script>
<div id="doc-content">
<!-- window showing the filter options -->
<div id="MSearchSelectWindow"
     onmouseover="return searchBox.OnSearchSelectShow()"
     onmouseout="return searchBox.OnSearchSelectHide()"
     onkeydown="return searchBox.OnSearchSelectKey(event)">
</div>

<!-- iframe showing the search results (closed by default) -->
<div id="MSearchResultsWindow">
<iframe src="javascript:void(0)" frameborder="0" 
        name="MSearchResults" id="MSearchResults">
</iframe>
</div>

<div class="header">
  <div class="headertitle">
<div class="title">README.md</div>  </div>
</div><!--header-->
<div class="contents">
<a href="_r_e_a_d_m_e_8md.html">Go to the documentation of this file.</a><div class="fragment"><div class="line"><a name="l00001"></a><span class="lineno">    1</span>&#160;Evolution of MHC repertoire and number of copies:</div><div class="line"><a name="l00002"></a><span class="lineno">    2</span>&#160;==============================</div><div class="line"><a name="l00003"></a><span class="lineno">    3</span>&#160;</div><div class="line"><a name="l00004"></a><span class="lineno">    4</span>&#160;This work is part of the project [**Evolution of the number of copies of MHC genes: testing the optimality hypothesis and exploring alternatives.**](https://sites.google.com/site/evobiolab/projects) supported by the Polish Science Centre (NCN).</div><div class="line"><a name="l00005"></a><span class="lineno">    5</span>&#160;</div><div class="line"><a name="l00006"></a><span class="lineno">    6</span>&#160;This branch **MhcEvo2000** has the ability to be run parellel on multi-core CPUs using openMP with *#pragma omp parallel* preprocessor calls. **It aims at simulating the sexual selection component in MHC gene copy number variation**.</div><div class="line"><a name="l00007"></a><span class="lineno">    7</span>&#160;</div><div class="line"><a name="l00008"></a><span class="lineno">    8</span>&#160;This is a new development - the program used for research and papers is in the branch **InDeapthInfection**.</div><div class="line"><a name="l00009"></a><span class="lineno">    9</span>&#160;</div><div class="line"><a name="l00010"></a><span class="lineno">   10</span>&#160;Documentation:</div><div class="line"><a name="l00011"></a><span class="lineno">   11</span>&#160;-----------</div><div class="line"><a name="l00012"></a><span class="lineno">   12</span>&#160;</div><div class="line"><a name="l00013"></a><span class="lineno">   13</span>&#160;Go from here to *./Doc/html/* and launch the *./Doc/html/index.html* file in your web browser.</div><div class="line"><a name="l00014"></a><span class="lineno">   14</span>&#160;The documentation was created with Doxygen.</div><div class="line"><a name="l00015"></a><span class="lineno">   15</span>&#160;</div><div class="line"><a name="l00016"></a><span class="lineno">   16</span>&#160;Authors:</div><div class="line"><a name="l00017"></a><span class="lineno">   17</span>&#160;--------</div><div class="line"><a name="l00018"></a><span class="lineno">   18</span>&#160;Coded by **Piotr Bentkowski** - bentkowski.piotr@gmail.com  </div><div class="line"><a name="l00019"></a><span class="lineno">   19</span>&#160;in collaboration with **Jacek Radwan** - jradwan@amu.edu.pl</div><div class="line"><a name="l00020"></a><span class="lineno">   20</span>&#160;</div><div class="line"><a name="l00021"></a><span class="lineno">   21</span>&#160;Evolutionary Biology Group  </div><div class="line"><a name="l00022"></a><span class="lineno">   22</span>&#160;Faculty of Biology  </div><div class="line"><a name="l00023"></a><span class="lineno">   23</span>&#160;Adam Mickiewicz University  </div><div class="line"><a name="l00024"></a><span class="lineno">   24</span>&#160;ul. Umultowska 89  </div><div class="line"><a name="l00025"></a><span class="lineno">   25</span>&#160;61-614 Poznań  </div><div class="line"><a name="l00026"></a><span class="lineno">   26</span>&#160;Poland  </div><div class="line"><a name="l00027"></a><span class="lineno">   27</span>&#160;</div><div class="line"><a name="l00028"></a><span class="lineno">   28</span>&#160;www: https://sites.google.com/site/evobiolab/home  </div><div class="line"><a name="l00029"></a><span class="lineno">   29</span>&#160;PI : **Jacek Radwan** - jradwan@amu.edu.pl</div><div class="line"><a name="l00030"></a><span class="lineno">   30</span>&#160;</div><div class="line"><a name="l00031"></a><span class="lineno">   31</span>&#160;How to compile:</div><div class="line"><a name="l00032"></a><span class="lineno">   32</span>&#160;-----------</div><div class="line"><a name="l00033"></a><span class="lineno">   33</span>&#160;The program was written in [C++14 standard](https://en.wikipedia.org/wiki/C%2B%2B14) so if you are using GCC, then version gcc 4.8 seems to be the minimum requirement (I had 5.4 and 6.2). This program has some serious dependencies on [C++ Boost Libraries](http://www.boost.org/). Should compile smoothly on most modern GNU/Linux distros with Boost Libs installed. Having [Scons build tool](http://www.scons.org/) might be useful too. Basic compilation works fine on Ubuntu 16.04 LTS with mentioned packages installed by running the command:</div><div class="line"><a name="l00034"></a><span class="lineno">   34</span>&#160;```bash</div><div class="line"><a name="l00035"></a><span class="lineno">   35</span>&#160;g++ -O3 -o MHC_model main.cpp Gene.cpp Antigen.cpp Host.cpp Pathogen.cpp H2Pinteraction.cpp Random.cpp Tagging_system.cpp Environment.cpp DataHandler.cpp -fopenmp -std=c++14</div><div class="line"><a name="l00036"></a><span class="lineno">   36</span>&#160;```</div><div class="line"><a name="l00037"></a><span class="lineno">   37</span>&#160;</div><div class="line"><a name="l00038"></a><span class="lineno">   38</span>&#160;The code here can be also used as a toolbox for your research. You can stitch your own *main_yourown.cpp* file with your scenario and tailored procedures (we did so for our research) and compile it using Scons script which is part of this code bundle. To do so run:</div><div class="line"><a name="l00039"></a><span class="lineno">   39</span>&#160;```Bash</div><div class="line"><a name="l00040"></a><span class="lineno">   40</span>&#160;scons -Q scenario=&quot;Scenarios/main_default.cpp&quot;</div><div class="line"><a name="l00041"></a><span class="lineno">   41</span>&#160;```</div><div class="line"><a name="l00042"></a><span class="lineno">   42</span>&#160;replacing *Scenarios/main_default.cpp* with the path to your own custom *main_xyz.cpp*. Use our *main_core.cpp* file as the base to develop your scenarios of MHC evolution dynamics. WARNING! Using our Scons script overwrites the *main.cpp* file in the main source directory - always store your scenarios somewhere else than the *main.cpp* file!</div><div class="line"><a name="l00043"></a><span class="lineno">   43</span>&#160;</div><div class="line"><a name="l00044"></a><span class="lineno">   44</span>&#160;Sometimes your HPC Cluster is lame and old and it has fairly outdated compiler (e.g. gcc &lt; 4.8). Then you can statically link the libraries on your fancy brand new PC running the latest Linux distro and send the no-dependencies executable to cluster. Compile like this:</div><div class="line"><a name="l00045"></a><span class="lineno">   45</span>&#160;```bash</div><div class="line"><a name="l00046"></a><span class="lineno">   46</span>&#160;g++ -static -O3 -o MHC_model main.cpp Gene.cpp Antigen.cpp Host.cpp Pathogen.cpp H2Pinteraction.cpp Random.cpp Tagging_system.cpp Environment.cpp DataHandler.cpp -fopenmp -std=c++14</div><div class="line"><a name="l00047"></a><span class="lineno">   47</span>&#160;```</div><div class="line"><a name="l00048"></a><span class="lineno">   48</span>&#160;</div><div class="line"><a name="l00049"></a><span class="lineno">   49</span>&#160;Or run the Scons script:</div><div class="line"><a name="l00050"></a><span class="lineno">   50</span>&#160;```shell</div><div class="line"><a name="l00051"></a><span class="lineno">   51</span>&#160;scons -Q scenario=&quot;Scenarios/main_default.cpp&quot; linking=&quot;static&quot;</div><div class="line"><a name="l00052"></a><span class="lineno">   52</span>&#160;```</div><div class="line"><a name="l00053"></a><span class="lineno">   53</span>&#160;But it may throw some warnings… e.g. */usr/lib/gcc/x86_64-linux-gnu/5/libgomp.a(target.o): In function `gomp_target_init`: (.text+0xba): warning: Using &#39;dlopen&#39; in statically linked applications requires at runtime the shared libraries from the glibc version used for linking*.</div><div class="line"><a name="l00054"></a><span class="lineno">   54</span>&#160;</div><div class="line"><a name="l00055"></a><span class="lineno">   55</span>&#160;How to run:</div><div class="line"><a name="l00056"></a><span class="lineno">   56</span>&#160;-----------</div><div class="line"><a name="l00057"></a><span class="lineno">   57</span>&#160;</div><div class="line"><a name="l00058"></a><span class="lineno">   58</span>&#160;The program takes exactly 17 parameters. These are:</div><div class="line"><a name="l00059"></a><span class="lineno">   59</span>&#160;</div><div class="line"><a name="l00060"></a><span class="lineno">   60</span>&#160;*  **00** - Program&#39;s name</div><div class="line"><a name="l00061"></a><span class="lineno">   61</span>&#160;*  **01** - Number of threads program will try to use on a multi-core CPU. Giving 0 will make the program use all CPU cores available.</div><div class="line"><a name="l00062"></a><span class="lineno">   62</span>&#160;*  **02** - Number of bits in a gene.</div><div class="line"><a name="l00063"></a><span class="lineno">   63</span>&#160;*  **03** - Number of bits in an antigen.</div><div class="line"><a name="l00064"></a><span class="lineno">   64</span>&#160;*  **04** - Host population size.</div><div class="line"><a name="l00065"></a><span class="lineno">   65</span>&#160;*  **05** - Pathogen population size.</div><div class="line"><a name="l00066"></a><span class="lineno">   66</span>&#160;*  **06** - Number of pathogen species.</div><div class="line"><a name="l00067"></a><span class="lineno">   67</span>&#160;*  **07** - Number of genes in one host chromosome (they have two chromosomes) when the model is being initialised.</div><div class="line"><a name="l00068"></a><span class="lineno">   68</span>&#160;*  **08** - Number of pathogen generations per one host generation.</div><div class="line"><a name="l00069"></a><span class="lineno">   69</span>&#160;*  **09** - Number of host generations (effective length of model run).</div><div class="line"><a name="l00070"></a><span class="lineno">   70</span>&#160;*  **10** - Probability of mutation in hosts ([0,1] range).</div><div class="line"><a name="l00071"></a><span class="lineno">   71</span>&#160;*  **11** - Probability of mutation in pathogens ([0,1] range).</div><div class="line"><a name="l00072"></a><span class="lineno">   72</span>&#160;*  **12** - The heterozygote advantage / lack of advantage mode. It has to be 10 for heterozygote advantage or 11 for lack of thereof.</div><div class="line"><a name="l00073"></a><span class="lineno">   73</span>&#160;*  **13** - Probability of deleting a gene in the host ([0,1] range).</div><div class="line"><a name="l00074"></a><span class="lineno">   74</span>&#160;*  **14** - Probability of duplicating a gene in the host ([0,1] range).</div><div class="line"><a name="l00075"></a><span class="lineno">   75</span>&#160;*  **15** - Maximal number of genes permitted in one host chromosome.</div><div class="line"><a name="l00076"></a><span class="lineno">   76</span>&#160;*  **16** - Number of sexual partners an individual checks out before selecting one for mating.</div><div class="line"><a name="l00077"></a><span class="lineno">   77</span>&#160;*  **17** - Alpha factor for the host fitness function ([0,1] range).</div><div class="line"><a name="l00078"></a><span class="lineno">   78</span>&#160;</div><div class="line"><a name="l00079"></a><span class="lineno">   79</span>&#160; Historically we also had one more argument, that can be added if you want:</div><div class="line"><a name="l00080"></a><span class="lineno">   80</span>&#160; - 18 - Fraction of the antigen&#39;s bits which are forbidden from changing (a.k.a. *No Mutation Bits*).</div><div class="line"><a name="l00081"></a><span class="lineno">   81</span>&#160;</div><div class="line"><a name="l00082"></a><span class="lineno">   82</span>&#160;This program can recognise simple errors in the argument list (a probability value out of [0,1] range, negative</div><div class="line"><a name="l00083"></a><span class="lineno">   83</span>&#160;values when only positive are allowed etc.), but will not recognise when they don&#39;t make a &#39;biological&#39; sense.</div><div class="line"><a name="l00084"></a><span class="lineno">   84</span>&#160;</div><div class="line"><a name="l00085"></a><span class="lineno">   85</span>&#160;Example run call:</div><div class="line"><a name="l00086"></a><span class="lineno">   86</span>&#160;```shell</div><div class="line"><a name="l00087"></a><span class="lineno">   87</span>&#160;./main_default 0 16 6000 1000 8000 8 1 10 3000 0.0001 5e-05 10 0.001 0.001 50 10 0.02</div><div class="line"><a name="l00088"></a><span class="lineno">   88</span>&#160;```</div><div class="line"><a name="l00089"></a><span class="lineno">   89</span>&#160;There are 8 pathogen species here, each has 1000 individuals, hence the total pathogen population size is 8000.</div><div class="line"><a name="l00090"></a><span class="lineno">   90</span>&#160;</div><div class="line"><a name="l00091"></a><span class="lineno">   91</span>&#160;The output and data visualisation:</div><div class="line"><a name="l00092"></a><span class="lineno">   92</span>&#160;-----------</div><div class="line"><a name="l00093"></a><span class="lineno">   93</span>&#160;</div><div class="line"><a name="l00094"></a><span class="lineno">   94</span>&#160;Program produces a number of text files containing desired data. The file *InputParameters.json* contains the values of parameters fed to the model. The file *HostsGeneDivers.csv* contains the main statistics of the model, including the time stamp used in the visualisation process of the other files with data. This two files are the basics to analyse the simulation output. Other files are optional and they can be switched off by commenting out some lines in the main function (*main.cpp* file) what should speed up the program runtime.</div><div class="line"><a name="l00095"></a><span class="lineno">   95</span>&#160;</div><div class="line"><a name="l00096"></a><span class="lineno">   96</span>&#160;All the files are:</div><div class="line"><a name="l00097"></a><span class="lineno">   97</span>&#160;</div><div class="line"><a name="l00098"></a><span class="lineno">   98</span>&#160;*  ***InputParameters.json*** - contains run&#39;s parametrisation in JSON text format</div><div class="line"><a name="l00099"></a><span class="lineno">   99</span>&#160;*  ***HostsGeneDivers.csv*** - contain basic statistics of the output for host genomes</div><div class="line"><a name="l00100"></a><span class="lineno">  100</span>&#160;*  ***HostGenomesFile.[t].csv*** - contains all the genomes of all the cells for host population in time *t*. There can be more then one file like this for different time snapshots.</div><div class="line"><a name="l00101"></a><span class="lineno">  101</span>&#160;*  ***PathoGenomesFile.[t].csv*** - contains all the genomes of all the cells for pathogen population in time *t*. There can be more then one file like this for different time snapshots.</div><div class="line"><a name="l00102"></a><span class="lineno">  102</span>&#160;*  ***HostGeneNumbTotal.csv*** - total number of genes in each individual host in each time step (linked to *HostMHCsNumbUniq.csv*, that each column in *HostGeneNumbTotal.csv* is represents the same cells as in *HostMHCsNumbUniq.csv*).</div><div class="line"><a name="l00103"></a><span class="lineno">  103</span>&#160;*  ***HostMHCsNumbUniq.csv*** - number of unique MHC alleles in each individual host in each time step (linked to*HostGeneNumbTotal.csv*, that each column in *HostGeneNumbTotal.csv* is represents the same cells as in *HostMHCsNumbUniq.csv*).</div><div class="line"><a name="l00104"></a><span class="lineno">  104</span>&#160;*  ***NoMutationInPathoList.csv*** - list of conserved antigen sites. Each line contains one pathogen species, each number indicates the index of a &quot;no-mutation&quot; site in this species antigen. When empty it means there is no mutation restrictions.</div><div class="line"><a name="l00105"></a><span class="lineno">  105</span>&#160;*  ***NumberOfMhcAfterMating.csv*** -number of the unique MHC types in each individual host in each time step after the mating procedure took place.</div><div class="line"><a name="l00106"></a><span class="lineno">  106</span>&#160;*  ***NumberOfMhcBeforeMating.csv*** - number of the unique MHC types in each individual host in each time step before the mating procedure took place.</div><div class="line"><a name="l00107"></a><span class="lineno">  107</span>&#160;*  ***NumberOfMhcInMother.csv*** - number of the unique MHC types in each individual host that is selecting a partner (a.k.a. &quot;mother&quot;) in each time step during mating procedure. Each individual has a corresponding partner at the same index in the file *NumberOfMhcInFather.csv*.</div><div class="line"><a name="l00108"></a><span class="lineno">  108</span>&#160;*  ***NumberOfMhcInFather.csv*** - number of the unique MHC types in each individual host that has been selected as a mating (a.k.a. &quot;father&quot;) in each time step during mating procedure. Each individual has a corresponding partner at the same index in the file *NumberOfMhcInMother.csv*.</div><div class="line"><a name="l00109"></a><span class="lineno">  109</span>&#160;*  ***PresentedPathogenNumbers.csv*** - number of presented pathogens by each individual in each time step.</div><div class="line"><a name="l00110"></a><span class="lineno">  110</span>&#160;</div><div class="line"><a name="l00111"></a><span class="lineno">  111</span>&#160;</div><div class="line"><a name="l00112"></a><span class="lineno">  112</span>&#160;Visualisation is done using Python 3.6 scripts containing a a lot of calls to Numpy, Matplotlib and other scientific Python libraries. You may wish to consider using the [Python Anaconda](https://www.anaconda.com/download/) for your Pythonic endeavours. Visualisation and stats scripts can be found in *PyScripts* directory.</div><div class="line"><a name="l00113"></a><span class="lineno">  113</span>&#160;</div><div class="line"><a name="l00114"></a><span class="lineno">  114</span>&#160;To load the environment for output analysis run:</div><div class="line"><a name="l00115"></a><span class="lineno">  115</span>&#160;```shell</div><div class="line"><a name="l00116"></a><span class="lineno">  116</span>&#160;conda env create -f mhcEvoCondaEnv.yaml</div><div class="line"><a name="l00117"></a><span class="lineno">  117</span>&#160;```</div><div class="line"><a name="l00118"></a><span class="lineno">  118</span>&#160;file `mhcEvoCondaEnv.yaml` can be found in *PyScripts* directory.</div></div><!-- fragment --></div><!-- contents -->
</div><!-- doc-content -->
<!-- start footer part -->
<div id="nav-path" class="navpath"><!-- id is needed for treeview function! -->
  <ul>
    <li class="navelem"><a class="el" href="_r_e_a_d_m_e_8md.html">README.md</a></li>
    <li class="footer">Generated on Sun Jan 20 2019 23:54:13 for Modelling MHC Evolutionary Dynamics by
    <a href="http://www.doxygen.org/index.html">
    <img class="footer" src="doxygen.png" alt="doxygen"/></a> 1.8.11 </li>
  </ul>
</div>
</body>
</html>
