<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml">
<head>
<meta http-equiv="Content-Type" content="text/xhtml;charset=UTF-8"/>
<meta http-equiv="X-UA-Compatible" content="IE=9"/>
<meta name="generator" content="Doxygen 1.8.7"/>
<title>Modelling MHC Evolutionary Dynamics: Environment Class Reference</title>
<link href="tabs.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="jquery.js"></script>
<script type="text/javascript" src="dynsections.js"></script>
<link href="navtree.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="resize.js"></script>
<script type="text/javascript" src="navtree.js"></script>
<script type="text/javascript">
  $(document).ready(initResizable);
  $(window).load(resizeHeight);
</script>
<link href="search/search.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="search/search.js"></script>
<script type="text/javascript">
  $(document).ready(function() { searchBox.OnSelectItem(0); });
</script>
<link href="doxygen.css" rel="stylesheet" type="text/css" />
</head>
<body>
<div id="top"><!-- do not remove this div, it is closed by doxygen! -->
<div id="titlearea">
<table cellspacing="0" cellpadding="0">
 <tbody>
 <tr style="height: 56px;">
  <td style="padding-left: 0.5em;">
   <div id="projectname">Modelling MHC Evolutionary Dynamics
   &#160;<span id="projectnumber">0.9</span>
   </div>
  </td>
 </tr>
 </tbody>
</table>
</div>
<!-- end header part -->
<!-- Generated by Doxygen 1.8.7 -->
<script type="text/javascript">
var searchBox = new SearchBox("searchBox", "search",false,'Search');
</script>
  <div id="navrow1" class="tabs">
    <ul class="tablist">
      <li><a href="index.html"><span>Main&#160;Page</span></a></li>
      <li><a href="pages.html"><span>Related&#160;Pages</span></a></li>
      <li class="current"><a href="annotated.html"><span>Classes</span></a></li>
      <li><a href="files.html"><span>Files</span></a></li>
      <li>
        <div id="MSearchBox" class="MSearchBoxInactive">
        <span class="left">
          <img id="MSearchSelect" src="search/mag_sel.png"
               onmouseover="return searchBox.OnSearchSelectShow()"
               onmouseout="return searchBox.OnSearchSelectHide()"
               alt=""/>
          <input type="text" id="MSearchField" value="Search" accesskey="S"
               onfocus="searchBox.OnSearchFieldFocus(true)" 
               onblur="searchBox.OnSearchFieldFocus(false)" 
               onkeyup="searchBox.OnSearchFieldChange(event)"/>
          </span><span class="right">
            <a id="MSearchClose" href="javascript:searchBox.CloseResultsWindow()"><img id="MSearchCloseImg" border="0" src="search/close.png" alt=""/></a>
          </span>
        </div>
      </li>
    </ul>
  </div>
  <div id="navrow2" class="tabs2">
    <ul class="tablist">
      <li><a href="annotated.html"><span>Class&#160;List</span></a></li>
      <li><a href="classes.html"><span>Class&#160;Index</span></a></li>
      <li><a href="functions.html"><span>Class&#160;Members</span></a></li>
    </ul>
  </div>
</div><!-- top -->
<div id="side-nav" class="ui-resizable side-nav-resizable">
  <div id="nav-tree">
    <div id="nav-tree-contents">
      <div id="nav-sync" class="sync"></div>
    </div>
  </div>
  <div id="splitbar" style="-moz-user-select:none;" 
       class="ui-resizable-handle">
  </div>
</div>
<script type="text/javascript">
$(document).ready(function(){initNavTree('class_environment.html','');});
</script>
<div id="doc-content">
<!-- window showing the filter options -->
<div id="MSearchSelectWindow"
     onmouseover="return searchBox.OnSearchSelectShow()"
     onmouseout="return searchBox.OnSearchSelectHide()"
     onkeydown="return searchBox.OnSearchSelectKey(event)">
<a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(0)"><span class="SelectionMark">&#160;</span>All</a><a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(1)"><span class="SelectionMark">&#160;</span>Classes</a><a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(2)"><span class="SelectionMark">&#160;</span>Files</a><a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(3)"><span class="SelectionMark">&#160;</span>Functions</a><a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(4)"><span class="SelectionMark">&#160;</span>Variables</a><a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(5)"><span class="SelectionMark">&#160;</span>Typedefs</a><a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(6)"><span class="SelectionMark">&#160;</span>Pages</a></div>

<!-- iframe showing the search results (closed by default) -->
<div id="MSearchResultsWindow">
<iframe src="javascript:void(0)" frameborder="0" 
        name="MSearchResults" id="MSearchResults">
</iframe>
</div>

<div class="header">
  <div class="summary">
<a href="#pub-methods">Public Member Functions</a> &#124;
<a href="class_environment-members.html">List of all members</a>  </div>
  <div class="headertitle">
<div class="title">Environment Class Reference</div>  </div>
</div><!--header-->
<div class="contents">

<p>Core class. Stores and handles the environment object that is the place where all the things are happening. It contains host and pathogen populations stored in separate STL Vectors.  
 <a href="class_environment.html#details">More...</a></p>

<p><code>#include &lt;<a class="el" href="_environment_8h_source.html">Environment.h</a>&gt;</code></p>
<table class="memberdecls">
<tr class="heading"><td colspan="2"><h2 class="groupheader"><a name="pub-methods"></a>
Public Member Functions</h2></td></tr>
<tr class="memitem:a8b427c4448d8b7536666837521b9e83d"><td class="memItemLeft" align="right" valign="top">&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_environment.html#a8b427c4448d8b7536666837521b9e83d">Environment</a> ()</td></tr>
<tr class="separator:a8b427c4448d8b7536666837521b9e83d"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a8e294735187880dd3d59be10c425b29d"><td class="memItemLeft" align="right" valign="top">virtual&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_environment.html#a8e294735187880dd3d59be10c425b29d">~Environment</a> ()</td></tr>
<tr class="separator:a8e294735187880dd3d59be10c425b29d"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a559d186796a3f21fcaccce6cbbc6fe89"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_environment.html#a559d186796a3f21fcaccce6cbbc6fe89">setNoMutsVector</a> (int numb_of_species, int antigen_size, double fixedAntigenFrac)</td></tr>
<tr class="memdesc:a559d186796a3f21fcaccce6cbbc6fe89"><td class="mdescLeft">&#160;</td><td class="mdescRight">Core method. It defines "no mutation sites" of the antigen for all individual pathogen species in the simulation. It should be run only ones per simulation.  <a href="#a559d186796a3f21fcaccce6cbbc6fe89">More...</a><br /></td></tr>
<tr class="separator:a559d186796a3f21fcaccce6cbbc6fe89"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a95cc0c7ecaa8fad54c2d62328134c73e"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_environment.html#a95cc0c7ecaa8fad54c2d62328134c73e">setNoMutsVecInFours</a> (int numb_of_species, int antigen_size, double fixedAntigenFrac)</td></tr>
<tr class="memdesc:a95cc0c7ecaa8fad54c2d62328134c73e"><td class="mdescLeft">&#160;</td><td class="mdescRight">Core method. It defines "no mutation sites" in 4-bit-long packages of the antigen for all individual pathogen species in the simulation. It should be run only ones per simulation.  <a href="#a95cc0c7ecaa8fad54c2d62328134c73e">More...</a><br /></td></tr>
<tr class="separator:a95cc0c7ecaa8fad54c2d62328134c73e"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a558656ec1b4948fbbfee1358c79c67d0"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_environment.html#a558656ec1b4948fbbfee1358c79c67d0">setNoMutsVecFourClads</a> (int numb_of_species, int antigen_size, double fixedAntigenFrac)</td></tr>
<tr class="memdesc:a558656ec1b4948fbbfee1358c79c67d0"><td class="mdescLeft">&#160;</td><td class="mdescRight">Core method. It defines "no mutation sites" of the antigen for all individual pathogen species in the simulation. It should be run only ones per simulation.  <a href="#a558656ec1b4948fbbfee1358c79c67d0">More...</a><br /></td></tr>
<tr class="separator:a558656ec1b4948fbbfee1358c79c67d0"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ae6c1788a9b69c76b657d02e38548e5ce"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_environment.html#ae6c1788a9b69c76b657d02e38548e5ce">setHostRandomPopulation</a> (int pop_size, int gene_size, int chrom_size, int timeStamp)</td></tr>
<tr class="memdesc:ae6c1788a9b69c76b657d02e38548e5ce"><td class="mdescLeft">&#160;</td><td class="mdescRight">Core method. Initializes a vector containing host population.  <a href="#ae6c1788a9b69c76b657d02e38548e5ce">More...</a><br /></td></tr>
<tr class="separator:ae6c1788a9b69c76b657d02e38548e5ce"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a3b828fcd48c6ef8b0d8d7ea61de3817b"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_environment.html#a3b828fcd48c6ef8b0d8d7ea61de3817b">setHostRandomPopulation</a> (int pop_size, int gene_size, int chrom_size_lower, int chrom_size_uper, int timeStamp)</td></tr>
<tr class="memdesc:a3b828fcd48c6ef8b0d8d7ea61de3817b"><td class="mdescLeft">&#160;</td><td class="mdescRight">Core method. Initializes a vector containing host population.  <a href="#a3b828fcd48c6ef8b0d8d7ea61de3817b">More...</a><br /></td></tr>
<tr class="separator:a3b828fcd48c6ef8b0d8d7ea61de3817b"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:aeadba01513bfdaef3f247558a6e5ce1d"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_environment.html#aeadba01513bfdaef3f247558a6e5ce1d">setHostClonalPopulation</a> (int pop_size, int gene_size, int chrom_size, int timeStamp)</td></tr>
<tr class="memdesc:aeadba01513bfdaef3f247558a6e5ce1d"><td class="mdescLeft">&#160;</td><td class="mdescRight">Core method. Initializes a vector containing clonal host population.  <a href="#aeadba01513bfdaef3f247558a6e5ce1d">More...</a><br /></td></tr>
<tr class="separator:aeadba01513bfdaef3f247558a6e5ce1d"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a03d6f594caf898481ef784aafa9b6ed8"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_environment.html#a03d6f594caf898481ef784aafa9b6ed8">setPathoPopulatioUniformGenome</a> (int pop_size, int gene_size, int chrom_size, int numb_of_species, int mhcSize, int timeStamp, double fixedAntigenFrac)</td></tr>
<tr class="memdesc:a03d6f594caf898481ef784aafa9b6ed8"><td class="mdescLeft">&#160;</td><td class="mdescRight">Core method. Initializes the pathogen population.  <a href="#a03d6f594caf898481ef784aafa9b6ed8">More...</a><br /></td></tr>
<tr class="separator:a03d6f594caf898481ef784aafa9b6ed8"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a34b326e2e2880881e48ddf9312ff0bc0"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_environment.html#a34b326e2e2880881e48ddf9312ff0bc0">setPathoPopulatioDivSpecies</a> (int pop_size, int gene_size, int chrom_size, int numb_of_species, int mhcSize, int timeStamp, double fixedAntigenFrac)</td></tr>
<tr class="memdesc:a34b326e2e2880881e48ddf9312ff0bc0"><td class="mdescLeft">&#160;</td><td class="mdescRight">Core method. Initializes the pathogen population.  <a href="#a34b326e2e2880881e48ddf9312ff0bc0">More...</a><br /></td></tr>
<tr class="separator:a34b326e2e2880881e48ddf9312ff0bc0"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:aa96a77a7b6544503d61e26149c0d4cca"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_environment.html#aa96a77a7b6544503d61e26149c0d4cca">setPathoPopulatioDistincSpp</a> (int pop_size, int antigenSize, int chrom_size, int numb_of_species, int mhcSize, int timeStamp, double fixedAntigenFrac)</td></tr>
<tr class="memdesc:aa96a77a7b6544503d61e26149c0d4cca"><td class="mdescLeft">&#160;</td><td class="mdescRight">Core method. Initializes the pathogen population.  <a href="#aa96a77a7b6544503d61e26149c0d4cca">More...</a><br /></td></tr>
<tr class="separator:aa96a77a7b6544503d61e26149c0d4cca"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ab8e36a6f12db407caf9a63cbeae62a16"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_environment.html#ab8e36a6f12db407caf9a63cbeae62a16">setPathoPopulationFourClades</a> (int pop_size, int antigenSize, int chrom_size, int numb_of_species, int mhcSize, int timeStamp, double fixedAntigenFrac)</td></tr>
<tr class="memdesc:ab8e36a6f12db407caf9a63cbeae62a16"><td class="mdescLeft">&#160;</td><td class="mdescRight">Core method. Initializes the pathogen population.  <a href="#ab8e36a6f12db407caf9a63cbeae62a16">More...</a><br /></td></tr>
<tr class="separator:ab8e36a6f12db407caf9a63cbeae62a16"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:aecd592b734359557235bb0a6980f77f9"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_environment.html#aecd592b734359557235bb0a6980f77f9">infectOneFromOneSpecHetero</a> ()</td></tr>
<tr class="memdesc:aecd592b734359557235bb0a6980f77f9"><td class="mdescLeft">&#160;</td><td class="mdescRight">Core method. Iterates through the host population and the parasite population to "infect" the hosts with parasites. With heterozygote advantage and it permits a pathogen species to infect a host only ONES.  <a href="#aecd592b734359557235bb0a6980f77f9">More...</a><br /></td></tr>
<tr class="separator:aecd592b734359557235bb0a6980f77f9"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a670a9eaf38f556ab8ca096bc9a1433a5"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_environment.html#a670a9eaf38f556ab8ca096bc9a1433a5">infectEveryOne</a> (int simil_mesure)</td></tr>
<tr class="separator:a670a9eaf38f556ab8ca096bc9a1433a5"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ae63fad9a51f8700cf96cd3019f5a3bec"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_environment.html#ae63fad9a51f8700cf96cd3019f5a3bec">calculateHostsFitnessPerGene</a> ()</td></tr>
<tr class="memdesc:ae63fad9a51f8700cf96cd3019f5a3bec"><td class="mdescLeft">&#160;</td><td class="mdescRight">Core method. Iterates through the host population and calculates the Fitness for each single individual by calling <a class="el" href="class_host.html#a6079285b3b8bb5280d52500e6b7fc7ef" title="Core method. Calculates host individual fitness as the number of exposed pathogens divided by the gen...">Host::calculateFitnessAccChromSize()</a>, which takes the number of MHC genes under account.  <a href="#ae63fad9a51f8700cf96cd3019f5a3bec">More...</a><br /></td></tr>
<tr class="separator:ae63fad9a51f8700cf96cd3019f5a3bec"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a2d11b6356c2d62e7ebf4c97115118c5c"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_environment.html#a2d11b6356c2d62e7ebf4c97115118c5c">calculateHostsFitnessPlainPresent</a> ()</td></tr>
<tr class="memdesc:a2d11b6356c2d62e7ebf4c97115118c5c"><td class="mdescLeft">&#160;</td><td class="mdescRight">Core method. Iterates through the host population and calculates the Fitness for each single individual by calling Host::calculateHostsFitnessPlainInfect(), which is the plain-and-lame sum of presented pathogens.  <a href="#a2d11b6356c2d62e7ebf4c97115118c5c">More...</a><br /></td></tr>
<tr class="separator:a2d11b6356c2d62e7ebf4c97115118c5c"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a596cf4db11a6e6356211a45540e04c7d"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_environment.html#a596cf4db11a6e6356211a45540e04c7d">calculateHostsFitnessForDrift</a> ()</td></tr>
<tr class="memdesc:a596cf4db11a6e6356211a45540e04c7d"><td class="mdescLeft">&#160;</td><td class="mdescRight">Core method. Iterates through the host population and calculates the Fitness for each single individual by calling <a class="el" href="class_host.html#a614430aea49534f29b63232414cf4fcd" title="Core method. Sets fitness as 1, a fixed value to simulate genetic drift. ">Host::calculateFitnessForDrift()</a>, which assigns "1" for each cell to make the genetic driff work.  <a href="#a596cf4db11a6e6356211a45540e04c7d">More...</a><br /></td></tr>
<tr class="separator:a596cf4db11a6e6356211a45540e04c7d"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a0994a6e48ab6a97cbff9feb68ce9dcd3"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_environment.html#a0994a6e48ab6a97cbff9feb68ce9dcd3">calculateHostsFitnessAlphaXsqr</a> (double alpha)</td></tr>
<tr class="memdesc:a0994a6e48ab6a97cbff9feb68ce9dcd3"><td class="mdescLeft">&#160;</td><td class="mdescRight">Core method. Iterates through the host population and calculates the Fitness for each single individual by calling Host::calculateHostsFitnessAlphaXsqr(), which uses one over the square on number of genes as a fitness cost.  <a href="#a0994a6e48ab6a97cbff9feb68ce9dcd3">More...</a><br /></td></tr>
<tr class="separator:a0994a6e48ab6a97cbff9feb68ce9dcd3"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a5f156d009868bce000e653b25fe53967"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_environment.html#a5f156d009868bce000e653b25fe53967">calculateHostsFitnessExpScaling</a> (double alpha)</td></tr>
<tr class="memdesc:a5f156d009868bce000e653b25fe53967"><td class="mdescLeft">&#160;</td><td class="mdescRight">Core method. Iterates through the host population and calculates the Fitness for each single individual by calling <a class="el" href="class_host.html#a9134661842de046453acc0342ef74eb6" title="Core method. Calculates host individual fitness in proportion to the Gaussian function of the number ...">Host::calculateFitnessExpFunc()</a>, which uses a Gaussian function to accommodate the costs of having lots of genes.  <a href="#a5f156d009868bce000e653b25fe53967">More...</a><br /></td></tr>
<tr class="separator:a5f156d009868bce000e653b25fe53967"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:aad5bb2197b43d307428356c123e34f43"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_environment.html#aad5bb2197b43d307428356c123e34f43">calculateHostsFitnessExpScalingUniqAlleles</a> (double alpha)</td></tr>
<tr class="memdesc:aad5bb2197b43d307428356c123e34f43"><td class="mdescLeft">&#160;</td><td class="mdescRight">Core method. Iterates through the host population and calculates the Fitness for each single individual by calling <a class="el" href="class_host.html#ad185a5e02c8ed0930e4d400e9c0e6a9a" title="Core method. Calculates host individual fitness in proportion to the Gaussian function of the number ...">Host::calculateFitnessExpFuncUniqAlleles()</a>, which uses a Gaussian function to accommodate the costs of having lots of unique MHC alleles in chromosomes.  <a href="#aad5bb2197b43d307428356c123e34f43">More...</a><br /></td></tr>
<tr class="separator:aad5bb2197b43d307428356c123e34f43"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a56c39a47a881880bd36c3b8c5f0dbe8a"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_environment.html#a56c39a47a881880bd36c3b8c5f0dbe8a">selectAndReprodHostsAddOffspring</a> ()</td></tr>
<tr class="memdesc:a56c39a47a881880bd36c3b8c5f0dbe8a"><td class="mdescLeft">&#160;</td><td class="mdescRight">Core method. Forms the next generation of hosts using the fitness proportionate selection method.  <a href="#a56c39a47a881880bd36c3b8c5f0dbe8a">More...</a><br /></td></tr>
<tr class="separator:a56c39a47a881880bd36c3b8c5f0dbe8a"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a09af2890db9eb1a02069adfb8836aa47"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_environment.html#a09af2890db9eb1a02069adfb8836aa47">selectAndReprodHostsReplace</a> ()</td></tr>
<tr class="memdesc:a09af2890db9eb1a02069adfb8836aa47"><td class="mdescLeft">&#160;</td><td class="mdescRight">Core method. Forms the next generation of hosts using the fitness proportionate selection method. Replaces the old population with a new one.  <a href="#a09af2890db9eb1a02069adfb8836aa47">More...</a><br /></td></tr>
<tr class="separator:a09af2890db9eb1a02069adfb8836aa47"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ad7f96a8620fa9df1601c613130b275e0"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_environment.html#ad7f96a8620fa9df1601c613130b275e0">selectAndReproducePathoFlexPopSizes</a> ()</td></tr>
<tr class="memdesc:ad7f96a8620fa9df1601c613130b275e0"><td class="mdescLeft">&#160;</td><td class="mdescRight">Core method. Forms the next generation of hosts using the fitness proportionate selection method. It can adjust species population sizes in proportion to its individuals fitness values keeping the total number of pathogens fixed.  <a href="#ad7f96a8620fa9df1601c613130b275e0">More...</a><br /></td></tr>
<tr class="separator:ad7f96a8620fa9df1601c613130b275e0"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a90177792fb03087e1fc64dca78f7e312"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_environment.html#a90177792fb03087e1fc64dca78f7e312">selectAndReproducePathoFixedPopSizes</a> ()</td></tr>
<tr class="memdesc:a90177792fb03087e1fc64dca78f7e312"><td class="mdescLeft">&#160;</td><td class="mdescRight">Core method. Forms the next generation of hosts using the fitness proportionate selection method. It keeps population sizes of different pathogens species at a fixed number.  <a href="#a90177792fb03087e1fc64dca78f7e312">More...</a><br /></td></tr>
<tr class="separator:a90177792fb03087e1fc64dca78f7e312"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:adff3e1c9799ef7a3bc8d7b988bda3381"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_environment.html#adff3e1c9799ef7a3bc8d7b988bda3381">clearHostInfectionsData</a> ()</td></tr>
<tr class="memdesc:adff3e1c9799ef7a3bc8d7b988bda3381"><td class="mdescLeft">&#160;</td><td class="mdescRight">. Core method. Clears information about infection and fitness in the whole host population.  <a href="#adff3e1c9799ef7a3bc8d7b988bda3381">More...</a><br /></td></tr>
<tr class="separator:adff3e1c9799ef7a3bc8d7b988bda3381"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:af43ff3d50276e23f42895c6c7ddac69c"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_environment.html#af43ff3d50276e23f42895c6c7ddac69c">clearPathoInfectionData</a> ()</td></tr>
<tr class="memdesc:af43ff3d50276e23f42895c6c7ddac69c"><td class="mdescLeft">&#160;</td><td class="mdescRight">. Core method. Clears information about infection and fitness in the whole pathogen population.  <a href="#af43ff3d50276e23f42895c6c7ddac69c">More...</a><br /></td></tr>
<tr class="separator:af43ff3d50276e23f42895c6c7ddac69c"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a54bc672882bea05101d65942b35cdc63"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_environment.html#a54bc672882bea05101d65942b35cdc63">mutatePathogens</a> (double mut_probabl, int mhcSize, int timeStamp)</td></tr>
<tr class="memdesc:a54bc672882bea05101d65942b35cdc63"><td class="mdescLeft">&#160;</td><td class="mdescRight">Core method. Iterates through the all genes of the pathogen population and performs mutations in genes with a given probability.  <a href="#a54bc672882bea05101d65942b35cdc63">More...</a><br /></td></tr>
<tr class="separator:a54bc672882bea05101d65942b35cdc63"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:af56f92817b21be862325edb48f291037"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_environment.html#af56f92817b21be862325edb48f291037">mutatePathogensWithRestric</a> (double mut_probabl, int mhcSize, int timeStamp)</td></tr>
<tr class="memdesc:af56f92817b21be862325edb48f291037"><td class="mdescLeft">&#160;</td><td class="mdescRight">Core method. Iterates through the all genes of the pathogen population and performs mutations in genes with a given probability. But some positions in the bit-string (gene) are not allowed to change.  <a href="#af56f92817b21be862325edb48f291037">More...</a><br /></td></tr>
<tr class="separator:af56f92817b21be862325edb48f291037"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ad3d9445c714e728df229eabbe425fff1"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_environment.html#ad3d9445c714e728df229eabbe425fff1">mutateHosts</a> (double mut_probabl, int timeStamp)</td></tr>
<tr class="memdesc:ad3d9445c714e728df229eabbe425fff1"><td class="mdescLeft">&#160;</td><td class="mdescRight">Core method. Iterates through the all genes of the host population and performs mutations in genes with a given probability.  <a href="#ad3d9445c714e728df229eabbe425fff1">More...</a><br /></td></tr>
<tr class="separator:ad3d9445c714e728df229eabbe425fff1"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:af7fc9c6750f847a8260df37927646583"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_environment.html#af7fc9c6750f847a8260df37927646583">mutateHostsWithDelDupl</a> (double mut_probabl, double del, double dupl, unsigned int maxGene, int timeStamp)</td></tr>
<tr class="memdesc:af7fc9c6750f847a8260df37927646583"><td class="mdescLeft">&#160;</td><td class="mdescRight">Core method. Iterates through the all genes of the host population and performs mutations in genes with a given probability. Also deletions and duplications of genes.  <a href="#af7fc9c6750f847a8260df37927646583">More...</a><br /></td></tr>
<tr class="separator:af7fc9c6750f847a8260df37927646583"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a11ca1135314c732332eaf55b5890b57e"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_environment.html#a11ca1135314c732332eaf55b5890b57e">mutateHostsWithDelDuplPointMuts</a> (double mut_probabl, double del, double dupl, unsigned int maxGene, int timeStamp)</td></tr>
<tr class="memdesc:a11ca1135314c732332eaf55b5890b57e"><td class="mdescLeft">&#160;</td><td class="mdescRight">Core method. Iterates through the all genes of the host population and performs POINT mutations in genes with a given probability. Also deletions and duplications of genes.  <a href="#a11ca1135314c732332eaf55b5890b57e">More...</a><br /></td></tr>
<tr class="separator:a11ca1135314c732332eaf55b5890b57e"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a13b07a4255b87d29541d0f78af6f49f7"><td class="memItemLeft" align="right" valign="top">double&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_environment.html#a13b07a4255b87d29541d0f78af6f49f7">MMtoPMscaling</a> (double MM_prob_mut, int geneLength)</td></tr>
<tr class="memdesc:a13b07a4255b87d29541d0f78af6f49f7"><td class="mdescLeft">&#160;</td><td class="mdescRight">Core method. When given micro-recombination mutation probability it returns point-mutation probability calculated the way that it the average number of new MHC match in both mutation scenarios.  <a href="#a13b07a4255b87d29541d0f78af6f49f7">More...</a><br /></td></tr>
<tr class="separator:a13b07a4255b87d29541d0f78af6f49f7"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a00d8b158811c0fa36a3fdeccdfebea91"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_environment.html#a00d8b158811c0fa36a3fdeccdfebea91">matingWithNoCommonMHC</a> ()</td></tr>
<tr class="memdesc:a00d8b158811c0fa36a3fdeccdfebea91"><td class="mdescLeft">&#160;</td><td class="mdescRight">Core method. Creates a new generation of hosts by sexual reproduction with strong negative preference towards MHC similarity between mates.  <a href="#a00d8b158811c0fa36a3fdeccdfebea91">More...</a><br /></td></tr>
<tr class="separator:a00d8b158811c0fa36a3fdeccdfebea91"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a7d3359f2eaf7379600239d18506aafe0"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_environment.html#a7d3359f2eaf7379600239d18506aafe0">matingWithOneDifferentMHC</a> ()</td></tr>
<tr class="memdesc:a7d3359f2eaf7379600239d18506aafe0"><td class="mdescLeft">&#160;</td><td class="mdescRight">Core method. Creates a new generation of hosts by sexual reproduction with weak negative preference towards MHC similarity between mates.  <a href="#a7d3359f2eaf7379600239d18506aafe0">More...</a><br /></td></tr>
<tr class="separator:a7d3359f2eaf7379600239d18506aafe0"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ae2adc5e02f72005d7854a4d597c4c294"><td class="memItemLeft" align="right" valign="top">unsigned&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_environment.html#ae2adc5e02f72005d7854a4d597c4c294">getPathoNumOfSpecies</a> ()</td></tr>
<tr class="memdesc:ae2adc5e02f72005d7854a4d597c4c294"><td class="mdescLeft">&#160;</td><td class="mdescRight">Core method. Gets the number of species of pathogens.  <a href="#ae2adc5e02f72005d7854a4d597c4c294">More...</a><br /></td></tr>
<tr class="separator:ae2adc5e02f72005d7854a4d597c4c294"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:aefe3e9cb58d92c0247cb56dda5eb57ac"><td class="memItemLeft" align="right" valign="top">unsigned&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_environment.html#aefe3e9cb58d92c0247cb56dda5eb57ac">getPathoSpeciesPopSize</a> (unsigned spec_numb)</td></tr>
<tr class="memdesc:aefe3e9cb58d92c0247cb56dda5eb57ac"><td class="mdescLeft">&#160;</td><td class="mdescRight">Core method. Gets a number of individuals in a selected species of pathogen.  <a href="#aefe3e9cb58d92c0247cb56dda5eb57ac">More...</a><br /></td></tr>
<tr class="separator:aefe3e9cb58d92c0247cb56dda5eb57ac"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a09b16d159ccf1ff7210c339e7c2b9408"><td class="memItemLeft" align="right" valign="top">unsigned&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_environment.html#a09b16d159ccf1ff7210c339e7c2b9408">getHostsPopSize</a> ()</td></tr>
<tr class="memdesc:a09b16d159ccf1ff7210c339e7c2b9408"><td class="mdescLeft">&#160;</td><td class="mdescRight">Core method. Gets the host population size.  <a href="#a09b16d159ccf1ff7210c339e7c2b9408">More...</a><br /></td></tr>
<tr class="separator:a09b16d159ccf1ff7210c339e7c2b9408"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a93a98eb73207ed37097a5b398579b44a"><td class="memItemLeft" align="right" valign="top">std::string&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_environment.html#a93a98eb73207ed37097a5b398579b44a">getHostsTags</a> ()</td></tr>
<tr class="separator:a93a98eb73207ed37097a5b398579b44a"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a9bbb94fdf32fa836c5b76b350bb2df4c"><td class="memItemLeft" align="right" valign="top">std::string&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_environment.html#a9bbb94fdf32fa836c5b76b350bb2df4c">getPathoGenesToString</a> (int i, int j)</td></tr>
<tr class="memdesc:a9bbb94fdf32fa836c5b76b350bb2df4c"><td class="mdescLeft">&#160;</td><td class="mdescRight">Data harvesting method. Gets the genome of selected pathogen in human-readable format.  <a href="#a9bbb94fdf32fa836c5b76b350bb2df4c">More...</a><br /></td></tr>
<tr class="separator:a9bbb94fdf32fa836c5b76b350bb2df4c"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a71dc8a1828a2830ffefa901d0974b2c1"><td class="memItemLeft" align="right" valign="top">std::string&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_environment.html#a71dc8a1828a2830ffefa901d0974b2c1">getHostGenesToString</a> (int i)</td></tr>
<tr class="memdesc:a71dc8a1828a2830ffefa901d0974b2c1"><td class="mdescLeft">&#160;</td><td class="mdescRight">Data harvesting method. Gets the genome of selected host in human-readable format.  <a href="#a71dc8a1828a2830ffefa901d0974b2c1">More...</a><br /></td></tr>
<tr class="separator:a71dc8a1828a2830ffefa901d0974b2c1"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a7ca87db3a1ceb9e27670538fce828675"><td class="memItemLeft" align="right" valign="top">std::string&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_environment.html#a7ca87db3a1ceb9e27670538fce828675">getFixedBitsInAntigens</a> ()</td></tr>
<tr class="memdesc:a7ca87db3a1ceb9e27670538fce828675"><td class="mdescLeft">&#160;</td><td class="mdescRight">Data harvesting method. Prepares a string with the list of fixed "no mutation sites" in antigens used by the DataHarvester to write these data to file.  <a href="#a7ca87db3a1ceb9e27670538fce828675">More...</a><br /></td></tr>
<tr class="separator:a7ca87db3a1ceb9e27670538fce828675"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ae83ec6e53b3b88d1e7e02050f81a9185"><td class="memItemLeft" align="right" valign="top">int&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_environment.html#ae83ec6e53b3b88d1e7e02050f81a9185">getSingleHostGenomeSize</a> (int indx)</td></tr>
<tr class="separator:ae83ec6e53b3b88d1e7e02050f81a9185"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a7d23e5f51b1c1feed367da44852b2dba"><td class="memItemLeft" align="right" valign="top">int&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_environment.html#a7d23e5f51b1c1feed367da44852b2dba">getSingleHostChromoOneSize</a> (int indx)</td></tr>
<tr class="separator:a7d23e5f51b1c1feed367da44852b2dba"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ac98907c5bc18780a611a743fb4afa479"><td class="memItemLeft" align="right" valign="top">int&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_environment.html#ac98907c5bc18780a611a743fb4afa479">getSingleHostChromoTwoSize</a> (int indx)</td></tr>
<tr class="separator:ac98907c5bc18780a611a743fb4afa479"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a86d8f067f5576f39047e360dd717b1fe"><td class="memItemLeft" align="right" valign="top">int&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_environment.html#a86d8f067f5576f39047e360dd717b1fe">getSingleHostRealGeneOne</a> (int i, int j)</td></tr>
<tr class="separator:a86d8f067f5576f39047e360dd717b1fe"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a20b1bdddb432ef84ee08fd1cd77d5fd9"><td class="memItemLeft" align="right" valign="top">int&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_environment.html#a20b1bdddb432ef84ee08fd1cd77d5fd9">getSingleHostRealGeneTwo</a> (int i, int j)</td></tr>
<tr class="separator:a20b1bdddb432ef84ee08fd1cd77d5fd9"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ab70e6fda860c35a3ac222cfdef2dff44"><td class="memItemLeft" align="right" valign="top">double&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_environment.html#ab70e6fda860c35a3ac222cfdef2dff44">getHostFitness</a> (int indx)</td></tr>
<tr class="separator:ab70e6fda860c35a3ac222cfdef2dff44"><td class="memSeparator" colspan="2">&#160;</td></tr>
</table>
<a name="details" id="details"></a><h2 class="groupheader">Detailed Description</h2>
<div class="textblock"><p>Core class. Stores and handles the environment object that is the place where all the things are happening. It contains host and pathogen populations stored in separate STL Vectors. </p>

<p>Definition at line <a class="el" href="_environment_8h_source.html#l00037">37</a> of file <a class="el" href="_environment_8h_source.html">Environment.h</a>.</p>
</div><h2 class="groupheader">Constructor &amp; Destructor Documentation</h2>
<a class="anchor" id="a8b427c4448d8b7536666837521b9e83d"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">Environment::Environment </td>
          <td>(</td>
          <td class="paramname"></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Definition at line <a class="el" href="_environment_8cpp_source.html#l00043">43</a> of file <a class="el" href="_environment_8cpp_source.html">Environment.cpp</a>.</p>

</div>
</div>
<a class="anchor" id="a8e294735187880dd3d59be10c425b29d"></a>
<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">Environment::~Environment </td>
          <td>(</td>
          <td class="paramname"></td><td>)</td>
          <td></td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">virtual</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">

<p>Definition at line <a class="el" href="_environment_8cpp_source.html#l00049">49</a> of file <a class="el" href="_environment_8cpp_source.html">Environment.cpp</a>.</p>

</div>
</div>
<h2 class="groupheader">Member Function Documentation</h2>
<a class="anchor" id="a0994a6e48ab6a97cbff9feb68ce9dcd3"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void Environment::calculateHostsFitnessAlphaXsqr </td>
          <td>(</td>
          <td class="paramtype">double&#160;</td>
          <td class="paramname"><em>alpha</em></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Core method. Iterates through the host population and calculates the Fitness for each single individual by calling Host::calculateHostsFitnessAlphaXsqr(), which uses one over the square on number of genes as a fitness cost. </p>

<p>Definition at line <a class="el" href="_environment_8cpp_source.html#l00518">518</a> of file <a class="el" href="_environment_8cpp_source.html">Environment.cpp</a>.</p>

</div>
</div>
<a class="anchor" id="a5f156d009868bce000e653b25fe53967"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void Environment::calculateHostsFitnessExpScaling </td>
          <td>(</td>
          <td class="paramtype">double&#160;</td>
          <td class="paramname"><em>alpha</em></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Core method. Iterates through the host population and calculates the Fitness for each single individual by calling <a class="el" href="class_host.html#a9134661842de046453acc0342ef74eb6" title="Core method. Calculates host individual fitness in proportion to the Gaussian function of the number ...">Host::calculateFitnessExpFunc()</a>, which uses a Gaussian function to accommodate the costs of having lots of genes. </p>

<p>Definition at line <a class="el" href="_environment_8cpp_source.html#l00531">531</a> of file <a class="el" href="_environment_8cpp_source.html">Environment.cpp</a>.</p>

</div>
</div>
<a class="anchor" id="aad5bb2197b43d307428356c123e34f43"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void Environment::calculateHostsFitnessExpScalingUniqAlleles </td>
          <td>(</td>
          <td class="paramtype">double&#160;</td>
          <td class="paramname"><em>alpha</em></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Core method. Iterates through the host population and calculates the Fitness for each single individual by calling <a class="el" href="class_host.html#ad185a5e02c8ed0930e4d400e9c0e6a9a" title="Core method. Calculates host individual fitness in proportion to the Gaussian function of the number ...">Host::calculateFitnessExpFuncUniqAlleles()</a>, which uses a Gaussian function to accommodate the costs of having lots of unique MHC alleles in chromosomes. </p>

<p>Definition at line <a class="el" href="_environment_8cpp_source.html#l00544">544</a> of file <a class="el" href="_environment_8cpp_source.html">Environment.cpp</a>.</p>

</div>
</div>
<a class="anchor" id="a596cf4db11a6e6356211a45540e04c7d"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void Environment::calculateHostsFitnessForDrift </td>
          <td>(</td>
          <td class="paramname"></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Core method. Iterates through the host population and calculates the Fitness for each single individual by calling <a class="el" href="class_host.html#a614430aea49534f29b63232414cf4fcd" title="Core method. Sets fitness as 1, a fixed value to simulate genetic drift. ">Host::calculateFitnessForDrift()</a>, which assigns "1" for each cell to make the genetic driff work. </p>

<p>Definition at line <a class="el" href="_environment_8cpp_source.html#l00505">505</a> of file <a class="el" href="_environment_8cpp_source.html">Environment.cpp</a>.</p>

</div>
</div>
<a class="anchor" id="ae63fad9a51f8700cf96cd3019f5a3bec"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void Environment::calculateHostsFitnessPerGene </td>
          <td>(</td>
          <td class="paramname"></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Core method. Iterates through the host population and calculates the Fitness for each single individual by calling <a class="el" href="class_host.html#a6079285b3b8bb5280d52500e6b7fc7ef" title="Core method. Calculates host individual fitness as the number of exposed pathogens divided by the gen...">Host::calculateFitnessAccChromSize()</a>, which takes the number of MHC genes under account. </p>

<p>Definition at line <a class="el" href="_environment_8cpp_source.html#l00479">479</a> of file <a class="el" href="_environment_8cpp_source.html">Environment.cpp</a>.</p>

</div>
</div>
<a class="anchor" id="a2d11b6356c2d62e7ebf4c97115118c5c"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void Environment::calculateHostsFitnessPlainPresent </td>
          <td>(</td>
          <td class="paramname"></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Core method. Iterates through the host population and calculates the Fitness for each single individual by calling Host::calculateHostsFitnessPlainInfect(), which is the plain-and-lame sum of presented pathogens. </p>

<p>Definition at line <a class="el" href="_environment_8cpp_source.html#l00492">492</a> of file <a class="el" href="_environment_8cpp_source.html">Environment.cpp</a>.</p>

</div>
</div>
<a class="anchor" id="adff3e1c9799ef7a3bc8d7b988bda3381"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void Environment::clearHostInfectionsData </td>
          <td>(</td>
          <td class="paramname"></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>. Core method. Clears information about infection and fitness in the whole host population. </p>

<p>Definition at line <a class="el" href="_environment_8cpp_source.html#l00838">838</a> of file <a class="el" href="_environment_8cpp_source.html">Environment.cpp</a>.</p>

<p><div class="dynheader">
Here is the caller graph for this function:</div>
<div class="dyncontent">
<div class="center"><img src="class_environment_adff3e1c9799ef7a3bc8d7b988bda3381_icgraph.png" border="0" usemap="#class_environment_adff3e1c9799ef7a3bc8d7b988bda3381_icgraph" alt=""/></div>
<map name="class_environment_adff3e1c9799ef7a3bc8d7b988bda3381_icgraph" id="class_environment_adff3e1c9799ef7a3bc8d7b988bda3381_icgraph">
<area shape="rect" id="node2" href="main_8cpp.html#a3c04138a5bfe5d72780bb7e82a18e627" title="The main function. Things are happening here. " alt="" coords="299,5,349,32"/></map>
</div>
</p>

</div>
</div>
<a class="anchor" id="af43ff3d50276e23f42895c6c7ddac69c"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void Environment::clearPathoInfectionData </td>
          <td>(</td>
          <td class="paramname"></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>. Core method. Clears information about infection and fitness in the whole pathogen population. </p>

<p>Definition at line <a class="el" href="_environment_8cpp_source.html#l00823">823</a> of file <a class="el" href="_environment_8cpp_source.html">Environment.cpp</a>.</p>

</div>
</div>
<a class="anchor" id="a7ca87db3a1ceb9e27670538fce828675"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">std::string Environment::getFixedBitsInAntigens </td>
          <td>(</td>
          <td class="paramname"></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Data harvesting method. Prepares a string with the list of fixed "no mutation sites" in antigens used by the DataHarvester to write these data to file. </p>
<dl class="section return"><dt>Returns</dt><dd>a string listing all the fixed sites in all pathogen species </dd></dl>

<p>Definition at line <a class="el" href="_environment_8cpp_source.html#l01220">1220</a> of file <a class="el" href="_environment_8cpp_source.html">Environment.cpp</a>.</p>

<p><div class="dynheader">
Here is the caller graph for this function:</div>
<div class="dyncontent">
<div class="center"><img src="class_environment_a7ca87db3a1ceb9e27670538fce828675_icgraph.png" border="0" usemap="#class_environment_a7ca87db3a1ceb9e27670538fce828675_icgraph" alt=""/></div>
<map name="class_environment_a7ca87db3a1ceb9e27670538fce828675_icgraph" id="class_environment_a7ca87db3a1ceb9e27670538fce828675_icgraph">
<area shape="rect" id="node2" href="class_data_handler.html#a6378c36aa8e4daf8f525c0fee446897e" title="Data harvesting method. Record the indices of fixed bits in all pathogen species. Run this just ones..." alt="" coords="230,5,413,46"/></map>
</div>
</p>

</div>
</div>
<a class="anchor" id="ab70e6fda860c35a3ac222cfdef2dff44"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">double Environment::getHostFitness </td>
          <td>(</td>
          <td class="paramtype">int&#160;</td>
          <td class="paramname"><em>indx</em></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Definition at line <a class="el" href="_environment_8cpp_source.html#l01254">1254</a> of file <a class="el" href="_environment_8cpp_source.html">Environment.cpp</a>.</p>

<p><div class="dynheader">
Here is the caller graph for this function:</div>
<div class="dyncontent">
<div class="center"><img src="class_environment_ab70e6fda860c35a3ac222cfdef2dff44_icgraph.png" border="0" usemap="#class_environment_ab70e6fda860c35a3ac222cfdef2dff44_icgraph" alt=""/></div>
<map name="class_environment_ab70e6fda860c35a3ac222cfdef2dff44_icgraph" id="class_environment_ab70e6fda860c35a3ac222cfdef2dff44_icgraph">
<area shape="rect" id="node2" href="class_data_handler.html#a23ad0de50b3da065674e92b150e12d9f" title="Data harvesting method. Calculates and writes to a file some stats about the hosts population genetic..." alt="" coords="246,5,450,46"/><area shape="rect" id="node3" href="main_8cpp.html#a3c04138a5bfe5d72780bb7e82a18e627" title="The main function. Things are happening here. " alt="" coords="499,12,549,39"/></map>
</div>
</p>

</div>
</div>
<a class="anchor" id="a71dc8a1828a2830ffefa901d0974b2c1"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">std::string Environment::getHostGenesToString </td>
          <td>(</td>
          <td class="paramtype">int&#160;</td>
          <td class="paramname"><em>i</em></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Data harvesting method. Gets the genome of selected host in human-readable format. </p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">i</td><td>individual's index within the host vector </td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>a string of the host's chromosome in a human-readable format. </dd></dl>

<p>Definition at line <a class="el" href="_environment_8cpp_source.html#l01209">1209</a> of file <a class="el" href="_environment_8cpp_source.html">Environment.cpp</a>.</p>

<p><div class="dynheader">
Here is the caller graph for this function:</div>
<div class="dyncontent">
<div class="center"><img src="class_environment_a71dc8a1828a2830ffefa901d0974b2c1_icgraph.png" border="0" usemap="#class_environment_a71dc8a1828a2830ffefa901d0974b2c1_icgraph" alt=""/></div>
<map name="class_environment_a71dc8a1828a2830ffefa901d0974b2c1_icgraph" id="class_environment_a71dc8a1828a2830ffefa901d0974b2c1_icgraph">
<area shape="rect" id="node2" href="class_data_handler.html#a46129d74eac82af0caae37a9e20e3e2e" title="Data harvesting method. Writes to a file all hosts with their genomes in a human&#45;readable format..." alt="" coords="240,5,432,46"/><area shape="rect" id="node3" href="main_8cpp.html#a3c04138a5bfe5d72780bb7e82a18e627" title="The main function. Things are happening here. " alt="" coords="480,12,531,39"/></map>
</div>
</p>

</div>
</div>
<a class="anchor" id="a09b16d159ccf1ff7210c339e7c2b9408"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">unsigned Environment::getHostsPopSize </td>
          <td>(</td>
          <td class="paramname"></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Core method. Gets the host population size. </p>
<dl class="section return"><dt>Returns</dt><dd>host population size </dd></dl>

<p>Definition at line <a class="el" href="_environment_8cpp_source.html#l01002">1002</a> of file <a class="el" href="_environment_8cpp_source.html">Environment.cpp</a>.</p>

<p><div class="dynheader">
Here is the caller graph for this function:</div>
<div class="dyncontent">
<div class="center"><img src="class_environment_a09b16d159ccf1ff7210c339e7c2b9408_icgraph.png" border="0" usemap="#class_environment_a09b16d159ccf1ff7210c339e7c2b9408_icgraph" alt=""/></div>
<map name="class_environment_a09b16d159ccf1ff7210c339e7c2b9408_icgraph" id="class_environment_a09b16d159ccf1ff7210c339e7c2b9408_icgraph">
<area shape="rect" id="node2" href="class_data_handler.html#a46129d74eac82af0caae37a9e20e3e2e" title="Data harvesting method. Writes to a file all hosts with their genomes in a human&#45;readable format..." alt="" coords="265,5,457,46"/><area shape="rect" id="node4" href="class_data_handler.html#a23ad0de50b3da065674e92b150e12d9f" title="Data harvesting method. Calculates and writes to a file some stats about the hosts population genetic..." alt="" coords="259,70,463,111"/><area shape="rect" id="node5" href="class_data_handler.html#a5661a45f0a9965bcd041c5617f1a1e8b" title="Data harvesting method. Record the total number of genes and types of MHCs in hosts. All the freaking hosts! " alt="" coords="266,135,457,177"/><area shape="rect" id="node3" href="main_8cpp.html#a3c04138a5bfe5d72780bb7e82a18e627" title="The main function. Things are happening here. " alt="" coords="512,77,563,104"/></map>
</div>
</p>

</div>
</div>
<a class="anchor" id="a93a98eb73207ed37097a5b398579b44a"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">std::string Environment::getHostsTags </td>
          <td>(</td>
          <td class="paramname"></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">

</div>
</div>
<a class="anchor" id="a9bbb94fdf32fa836c5b76b350bb2df4c"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">std::string Environment::getPathoGenesToString </td>
          <td>(</td>
          <td class="paramtype">int&#160;</td>
          <td class="paramname"><em>i</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">int&#160;</td>
          <td class="paramname"><em>j</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Data harvesting method. Gets the genome of selected pathogen in human-readable format. </p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">i</td><td>pathogen species number </td></tr>
    <tr><td class="paramname">j</td><td>individual's index within the pathogen species vector </td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>a string of the pathogen's chromosome in a human-readable format. </dd></dl>

<p>Definition at line <a class="el" href="_environment_8cpp_source.html#l01198">1198</a> of file <a class="el" href="_environment_8cpp_source.html">Environment.cpp</a>.</p>

<p><div class="dynheader">
Here is the caller graph for this function:</div>
<div class="dyncontent">
<div class="center"><img src="class_environment_a9bbb94fdf32fa836c5b76b350bb2df4c_icgraph.png" border="0" usemap="#class_environment_a9bbb94fdf32fa836c5b76b350bb2df4c_icgraph" alt=""/></div>
<map name="class_environment_a9bbb94fdf32fa836c5b76b350bb2df4c_icgraph" id="class_environment_a9bbb94fdf32fa836c5b76b350bb2df4c_icgraph">
<area shape="rect" id="node2" href="class_data_handler.html#a877ed5bfd00f0cbeef0413bf1d47f865" title="Data harvesting method. Writes to a file all pathogens with their genomes in a human&#45;readable format..." alt="" coords="249,5,447,46"/></map>
</div>
</p>

</div>
</div>
<a class="anchor" id="ae2adc5e02f72005d7854a4d597c4c294"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">unsigned Environment::getPathoNumOfSpecies </td>
          <td>(</td>
          <td class="paramname"></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Core method. Gets the number of species of pathogens. </p>
<dl class="section return"><dt>Returns</dt><dd>number of pathogen species </dd></dl>

<p>Definition at line <a class="el" href="_environment_8cpp_source.html#l00982">982</a> of file <a class="el" href="_environment_8cpp_source.html">Environment.cpp</a>.</p>

<p><div class="dynheader">
Here is the caller graph for this function:</div>
<div class="dyncontent">
<div class="center"><img src="class_environment_ae2adc5e02f72005d7854a4d597c4c294_icgraph.png" border="0" usemap="#class_environment_ae2adc5e02f72005d7854a4d597c4c294_icgraph" alt=""/></div>
<map name="class_environment_ae2adc5e02f72005d7854a4d597c4c294_icgraph" id="class_environment_ae2adc5e02f72005d7854a4d597c4c294_icgraph">
<area shape="rect" id="node2" href="class_data_handler.html#a89f86a67a11f719ec282f37740663f95" title="Data harvesting method. Writes to a file population sizes of all pathogen species in a given time..." alt="" coords="238,5,445,46"/><area shape="rect" id="node3" href="class_data_handler.html#a877ed5bfd00f0cbeef0413bf1d47f865" title="Data harvesting method. Writes to a file all pathogens with their genomes in a human&#45;readable format..." alt="" coords="242,70,441,111"/></map>
</div>
</p>

</div>
</div>
<a class="anchor" id="aefe3e9cb58d92c0247cb56dda5eb57ac"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">unsigned Environment::getPathoSpeciesPopSize </td>
          <td>(</td>
          <td class="paramtype">unsigned&#160;</td>
          <td class="paramname"><em>spec_numb</em></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Core method. Gets a number of individuals in a selected species of pathogen. </p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">spec_numb</td><td>- number of selected species. </td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>number of individuals of a selected species. </dd></dl>

<p>Definition at line <a class="el" href="_environment_8cpp_source.html#l00993">993</a> of file <a class="el" href="_environment_8cpp_source.html">Environment.cpp</a>.</p>

<p><div class="dynheader">
Here is the caller graph for this function:</div>
<div class="dyncontent">
<div class="center"><img src="class_environment_aefe3e9cb58d92c0247cb56dda5eb57ac_icgraph.png" border="0" usemap="#class_environment_aefe3e9cb58d92c0247cb56dda5eb57ac_icgraph" alt=""/></div>
<map name="class_environment_aefe3e9cb58d92c0247cb56dda5eb57ac_icgraph" id="class_environment_aefe3e9cb58d92c0247cb56dda5eb57ac_icgraph">
<area shape="rect" id="node2" href="class_data_handler.html#a89f86a67a11f719ec282f37740663f95" title="Data harvesting method. Writes to a file population sizes of all pathogen species in a given time..." alt="" coords="257,5,463,46"/><area shape="rect" id="node3" href="class_data_handler.html#a877ed5bfd00f0cbeef0413bf1d47f865" title="Data harvesting method. Writes to a file all pathogens with their genomes in a human&#45;readable format..." alt="" coords="261,70,459,111"/></map>
</div>
</p>

</div>
</div>
<a class="anchor" id="a7d23e5f51b1c1feed367da44852b2dba"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">int Environment::getSingleHostChromoOneSize </td>
          <td>(</td>
          <td class="paramtype">int&#160;</td>
          <td class="paramname"><em>indx</em></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Definition at line <a class="el" href="_environment_8cpp_source.html#l01236">1236</a> of file <a class="el" href="_environment_8cpp_source.html">Environment.cpp</a>.</p>

<p><div class="dynheader">
Here is the caller graph for this function:</div>
<div class="dyncontent">
<div class="center"><img src="class_environment_a7d23e5f51b1c1feed367da44852b2dba_icgraph.png" border="0" usemap="#class_environment_a7d23e5f51b1c1feed367da44852b2dba_icgraph" alt=""/></div>
<map name="class_environment_a7d23e5f51b1c1feed367da44852b2dba_icgraph" id="class_environment_a7d23e5f51b1c1feed367da44852b2dba_icgraph">
<area shape="rect" id="node2" href="class_data_handler.html#a23ad0de50b3da065674e92b150e12d9f" title="Data harvesting method. Calculates and writes to a file some stats about the hosts population genetic..." alt="" coords="241,5,445,46"/><area shape="rect" id="node4" href="class_data_handler.html#a5661a45f0a9965bcd041c5617f1a1e8b" title="Data harvesting method. Record the total number of genes and types of MHCs in hosts. All the freaking hosts! " alt="" coords="247,70,438,111"/><area shape="rect" id="node3" href="main_8cpp.html#a3c04138a5bfe5d72780bb7e82a18e627" title="The main function. Things are happening here. " alt="" coords="493,45,544,72"/></map>
</div>
</p>

</div>
</div>
<a class="anchor" id="ac98907c5bc18780a611a743fb4afa479"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">int Environment::getSingleHostChromoTwoSize </td>
          <td>(</td>
          <td class="paramtype">int&#160;</td>
          <td class="paramname"><em>indx</em></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Definition at line <a class="el" href="_environment_8cpp_source.html#l01241">1241</a> of file <a class="el" href="_environment_8cpp_source.html">Environment.cpp</a>.</p>

<p><div class="dynheader">
Here is the caller graph for this function:</div>
<div class="dyncontent">
<div class="center"><img src="class_environment_ac98907c5bc18780a611a743fb4afa479_icgraph.png" border="0" usemap="#class_environment_ac98907c5bc18780a611a743fb4afa479_icgraph" alt=""/></div>
<map name="class_environment_ac98907c5bc18780a611a743fb4afa479_icgraph" id="class_environment_ac98907c5bc18780a611a743fb4afa479_icgraph">
<area shape="rect" id="node2" href="class_data_handler.html#a23ad0de50b3da065674e92b150e12d9f" title="Data harvesting method. Calculates and writes to a file some stats about the hosts population genetic..." alt="" coords="241,5,445,46"/><area shape="rect" id="node3" href="main_8cpp.html#a3c04138a5bfe5d72780bb7e82a18e627" title="The main function. Things are happening here. " alt="" coords="493,12,544,39"/></map>
</div>
</p>

</div>
</div>
<a class="anchor" id="ae83ec6e53b3b88d1e7e02050f81a9185"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">int Environment::getSingleHostGenomeSize </td>
          <td>(</td>
          <td class="paramtype">int&#160;</td>
          <td class="paramname"><em>indx</em></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Definition at line <a class="el" href="_environment_8cpp_source.html#l01231">1231</a> of file <a class="el" href="_environment_8cpp_source.html">Environment.cpp</a>.</p>

</div>
</div>
<a class="anchor" id="a86d8f067f5576f39047e360dd717b1fe"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">int Environment::getSingleHostRealGeneOne </td>
          <td>(</td>
          <td class="paramtype">int&#160;</td>
          <td class="paramname"><em>i</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">int&#160;</td>
          <td class="paramname"><em>j</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Definition at line <a class="el" href="_environment_8cpp_source.html#l01246">1246</a> of file <a class="el" href="_environment_8cpp_source.html">Environment.cpp</a>.</p>

<p><div class="dynheader">
Here is the caller graph for this function:</div>
<div class="dyncontent">
<div class="center"><img src="class_environment_a86d8f067f5576f39047e360dd717b1fe_icgraph.png" border="0" usemap="#class_environment_a86d8f067f5576f39047e360dd717b1fe_icgraph" alt=""/></div>
<map name="class_environment_a86d8f067f5576f39047e360dd717b1fe_icgraph" id="class_environment_a86d8f067f5576f39047e360dd717b1fe_icgraph">
<area shape="rect" id="node2" href="class_data_handler.html#a23ad0de50b3da065674e92b150e12d9f" title="Data harvesting method. Calculates and writes to a file some stats about the hosts population genetic..." alt="" coords="241,5,445,46"/><area shape="rect" id="node4" href="class_data_handler.html#a5661a45f0a9965bcd041c5617f1a1e8b" title="Data harvesting method. Record the total number of genes and types of MHCs in hosts. All the freaking hosts! " alt="" coords="247,70,438,111"/><area shape="rect" id="node3" href="main_8cpp.html#a3c04138a5bfe5d72780bb7e82a18e627" title="The main function. Things are happening here. " alt="" coords="493,45,544,72"/></map>
</div>
</p>

</div>
</div>
<a class="anchor" id="a20b1bdddb432ef84ee08fd1cd77d5fd9"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">int Environment::getSingleHostRealGeneTwo </td>
          <td>(</td>
          <td class="paramtype">int&#160;</td>
          <td class="paramname"><em>i</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">int&#160;</td>
          <td class="paramname"><em>j</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Definition at line <a class="el" href="_environment_8cpp_source.html#l01250">1250</a> of file <a class="el" href="_environment_8cpp_source.html">Environment.cpp</a>.</p>

<p><div class="dynheader">
Here is the caller graph for this function:</div>
<div class="dyncontent">
<div class="center"><img src="class_environment_a20b1bdddb432ef84ee08fd1cd77d5fd9_icgraph.png" border="0" usemap="#class_environment_a20b1bdddb432ef84ee08fd1cd77d5fd9_icgraph" alt=""/></div>
<map name="class_environment_a20b1bdddb432ef84ee08fd1cd77d5fd9_icgraph" id="class_environment_a20b1bdddb432ef84ee08fd1cd77d5fd9_icgraph">
<area shape="rect" id="node2" href="class_data_handler.html#a23ad0de50b3da065674e92b150e12d9f" title="Data harvesting method. Calculates and writes to a file some stats about the hosts population genetic..." alt="" coords="241,5,445,46"/><area shape="rect" id="node3" href="main_8cpp.html#a3c04138a5bfe5d72780bb7e82a18e627" title="The main function. Things are happening here. " alt="" coords="493,12,544,39"/></map>
</div>
</p>

</div>
</div>
<a class="anchor" id="a670a9eaf38f556ab8ca096bc9a1433a5"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void Environment::infectEveryOne </td>
          <td>(</td>
          <td class="paramtype">int&#160;</td>
          <td class="paramname"><em>simil_mesure</em></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">

</div>
</div>
<a class="anchor" id="aecd592b734359557235bb0a6980f77f9"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void Environment::infectOneFromOneSpecHetero </td>
          <td>(</td>
          <td class="paramname"></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Core method. Iterates through the host population and the parasite population to "infect" the hosts with parasites. With heterozygote advantage and it permits a pathogen species to infect a host only ONES. </p>
<p>Each host is exposed to one, randomly selected individual from a pathogen species. Exposition procedure for a single host is repeated for all pathogen species. Maximum number of pathogens a host can contract in one go equals to the number of pathogen species. Fitness is calculated with heterozygote advantage added (antigen recognition just by one allele gives a full advantage). One species can infect a host only ONES.</p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">simil_mesure</td><td>- number of bits which have to be similar, to expose a pathogen. It's passed to H2Pinteraction::doesInfected() method. </td></tr>
  </table>
  </dd>
</dl>

<p>Definition at line <a class="el" href="_environment_8cpp_source.html#l00457">457</a> of file <a class="el" href="_environment_8cpp_source.html">Environment.cpp</a>.</p>

<p><div class="dynheader">
Here is the call graph for this function:</div>
<div class="dyncontent">
<div class="center"><img src="class_environment_aecd592b734359557235bb0a6980f77f9_cgraph.png" border="0" usemap="#class_environment_aecd592b734359557235bb0a6980f77f9_cgraph" alt=""/></div>
<map name="class_environment_aecd592b734359557235bb0a6980f77f9_cgraph" id="class_environment_aecd592b734359557235bb0a6980f77f9_cgraph">
<area shape="rect" id="node2" href="class_random_numbs.html#a364c415e5105061274013a1e68576a64" title="Core method. Gets the instance of a singleton class. " alt="" coords="244,5,433,32"/><area shape="rect" id="node3" href="class_random_numbs.html#a301ebb1a7c447763903f925a634da605" title="Core method. Generates one integer random number from a given interval with an uniform distribution..." alt="" coords="257,56,421,83"/><area shape="rect" id="node4" href="class_h2_pinteraction.html#a46703648dcef05c6478d6249aeadd2c8" title="Core method. Checks if a host gets infected with a pathogen. Heterozygote has an advantage here over ..." alt="" coords="243,107,434,149"/><area shape="rect" id="node5" href="class_pathogen.html#a87ddcd5dc3951338533cc35b5b2d059c" title="Core method. Fetches the pathogene&#39;s antigens. " alt="" coords="499,13,669,40"/><area shape="rect" id="node6" href="class_host.html#ac387176ba203f6aa1f20610f5679385a" title="Core method. Returns the first chromosome. " alt="" coords="495,64,673,91"/><area shape="rect" id="node7" href="class_pathogen.html#ada9e5929d6736f63377f0625153b224b" title="Core method. Fetches a species tag. " alt="" coords="498,115,670,141"/><area shape="rect" id="node8" href="class_h2_pinteraction.html#ae63bc3afa414686c4c7d9a500f3d4a53" title="Core method. Checks if the antigen is presented by a given MHC. " alt="" coords="483,165,685,192"/><area shape="rect" id="node9" href="class_host.html#aa5afa38408ebba644f56667364213f0d" title="Core method. Returns the first chromosome. " alt="" coords="495,216,673,243"/></map>
</div>
</p>

</div>
</div>
<a class="anchor" id="a00d8b158811c0fa36a3fdeccdfebea91"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void Environment::matingWithNoCommonMHC </td>
          <td>(</td>
          <td class="paramname"></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Core method. Creates a new generation of hosts by sexual reproduction with strong negative preference towards MHC similarity between mates. </p>
<p>There are no sexes as the host species is assumed a hermaphrodite. Each individual checks out all other individuals in the population in a random order and mates with the first individual that has 0 same MHCs. If 0-similarity-rule will not provide enough offspring to match the old population size, then 1 same MHC is accepted, after that 2 same MHC are OK, etc. until the algorithm will recreate a population of the same size as the original one. </p>

<p>Definition at line <a class="el" href="_environment_8cpp_source.html#l01018">1018</a> of file <a class="el" href="_environment_8cpp_source.html">Environment.cpp</a>.</p>

</div>
</div>
<a class="anchor" id="a7d3359f2eaf7379600239d18506aafe0"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void Environment::matingWithOneDifferentMHC </td>
          <td>(</td>
          <td class="paramname"></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Core method. Creates a new generation of hosts by sexual reproduction with weak negative preference towards MHC similarity between mates. </p>
<p>There are no sexes as the host species is assumed a hermaphrodite. Each individual checks out all other individuals in the population in a random order and mates with the first individual that has at least one different MHC. If one-similarity-rule will not provide enough offspring to match the old population size, then partners are matched at random until the algorithm will recreate a population of the same size as the original one. </p>

<p>Definition at line <a class="el" href="_environment_8cpp_source.html#l01103">1103</a> of file <a class="el" href="_environment_8cpp_source.html">Environment.cpp</a>.</p>

<p><div class="dynheader">
Here is the call graph for this function:</div>
<div class="dyncontent">
<div class="center"><img src="class_environment_a7d3359f2eaf7379600239d18506aafe0_cgraph.png" border="0" usemap="#class_environment_a7d3359f2eaf7379600239d18506aafe0_cgraph" alt=""/></div>
<map name="class_environment_a7d3359f2eaf7379600239d18506aafe0_cgraph" id="class_environment_a7d3359f2eaf7379600239d18506aafe0_cgraph">
<area shape="rect" id="node2" href="class_random_numbs.html#a364c415e5105061274013a1e68576a64" title="Core method. Gets the instance of a singleton class. " alt="" coords="248,12,437,39"/></map>
</div>
</p>

<p><div class="dynheader">
Here is the caller graph for this function:</div>
<div class="dyncontent">
<div class="center"><img src="class_environment_a7d3359f2eaf7379600239d18506aafe0_icgraph.png" border="0" usemap="#class_environment_a7d3359f2eaf7379600239d18506aafe0_icgraph" alt=""/></div>
<map name="class_environment_a7d3359f2eaf7379600239d18506aafe0_icgraph" id="class_environment_a7d3359f2eaf7379600239d18506aafe0_icgraph">
<area shape="rect" id="node2" href="main_8cpp.html#a3c04138a5bfe5d72780bb7e82a18e627" title="The main function. Things are happening here. " alt="" coords="248,12,299,39"/></map>
</div>
</p>

</div>
</div>
<a class="anchor" id="a13b07a4255b87d29541d0f78af6f49f7"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">double Environment::MMtoPMscaling </td>
          <td>(</td>
          <td class="paramtype">double&#160;</td>
          <td class="paramname"><em>MM_prob_mut</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">int&#160;</td>
          <td class="paramname"><em>geneLength</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Core method. When given micro-recombination mutation probability it returns point-mutation probability calculated the way that it the average number of new MHC match in both mutation scenarios. </p>
<p>It implements the way in Ejsmond MJ, Radwan J. 2015. <em>Red Queen drives positive selection on Major Histocompatibility Complex genes (MHC)</em> [not published yet] micro-recombination mutation is transformed to point-mutation probability:</p>
<p><img class="formulaInl" alt="$ p_{t} = 1 - \left[1 - p_{m} (1 - 0.5^{b})\right]^{1/b} $" src="form_0.png"/></p>
<p>where <img class="formulaInl" alt="$b$" src="form_1.png"/> is the number of bits, <img class="formulaInl" alt="$p_{m}$" src="form_2.png"/> is the old-style mutation probability. Only to avoid problems with numerical stability when calculating the power of <img class="formulaInl" alt="$1/b$" src="form_3.png"/> when <img class="formulaInl" alt="$b$" src="form_1.png"/> is large and the base is small we re-phrased it into:</p>
<p><img class="formulaInl" alt="$ p_{t} = 1 - \exp\left( 1/b \cdot \ln [1 - p_{m} (1 - 0.5^{b})]\right) $" src="form_7.png"/></p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">MM_prob_mut</td><td>- micro-recombination mutation ('old-style' mutation) </td></tr>
    <tr><td class="paramname">geneLength</td><td>- number of bits in a gene representation </td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>- calculated corresponding point mutation </dd></dl>

<p>Definition at line <a class="el" href="_environment_8cpp_source.html#l00924">924</a> of file <a class="el" href="_environment_8cpp_source.html">Environment.cpp</a>.</p>

<p><div class="dynheader">
Here is the caller graph for this function:</div>
<div class="dyncontent">
<div class="center"><img src="class_environment_a13b07a4255b87d29541d0f78af6f49f7_icgraph.png" border="0" usemap="#class_environment_a13b07a4255b87d29541d0f78af6f49f7_icgraph" alt=""/></div>
<map name="class_environment_a13b07a4255b87d29541d0f78af6f49f7_icgraph" id="class_environment_a13b07a4255b87d29541d0f78af6f49f7_icgraph">
<area shape="rect" id="node2" href="main_8cpp.html#a3c04138a5bfe5d72780bb7e82a18e627" title="The main function. Things are happening here. " alt="" coords="251,5,301,32"/></map>
</div>
</p>

</div>
</div>
<a class="anchor" id="ad3d9445c714e728df229eabbe425fff1"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void Environment::mutateHosts </td>
          <td>(</td>
          <td class="paramtype">double&#160;</td>
          <td class="paramname"><em>mut_probabl</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">int&#160;</td>
          <td class="paramname"><em>timeStamp</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Core method. Iterates through the all genes of the host population and performs mutations in genes with a given probability. </p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">mut_probabl</td><td>- probability of a mutation in a single gene. </td></tr>
    <tr><td class="paramname">timeStamp</td><td>- current time (number of the model iteration) </td></tr>
  </table>
  </dd>
</dl>

<p>Definition at line <a class="el" href="_environment_8cpp_source.html#l00853">853</a> of file <a class="el" href="_environment_8cpp_source.html">Environment.cpp</a>.</p>

</div>
</div>
<a class="anchor" id="af7fc9c6750f847a8260df37927646583"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void Environment::mutateHostsWithDelDupl </td>
          <td>(</td>
          <td class="paramtype">double&#160;</td>
          <td class="paramname"><em>mut_probabl</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">double&#160;</td>
          <td class="paramname"><em>del</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">double&#160;</td>
          <td class="paramname"><em>dupl</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">unsigned int&#160;</td>
          <td class="paramname"><em>maxGene</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">int&#160;</td>
          <td class="paramname"><em>timeStamp</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Core method. Iterates through the all genes of the host population and performs mutations in genes with a given probability. Also deletions and duplications of genes. </p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">mut_probabl</td><td>- probability of a mutation in a single gene. </td></tr>
    <tr><td class="paramname">del</td><td>- mutation probability, probability a gene will be deleted </td></tr>
    <tr><td class="paramname">dupli</td><td>- mutation probability, probability a gene will be duplicated (and added at the end of the Chromosome vector) </td></tr>
    <tr><td class="paramname">maxGene</td><td>- maximal allowed number of genes in a chromosome </td></tr>
    <tr><td class="paramname">timeStamp</td><td>- current time (number of the model iteration) </td></tr>
  </table>
  </dd>
</dl>

<p>Definition at line <a class="el" href="_environment_8cpp_source.html#l00872">872</a> of file <a class="el" href="_environment_8cpp_source.html">Environment.cpp</a>.</p>

<p><div class="dynheader">
Here is the caller graph for this function:</div>
<div class="dyncontent">
<div class="center"><img src="class_environment_af7fc9c6750f847a8260df37927646583_icgraph.png" border="0" usemap="#class_environment_af7fc9c6750f847a8260df37927646583_icgraph" alt=""/></div>
<map name="class_environment_af7fc9c6750f847a8260df37927646583_icgraph" id="class_environment_af7fc9c6750f847a8260df37927646583_icgraph">
<area shape="rect" id="node2" href="main_8cpp.html#a3c04138a5bfe5d72780bb7e82a18e627" title="The main function. Things are happening here. " alt="" coords="232,12,283,39"/></map>
</div>
</p>

</div>
</div>
<a class="anchor" id="a11ca1135314c732332eaf55b5890b57e"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void Environment::mutateHostsWithDelDuplPointMuts </td>
          <td>(</td>
          <td class="paramtype">double&#160;</td>
          <td class="paramname"><em>pm_mut_probabl</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">double&#160;</td>
          <td class="paramname"><em>del</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">double&#160;</td>
          <td class="paramname"><em>dupl</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">unsigned int&#160;</td>
          <td class="paramname"><em>maxGene</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">int&#160;</td>
          <td class="paramname"><em>timeStamp</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Core method. Iterates through the all genes of the host population and performs POINT mutations in genes with a given probability. Also deletions and duplications of genes. </p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">pm_mut_probabl</td><td>- probability of a point mutation in a single gene. </td></tr>
    <tr><td class="paramname">del</td><td>- mutation probability, probability a gene will be deleted </td></tr>
    <tr><td class="paramname">dupli</td><td>- mutation probability, probability a gene will be duplicated (and added at the end of the Chromosome vector) </td></tr>
    <tr><td class="paramname">timeStamp</td><td>- current time (number of the model iteration) </td></tr>
  </table>
  </dd>
</dl>

<p>Definition at line <a class="el" href="_environment_8cpp_source.html#l00892">892</a> of file <a class="el" href="_environment_8cpp_source.html">Environment.cpp</a>.</p>

</div>
</div>
<a class="anchor" id="a54bc672882bea05101d65942b35cdc63"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void Environment::mutatePathogens </td>
          <td>(</td>
          <td class="paramtype">double&#160;</td>
          <td class="paramname"><em>mut_probabl</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">int&#160;</td>
          <td class="paramname"><em>mhcSize</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">int&#160;</td>
          <td class="paramname"><em>timeStamp</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Core method. Iterates through the all genes of the pathogen population and performs mutations in genes with a given probability. </p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">mut_probabl</td><td>- probability of a mutation in a single gene. </td></tr>
    <tr><td class="paramname">mhcSize</td><td>- number of bits in MHC protein </td></tr>
    <tr><td class="paramname">timeStamp</td><td>- current time (number of the model iteration) </td></tr>
  </table>
  </dd>
</dl>

<p>Definition at line <a class="el" href="_environment_8cpp_source.html#l00938">938</a> of file <a class="el" href="_environment_8cpp_source.html">Environment.cpp</a>.</p>

</div>
</div>
<a class="anchor" id="af56f92817b21be862325edb48f291037"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void Environment::mutatePathogensWithRestric </td>
          <td>(</td>
          <td class="paramtype">double&#160;</td>
          <td class="paramname"><em>mut_probabl</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">int&#160;</td>
          <td class="paramname"><em>mhcSize</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">int&#160;</td>
          <td class="paramname"><em>timeStamp</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Core method. Iterates through the all genes of the pathogen population and performs mutations in genes with a given probability. But some positions in the bit-string (gene) are not allowed to change. </p>
<p>The "No-Mutation Vector" is defined within the <a class="el" href="class_environment.html" title="Core class. Stores and handles the environment object that is the place where all the things are happ...">Environment</a> object.</p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">mut_probabl</td><td>- probability of a mutation in a single gene. </td></tr>
    <tr><td class="paramname">mhcSize</td><td>- number of bits in MHC protein </td></tr>
    <tr><td class="paramname">timeStamp</td><td>- current time (number of the model iteration) </td></tr>
  </table>
  </dd>
</dl>

<p>Definition at line <a class="el" href="_environment_8cpp_source.html#l00959">959</a> of file <a class="el" href="_environment_8cpp_source.html">Environment.cpp</a>.</p>

</div>
</div>
<a class="anchor" id="a56c39a47a881880bd36c3b8c5f0dbe8a"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void Environment::selectAndReprodHostsAddOffspring </td>
          <td>(</td>
          <td class="paramname"></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Core method. Forms the next generation of hosts using the fitness proportionate selection method. </p>
<p>The new generation takes place of individuals which got removed from population due to too low fitness.</p>
<p>Iterates through the host population selecting the next generation for reproduction using <a href="http://en.wikipedia.org/wiki/Fitness_proportionate_selection">fitness proportionate selection method</a> (also known as the roulette wheel selection). </p>

<p>Definition at line <a class="el" href="_environment_8cpp_source.html#l00564">564</a> of file <a class="el" href="_environment_8cpp_source.html">Environment.cpp</a>.</p>

<p><div class="dynheader">
Here is the call graph for this function:</div>
<div class="dyncontent">
<div class="center"><img src="class_environment_a56c39a47a881880bd36c3b8c5f0dbe8a_cgraph.png" border="0" usemap="#class_environment_a56c39a47a881880bd36c3b8c5f0dbe8a_cgraph" alt=""/></div>
<map name="class_environment_a56c39a47a881880bd36c3b8c5f0dbe8a_cgraph" id="class_environment_a56c39a47a881880bd36c3b8c5f0dbe8a_cgraph">
<area shape="rect" id="node2" href="class_random_numbs.html#a364c415e5105061274013a1e68576a64" title="Core method. Gets the instance of a singleton class. " alt="" coords="256,5,445,32"/><area shape="rect" id="node3" href="class_random_numbs.html#aaae77654579060738e55b46c8caba9a6" title="Core method. Generates one real random number from a given interval with an uniform distribution..." alt="" coords="263,56,439,83"/><area shape="rect" id="node4" href="class_random_numbs.html#a301ebb1a7c447763903f925a634da605" title="Core method. Generates one integer random number from a given interval with an uniform distribution..." alt="" coords="269,107,433,133"/></map>
</div>
</p>

</div>
</div>
<a class="anchor" id="a09af2890db9eb1a02069adfb8836aa47"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void Environment::selectAndReprodHostsReplace </td>
          <td>(</td>
          <td class="paramname"></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Core method. Forms the next generation of hosts using the fitness proportionate selection method. Replaces the old population with a new one. </p>
<p>Iterates through the host population selecting the next generation for reproduction using <a href="http://en.wikipedia.org/wiki/Fitness_proportionate_selection">fitness proportionate selection method</a> (also known as the roulette wheel selection). Simulates random mating of hermaphrodites with no difference between sexes. </p>

<p>Definition at line <a class="el" href="_environment_8cpp_source.html#l00633">633</a> of file <a class="el" href="_environment_8cpp_source.html">Environment.cpp</a>.</p>

<p><div class="dynheader">
Here is the call graph for this function:</div>
<div class="dyncontent">
<div class="center"><img src="class_environment_a09af2890db9eb1a02069adfb8836aa47_cgraph.png" border="0" usemap="#class_environment_a09af2890db9eb1a02069adfb8836aa47_cgraph" alt=""/></div>
<map name="class_environment_a09af2890db9eb1a02069adfb8836aa47_cgraph" id="class_environment_a09af2890db9eb1a02069adfb8836aa47_cgraph">
<area shape="rect" id="node2" href="class_random_numbs.html#a364c415e5105061274013a1e68576a64" title="Core method. Gets the instance of a singleton class. " alt="" coords="256,5,445,32"/><area shape="rect" id="node3" href="class_random_numbs.html#aaae77654579060738e55b46c8caba9a6" title="Core method. Generates one real random number from a given interval with an uniform distribution..." alt="" coords="263,56,439,83"/></map>
</div>
</p>

</div>
</div>
<a class="anchor" id="a90177792fb03087e1fc64dca78f7e312"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void Environment::selectAndReproducePathoFixedPopSizes </td>
          <td>(</td>
          <td class="paramname"></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Core method. Forms the next generation of hosts using the fitness proportionate selection method. It keeps population sizes of different pathogens species at a fixed number. </p>
<p>Iterates through the pathogen population selecting the next generation for reproduction using fitness proportionate selection method (roulette wheel selection). </p>

<p>Definition at line <a class="el" href="_environment_8cpp_source.html#l00770">770</a> of file <a class="el" href="_environment_8cpp_source.html">Environment.cpp</a>.</p>

<p><div class="dynheader">
Here is the call graph for this function:</div>
<div class="dyncontent">
<div class="center"><img src="class_environment_a90177792fb03087e1fc64dca78f7e312_cgraph.png" border="0" usemap="#class_environment_a90177792fb03087e1fc64dca78f7e312_cgraph" alt=""/></div>
<map name="class_environment_a90177792fb03087e1fc64dca78f7e312_cgraph" id="class_environment_a90177792fb03087e1fc64dca78f7e312_cgraph">
<area shape="rect" id="node2" href="class_random_numbs.html#a364c415e5105061274013a1e68576a64" title="Core method. Gets the instance of a singleton class. " alt="" coords="277,5,467,32"/><area shape="rect" id="node3" href="class_random_numbs.html#a301ebb1a7c447763903f925a634da605" title="Core method. Generates one integer random number from a given interval with an uniform distribution..." alt="" coords="290,56,454,83"/></map>
</div>
</p>

</div>
</div>
<a class="anchor" id="ad7f96a8620fa9df1601c613130b275e0"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void Environment::selectAndReproducePathoFlexPopSizes </td>
          <td>(</td>
          <td class="paramname"></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Core method. Forms the next generation of hosts using the fitness proportionate selection method. It can adjust species population sizes in proportion to its individuals fitness values keeping the total number of pathogens fixed. </p>
<p>Iterates through the pathogen population selecting the next generation for reproduction using fitness proportionate selection method (roulette wheel selection). </p>

<p>Definition at line <a class="el" href="_environment_8cpp_source.html#l00699">699</a> of file <a class="el" href="_environment_8cpp_source.html">Environment.cpp</a>.</p>

<p><div class="dynheader">
Here is the call graph for this function:</div>
<div class="dyncontent">
<div class="center"><img src="class_environment_ad7f96a8620fa9df1601c613130b275e0_cgraph.png" border="0" usemap="#class_environment_ad7f96a8620fa9df1601c613130b275e0_cgraph" alt=""/></div>
<map name="class_environment_ad7f96a8620fa9df1601c613130b275e0_cgraph" id="class_environment_ad7f96a8620fa9df1601c613130b275e0_cgraph">
<area shape="rect" id="node2" href="class_random_numbs.html#a364c415e5105061274013a1e68576a64" title="Core method. Gets the instance of a singleton class. " alt="" coords="277,5,467,32"/><area shape="rect" id="node3" href="class_random_numbs.html#a301ebb1a7c447763903f925a634da605" title="Core method. Generates one integer random number from a given interval with an uniform distribution..." alt="" coords="290,56,454,83"/></map>
</div>
</p>

</div>
</div>
<a class="anchor" id="aeadba01513bfdaef3f247558a6e5ce1d"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void Environment::setHostClonalPopulation </td>
          <td>(</td>
          <td class="paramtype">int&#160;</td>
          <td class="paramname"><em>pop_size</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">int&#160;</td>
          <td class="paramname"><em>gene_size</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">int&#160;</td>
          <td class="paramname"><em>chrom_size</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">int&#160;</td>
          <td class="paramname"><em>timeStamp</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Core method. Initializes a vector containing clonal host population. </p>
<p>Sets and fills a vector containing the host population consisting of a single homozygous clone. Parameters like size of the population, length of genes represented by bit-strings, number of genes in a chromosome (remember that there are two chromosomes) are user-defined, the rest is set at random (e.g. actual gene values).</p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">pop_size</td><td>- number of host individuals in a simulation </td></tr>
    <tr><td class="paramname">gene_size</td><td>- number of bits in bit-represented genes </td></tr>
    <tr><td class="paramname">chrom_size</td><td>- number of genes in a chromosome </td></tr>
    <tr><td class="paramname">timeStamp</td><td>- current time (number of the model iteration) </td></tr>
  </table>
  </dd>
</dl>

<p>Definition at line <a class="el" href="_environment_8cpp_source.html#l00223">223</a> of file <a class="el" href="_environment_8cpp_source.html">Environment.cpp</a>.</p>

<p><div class="dynheader">
Here is the caller graph for this function:</div>
<div class="dyncontent">
<div class="center"><img src="class_environment_aeadba01513bfdaef3f247558a6e5ce1d_icgraph.png" border="0" usemap="#class_environment_aeadba01513bfdaef3f247558a6e5ce1d_icgraph" alt=""/></div>
<map name="class_environment_aeadba01513bfdaef3f247558a6e5ce1d_icgraph" id="class_environment_aeadba01513bfdaef3f247558a6e5ce1d_icgraph">
<area shape="rect" id="node2" href="main_8cpp.html#a3c04138a5bfe5d72780bb7e82a18e627" title="The main function. Things are happening here. " alt="" coords="240,12,291,39"/></map>
</div>
</p>

</div>
</div>
<a class="anchor" id="ae6c1788a9b69c76b657d02e38548e5ce"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void Environment::setHostRandomPopulation </td>
          <td>(</td>
          <td class="paramtype">int&#160;</td>
          <td class="paramname"><em>pop_size</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">int&#160;</td>
          <td class="paramname"><em>gene_size</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">int&#160;</td>
          <td class="paramname"><em>chrom_size</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">int&#160;</td>
          <td class="paramname"><em>timeStamp</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Core method. Initializes a vector containing host population. </p>
<p>Sets and fills a vector containing the host population. Parameters like size of the population, length of genes represented by bit-strings, number of genes in a chromosome (remember that there are two chromosomes) are user-defined, the rest is set at random (e.g. actual gene values).</p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">pop_size</td><td>- number of host individuals in a simulation </td></tr>
    <tr><td class="paramname">gene_size</td><td>- number of bits in bit-represented genes </td></tr>
    <tr><td class="paramname">chrom_size</td><td>- number of genes in a chromosome </td></tr>
    <tr><td class="paramname">timeStamp</td><td>- current time (number of the model iteration) </td></tr>
  </table>
  </dd>
</dl>

<p>Definition at line <a class="el" href="_environment_8cpp_source.html#l00169">169</a> of file <a class="el" href="_environment_8cpp_source.html">Environment.cpp</a>.</p>

</div>
</div>
<a class="anchor" id="a3b828fcd48c6ef8b0d8d7ea61de3817b"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void Environment::setHostRandomPopulation </td>
          <td>(</td>
          <td class="paramtype">int&#160;</td>
          <td class="paramname"><em>pop_size</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">int&#160;</td>
          <td class="paramname"><em>gene_size</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">int&#160;</td>
          <td class="paramname"><em>chrom_size_lower</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">int&#160;</td>
          <td class="paramname"><em>chrom_size_uper</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">int&#160;</td>
          <td class="paramname"><em>timeStamp</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Core method. Initializes a vector containing host population. </p>
<p>Sets and fills a vector containing the host population. Parameters like size of the population and the length of genes represented by bit-strings are user-defined. Whereas the range of the number of genes in a chromosome (remember that there are two chromosomes) is selected randomly from a set range of possible values, the rest of parameters is set totally at random (e.g. actual gene values).</p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">pop_size</td><td>- number of host individuals in a simulation </td></tr>
    <tr><td class="paramname">gene_size</td><td>- number of bits in bit-represented genes </td></tr>
    <tr><td class="paramname">chrom_size_lower</td><td>- lower limit of the number of genes in a chromosome </td></tr>
    <tr><td class="paramname">chrom_size_uper</td><td>- upper limit of the number of genes in a chromosome </td></tr>
    <tr><td class="paramname">timeStamp</td><td>- current time (number of the model iteration) </td></tr>
  </table>
  </dd>
</dl>

<p>Definition at line <a class="el" href="_environment_8cpp_source.html#l00193">193</a> of file <a class="el" href="_environment_8cpp_source.html">Environment.cpp</a>.</p>

<p><div class="dynheader">
Here is the call graph for this function:</div>
<div class="dyncontent">
<div class="center"><img src="class_environment_a3b828fcd48c6ef8b0d8d7ea61de3817b_cgraph.png" border="0" usemap="#class_environment_a3b828fcd48c6ef8b0d8d7ea61de3817b_cgraph" alt=""/></div>
<map name="class_environment_a3b828fcd48c6ef8b0d8d7ea61de3817b_cgraph" id="class_environment_a3b828fcd48c6ef8b0d8d7ea61de3817b_cgraph">
<area shape="rect" id="node2" href="class_random_numbs.html#a364c415e5105061274013a1e68576a64" title="Core method. Gets the instance of a singleton class. " alt="" coords="253,5,443,32"/><area shape="rect" id="node3" href="class_random_numbs.html#a301ebb1a7c447763903f925a634da605" title="Core method. Generates one integer random number from a given interval with an uniform distribution..." alt="" coords="266,56,430,83"/></map>
</div>
</p>

</div>
</div>
<a class="anchor" id="a558656ec1b4948fbbfee1358c79c67d0"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void Environment::setNoMutsVecFourClads </td>
          <td>(</td>
          <td class="paramtype">int&#160;</td>
          <td class="paramname"><em>numb_of_species</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">int&#160;</td>
          <td class="paramname"><em>antigen_size</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">double&#160;</td>
          <td class="paramname"><em>fixedAntigenFrac</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Core method. It defines "no mutation sites" of the antigen for all individual pathogen species in the simulation. It should be run only ones per simulation. </p>
<p>Creates a vector of <a href="http://www.cplusplus.com/reference/set/set/">STD sets</a> containing indices of antigen bits in which mutations are not allowed. The length of the vector equals the number of pathogen species. Each species has its own vector. Number of fixed sites is a user-defined parameter. This version generates only four unique "no-mutation" vectors (all vectors are just copies of one of the four) to differentiate "clads" of pathogen species.</p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">numb_of_species</td><td>- number of pathogen species </td></tr>
    <tr><td class="paramname">antigen_size</td><td>- number of bits per antigen </td></tr>
    <tr><td class="paramname">fixedAntigenFrac</td><td>- fraction of bits in antigens which need to be fixed </td></tr>
  </table>
  </dd>
</dl>

<p>Definition at line <a class="el" href="_environment_8cpp_source.html#l00135">135</a> of file <a class="el" href="_environment_8cpp_source.html">Environment.cpp</a>.</p>

<p><div class="dynheader">
Here is the call graph for this function:</div>
<div class="dyncontent">
<div class="center"><img src="class_environment_a558656ec1b4948fbbfee1358c79c67d0_cgraph.png" border="0" usemap="#class_environment_a558656ec1b4948fbbfee1358c79c67d0_cgraph" alt=""/></div>
<map name="class_environment_a558656ec1b4948fbbfee1358c79c67d0_cgraph" id="class_environment_a558656ec1b4948fbbfee1358c79c67d0_cgraph">
<area shape="rect" id="node2" href="class_random_numbs.html#a364c415e5105061274013a1e68576a64" title="Core method. Gets the instance of a singleton class. " alt="" coords="245,5,435,32"/><area shape="rect" id="node3" href="class_random_numbs.html#aaae77654579060738e55b46c8caba9a6" title="Core method. Generates one real random number from a given interval with an uniform distribution..." alt="" coords="252,56,428,83"/></map>
</div>
</p>

<p><div class="dynheader">
Here is the caller graph for this function:</div>
<div class="dyncontent">
<div class="center"><img src="class_environment_a558656ec1b4948fbbfee1358c79c67d0_icgraph.png" border="0" usemap="#class_environment_a558656ec1b4948fbbfee1358c79c67d0_icgraph" alt=""/></div>
<map name="class_environment_a558656ec1b4948fbbfee1358c79c67d0_icgraph" id="class_environment_a558656ec1b4948fbbfee1358c79c67d0_icgraph">
<area shape="rect" id="node2" href="class_environment.html#ab8e36a6f12db407caf9a63cbeae62a16" title="Core method. Initializes the pathogen population. " alt="" coords="245,5,461,46"/></map>
</div>
</p>

</div>
</div>
<a class="anchor" id="a95cc0c7ecaa8fad54c2d62328134c73e"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void Environment::setNoMutsVecInFours </td>
          <td>(</td>
          <td class="paramtype">int&#160;</td>
          <td class="paramname"><em>numb_of_species</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">int&#160;</td>
          <td class="paramname"><em>antigen_size</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">double&#160;</td>
          <td class="paramname"><em>fixedAntigenFrac</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Core method. It defines "no mutation sites" in 4-bit-long packages of the antigen for all individual pathogen species in the simulation. It should be run only ones per simulation. </p>
<p>Creates a vector of <a href="http://www.cplusplus.com/reference/set/set/">STD sets</a> containing indices of antigen bits in which mutations are not allowed. The length of the vector equals the number of pathogen species. Each species has its own vector. Number of fixed sites is a user-defined parameter.</p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">numb_of_species</td><td>- number of pathogen species </td></tr>
    <tr><td class="paramname">antigen_size</td><td>- number of bits per antigen </td></tr>
    <tr><td class="paramname">fixedAntigenFrac</td><td>- fraction of bits in antigens which need to be fixed </td></tr>
  </table>
  </dd>
</dl>

<p>Definition at line <a class="el" href="_environment_8cpp_source.html#l00095">95</a> of file <a class="el" href="_environment_8cpp_source.html">Environment.cpp</a>.</p>

<p><div class="dynheader">
Here is the call graph for this function:</div>
<div class="dyncontent">
<div class="center"><img src="class_environment_a95cc0c7ecaa8fad54c2d62328134c73e_cgraph.png" border="0" usemap="#class_environment_a95cc0c7ecaa8fad54c2d62328134c73e_cgraph" alt=""/></div>
<map name="class_environment_a95cc0c7ecaa8fad54c2d62328134c73e_cgraph" id="class_environment_a95cc0c7ecaa8fad54c2d62328134c73e_cgraph">
<area shape="rect" id="node2" href="class_random_numbs.html#a364c415e5105061274013a1e68576a64" title="Core method. Gets the instance of a singleton class. " alt="" coords="245,5,435,32"/><area shape="rect" id="node3" href="class_random_numbs.html#aaae77654579060738e55b46c8caba9a6" title="Core method. Generates one real random number from a given interval with an uniform distribution..." alt="" coords="252,56,428,83"/></map>
</div>
</p>

</div>
</div>
<a class="anchor" id="a559d186796a3f21fcaccce6cbbc6fe89"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void Environment::setNoMutsVector </td>
          <td>(</td>
          <td class="paramtype">int&#160;</td>
          <td class="paramname"><em>numb_of_species</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">int&#160;</td>
          <td class="paramname"><em>antigen_size</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">double&#160;</td>
          <td class="paramname"><em>fixedAntigenFrac</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Core method. It defines "no mutation sites" of the antigen for all individual pathogen species in the simulation. It should be run only ones per simulation. </p>
<p>Creates a vector of <a href="http://www.cplusplus.com/reference/set/set/">STD sets</a> containing indices of antigen bits in which mutations are not allowed. The length of the vector equals the number of pathogen species. Each species has its own vector. Number of fixed sites is a user-defined parameter.</p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">numb_of_species</td><td>- number of pathogen species </td></tr>
    <tr><td class="paramname">antigen_size</td><td>- number of bits per antigen </td></tr>
    <tr><td class="paramname">fixedAntigenFrac</td><td>- fraction of bits in antigens which need to be fixed </td></tr>
  </table>
  </dd>
</dl>

<p>Definition at line <a class="el" href="_environment_8cpp_source.html#l00066">66</a> of file <a class="el" href="_environment_8cpp_source.html">Environment.cpp</a>.</p>

<p><div class="dynheader">
Here is the call graph for this function:</div>
<div class="dyncontent">
<div class="center"><img src="class_environment_a559d186796a3f21fcaccce6cbbc6fe89_cgraph.png" border="0" usemap="#class_environment_a559d186796a3f21fcaccce6cbbc6fe89_cgraph" alt=""/></div>
<map name="class_environment_a559d186796a3f21fcaccce6cbbc6fe89_cgraph" id="class_environment_a559d186796a3f21fcaccce6cbbc6fe89_cgraph">
<area shape="rect" id="node2" href="class_random_numbs.html#a364c415e5105061274013a1e68576a64" title="Core method. Gets the instance of a singleton class. " alt="" coords="259,5,448,32"/><area shape="rect" id="node3" href="class_random_numbs.html#aaae77654579060738e55b46c8caba9a6" title="Core method. Generates one real random number from a given interval with an uniform distribution..." alt="" coords="265,56,441,83"/></map>
</div>
</p>

<p><div class="dynheader">
Here is the caller graph for this function:</div>
<div class="dyncontent">
<div class="center"><img src="class_environment_a559d186796a3f21fcaccce6cbbc6fe89_icgraph.png" border="0" usemap="#class_environment_a559d186796a3f21fcaccce6cbbc6fe89_icgraph" alt=""/></div>
<map name="class_environment_a559d186796a3f21fcaccce6cbbc6fe89_icgraph" id="class_environment_a559d186796a3f21fcaccce6cbbc6fe89_icgraph">
<area shape="rect" id="node2" href="class_environment.html#a03d6f594caf898481ef784aafa9b6ed8" title="Core method. Initializes the pathogen population. " alt="" coords="259,5,469,46"/><area shape="rect" id="node3" href="class_environment.html#aa96a77a7b6544503d61e26149c0d4cca" title="Core method. Initializes the pathogen population. " alt="" coords="259,70,469,111"/><area shape="rect" id="node4" href="class_environment.html#a34b326e2e2880881e48ddf9312ff0bc0" title="Core method. Initializes the pathogen population. " alt="" coords="259,135,469,177"/></map>
</div>
</p>

</div>
</div>
<a class="anchor" id="aa96a77a7b6544503d61e26149c0d4cca"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void Environment::setPathoPopulatioDistincSpp </td>
          <td>(</td>
          <td class="paramtype">int&#160;</td>
          <td class="paramname"><em>pop_size</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">int&#160;</td>
          <td class="paramname"><em>antigenSize</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">int&#160;</td>
          <td class="paramname"><em>chrom_size</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">int&#160;</td>
          <td class="paramname"><em>numb_of_species</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">int&#160;</td>
          <td class="paramname"><em>mhcSize</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">int&#160;</td>
          <td class="paramname"><em>timeStamp</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">double&#160;</td>
          <td class="paramname"><em>fixedAntigenFrac</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Core method. Initializes the pathogen population. </p>
<p>Given the number of individuals, number of bit per gene, desired number of genes in a genome and desired number of pathogen species it generates random population of pathogens. Number of individuals will be evenly distributed between species and each species consists of identical clones. Antigens in species are not assigned at random, but are spread evenly in all possible bit strings space.</p>
<p>@param pop_size - total number of individuals </p><dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">antigenSize</td><td>- number of bits per gene </td></tr>
    <tr><td class="paramname">chrom_size</td><td>- number of genes per genome </td></tr>
    <tr><td class="paramname">numb_of_species</td><td>- number of pathogen species </td></tr>
    <tr><td class="paramname">mhcSize</td><td>- number of bits in MHC protein </td></tr>
    <tr><td class="paramname">timeStamp</td><td>- current time (number of the model iteration) </td></tr>
    <tr><td class="paramname">fixedAntigenFrac</td><td>- fraction of bits in antigens which need to be fixed </td></tr>
  </table>
  </dd>
</dl>

<p>Definition at line <a class="el" href="_environment_8cpp_source.html#l00299">299</a> of file <a class="el" href="_environment_8cpp_source.html">Environment.cpp</a>.</p>

<p><div class="dynheader">
Here is the call graph for this function:</div>
<div class="dyncontent">
<div class="center"><img src="class_environment_aa96a77a7b6544503d61e26149c0d4cca_cgraph.png" border="0" usemap="#class_environment_aa96a77a7b6544503d61e26149c0d4cca_cgraph" alt=""/></div>
<map name="class_environment_aa96a77a7b6544503d61e26149c0d4cca_cgraph" id="class_environment_aa96a77a7b6544503d61e26149c0d4cca_cgraph">
<area shape="rect" id="node2" href="class_environment.html#a559d186796a3f21fcaccce6cbbc6fe89" title="Core method. It defines &quot;no mutation sites&quot; of the antigen for all individual pathogen species in the..." alt="" coords="264,31,469,57"/><area shape="rect" id="node3" href="class_random_numbs.html#a364c415e5105061274013a1e68576a64" title="Core method. Gets the instance of a singleton class. " alt="" coords="517,5,707,32"/><area shape="rect" id="node4" href="class_random_numbs.html#aaae77654579060738e55b46c8caba9a6" title="Core method. Generates one real random number from a given interval with an uniform distribution..." alt="" coords="524,56,700,83"/></map>
</div>
</p>

</div>
</div>
<a class="anchor" id="a34b326e2e2880881e48ddf9312ff0bc0"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void Environment::setPathoPopulatioDivSpecies </td>
          <td>(</td>
          <td class="paramtype">int&#160;</td>
          <td class="paramname"><em>pop_size</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">int&#160;</td>
          <td class="paramname"><em>antigenSize</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">int&#160;</td>
          <td class="paramname"><em>chrom_size</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">int&#160;</td>
          <td class="paramname"><em>numb_of_species</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">int&#160;</td>
          <td class="paramname"><em>mhcSize</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">int&#160;</td>
          <td class="paramname"><em>timeStamp</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">double&#160;</td>
          <td class="paramname"><em>fixedAntigenFrac</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Core method. Initializes the pathogen population. </p>
<p>Given the number of individuals, number of bit per gene, desired number of genes in a genome and desired number of pathogen species it generates random population of pathogens. Number of individuals will be evenly distributed between species and each species consists of identical clones.</p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">pop_size</td><td>- total number of individuals </td></tr>
    <tr><td class="paramname">antigenSize</td><td>- number of bits per gene </td></tr>
    <tr><td class="paramname">chrom_size</td><td>- number of genes per genome </td></tr>
    <tr><td class="paramname">numb_of_species</td><td>- number of pathogen species </td></tr>
    <tr><td class="paramname">mhcSize</td><td>- number of bits in MHC protein </td></tr>
    <tr><td class="paramname">timeStamp</td><td>- current time (number of the model iteration) </td></tr>
    <tr><td class="paramname">fixedAntigenFrac</td><td>- fraction of bits in antigens which need to be fixed </td></tr>
  </table>
  </dd>
</dl>

<p>Definition at line <a class="el" href="_environment_8cpp_source.html#l00354">354</a> of file <a class="el" href="_environment_8cpp_source.html">Environment.cpp</a>.</p>

<p><div class="dynheader">
Here is the call graph for this function:</div>
<div class="dyncontent">
<div class="center"><img src="class_environment_a34b326e2e2880881e48ddf9312ff0bc0_cgraph.png" border="0" usemap="#class_environment_a34b326e2e2880881e48ddf9312ff0bc0_cgraph" alt=""/></div>
<map name="class_environment_a34b326e2e2880881e48ddf9312ff0bc0_cgraph" id="class_environment_a34b326e2e2880881e48ddf9312ff0bc0_cgraph">
<area shape="rect" id="node2" href="class_environment.html#a559d186796a3f21fcaccce6cbbc6fe89" title="Core method. It defines &quot;no mutation sites&quot; of the antigen for all individual pathogen species in the..." alt="" coords="264,31,469,57"/><area shape="rect" id="node3" href="class_random_numbs.html#a364c415e5105061274013a1e68576a64" title="Core method. Gets the instance of a singleton class. " alt="" coords="517,5,707,32"/><area shape="rect" id="node4" href="class_random_numbs.html#aaae77654579060738e55b46c8caba9a6" title="Core method. Generates one real random number from a given interval with an uniform distribution..." alt="" coords="524,56,700,83"/></map>
</div>
</p>

</div>
</div>
<a class="anchor" id="ab8e36a6f12db407caf9a63cbeae62a16"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void Environment::setPathoPopulationFourClades </td>
          <td>(</td>
          <td class="paramtype">int&#160;</td>
          <td class="paramname"><em>pop_size</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">int&#160;</td>
          <td class="paramname"><em>antigenSize</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">int&#160;</td>
          <td class="paramname"><em>chrom_size</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">int&#160;</td>
          <td class="paramname"><em>numb_of_species</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">int&#160;</td>
          <td class="paramname"><em>mhcSize</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">int&#160;</td>
          <td class="paramname"><em>timeStamp</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">double&#160;</td>
          <td class="paramname"><em>fixedAntigenFrac</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Core method. Initializes the pathogen population. </p>
<p>Given the number of individuals, number of bit per gene, desired number of genes in a genome and desired number of pathogen species it generates random population of pathogens. Number of individuals will be evenly distributed between species, each species draws its genes from the same pool of possible bit strings and there are only 4 possible "clads" - groups of species initialized with the same antigens and same mutations restrictions.</p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">pop_size</td><td>- total number of individuals </td></tr>
    <tr><td class="paramname">antigenSize</td><td>- number of bits per gene </td></tr>
    <tr><td class="paramname">chrom_size</td><td>- number of genes per genome </td></tr>
    <tr><td class="paramname">numb_of_species</td><td>- number of pathogen species </td></tr>
    <tr><td class="paramname">mhcSize</td><td>- number of bits in MHC protein </td></tr>
    <tr><td class="paramname">timeStamp</td><td>- current time (number of the model iteration) </td></tr>
    <tr><td class="paramname">fixedAntigenFrac</td><td>- fraction of bits in antigens which need to be fixed </td></tr>
  </table>
  </dd>
</dl>

<p>Definition at line <a class="el" href="_environment_8cpp_source.html#l00399">399</a> of file <a class="el" href="_environment_8cpp_source.html">Environment.cpp</a>.</p>

<p><div class="dynheader">
Here is the call graph for this function:</div>
<div class="dyncontent">
<div class="center"><img src="class_environment_ab8e36a6f12db407caf9a63cbeae62a16_cgraph.png" border="0" usemap="#class_environment_ab8e36a6f12db407caf9a63cbeae62a16_cgraph" alt=""/></div>
<map name="class_environment_ab8e36a6f12db407caf9a63cbeae62a16_cgraph" id="class_environment_ab8e36a6f12db407caf9a63cbeae62a16_cgraph">
<area shape="rect" id="node2" href="class_environment.html#a558656ec1b4948fbbfee1358c79c67d0" title="Core method. It defines &quot;no mutation sites&quot; of the antigen for all individual pathogen species in the..." alt="" coords="270,23,461,65"/><area shape="rect" id="node3" href="class_random_numbs.html#a364c415e5105061274013a1e68576a64" title="Core method. Gets the instance of a singleton class. " alt="" coords="509,5,699,32"/><area shape="rect" id="node4" href="class_random_numbs.html#aaae77654579060738e55b46c8caba9a6" title="Core method. Generates one real random number from a given interval with an uniform distribution..." alt="" coords="516,56,692,83"/></map>
</div>
</p>

</div>
</div>
<a class="anchor" id="a03d6f594caf898481ef784aafa9b6ed8"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void Environment::setPathoPopulatioUniformGenome </td>
          <td>(</td>
          <td class="paramtype">int&#160;</td>
          <td class="paramname"><em>pop_size</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">int&#160;</td>
          <td class="paramname"><em>antigenSize</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">int&#160;</td>
          <td class="paramname"><em>chrom_size</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">int&#160;</td>
          <td class="paramname"><em>numb_of_species</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">int&#160;</td>
          <td class="paramname"><em>mhcSize</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">int&#160;</td>
          <td class="paramname"><em>timeStamp</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">double&#160;</td>
          <td class="paramname"><em>fixedAntigenFrac</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Core method. Initializes the pathogen population. </p>
<p>Given the number of individuals, number of bit per gene, desired number of genes in a genome and desired number of pathogen species it generates random population of pathogens. Number of individuals will be evenly distributed between species and each species draws its genes from the same pool of possible bit strings.</p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">pop_size</td><td>- total number of individuals </td></tr>
    <tr><td class="paramname">antigenSize</td><td>- number of bits per gene </td></tr>
    <tr><td class="paramname">chrom_size</td><td>- number of genes per genome </td></tr>
    <tr><td class="paramname">numb_of_species</td><td>- number of pathogen species </td></tr>
    <tr><td class="paramname">mhcSize</td><td>- number of bits in MHC protein </td></tr>
    <tr><td class="paramname">timeStamp</td><td>- current time (number of the model iteration) </td></tr>
    <tr><td class="paramname">fixedAntigenFrac</td><td>- fraction of bits in antigens which need to be fixed </td></tr>
  </table>
  </dd>
</dl>

<p>Definition at line <a class="el" href="_environment_8cpp_source.html#l00251">251</a> of file <a class="el" href="_environment_8cpp_source.html">Environment.cpp</a>.</p>

<p><div class="dynheader">
Here is the call graph for this function:</div>
<div class="dyncontent">
<div class="center"><img src="class_environment_a03d6f594caf898481ef784aafa9b6ed8_cgraph.png" border="0" usemap="#class_environment_a03d6f594caf898481ef784aafa9b6ed8_cgraph" alt=""/></div>
<map name="class_environment_a03d6f594caf898481ef784aafa9b6ed8_cgraph" id="class_environment_a03d6f594caf898481ef784aafa9b6ed8_cgraph">
<area shape="rect" id="node2" href="class_environment.html#a559d186796a3f21fcaccce6cbbc6fe89" title="Core method. It defines &quot;no mutation sites&quot; of the antigen for all individual pathogen species in the..." alt="" coords="264,31,469,57"/><area shape="rect" id="node3" href="class_random_numbs.html#a364c415e5105061274013a1e68576a64" title="Core method. Gets the instance of a singleton class. " alt="" coords="517,5,707,32"/><area shape="rect" id="node4" href="class_random_numbs.html#aaae77654579060738e55b46c8caba9a6" title="Core method. Generates one real random number from a given interval with an uniform distribution..." alt="" coords="524,56,700,83"/></map>
</div>
</p>

</div>
</div>
<hr/>The documentation for this class was generated from the following files:<ul>
<li><a class="el" href="_environment_8h_source.html">Environment.h</a></li>
<li><a class="el" href="_environment_8cpp_source.html">Environment.cpp</a></li>
</ul>
</div><!-- contents -->
</div><!-- doc-content -->
<!-- start footer part -->
<div id="nav-path" class="navpath"><!-- id is needed for treeview function! -->
  <ul>
    <li class="navelem"><a class="el" href="class_environment.html">Environment</a></li>
    <li class="footer">Generated on Fri Aug 19 2016 19:51:24 for Modelling MHC Evolutionary Dynamics by
    <a href="http://www.doxygen.org/index.html">
    <img class="footer" src="doxygen.png" alt="doxygen"/></a> 1.8.7 </li>
  </ul>
</div>
</body>
</html>
