<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml">
<head>
<meta http-equiv="Content-Type" content="text/xhtml;charset=UTF-8"/>
<meta http-equiv="X-UA-Compatible" content="IE=9"/>
<meta name="generator" content="Doxygen 1.8.13"/>
<meta name="viewport" content="width=device-width, initial-scale=1"/>
<title>Modelling MHC Evolutionary Dynamics: DataHandler Class Reference</title>
<link href="tabs.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="jquery.js"></script>
<script type="text/javascript" src="dynsections.js"></script>
<link href="navtree.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="resize.js"></script>
<script type="text/javascript" src="navtreedata.js"></script>
<script type="text/javascript" src="navtree.js"></script>
<script type="text/javascript">
  $(document).ready(initResizable);
</script>
<link href="search/search.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="search/searchdata.js"></script>
<script type="text/javascript" src="search/search.js"></script>
<link href="doxygen.css" rel="stylesheet" type="text/css" />
</head>
<body>
<div id="top"><!-- do not remove this div, it is closed by doxygen! -->
<div id="titlearea">
<table cellspacing="0" cellpadding="0">
 <tbody>
 <tr style="height: 56px;">
  <td id="projectalign" style="padding-left: 0.5em;">
   <div id="projectname">Modelling MHC Evolutionary Dynamics
   &#160;<span id="projectnumber">0.9.3</span>
   </div>
  </td>
 </tr>
 </tbody>
</table>
</div>
<!-- end header part -->
<!-- Generated by Doxygen 1.8.13 -->
<script type="text/javascript">
var searchBox = new SearchBox("searchBox", "search",false,'Search');
</script>
<script type="text/javascript" src="menudata.js"></script>
<script type="text/javascript" src="menu.js"></script>
<script type="text/javascript">
$(function() {
  initMenu('',true,false,'search.php','Search');
  $(document).ready(function() { init_search(); });
});
</script>
<div id="main-nav"></div>
</div><!-- top -->
<div id="side-nav" class="ui-resizable side-nav-resizable">
  <div id="nav-tree">
    <div id="nav-tree-contents">
      <div id="nav-sync" class="sync"></div>
    </div>
  </div>
  <div id="splitbar" style="-moz-user-select:none;" 
       class="ui-resizable-handle">
  </div>
</div>
<script type="text/javascript">
$(document).ready(function(){initNavTree('class_data_handler.html','');});
</script>
<div id="doc-content">
<!-- window showing the filter options -->
<div id="MSearchSelectWindow"
     onmouseover="return searchBox.OnSearchSelectShow()"
     onmouseout="return searchBox.OnSearchSelectHide()"
     onkeydown="return searchBox.OnSearchSelectKey(event)">
</div>

<!-- iframe showing the search results (closed by default) -->
<div id="MSearchResultsWindow">
<iframe src="javascript:void(0)" frameborder="0" 
        name="MSearchResults" id="MSearchResults">
</iframe>
</div>

<div class="header">
  <div class="summary">
<a href="#pub-methods">Public Member Functions</a> &#124;
<a href="class_data_handler-members.html">List of all members</a>  </div>
  <div class="headertitle">
<div class="title">DataHandler Class Reference</div>  </div>
</div><!--header-->
<div class="contents">

<p>Data harvesting class. A class that has methods to collect data and to write them to output files. It heavily relies on the <a class="el" href="class_environment.html" title="Core class. Stores and handles the environment object that is the place where all the things are happ...">Environment</a> class that stores some of the data and has methods to access other kind of data.  
 <a href="class_data_handler.html#details">More...</a></p>

<p><code>#include &lt;<a class="el" href="_data_handler_8h_source.html">DataHandler.h</a>&gt;</code></p>
<table class="memberdecls">
<tr class="heading"><td colspan="2"><h2 class="groupheader"><a name="pub-methods"></a>
Public Member Functions</h2></td></tr>
<tr class="memitem:a51d845c1d9bbeef44607ec7592ba3c1d"><td class="memItemLeft" align="right" valign="top">&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_data_handler.html#a51d845c1d9bbeef44607ec7592ba3c1d">DataHandler</a> ()</td></tr>
<tr class="separator:a51d845c1d9bbeef44607ec7592ba3c1d"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a25f2ab01e70edd3a1c8b0ac4f25e0cd4"><td class="memItemLeft" align="right" valign="top">virtual&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_data_handler.html#a25f2ab01e70edd3a1c8b0ac4f25e0cd4">~DataHandler</a> ()</td></tr>
<tr class="separator:a25f2ab01e70edd3a1c8b0ac4f25e0cd4"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ad6e9cf11f43ceae1a20908f291470c0a"><td class="memItemLeft" align="right" valign="top">bool&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_data_handler.html#ad6e9cf11f43ceae1a20908f291470c0a">checkParamsIfWrong</a> (unsigned int numberOfThreads, unsigned long geneLength, unsigned long antigenLength, int hostPopSize, int pathoPopSize, int patho_sp, unsigned long hostGeneNumbb, int patoPerHostGeneration, int numOfHostGenerations, double hostMutationProb, double pathoMutationProb, int HeteroHomo, double hostDeletion, double hostDuplication, unsigned long maxGene, double alpha, int numberOfMates)</td></tr>
<tr class="memdesc:ad6e9cf11f43ceae1a20908f291470c0a"><td class="mdescLeft">&#160;</td><td class="mdescRight">Input params validation method. Does the basic check if the entered parameters are free of total nonsense. This one is for sex optimality hypothesis and alpha-based selection.  <a href="#ad6e9cf11f43ceae1a20908f291470c0a">More...</a><br /></td></tr>
<tr class="separator:ad6e9cf11f43ceae1a20908f291470c0a"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a39601b65461d1d628ce87a2a7be8d926"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_data_handler.html#a39601b65461d1d628ce87a2a7be8d926">inputParamsToFile</a> (unsigned int numberOfThreads, unsigned long geneLength, unsigned long antigenLength, int hostPopSize, int pathoPopSize, int patho_sp, unsigned long hostGeneNumbb, int patoPerHostGeneration, int numOfHostGenerations, double hostMutationProb, double pathoMutationProb, int HeteroHomo, double hostDeletion, double hostDuplication, unsigned long maxGene, double alpha, int numberOfMates)</td></tr>
<tr class="memdesc:a39601b65461d1d628ce87a2a7be8d926"><td class="mdescLeft">&#160;</td><td class="mdescRight">Saves given parametrisation of the model to a JSON file. Make sure you always have this JSON file, or you will loose track what simulation did what!  <a href="#a39601b65461d1d628ce87a2a7be8d926">More...</a><br /></td></tr>
<tr class="separator:a39601b65461d1d628ce87a2a7be8d926"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a049645bbd6069ad5d6a90c8469390811"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_data_handler.html#a049645bbd6069ad5d6a90c8469390811">setAllFilesAsFirtsTimers</a> ()</td></tr>
<tr class="memdesc:a049645bbd6069ad5d6a90c8469390811"><td class="mdescLeft">&#160;</td><td class="mdescRight">Data harvesting method. Sets status of all data files as "brand new".  <a href="#a049645bbd6069ad5d6a90c8469390811">More...</a><br /></td></tr>
<tr class="separator:a049645bbd6069ad5d6a90c8469390811"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a89f86a67a11f719ec282f37740663f95"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_data_handler.html#a89f86a67a11f719ec282f37740663f95">saveNumOfPathoSpeciesToFile</a> (<a class="el" href="class_environment.html">Environment</a> &amp;EnvObj, int tayme)</td></tr>
<tr class="memdesc:a89f86a67a11f719ec282f37740663f95"><td class="mdescLeft">&#160;</td><td class="mdescRight">Data harvesting method. Writes to a file population sizes of all pathogen species in a given time.  <a href="#a89f86a67a11f719ec282f37740663f95">More...</a><br /></td></tr>
<tr class="separator:a89f86a67a11f719ec282f37740663f95"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a877ed5bfd00f0cbeef0413bf1d47f865"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_data_handler.html#a877ed5bfd00f0cbeef0413bf1d47f865">savePathoPopulToFile</a> (<a class="el" href="class_environment.html">Environment</a> &amp;EnvObj, int tayme)</td></tr>
<tr class="memdesc:a877ed5bfd00f0cbeef0413bf1d47f865"><td class="mdescLeft">&#160;</td><td class="mdescRight">Data harvesting method. Writes to a file all pathogens with their genomes in a human-readable format.  <a href="#a877ed5bfd00f0cbeef0413bf1d47f865">More...</a><br /></td></tr>
<tr class="separator:a877ed5bfd00f0cbeef0413bf1d47f865"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a46129d74eac82af0caae37a9e20e3e2e"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_data_handler.html#a46129d74eac82af0caae37a9e20e3e2e">saveHostPopulToFile</a> (<a class="el" href="class_environment.html">Environment</a> &amp;EnvObj, int tayme)</td></tr>
<tr class="memdesc:a46129d74eac82af0caae37a9e20e3e2e"><td class="mdescLeft">&#160;</td><td class="mdescRight">Data harvesting method. Writes to a file all hosts with their genomes in a human-readable format.  <a href="#a46129d74eac82af0caae37a9e20e3e2e">More...</a><br /></td></tr>
<tr class="separator:a46129d74eac82af0caae37a9e20e3e2e"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a23ad0de50b3da065674e92b150e12d9f"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_data_handler.html#a23ad0de50b3da065674e92b150e12d9f">saveHostGeneticDivers</a> (<a class="el" href="class_environment.html">Environment</a> &amp;EnvObj, int tayme)</td></tr>
<tr class="memdesc:a23ad0de50b3da065674e92b150e12d9f"><td class="mdescLeft">&#160;</td><td class="mdescRight">Data harvesting method. Calculates and writes to a file some stats about the hosts population genetic. Will create one file per run. First column is time stamp (hosts generation number).  <a href="#a23ad0de50b3da065674e92b150e12d9f">More...</a><br /></td></tr>
<tr class="separator:a23ad0de50b3da065674e92b150e12d9f"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a5661a45f0a9965bcd041c5617f1a1e8b"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_data_handler.html#a5661a45f0a9965bcd041c5617f1a1e8b">saveHostGeneNumbers</a> (<a class="el" href="class_environment.html">Environment</a> &amp;EnvObj, int tayme)</td></tr>
<tr class="memdesc:a5661a45f0a9965bcd041c5617f1a1e8b"><td class="mdescLeft">&#160;</td><td class="mdescRight">Data harvesting method. Saves the number of genes in Chromosome ONE and the number of unique MHC types. You get a pair of files with corresponding data in each row-and-column. First column is time stamp (hosts generation number).  <a href="#a5661a45f0a9965bcd041c5617f1a1e8b">More...</a><br /></td></tr>
<tr class="separator:a5661a45f0a9965bcd041c5617f1a1e8b"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a6378c36aa8e4daf8f525c0fee446897e"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_data_handler.html#a6378c36aa8e4daf8f525c0fee446897e">savePathoNoMuttList</a> (<a class="el" href="class_environment.html">Environment</a> &amp;EnvObj)</td></tr>
<tr class="memdesc:a6378c36aa8e4daf8f525c0fee446897e"><td class="mdescLeft">&#160;</td><td class="mdescRight">Data harvesting method. Record the indices of fixed bits that are not allow to mutate in all pathogen species. Run this just ones.  <a href="#a6378c36aa8e4daf8f525c0fee446897e">More...</a><br /></td></tr>
<tr class="separator:a6378c36aa8e4daf8f525c0fee446897e"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ac0bd3c1ba0fa18e132ea9095db9747cd"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_data_handler.html#ac0bd3c1ba0fa18e132ea9095db9747cd">savePresentedPathos</a> (<a class="el" href="class_environment.html">Environment</a> &amp;EnvObj, int tayme)</td></tr>
<tr class="memdesc:ac0bd3c1ba0fa18e132ea9095db9747cd"><td class="mdescLeft">&#160;</td><td class="mdescRight">Data harvesting method. Saves the number of presented pathogens by each indrividual host in each time step. First column is time stamp (hosts generation number).  <a href="#ac0bd3c1ba0fa18e132ea9095db9747cd">More...</a><br /></td></tr>
<tr class="separator:ac0bd3c1ba0fa18e132ea9095db9747cd"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ab7b38c92d0d564999a1159ceb56494e4"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_data_handler.html#ab7b38c92d0d564999a1159ceb56494e4">saveMhcNumbersWhenMating</a> (<a class="el" href="class_environment.html">Environment</a> &amp;EnvObj, int tayme)</td></tr>
<tr class="memdesc:ab7b38c92d0d564999a1159ceb56494e4"><td class="mdescLeft">&#160;</td><td class="mdescRight">Data harvesting method. Records the number of unique MHCs a "mother" and a "farther" had. First column is time stamp (hosts generation number).  <a href="#ab7b38c92d0d564999a1159ceb56494e4">More...</a><br /></td></tr>
<tr class="separator:ab7b38c92d0d564999a1159ceb56494e4"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a791581a6ac572d0daabd2c5bf8a53033"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_data_handler.html#a791581a6ac572d0daabd2c5bf8a53033">saveMhcNumbersBeforeMating</a> (<a class="el" href="class_environment.html">Environment</a> &amp;EnvObj, int tayme)</td></tr>
<tr class="memdesc:a791581a6ac572d0daabd2c5bf8a53033"><td class="mdescLeft">&#160;</td><td class="mdescRight">Saves the number of unique MHC types of each individual in each time step. Should be used BEFORE mating occurs. First column is time stamp (hosts generation number).  <a href="#a791581a6ac572d0daabd2c5bf8a53033">More...</a><br /></td></tr>
<tr class="separator:a791581a6ac572d0daabd2c5bf8a53033"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a95cd188763d53b65e7e2e658bb76760b"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_data_handler.html#a95cd188763d53b65e7e2e658bb76760b">saveMhcNumbersAfterMating</a> (<a class="el" href="class_environment.html">Environment</a> &amp;EnvObj, int tayme)</td></tr>
<tr class="memdesc:a95cd188763d53b65e7e2e658bb76760b"><td class="mdescLeft">&#160;</td><td class="mdescRight">Saves the number of unique MHC types of each individual in each time step. Should be used AFTER mating occurs.  <a href="#a95cd188763d53b65e7e2e658bb76760b">More...</a><br /></td></tr>
<tr class="separator:a95cd188763d53b65e7e2e658bb76760b"><td class="memSeparator" colspan="2">&#160;</td></tr>
</table>
<a name="details" id="details"></a><h2 class="groupheader">Detailed Description</h2>
<div class="textblock"><p>Data harvesting class. A class that has methods to collect data and to write them to output files. It heavily relies on the <a class="el" href="class_environment.html" title="Core class. Stores and handles the environment object that is the place where all the things are happ...">Environment</a> class that stores some of the data and has methods to access other kind of data. </p>

<p class="definition">Definition at line <a class="el" href="_data_handler_8h_source.html#l00038">38</a> of file <a class="el" href="_data_handler_8h_source.html">DataHandler.h</a>.</p>
</div><h2 class="groupheader">Constructor &amp; Destructor Documentation</h2>
<a id="a51d845c1d9bbeef44607ec7592ba3c1d"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a51d845c1d9bbeef44607ec7592ba3c1d">&#9670;&nbsp;</a></span>DataHandler()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">DataHandler::DataHandler </td>
          <td>(</td>
          <td class="paramname"></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p class="definition">Definition at line <a class="el" href="_data_handler_8cpp_source.html#l00034">34</a> of file <a class="el" href="_data_handler_8cpp_source.html">DataHandler.cpp</a>.</p>

</div>
</div>
<a id="a25f2ab01e70edd3a1c8b0ac4f25e0cd4"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a25f2ab01e70edd3a1c8b0ac4f25e0cd4">&#9670;&nbsp;</a></span>~DataHandler()</h2>

<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">DataHandler::~DataHandler </td>
          <td>(</td>
          <td class="paramname"></td><td>)</td>
          <td></td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">virtual</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">

<p class="definition">Definition at line <a class="el" href="_data_handler_8cpp_source.html#l00040">40</a> of file <a class="el" href="_data_handler_8cpp_source.html">DataHandler.cpp</a>.</p>

</div>
</div>
<h2 class="groupheader">Member Function Documentation</h2>
<a id="ad6e9cf11f43ceae1a20908f291470c0a"></a>
<h2 class="memtitle"><span class="permalink"><a href="#ad6e9cf11f43ceae1a20908f291470c0a">&#9670;&nbsp;</a></span>checkParamsIfWrong()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">bool DataHandler::checkParamsIfWrong </td>
          <td>(</td>
          <td class="paramtype">unsigned int&#160;</td>
          <td class="paramname"><em>numberOfThreads</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">unsigned long&#160;</td>
          <td class="paramname"><em>geneLength</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">unsigned long&#160;</td>
          <td class="paramname"><em>antigenLength</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">int&#160;</td>
          <td class="paramname"><em>hostPopSize</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">int&#160;</td>
          <td class="paramname"><em>pathoPopSize</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">int&#160;</td>
          <td class="paramname"><em>patho_sp</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">unsigned long&#160;</td>
          <td class="paramname"><em>hostGeneNumbb</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">int&#160;</td>
          <td class="paramname"><em>patoPerHostGeneration</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">int&#160;</td>
          <td class="paramname"><em>numOfHostGenerations</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">double&#160;</td>
          <td class="paramname"><em>hostMutationProb</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">double&#160;</td>
          <td class="paramname"><em>pathoMutationProb</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">int&#160;</td>
          <td class="paramname"><em>HeteroHomo</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">double&#160;</td>
          <td class="paramname"><em>hostDeletion</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">double&#160;</td>
          <td class="paramname"><em>hostDuplication</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">unsigned long&#160;</td>
          <td class="paramname"><em>maxGene</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">double&#160;</td>
          <td class="paramname"><em>alpha</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">int&#160;</td>
          <td class="paramname"><em>numberOfMates</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Input params validation method. Does the basic check if the entered parameters are free of total nonsense. This one is for sex optimality hypothesis and alpha-based selection. </p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">rndSeed</td><td></td></tr>
    <tr><td class="paramname">geneLength</td><td></td></tr>
    <tr><td class="paramname">antigenLength</td><td></td></tr>
    <tr><td class="paramname">hostPopSize</td><td></td></tr>
    <tr><td class="paramname">pathoPopSize</td><td></td></tr>
    <tr><td class="paramname">patho_sp</td><td></td></tr>
    <tr><td class="paramname">hostGeneNumbb</td><td></td></tr>
    <tr><td class="paramname">pathoGeneNumb</td><td></td></tr>
    <tr><td class="paramname">patoPerHostGeneration</td><td></td></tr>
    <tr><td class="paramname">numOfHostGenerations</td><td></td></tr>
    <tr><td class="paramname">hostMutationProb</td><td></td></tr>
    <tr><td class="paramname">pathoMutationProb</td><td></td></tr>
    <tr><td class="paramname">HeteroHomo</td><td></td></tr>
    <tr><td class="paramname">hostDeletion</td><td></td></tr>
    <tr><td class="paramname">hostDuplication</td><td></td></tr>
    <tr><td class="paramname">maxGene</td><td></td></tr>
    <tr><td class="paramname">alpha</td><td></td></tr>
    <tr><td class="paramname">numberOfMates</td><td></td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>'true' if something is wrong, 'false' if no errors were found. </dd></dl>

<p class="definition">Definition at line <a class="el" href="_data_handler_8cpp_source.html#l00142">142</a> of file <a class="el" href="_data_handler_8cpp_source.html">DataHandler.cpp</a>.</p>

</div>
</div>
<a id="a39601b65461d1d628ce87a2a7be8d926"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a39601b65461d1d628ce87a2a7be8d926">&#9670;&nbsp;</a></span>inputParamsToFile()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void DataHandler::inputParamsToFile </td>
          <td>(</td>
          <td class="paramtype">unsigned int&#160;</td>
          <td class="paramname"><em>numberOfThreads</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">unsigned long&#160;</td>
          <td class="paramname"><em>geneLength</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">unsigned long&#160;</td>
          <td class="paramname"><em>antigenLength</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">int&#160;</td>
          <td class="paramname"><em>hostPopSize</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">int&#160;</td>
          <td class="paramname"><em>pathoPopSize</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">int&#160;</td>
          <td class="paramname"><em>patho_sp</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">unsigned long&#160;</td>
          <td class="paramname"><em>hostGeneNumbb</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">int&#160;</td>
          <td class="paramname"><em>patoPerHostGeneration</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">int&#160;</td>
          <td class="paramname"><em>numOfHostGenerations</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">double&#160;</td>
          <td class="paramname"><em>hostMutationProb</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">double&#160;</td>
          <td class="paramname"><em>pathoMutationProb</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">int&#160;</td>
          <td class="paramname"><em>HeteroHomo</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">double&#160;</td>
          <td class="paramname"><em>hostDeletion</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">double&#160;</td>
          <td class="paramname"><em>hostDuplication</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">unsigned long&#160;</td>
          <td class="paramname"><em>maxGene</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">double&#160;</td>
          <td class="paramname"><em>alpha</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">int&#160;</td>
          <td class="paramname"><em>numberOfMates</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Saves given parametrisation of the model to a JSON file. Make sure you always have this JSON file, or you will loose track what simulation did what! </p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">numberOfThreads</td><td></td></tr>
    <tr><td class="paramname">geneLength</td><td></td></tr>
    <tr><td class="paramname">antigenLength</td><td></td></tr>
    <tr><td class="paramname">hostPopSize</td><td></td></tr>
    <tr><td class="paramname">pathoPopSize</td><td></td></tr>
    <tr><td class="paramname">patho_sp</td><td></td></tr>
    <tr><td class="paramname">hostGeneNumbb</td><td></td></tr>
    <tr><td class="paramname">patoPerHostGeneration</td><td></td></tr>
    <tr><td class="paramname">numOfHostGenerations</td><td></td></tr>
    <tr><td class="paramname">hostMutationProb</td><td></td></tr>
    <tr><td class="paramname">pathoMutationProb</td><td></td></tr>
    <tr><td class="paramname">HeteroHomo</td><td></td></tr>
    <tr><td class="paramname">hostDeletion</td><td></td></tr>
    <tr><td class="paramname">hostDuplication</td><td></td></tr>
    <tr><td class="paramname">maxGene</td><td></td></tr>
    <tr><td class="paramname">alpha</td><td></td></tr>
    <tr><td class="paramname">numberOfMates</td><td></td></tr>
  </table>
  </dd>
</dl>

<p class="definition">Definition at line <a class="el" href="_data_handler_8cpp_source.html#l00220">220</a> of file <a class="el" href="_data_handler_8cpp_source.html">DataHandler.cpp</a>.</p>
<div class="dynheader">
Here is the call graph for this function:</div>
<div class="dyncontent">
<div class="center"><img src="class_data_handler_a39601b65461d1d628ce87a2a7be8d926_cgraph.png" border="0" usemap="#class_data_handler_a39601b65461d1d628ce87a2a7be8d926_cgraph" alt=""/></div>
<map name="class_data_handler_a39601b65461d1d628ce87a2a7be8d926_cgraph" id="class_data_handler_a39601b65461d1d628ce87a2a7be8d926_cgraph">
<area shape="rect" id="node2" href="_data_handler_8cpp.html#aa46369f3c8adbff876c82270346fffa2" title="Data harvesting method. Gets current date/time, format is YYYY&#45;MM&#45;DD.HH:mm:ss. " alt="" coords="229,13,348,39"/>
</map>
</div>

</div>
</div>
<a id="a5661a45f0a9965bcd041c5617f1a1e8b"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a5661a45f0a9965bcd041c5617f1a1e8b">&#9670;&nbsp;</a></span>saveHostGeneNumbers()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void DataHandler::saveHostGeneNumbers </td>
          <td>(</td>
          <td class="paramtype"><a class="el" href="class_environment.html">Environment</a> &amp;&#160;</td>
          <td class="paramname"><em>EnvObj</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">int&#160;</td>
          <td class="paramname"><em>tayme</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Data harvesting method. Saves the number of genes in Chromosome ONE and the number of unique MHC types. You get a pair of files with corresponding data in each row-and-column. First column is time stamp (hosts generation number). </p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">EnvObj</td><td>- the <a class="el" href="class_environment.html" title="Core class. Stores and handles the environment object that is the place where all the things are happ...">Environment</a> class object </td></tr>
    <tr><td class="paramname">tayme</td><td>- time stamp (hosts generation number) </td></tr>
  </table>
  </dd>
</dl>

<p class="definition">Definition at line <a class="el" href="_data_handler_8cpp_source.html#l00399">399</a> of file <a class="el" href="_data_handler_8cpp_source.html">DataHandler.cpp</a>.</p>
<div class="dynheader">
Here is the call graph for this function:</div>
<div class="dyncontent">
<div class="center"><img src="class_data_handler_a5661a45f0a9965bcd041c5617f1a1e8b_cgraph.png" border="0" usemap="#class_data_handler_a5661a45f0a9965bcd041c5617f1a1e8b_cgraph" alt=""/></div>
<map name="class_data_handler_a5661a45f0a9965bcd041c5617f1a1e8b_cgraph" id="class_data_handler_a5661a45f0a9965bcd041c5617f1a1e8b_cgraph">
<area shape="rect" id="node2" href="class_environment.html#a45aeebf99b44c9b1e66ca16b923605ae" title="Core method. Gets the host population size. " alt="" coords="244,5,449,32"/>
<area shape="rect" id="node3" href="class_environment.html#af7fee696722912338dd1b540adf3dc02" title="Environment::getSingleHost\lChromoOneSize" alt="" coords="254,57,439,98"/>
<area shape="rect" id="node4" href="class_environment.html#ad5a2510424df27c93774fe0fde37e08b" title="Environment::getSingleHost\lRealGeneOne" alt="" coords="254,122,439,163"/>
<area shape="rect" id="node5" href="_data_handler_8cpp.html#a99feb0968fda949e69c4ff4aa9dd0a16" title="Data harvesting method. Calculates some stats of population genetics: Shannon&#39;s index, number of MHC/antigen types, total number of MHC copies. Runs internally within DataHarvester class. " alt="" coords="289,188,405,215"/>
</map>
</div>

</div>
</div>
<a id="a23ad0de50b3da065674e92b150e12d9f"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a23ad0de50b3da065674e92b150e12d9f">&#9670;&nbsp;</a></span>saveHostGeneticDivers()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void DataHandler::saveHostGeneticDivers </td>
          <td>(</td>
          <td class="paramtype"><a class="el" href="class_environment.html">Environment</a> &amp;&#160;</td>
          <td class="paramname"><em>EnvObj</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">int&#160;</td>
          <td class="paramname"><em>tayme</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Data harvesting method. Calculates and writes to a file some stats about the hosts population genetic. Will create one file per run. First column is time stamp (hosts generation number). </p>
<p>You call it ones per host generation iteration. In each call it adds a line with some statistics regarding population genetics of the host population. Data regarding individuals are not being stored here. We wrote custom Python scripts to visualize this dataset.</p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">EnvObj</td><td>- the <a class="el" href="class_environment.html" title="Core class. Stores and handles the environment object that is the place where all the things are happ...">Environment</a> object </td></tr>
    <tr><td class="paramname">tayme</td><td>- time stamp (hosts generation number) </td></tr>
  </table>
  </dd>
</dl>

<p class="definition">Definition at line <a class="el" href="_data_handler_8cpp_source.html#l00342">342</a> of file <a class="el" href="_data_handler_8cpp_source.html">DataHandler.cpp</a>.</p>
<div class="dynheader">
Here is the call graph for this function:</div>
<div class="dyncontent">
<div class="center"><img src="class_data_handler_a23ad0de50b3da065674e92b150e12d9f_cgraph.png" border="0" usemap="#class_data_handler_a23ad0de50b3da065674e92b150e12d9f_cgraph" alt=""/></div>
<map name="class_data_handler_a23ad0de50b3da065674e92b150e12d9f_cgraph" id="class_data_handler_a23ad0de50b3da065674e92b150e12d9f_cgraph">
<area shape="rect" id="node2" href="class_environment.html#a45aeebf99b44c9b1e66ca16b923605ae" title="Core method. Gets the host population size. " alt="" coords="257,5,463,32"/>
<area shape="rect" id="node3" href="class_environment.html#ab34c398994fe34c54bdb4239ac624ac0" title="Environment::getHostFitness" alt="" coords="264,56,456,83"/>
<area shape="rect" id="node4" href="class_environment.html#af7fee696722912338dd1b540adf3dc02" title="Environment::getSingleHost\lChromoOneSize" alt="" coords="267,107,453,149"/>
<area shape="rect" id="node5" href="class_environment.html#ad5a2510424df27c93774fe0fde37e08b" title="Environment::getSingleHost\lRealGeneOne" alt="" coords="267,173,453,214"/>
<area shape="rect" id="node6" href="class_environment.html#a15990da9399f9c49b7e4bcb84ccace45" title="Environment::getSingleHost\lChromoTwoSize" alt="" coords="267,238,453,279"/>
<area shape="rect" id="node7" href="class_environment.html#af838d055fa650935798fe466e19809e4" title="Environment::getSingleHost\lRealGeneTwo" alt="" coords="267,303,453,345"/>
<area shape="rect" id="node8" href="_data_handler_8cpp.html#a99feb0968fda949e69c4ff4aa9dd0a16" title="Data harvesting method. Calculates some stats of population genetics: Shannon&#39;s index, number of MHC/antigen types, total number of MHC copies. Runs internally within DataHarvester class. " alt="" coords="302,369,418,396"/>
</map>
</div>

</div>
</div>
<a id="a46129d74eac82af0caae37a9e20e3e2e"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a46129d74eac82af0caae37a9e20e3e2e">&#9670;&nbsp;</a></span>saveHostPopulToFile()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void DataHandler::saveHostPopulToFile </td>
          <td>(</td>
          <td class="paramtype"><a class="el" href="class_environment.html">Environment</a> &amp;&#160;</td>
          <td class="paramname"><em>EnvObj</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">int&#160;</td>
          <td class="paramname"><em>tayme</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Data harvesting method. Writes to a file all hosts with their genomes in a human-readable format. </p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">EnvObj</td><td>- the <a class="el" href="class_environment.html" title="Core class. Stores and handles the environment object that is the place where all the things are happ...">Environment</a> object </td></tr>
    <tr><td class="paramname">tayme</td><td>- time stamp (hosts generation number) </td></tr>
  </table>
  </dd>
</dl>

<p class="definition">Definition at line <a class="el" href="_data_handler_8cpp_source.html#l00314">314</a> of file <a class="el" href="_data_handler_8cpp_source.html">DataHandler.cpp</a>.</p>
<div class="dynheader">
Here is the call graph for this function:</div>
<div class="dyncontent">
<div class="center"><img src="class_data_handler_a46129d74eac82af0caae37a9e20e3e2e_cgraph.png" border="0" usemap="#class_data_handler_a46129d74eac82af0caae37a9e20e3e2e_cgraph" alt=""/></div>
<map name="class_data_handler_a46129d74eac82af0caae37a9e20e3e2e_cgraph" id="class_data_handler_a46129d74eac82af0caae37a9e20e3e2e_cgraph">
<area shape="rect" id="node2" href="class_environment.html#a45aeebf99b44c9b1e66ca16b923605ae" title="Core method. Gets the host population size. " alt="" coords="245,5,451,32"/>
<area shape="rect" id="node3" href="class_environment.html#a71dc8a1828a2830ffefa901d0974b2c1" title="Data harvesting method. Gets the genome of selected host in human&#45;readable format. " alt="" coords="255,57,441,98"/>
</map>
</div>

</div>
</div>
<a id="a95cd188763d53b65e7e2e658bb76760b"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a95cd188763d53b65e7e2e658bb76760b">&#9670;&nbsp;</a></span>saveMhcNumbersAfterMating()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void DataHandler::saveMhcNumbersAfterMating </td>
          <td>(</td>
          <td class="paramtype"><a class="el" href="class_environment.html">Environment</a> &amp;&#160;</td>
          <td class="paramname"><em>EnvObj</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">int&#160;</td>
          <td class="paramname"><em>tayme</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Saves the number of unique MHC types of each individual in each time step. Should be used AFTER mating occurs. </p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">EnvObj</td><td>- the <a class="el" href="class_environment.html" title="Core class. Stores and handles the environment object that is the place where all the things are happ...">Environment</a> class object </td></tr>
    <tr><td class="paramname">tayme</td><td>- time stamp (hosts generation number) </td></tr>
  </table>
  </dd>
</dl>

<p class="definition">Definition at line <a class="el" href="_data_handler_8cpp_source.html#l00581">581</a> of file <a class="el" href="_data_handler_8cpp_source.html">DataHandler.cpp</a>.</p>
<div class="dynheader">
Here is the call graph for this function:</div>
<div class="dyncontent">
<div class="center"><img src="class_data_handler_a95cd188763d53b65e7e2e658bb76760b_cgraph.png" border="0" usemap="#class_data_handler_a95cd188763d53b65e7e2e658bb76760b_cgraph" alt=""/></div>
<map name="class_data_handler_a95cd188763d53b65e7e2e658bb76760b_cgraph" id="class_data_handler_a95cd188763d53b65e7e2e658bb76760b_cgraph">
<area shape="rect" id="node2" href="class_environment.html#ab5cccffb4eaae7d47d8dc9078d7dc601" title="Environment::getNumbersOf\lUniqueMHCs" alt="" coords="263,5,451,47"/>
</map>
</div>

</div>
</div>
<a id="a791581a6ac572d0daabd2c5bf8a53033"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a791581a6ac572d0daabd2c5bf8a53033">&#9670;&nbsp;</a></span>saveMhcNumbersBeforeMating()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void DataHandler::saveMhcNumbersBeforeMating </td>
          <td>(</td>
          <td class="paramtype"><a class="el" href="class_environment.html">Environment</a> &amp;&#160;</td>
          <td class="paramname"><em>EnvObj</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">int&#160;</td>
          <td class="paramname"><em>tayme</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Saves the number of unique MHC types of each individual in each time step. Should be used BEFORE mating occurs. First column is time stamp (hosts generation number). </p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">EnvObj</td><td>- the <a class="el" href="class_environment.html" title="Core class. Stores and handles the environment object that is the place where all the things are happ...">Environment</a> class object </td></tr>
    <tr><td class="paramname">tayme</td><td>- time stamp (hosts generation number) </td></tr>
  </table>
  </dd>
</dl>

<p class="definition">Definition at line <a class="el" href="_data_handler_8cpp_source.html#l00556">556</a> of file <a class="el" href="_data_handler_8cpp_source.html">DataHandler.cpp</a>.</p>
<div class="dynheader">
Here is the call graph for this function:</div>
<div class="dyncontent">
<div class="center"><img src="class_data_handler_a791581a6ac572d0daabd2c5bf8a53033_cgraph.png" border="0" usemap="#class_data_handler_a791581a6ac572d0daabd2c5bf8a53033_cgraph" alt=""/></div>
<map name="class_data_handler_a791581a6ac572d0daabd2c5bf8a53033_cgraph" id="class_data_handler_a791581a6ac572d0daabd2c5bf8a53033_cgraph">
<area shape="rect" id="node2" href="class_environment.html#ab5cccffb4eaae7d47d8dc9078d7dc601" title="Environment::getNumbersOf\lUniqueMHCs" alt="" coords="263,5,451,47"/>
</map>
</div>

</div>
</div>
<a id="ab7b38c92d0d564999a1159ceb56494e4"></a>
<h2 class="memtitle"><span class="permalink"><a href="#ab7b38c92d0d564999a1159ceb56494e4">&#9670;&nbsp;</a></span>saveMhcNumbersWhenMating()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void DataHandler::saveMhcNumbersWhenMating </td>
          <td>(</td>
          <td class="paramtype"><a class="el" href="class_environment.html">Environment</a> &amp;&#160;</td>
          <td class="paramname"><em>EnvObj</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">int&#160;</td>
          <td class="paramname"><em>tayme</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Data harvesting method. Records the number of unique MHCs a "mother" and a "farther" had. First column is time stamp (hosts generation number). </p>
<p>Writes to two separate files in each time step the number of unique MHCs a seceting individual had (the "mather") in file NumberOfMhcInMother.csv and how many unique MHCs the selected individual had (the "father") in file NumberOfMhcInFather.csv. Both files are in sync - one line is same time stamp, one enty in both lines is the same individual.</p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">EnvObj</td><td>- the <a class="el" href="class_environment.html" title="Core class. Stores and handles the environment object that is the place where all the things are happ...">Environment</a> class object </td></tr>
    <tr><td class="paramname">tayme</td><td>- time stamp (hosts generation number) </td></tr>
  </table>
  </dd>
</dl>

<p class="definition">Definition at line <a class="el" href="_data_handler_8cpp_source.html#l00521">521</a> of file <a class="el" href="_data_handler_8cpp_source.html">DataHandler.cpp</a>.</p>
<div class="dynheader">
Here is the call graph for this function:</div>
<div class="dyncontent">
<div class="center"><img src="class_data_handler_ab7b38c92d0d564999a1159ceb56494e4_cgraph.png" border="0" usemap="#class_data_handler_ab7b38c92d0d564999a1159ceb56494e4_cgraph" alt=""/></div>
<map name="class_data_handler_ab7b38c92d0d564999a1159ceb56494e4_cgraph" id="class_data_handler_ab7b38c92d0d564999a1159ceb56494e4_cgraph">
<area shape="rect" id="node2" href="class_environment.html#aaaccca98aea69e0eede4010fec439e56" title="Data harvesting method. Prepares a string with the list of how many unique MHC a &quot;mother&quot; had (a sele..." alt="" coords="263,5,451,47"/>
<area shape="rect" id="node3" href="class_environment.html#abc923f56f8f796f81f05a41bea0b4c82" title="Data harvesting method. Prepares a string with the list of how many unique MHC a &quot;father&quot; had (the pa..." alt="" coords="263,71,451,112"/>
</map>
</div>

</div>
</div>
<a id="a89f86a67a11f719ec282f37740663f95"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a89f86a67a11f719ec282f37740663f95">&#9670;&nbsp;</a></span>saveNumOfPathoSpeciesToFile()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void DataHandler::saveNumOfPathoSpeciesToFile </td>
          <td>(</td>
          <td class="paramtype"><a class="el" href="class_environment.html">Environment</a> &amp;&#160;</td>
          <td class="paramname"><em>EnvObj</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">int&#160;</td>
          <td class="paramname"><em>tayme</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Data harvesting method. Writes to a file population sizes of all pathogen species in a given time. </p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">EnvObj</td><td>- the <a class="el" href="class_environment.html" title="Core class. Stores and handles the environment object that is the place where all the things are happ...">Environment</a> object </td></tr>
    <tr><td class="paramname">tayme</td><td>- time stamp (hosts generation number) </td></tr>
  </table>
  </dd>
</dl>

<p class="definition">Definition at line <a class="el" href="_data_handler_8cpp_source.html#l00266">266</a> of file <a class="el" href="_data_handler_8cpp_source.html">DataHandler.cpp</a>.</p>
<div class="dynheader">
Here is the call graph for this function:</div>
<div class="dyncontent">
<div class="center"><img src="class_data_handler_a89f86a67a11f719ec282f37740663f95_cgraph.png" border="0" usemap="#class_data_handler_a89f86a67a11f719ec282f37740663f95_cgraph" alt=""/></div>
<map name="class_data_handler_a89f86a67a11f719ec282f37740663f95_cgraph" id="class_data_handler_a89f86a67a11f719ec282f37740663f95_cgraph">
<area shape="rect" id="node2" href="class_environment.html#a9bbe18137bc9b931826ec18a036f5697" title="Core method. Gets the number of species of pathogens. " alt="" coords="270,5,453,47"/>
<area shape="rect" id="node3" href="class_environment.html#af8726165c8bdeb45dcc0e0f91e894c38" title="Core method. Gets a number of individuals in a selected species of pathogen. " alt="" coords="260,71,463,112"/>
</map>
</div>

</div>
</div>
<a id="a6378c36aa8e4daf8f525c0fee446897e"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a6378c36aa8e4daf8f525c0fee446897e">&#9670;&nbsp;</a></span>savePathoNoMuttList()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void DataHandler::savePathoNoMuttList </td>
          <td>(</td>
          <td class="paramtype"><a class="el" href="class_environment.html">Environment</a> &amp;&#160;</td>
          <td class="paramname"><em>EnvObj</em></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Data harvesting method. Record the indices of fixed bits that are not allow to mutate in all pathogen species. Run this just ones. </p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">EnvObj</td><td>- the <a class="el" href="class_environment.html" title="Core class. Stores and handles the environment object that is the place where all the things are happ...">Environment</a> class object </td></tr>
  </table>
  </dd>
</dl>

<p class="definition">Definition at line <a class="el" href="_data_handler_8cpp_source.html#l00467">467</a> of file <a class="el" href="_data_handler_8cpp_source.html">DataHandler.cpp</a>.</p>
<div class="dynheader">
Here is the call graph for this function:</div>
<div class="dyncontent">
<div class="center"><img src="class_data_handler_a6378c36aa8e4daf8f525c0fee446897e_cgraph.png" border="0" usemap="#class_data_handler_a6378c36aa8e4daf8f525c0fee446897e_cgraph" alt=""/></div>
<map name="class_data_handler_a6378c36aa8e4daf8f525c0fee446897e_cgraph" id="class_data_handler_a6378c36aa8e4daf8f525c0fee446897e_cgraph">
<area shape="rect" id="node2" href="class_environment.html#a7ca87db3a1ceb9e27670538fce828675" title="Data harvesting method. Prepares a string with the list of fixed &quot;no mutation sites&quot; in antigens used..." alt="" coords="236,5,412,47"/>
</map>
</div>

</div>
</div>
<a id="a877ed5bfd00f0cbeef0413bf1d47f865"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a877ed5bfd00f0cbeef0413bf1d47f865">&#9670;&nbsp;</a></span>savePathoPopulToFile()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void DataHandler::savePathoPopulToFile </td>
          <td>(</td>
          <td class="paramtype"><a class="el" href="class_environment.html">Environment</a> &amp;&#160;</td>
          <td class="paramname"><em>EnvObj</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">int&#160;</td>
          <td class="paramname"><em>tayme</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Data harvesting method. Writes to a file all pathogens with their genomes in a human-readable format. </p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">EnvObj</td><td>- the <a class="el" href="class_environment.html" title="Core class. Stores and handles the environment object that is the place where all the things are happ...">Environment</a> object </td></tr>
    <tr><td class="paramname">tayme</td><td>- time stamp (hosts generation number) </td></tr>
  </table>
  </dd>
</dl>

<p class="definition">Definition at line <a class="el" href="_data_handler_8cpp_source.html#l00292">292</a> of file <a class="el" href="_data_handler_8cpp_source.html">DataHandler.cpp</a>.</p>
<div class="dynheader">
Here is the call graph for this function:</div>
<div class="dyncontent">
<div class="center"><img src="class_data_handler_a877ed5bfd00f0cbeef0413bf1d47f865_cgraph.png" border="0" usemap="#class_data_handler_a877ed5bfd00f0cbeef0413bf1d47f865_cgraph" alt=""/></div>
<map name="class_data_handler_a877ed5bfd00f0cbeef0413bf1d47f865_cgraph" id="class_data_handler_a877ed5bfd00f0cbeef0413bf1d47f865_cgraph">
<area shape="rect" id="node2" href="class_environment.html#a9bbe18137bc9b931826ec18a036f5697" title="Core method. Gets the number of species of pathogens. " alt="" coords="262,5,445,47"/>
<area shape="rect" id="node3" href="class_environment.html#af8726165c8bdeb45dcc0e0f91e894c38" title="Core method. Gets a number of individuals in a selected species of pathogen. " alt="" coords="252,71,455,112"/>
<area shape="rect" id="node4" href="class_environment.html#aa0b7fddb5c105c65ef4817531593e64f" title="Data harvesting method. Gets the genome of selected pathogen in human&#45;readable format. " alt="" coords="257,136,450,177"/>
</map>
</div>

</div>
</div>
<a id="ac0bd3c1ba0fa18e132ea9095db9747cd"></a>
<h2 class="memtitle"><span class="permalink"><a href="#ac0bd3c1ba0fa18e132ea9095db9747cd">&#9670;&nbsp;</a></span>savePresentedPathos()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void DataHandler::savePresentedPathos </td>
          <td>(</td>
          <td class="paramtype"><a class="el" href="class_environment.html">Environment</a> &amp;&#160;</td>
          <td class="paramname"><em>EnvObj</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">int&#160;</td>
          <td class="paramname"><em>tayme</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Data harvesting method. Saves the number of presented pathogens by each indrividual host in each time step. First column is time stamp (hosts generation number). </p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">EnvObj</td><td>- the <a class="el" href="class_environment.html" title="Core class. Stores and handles the environment object that is the place where all the things are happ...">Environment</a> class object </td></tr>
    <tr><td class="paramname">tayme</td><td>- time stamp (hosts generation number) </td></tr>
  </table>
  </dd>
</dl>

<p class="definition">Definition at line <a class="el" href="_data_handler_8cpp_source.html#l00490">490</a> of file <a class="el" href="_data_handler_8cpp_source.html">DataHandler.cpp</a>.</p>
<div class="dynheader">
Here is the call graph for this function:</div>
<div class="dyncontent">
<div class="center"><img src="class_data_handler_ac0bd3c1ba0fa18e132ea9095db9747cd_cgraph.png" border="0" usemap="#class_data_handler_ac0bd3c1ba0fa18e132ea9095db9747cd_cgraph" alt=""/></div>
<map name="class_data_handler_ac0bd3c1ba0fa18e132ea9095db9747cd_cgraph" id="class_data_handler_ac0bd3c1ba0fa18e132ea9095db9747cd_cgraph">
<area shape="rect" id="node2" href="class_environment.html#a12173f6c8c70376c82bace9cc578a738" title="Data harvesting method. Prepares a string with the list of how many pathogen species each host had pr..." alt="" coords="244,5,432,47"/>
</map>
</div>

</div>
</div>
<a id="a049645bbd6069ad5d6a90c8469390811"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a049645bbd6069ad5d6a90c8469390811">&#9670;&nbsp;</a></span>setAllFilesAsFirtsTimers()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void DataHandler::setAllFilesAsFirtsTimers </td>
          <td>(</td>
          <td class="paramname"></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Data harvesting method. Sets status of all data files as "brand new". </p>

<p class="definition">Definition at line <a class="el" href="_data_handler_8cpp_source.html#l00087">87</a> of file <a class="el" href="_data_handler_8cpp_source.html">DataHandler.cpp</a>.</p>

</div>
</div>
<hr/>The documentation for this class was generated from the following files:<ul>
<li>src/<a class="el" href="_data_handler_8h_source.html">DataHandler.h</a></li>
<li>src/<a class="el" href="_data_handler_8cpp_source.html">DataHandler.cpp</a></li>
</ul>
</div><!-- contents -->
</div><!-- doc-content -->
<!-- start footer part -->
<div id="nav-path" class="navpath"><!-- id is needed for treeview function! -->
  <ul>
    <li class="navelem"><a class="el" href="class_data_handler.html">DataHandler</a></li>
    <li class="footer">Generated on Sun Oct 6 2019 18:26:47 for Modelling MHC Evolutionary Dynamics by
    <a href="http://www.doxygen.org/index.html">
    <img class="footer" src="doxygen.png" alt="doxygen"/></a> 1.8.13 </li>
  </ul>
</div>
</body>
</html>
