<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml">
<head>
<meta http-equiv="Content-Type" content="text/xhtml;charset=UTF-8"/>
<meta http-equiv="X-UA-Compatible" content="IE=9"/>
<meta name="generator" content="Doxygen 1.8.11"/>
<title>Modelling MHC Evolutionary Dynamics: Host Class Reference</title>
<link href="tabs.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="jquery.js"></script>
<script type="text/javascript" src="dynsections.js"></script>
<link href="navtree.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="resize.js"></script>
<script type="text/javascript" src="navtreedata.js"></script>
<script type="text/javascript" src="navtree.js"></script>
<script type="text/javascript">
  $(document).ready(initResizable);
  $(window).load(resizeHeight);
</script>
<link href="search/search.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="search/searchdata.js"></script>
<script type="text/javascript" src="search/search.js"></script>
<script type="text/javascript">
  $(document).ready(function() { init_search(); });
</script>
<link href="doxygen.css" rel="stylesheet" type="text/css" />
</head>
<body>
<div id="top"><!-- do not remove this div, it is closed by doxygen! -->
<div id="titlearea">
<table cellspacing="0" cellpadding="0">
 <tbody>
 <tr style="height: 56px;">
  <td id="projectalign" style="padding-left: 0.5em;">
   <div id="projectname">Modelling MHC Evolutionary Dynamics
   &#160;<span id="projectnumber">0.9</span>
   </div>
  </td>
 </tr>
 </tbody>
</table>
</div>
<!-- end header part -->
<!-- Generated by Doxygen 1.8.11 -->
<script type="text/javascript">
var searchBox = new SearchBox("searchBox", "search",false,'Search');
</script>
  <div id="navrow1" class="tabs">
    <ul class="tablist">
      <li><a href="index.html"><span>Main&#160;Page</span></a></li>
      <li><a href="pages.html"><span>Related&#160;Pages</span></a></li>
      <li class="current"><a href="annotated.html"><span>Classes</span></a></li>
      <li><a href="files.html"><span>Files</span></a></li>
      <li>
        <div id="MSearchBox" class="MSearchBoxInactive">
        <span class="left">
          <img id="MSearchSelect" src="search/mag_sel.png"
               onmouseover="return searchBox.OnSearchSelectShow()"
               onmouseout="return searchBox.OnSearchSelectHide()"
               alt=""/>
          <input type="text" id="MSearchField" value="Search" accesskey="S"
               onfocus="searchBox.OnSearchFieldFocus(true)" 
               onblur="searchBox.OnSearchFieldFocus(false)" 
               onkeyup="searchBox.OnSearchFieldChange(event)"/>
          </span><span class="right">
            <a id="MSearchClose" href="javascript:searchBox.CloseResultsWindow()"><img id="MSearchCloseImg" border="0" src="search/close.png" alt=""/></a>
          </span>
        </div>
      </li>
    </ul>
  </div>
  <div id="navrow2" class="tabs2">
    <ul class="tablist">
      <li><a href="annotated.html"><span>Class&#160;List</span></a></li>
      <li><a href="classes.html"><span>Class&#160;Index</span></a></li>
      <li><a href="functions.html"><span>Class&#160;Members</span></a></li>
    </ul>
  </div>
</div><!-- top -->
<div id="side-nav" class="ui-resizable side-nav-resizable">
  <div id="nav-tree">
    <div id="nav-tree-contents">
      <div id="nav-sync" class="sync"></div>
    </div>
  </div>
  <div id="splitbar" style="-moz-user-select:none;" 
       class="ui-resizable-handle">
  </div>
</div>
<script type="text/javascript">
$(document).ready(function(){initNavTree('class_host.html','');});
</script>
<div id="doc-content">
<!-- window showing the filter options -->
<div id="MSearchSelectWindow"
     onmouseover="return searchBox.OnSearchSelectShow()"
     onmouseout="return searchBox.OnSearchSelectHide()"
     onkeydown="return searchBox.OnSearchSelectKey(event)">
</div>

<!-- iframe showing the search results (closed by default) -->
<div id="MSearchResultsWindow">
<iframe src="javascript:void(0)" frameborder="0" 
        name="MSearchResults" id="MSearchResults">
</iframe>
</div>

<div class="header">
  <div class="summary">
<a href="#pub-methods">Public Member Functions</a> &#124;
<a href="#pub-attribs">Public Attributes</a> &#124;
<a href="class_host-members.html">List of all members</a>  </div>
  <div class="headertitle">
<div class="title">Host Class Reference</div>  </div>
</div><!--header-->
<div class="contents">

<p>Core class. Stores and handles a single host object. Each host has multiple instances of <a class="el" href="class_gene.html" title="Core class. Stores and handles a single gene object. Has methods to access and to mutate a gene...">Gene</a> class objects and stores them in 2 vectors called ChromosomeOne and ChromosomeTwo which simulate a diploid genome.  
 <a href="class_host.html#details">More...</a></p>

<p><code>#include &lt;<a class="el" href="_host_8h_source.html">Host.h</a>&gt;</code></p>
<table class="memberdecls">
<tr class="heading"><td colspan="2"><h2 class="groupheader"><a name="pub-methods"></a>
Public Member Functions</h2></td></tr>
<tr class="memitem:a369c12e7425c030528bab0784da40e4b"><td class="memItemLeft" align="right" valign="top">&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_host.html#a369c12e7425c030528bab0784da40e4b">Host</a> ()</td></tr>
<tr class="separator:a369c12e7425c030528bab0784da40e4b"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a6c14f16c39ae6bc5db2672e0141adccd"><td class="memItemLeft" align="right" valign="top">virtual&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_host.html#a6c14f16c39ae6bc5db2672e0141adccd">~Host</a> ()</td></tr>
<tr class="separator:a6c14f16c39ae6bc5db2672e0141adccd"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a4c017e042f00858c8ae173786e077bb4"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_host.html#a4c017e042f00858c8ae173786e077bb4">setNewHost</a> (unsigned long num_of_loci, unsigned long gene_size, int timeStamp)</td></tr>
<tr class="memdesc:a4c017e042f00858c8ae173786e077bb4"><td class="mdescLeft">&#160;</td><td class="mdescRight">Core method. Sets a new <a class="el" href="class_host.html" title="Core class. Stores and handles a single host object. Each host has multiple instances of Gene class o...">Host</a> object and assigns a user-defined number of genes of user-defined number of bits (as genes are bit-strings).  <a href="#a4c017e042f00858c8ae173786e077bb4">More...</a><br /></td></tr>
<tr class="separator:a4c017e042f00858c8ae173786e077bb4"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:afdfa6fd4262695491d83d3bea9cd29e7"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_host.html#afdfa6fd4262695491d83d3bea9cd29e7">setNewHomozygHost</a> (unsigned long num_of_loci, unsigned long gene_size, int timeStamp)</td></tr>
<tr class="memdesc:afdfa6fd4262695491d83d3bea9cd29e7"><td class="mdescLeft">&#160;</td><td class="mdescRight">Core method. Sets a new <a class="el" href="class_host.html" title="Core class. Stores and handles a single host object. Each host has multiple instances of Gene class o...">Host</a> object and assigns a user-defined number of genes of user-defined number of bits (as genes are bit-strings).  <a href="#afdfa6fd4262695491d83d3bea9cd29e7">More...</a><br /></td></tr>
<tr class="separator:afdfa6fd4262695491d83d3bea9cd29e7"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a5911792258205cae25a2a4796061fe49"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_host.html#a5911792258205cae25a2a4796061fe49">chromoMutProcess</a> (double mut_probabl, int timeStamp)</td></tr>
<tr class="memdesc:a5911792258205cae25a2a4796061fe49"><td class="mdescLeft">&#160;</td><td class="mdescRight">Core method. Decides (on a random basis) if there will be any mutations in the genome.  <a href="#a5911792258205cae25a2a4796061fe49">More...</a><br /></td></tr>
<tr class="separator:a5911792258205cae25a2a4796061fe49"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a03a4d86562ede11801244ee1242105e2"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_host.html#a03a4d86562ede11801244ee1242105e2">chromoMutProcessWithDelDupl</a> (double mut_probabl, double del, double dupli, unsigned long maxGene, int timeStamp)</td></tr>
<tr class="memdesc:a03a4d86562ede11801244ee1242105e2"><td class="mdescLeft">&#160;</td><td class="mdescRight">Core method. Decides (on a random basis) if there will be any mutations in the genome: deletion, duplication or change of a gene.  <a href="#a03a4d86562ede11801244ee1242105e2">More...</a><br /></td></tr>
<tr class="separator:a03a4d86562ede11801244ee1242105e2"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ad39c5fbaae3af86c44a312ebe9fb0baa"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_host.html#ad39c5fbaae3af86c44a312ebe9fb0baa">chromoMutProcessWithDelDuplPointMuts</a> (double mut_probabl, double del, double dupli, unsigned long maxGene, int timeStamp)</td></tr>
<tr class="memdesc:ad39c5fbaae3af86c44a312ebe9fb0baa"><td class="mdescLeft">&#160;</td><td class="mdescRight">Core method. Decides (on a random basis) if there will be any mutations in the genome.  <a href="#ad39c5fbaae3af86c44a312ebe9fb0baa">More...</a><br /></td></tr>
<tr class="separator:ad39c5fbaae3af86c44a312ebe9fb0baa"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a5da864a08370512a065cf5779795be5a"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_host.html#a5da864a08370512a065cf5779795be5a">chromoRecombination</a> (double recomb_prob)</td></tr>
<tr class="memdesc:a5da864a08370512a065cf5779795be5a"><td class="mdescLeft">&#160;</td><td class="mdescRight">Core method. Performs recombination between host's chromosomes by inserting a brake in randomly selected genes on Chromosome One and Two and swapping the resulting parts of the chromosomes.  <a href="#a5da864a08370512a065cf5779795be5a">More...</a><br /></td></tr>
<tr class="separator:a5da864a08370512a065cf5779795be5a"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a7c21e1938c61966dc2657a437aeec9b9"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_host.html#a7c21e1938c61966dc2657a437aeec9b9">clearInfections</a> ()</td></tr>
<tr class="memdesc:a7c21e1938c61966dc2657a437aeec9b9"><td class="mdescLeft">&#160;</td><td class="mdescRight">Core method. Zeroes all data regarding infections and fitness.  <a href="#a7c21e1938c61966dc2657a437aeec9b9">More...</a><br /></td></tr>
<tr class="separator:a7c21e1938c61966dc2657a437aeec9b9"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a505527db337d87670e5ab37eb9818fa3"><td class="memItemLeft" align="right" valign="top"><a class="el" href="_h2_pinteraction_8cpp.html#adcfab8f2ed9ed0c694cb95d35c91aa78">chromovector</a>&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_host.html#a505527db337d87670e5ab37eb9818fa3">doCrossAndMeiosis</a> (double corssing_prob)</td></tr>
<tr class="memdesc:a505527db337d87670e5ab37eb9818fa3"><td class="mdescLeft">&#160;</td><td class="mdescRight">Core method. Performing crossing over and chromosome selection before mating.  <a href="#a505527db337d87670e5ab37eb9818fa3">More...</a><br /></td></tr>
<tr class="separator:a505527db337d87670e5ab37eb9818fa3"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ac387176ba203f6aa1f20610f5679385a"><td class="memItemLeft" align="right" valign="top"><a class="el" href="_h2_pinteraction_8cpp.html#adcfab8f2ed9ed0c694cb95d35c91aa78">chromovector</a>&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_host.html#ac387176ba203f6aa1f20610f5679385a">getChromosomeOne</a> ()</td></tr>
<tr class="memdesc:ac387176ba203f6aa1f20610f5679385a"><td class="mdescLeft">&#160;</td><td class="mdescRight">Core method. Returns the first chromosome.  <a href="#ac387176ba203f6aa1f20610f5679385a">More...</a><br /></td></tr>
<tr class="separator:ac387176ba203f6aa1f20610f5679385a"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:aa5afa38408ebba644f56667364213f0d"><td class="memItemLeft" align="right" valign="top"><a class="el" href="_h2_pinteraction_8cpp.html#adcfab8f2ed9ed0c694cb95d35c91aa78">chromovector</a>&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_host.html#aa5afa38408ebba644f56667364213f0d">getChromosomeTwo</a> ()</td></tr>
<tr class="memdesc:aa5afa38408ebba644f56667364213f0d"><td class="mdescLeft">&#160;</td><td class="mdescRight">Core method. Returns the first chromosome.  <a href="#aa5afa38408ebba644f56667364213f0d">More...</a><br /></td></tr>
<tr class="separator:aa5afa38408ebba644f56667364213f0d"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a55edfefc42f7306c85983474500ef10b"><td class="memItemLeft" align="right" valign="top"><a class="el" href="_h2_pinteraction_8cpp.html#adcfab8f2ed9ed0c694cb95d35c91aa78">chromovector</a>&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_host.html#a55edfefc42f7306c85983474500ef10b">mergeChromosomes</a> ()</td></tr>
<tr class="memdesc:a55edfefc42f7306c85983474500ef10b"><td class="mdescLeft">&#160;</td><td class="mdescRight">Data harvesting method. Merges both chromosomes to create an object easier to handle is some situations.  <a href="#a55edfefc42f7306c85983474500ef10b">More...</a><br /></td></tr>
<tr class="separator:a55edfefc42f7306c85983474500ef10b"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a0d36e2b60e48e985e6e561fc46213564"><td class="memItemLeft" align="right" valign="top"><a class="el" href="_h2_pinteraction_8cpp.html#adcfab8f2ed9ed0c694cb95d35c91aa78">chromovector</a>&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_host.html#a0d36e2b60e48e985e6e561fc46213564">getUniqueMHCs</a> ()</td></tr>
<tr class="memdesc:a0d36e2b60e48e985e6e561fc46213564"><td class="mdescLeft">&#160;</td><td class="mdescRight">Core method. Returns unique MHC alleles from the host.  <a href="#a0d36e2b60e48e985e6e561fc46213564">More...</a><br /></td></tr>
<tr class="separator:a0d36e2b60e48e985e6e561fc46213564"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a8fd67f5e888387ddd188e1ba66d52727"><td class="memItemLeft" align="right" valign="top">unsigned long&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_host.html#a8fd67f5e888387ddd188e1ba66d52727">getGenomeSize</a> ()</td></tr>
<tr class="memdesc:a8fd67f5e888387ddd188e1ba66d52727"><td class="mdescLeft">&#160;</td><td class="mdescRight">Returns the total number of genes in the genome.  <a href="#a8fd67f5e888387ddd188e1ba66d52727">More...</a><br /></td></tr>
<tr class="separator:a8fd67f5e888387ddd188e1ba66d52727"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a51d5609864393846c577672908f824b0"><td class="memItemLeft" align="right" valign="top">unsigned long&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_host.html#a51d5609864393846c577672908f824b0">getChromoOneSize</a> ()</td></tr>
<tr class="memdesc:a51d5609864393846c577672908f824b0"><td class="mdescLeft">&#160;</td><td class="mdescRight">Returns the number of genes in the Chromosome One.  <a href="#a51d5609864393846c577672908f824b0">More...</a><br /></td></tr>
<tr class="separator:a51d5609864393846c577672908f824b0"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a5158ada636bdc6c9985abfeb51cc5642"><td class="memItemLeft" align="right" valign="top">unsigned long&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_host.html#a5158ada636bdc6c9985abfeb51cc5642">getChromoTwoSize</a> ()</td></tr>
<tr class="memdesc:a5158ada636bdc6c9985abfeb51cc5642"><td class="mdescLeft">&#160;</td><td class="mdescRight">Returns the number of genes in the Chromosome Two.  <a href="#a5158ada636bdc6c9985abfeb51cc5642">More...</a><br /></td></tr>
<tr class="separator:a5158ada636bdc6c9985abfeb51cc5642"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a6c8640b0cd873b9669983330b2c4c59e"><td class="memItemLeft" align="right" valign="top">unsigned long&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_host.html#a6c8640b0cd873b9669983330b2c4c59e">getUniqueMhcSize</a> ()</td></tr>
<tr class="memdesc:a6c8640b0cd873b9669983330b2c4c59e"><td class="mdescLeft">&#160;</td><td class="mdescRight">Core method. Returns the size of the vector containing unique MHCs.  <a href="#a6c8640b0cd873b9669983330b2c4c59e">More...</a><br /></td></tr>
<tr class="separator:a6c8640b0cd873b9669983330b2c4c59e"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a87feca6e68d17438266fe8e92468382f"><td class="memItemLeft" align="right" valign="top">double&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_host.html#a87feca6e68d17438266fe8e92468382f">getNumbOfUniqMHCgenes</a> ()</td></tr>
<tr class="memdesc:a87feca6e68d17438266fe8e92468382f"><td class="mdescLeft">&#160;</td><td class="mdescRight">Data harvesting method. Returns the number (double) of unique MHC in genome.  <a href="#a87feca6e68d17438266fe8e92468382f">More...</a><br /></td></tr>
<tr class="separator:a87feca6e68d17438266fe8e92468382f"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a9757c15b19a45468e3ce5921b5d2deab"><td class="memItemLeft" align="right" valign="top">double&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_host.html#a9757c15b19a45468e3ce5921b5d2deab">getNumbOfChromoOneUniqAlleles</a> ()</td></tr>
<tr class="memdesc:a9757c15b19a45468e3ce5921b5d2deab"><td class="mdescLeft">&#160;</td><td class="mdescRight">Returns the number of unique MHC alleles in the Chromosome One.  <a href="#a9757c15b19a45468e3ce5921b5d2deab">More...</a><br /></td></tr>
<tr class="separator:a9757c15b19a45468e3ce5921b5d2deab"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a533fcc0f3f715fe36abdb878b0491aaf"><td class="memItemLeft" align="right" valign="top">double&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_host.html#a533fcc0f3f715fe36abdb878b0491aaf">getNumbOfChromoTwoUniqAlleles</a> ()</td></tr>
<tr class="memdesc:a533fcc0f3f715fe36abdb878b0491aaf"><td class="mdescLeft">&#160;</td><td class="mdescRight">Returns the number of unique MHC alleles in the Chromosome Two.  <a href="#a533fcc0f3f715fe36abdb878b0491aaf">More...</a><br /></td></tr>
<tr class="separator:a533fcc0f3f715fe36abdb878b0491aaf"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ac6ecd452f531bbcaf469d5d9d42b8028"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_host.html#ac6ecd452f531bbcaf469d5d9d42b8028">assignChromOne</a> (<a class="el" href="_h2_pinteraction_8cpp.html#adcfab8f2ed9ed0c694cb95d35c91aa78">chromovector</a> One)</td></tr>
<tr class="memdesc:ac6ecd452f531bbcaf469d5d9d42b8028"><td class="mdescLeft">&#160;</td><td class="mdescRight">Core method. Assigns a new chromosome to host's Chromosome ONE.  <a href="#ac6ecd452f531bbcaf469d5d9d42b8028">More...</a><br /></td></tr>
<tr class="separator:ac6ecd452f531bbcaf469d5d9d42b8028"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ab3e04fdf569863ddeee00d6d2f5c990d"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_host.html#ab3e04fdf569863ddeee00d6d2f5c990d">assignChromTwo</a> (<a class="el" href="_h2_pinteraction_8cpp.html#adcfab8f2ed9ed0c694cb95d35c91aa78">chromovector</a> Two)</td></tr>
<tr class="memdesc:ab3e04fdf569863ddeee00d6d2f5c990d"><td class="mdescLeft">&#160;</td><td class="mdescRight">Core method. Assigns a new chromosome to host's Chromosome TWO.  <a href="#ab3e04fdf569863ddeee00d6d2f5c990d">More...</a><br /></td></tr>
<tr class="separator:ab3e04fdf569863ddeee00d6d2f5c990d"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a093534dcf4dc55de1bd817115676adb0"><td class="memItemLeft" align="right" valign="top"><a class="el" href="_gene_8cpp.html#a108efa7670466d15a5ecafb60bbc9c37">genestring</a>&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_host.html#a093534dcf4dc55de1bd817115676adb0">getSingleGeneFromOne</a> (unsigned long indx)</td></tr>
<tr class="memdesc:a093534dcf4dc55de1bd817115676adb0"><td class="mdescLeft">&#160;</td><td class="mdescRight">Core method. Fetches a gene (bit string) with a given position on the chromosome One.  <a href="#a093534dcf4dc55de1bd817115676adb0">More...</a><br /></td></tr>
<tr class="separator:a093534dcf4dc55de1bd817115676adb0"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a25421bd536db26a7e83cf3a67e7b4cbe"><td class="memItemLeft" align="right" valign="top"><a class="el" href="_gene_8cpp.html#a108efa7670466d15a5ecafb60bbc9c37">genestring</a>&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_host.html#a25421bd536db26a7e83cf3a67e7b4cbe">getSingleGeneFromTwo</a> (unsigned long indx)</td></tr>
<tr class="memdesc:a25421bd536db26a7e83cf3a67e7b4cbe"><td class="mdescLeft">&#160;</td><td class="mdescRight">Core method. Fetches a gene (bit string) with a given position on the chromosome Two.  <a href="#a25421bd536db26a7e83cf3a67e7b4cbe">More...</a><br /></td></tr>
<tr class="separator:a25421bd536db26a7e83cf3a67e7b4cbe"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ab30250ea6fdd47221d65914b3e6de4e0"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_host.html#ab30250ea6fdd47221d65914b3e6de4e0">swapChromosomes</a> ()</td></tr>
<tr class="memdesc:ab30250ea6fdd47221d65914b3e6de4e0"><td class="mdescLeft">&#160;</td><td class="mdescRight">Core method. Randomly swaps places of Chromosome One and Chromosome Two to avoid situation when they effectively become two separate populations.  <a href="#ab30250ea6fdd47221d65914b3e6de4e0">More...</a><br /></td></tr>
<tr class="separator:ab30250ea6fdd47221d65914b3e6de4e0"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a3ae32e9601a5c79da23b0ba4970ff0a6"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_host.html#a3ae32e9601a5c79da23b0ba4970ff0a6">evalUniqueMHCs</a> ()</td></tr>
<tr class="memdesc:a3ae32e9601a5c79da23b0ba4970ff0a6"><td class="mdescLeft">&#160;</td><td class="mdescRight">Core method. Creates a chromosome-like vector containing only the unique MHC genes from two chromosomes. This vector is useful when interacting with pathogens or evaluating fitness.  <a href="#a3ae32e9601a5c79da23b0ba4970ff0a6">More...</a><br /></td></tr>
<tr class="separator:a3ae32e9601a5c79da23b0ba4970ff0a6"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:aeb8564ae54f8bbf968a62f13c61083df"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_host.html#aeb8564ae54f8bbf968a62f13c61083df">calculateFitnessJustInfection</a> ()</td></tr>
<tr class="memdesc:aeb8564ae54f8bbf968a62f13c61083df"><td class="mdescLeft">&#160;</td><td class="mdescRight">Core method. Calculates host individual fitness simply as the number of exposed pathogens.  <a href="#aeb8564ae54f8bbf968a62f13c61083df">More...</a><br /></td></tr>
<tr class="separator:aeb8564ae54f8bbf968a62f13c61083df"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a6079285b3b8bb5280d52500e6b7fc7ef"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_host.html#a6079285b3b8bb5280d52500e6b7fc7ef">calculateFitnessAccChromSize</a> ()</td></tr>
<tr class="memdesc:a6079285b3b8bb5280d52500e6b7fc7ef"><td class="mdescLeft">&#160;</td><td class="mdescRight">Core method. Calculates host individual fitness as the number of exposed pathogens divided by the genome size.  <a href="#a6079285b3b8bb5280d52500e6b7fc7ef">More...</a><br /></td></tr>
<tr class="separator:a6079285b3b8bb5280d52500e6b7fc7ef"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a614430aea49534f29b63232414cf4fcd"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_host.html#a614430aea49534f29b63232414cf4fcd">calculateFitnessForDrift</a> ()</td></tr>
<tr class="memdesc:a614430aea49534f29b63232414cf4fcd"><td class="mdescLeft">&#160;</td><td class="mdescRight">Core method. Sets fitness as 1, a fixed value to simulate genetic drift.  <a href="#a614430aea49534f29b63232414cf4fcd">More...</a><br /></td></tr>
<tr class="separator:a614430aea49534f29b63232414cf4fcd"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:aca12c9829c01995669e842876478c7fb"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_host.html#aca12c9829c01995669e842876478c7fb">calculateFitnessAlphaXSqr</a> (double alpha)</td></tr>
<tr class="memdesc:aca12c9829c01995669e842876478c7fb"><td class="mdescLeft">&#160;</td><td class="mdescRight">Core method. Calculates host individual fitness in proportion to one over the square of the number of genes scaled to by factor <img class="formulaInl" alt="$ \alpha $" src="form_12.png"/>:  <a href="#aca12c9829c01995669e842876478c7fb">More...</a><br /></td></tr>
<tr class="separator:aca12c9829c01995669e842876478c7fb"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a9134661842de046453acc0342ef74eb6"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_host.html#a9134661842de046453acc0342ef74eb6">calculateFitnessExpFunc</a> (double alpha)</td></tr>
<tr class="memdesc:a9134661842de046453acc0342ef74eb6"><td class="mdescLeft">&#160;</td><td class="mdescRight">Core method. Calculates host individual fitness in proportion to the Gaussian function of the number of genes scaled to by factor <img class="formulaInl" alt="$ \alpha $" src="form_12.png"/>:  <a href="#a9134661842de046453acc0342ef74eb6">More...</a><br /></td></tr>
<tr class="separator:a9134661842de046453acc0342ef74eb6"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ad185a5e02c8ed0930e4d400e9c0e6a9a"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_host.html#ad185a5e02c8ed0930e4d400e9c0e6a9a">calculateFitnessExpFuncUniqAlleles</a> (double alpha)</td></tr>
<tr class="memdesc:ad185a5e02c8ed0930e4d400e9c0e6a9a"><td class="mdescLeft">&#160;</td><td class="mdescRight">Core method. Calculates host individual fitness in proportion to the Gaussian function of the number of unique MHC alleles in the chromosome scaled to by factor <img class="formulaInl" alt="$ \alpha $" src="form_12.png"/>:  <a href="#ad185a5e02c8ed0930e4d400e9c0e6a9a">More...</a><br /></td></tr>
<tr class="separator:ad185a5e02c8ed0930e4d400e9c0e6a9a"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a87d357c93783916037446bf8096ed45d"><td class="memItemLeft" align="right" valign="top">double&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_host.html#a87d357c93783916037446bf8096ed45d">getFitness</a> ()</td></tr>
<tr class="memdesc:a87d357c93783916037446bf8096ed45d"><td class="mdescLeft">&#160;</td><td class="mdescRight">Core method. Returns current value of host's fitness.  <a href="#a87d357c93783916037446bf8096ed45d">More...</a><br /></td></tr>
<tr class="separator:a87d357c93783916037446bf8096ed45d"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a481373a4981aad450aa666962f0c8252"><td class="memItemLeft" align="right" valign="top">std::string&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_host.html#a481373a4981aad450aa666962f0c8252">stringChromosomes</a> ()</td></tr>
<tr class="memdesc:a481373a4981aad450aa666962f0c8252"><td class="mdescLeft">&#160;</td><td class="mdescRight">Data harvesting method. Gives a host's genome in a human-readable format. With all the gene specs.  <a href="#a481373a4981aad450aa666962f0c8252">More...</a><br /></td></tr>
<tr class="separator:a481373a4981aad450aa666962f0c8252"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a118fc07706467cd5070dd7b8dd53a3a6"><td class="memItemLeft" align="right" valign="top">std::string&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_host.html#a118fc07706467cd5070dd7b8dd53a3a6">stringUniqMHCs</a> ()</td></tr>
<tr class="memdesc:a118fc07706467cd5070dd7b8dd53a3a6"><td class="mdescLeft">&#160;</td><td class="mdescRight">Data harvesting method. Gives a host's unique MHC genes in a human-readable format. With all the gene specs.  <a href="#a118fc07706467cd5070dd7b8dd53a3a6">More...</a><br /></td></tr>
<tr class="separator:a118fc07706467cd5070dd7b8dd53a3a6"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a48ffcaf86f00239f8dd8c56683478e2a"><td class="memItemLeft" align="right" valign="top">unsigned long int&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_host.html#a48ffcaf86f00239f8dd8c56683478e2a">getOneGeneFromOne</a> (unsigned long indx)</td></tr>
<tr class="memdesc:a48ffcaf86f00239f8dd8c56683478e2a"><td class="mdescLeft">&#160;</td><td class="mdescRight">Core method. Fetches a gene (as integer) with a given position on the Chromosome One.  <a href="#a48ffcaf86f00239f8dd8c56683478e2a">More...</a><br /></td></tr>
<tr class="separator:a48ffcaf86f00239f8dd8c56683478e2a"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a2540e6d2f6d139292a964b61f04449f0"><td class="memItemLeft" align="right" valign="top">unsigned long int&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_host.html#a2540e6d2f6d139292a964b61f04449f0">getOneGeneFromTwo</a> (unsigned long indx)</td></tr>
<tr class="memdesc:a2540e6d2f6d139292a964b61f04449f0"><td class="mdescLeft">&#160;</td><td class="mdescRight">Core method. Fetches a gene (as integer) with a given position on the Chromosome Two.  <a href="#a2540e6d2f6d139292a964b61f04449f0">More...</a><br /></td></tr>
<tr class="separator:a2540e6d2f6d139292a964b61f04449f0"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ab51c69cfe9d0f1b2bf4f3ca4225ae3d9"><td class="memItemLeft" align="right" valign="top">unsigned long int&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_host.html#ab51c69cfe9d0f1b2bf4f3ca4225ae3d9">getOneGeneFromUniqVect</a> (unsigned long indx)</td></tr>
<tr class="memdesc:ab51c69cfe9d0f1b2bf4f3ca4225ae3d9"><td class="mdescLeft">&#160;</td><td class="mdescRight">Core method. Fetches a gene (as integer) with a given position on the vector of unique MHCs.  <a href="#ab51c69cfe9d0f1b2bf4f3ca4225ae3d9">More...</a><br /></td></tr>
<tr class="separator:ab51c69cfe9d0f1b2bf4f3ca4225ae3d9"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a21a7fead03f3fb31005028f6575ea867"><td class="memItemLeft" align="right" valign="top">unsigned&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_host.html#a21a7fead03f3fb31005028f6575ea867">getNumberOfPresentedPatho</a> ()</td></tr>
<tr class="memdesc:a21a7fead03f3fb31005028f6575ea867"><td class="mdescLeft">&#160;</td><td class="mdescRight">Core method. Returns the number of pathogen species a host has presented.  <a href="#a21a7fead03f3fb31005028f6575ea867">More...</a><br /></td></tr>
<tr class="separator:a21a7fead03f3fb31005028f6575ea867"><td class="memSeparator" colspan="2">&#160;</td></tr>
</table><table class="memberdecls">
<tr class="heading"><td colspan="2"><h2 class="groupheader"><a name="pub-attribs"></a>
Public Attributes</h2></td></tr>
<tr class="memitem:a0d79bd02d6147f8e94ae975dc42ae026"><td class="memItemLeft" align="right" valign="top">std::vector&lt; int &gt;&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_host.html#a0d79bd02d6147f8e94ae975dc42ae026">PathoSpecInfecting</a></td></tr>
<tr class="separator:a0d79bd02d6147f8e94ae975dc42ae026"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ac8b084352039979629a7896e2aaefb45"><td class="memItemLeft" align="right" valign="top">std::vector&lt; int &gt;&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_host.html#ac8b084352039979629a7896e2aaefb45">PathogesPresented</a></td></tr>
<tr class="separator:ac8b084352039979629a7896e2aaefb45"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ac173100d13da6dfeb93387f0d689c349"><td class="memItemLeft" align="right" valign="top">unsigned&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_host.html#ac173100d13da6dfeb93387f0d689c349">NumOfPathogesInfecting</a></td></tr>
<tr class="separator:ac173100d13da6dfeb93387f0d689c349"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ad0411c059e63ed72507b0c370880b5ab"><td class="memItemLeft" align="right" valign="top">unsigned&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_host.html#ad0411c059e63ed72507b0c370880b5ab">NumOfPathogesPresented</a></td></tr>
<tr class="separator:ad0411c059e63ed72507b0c370880b5ab"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a50bc219031501dd2c19a69d4e3f19533"><td class="memItemLeft" align="right" valign="top">int&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_host.html#a50bc219031501dd2c19a69d4e3f19533">SelectedForReproduction</a></td></tr>
<tr class="separator:a50bc219031501dd2c19a69d4e3f19533"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:af4c460c65f085a6a7c9b73b46e83e7f6"><td class="memItemLeft" align="right" valign="top">std::vector&lt; <a class="el" href="class_gene.html">Gene</a> &gt;&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_host.html#af4c460c65f085a6a7c9b73b46e83e7f6">ChromosomeOne</a></td></tr>
<tr class="separator:af4c460c65f085a6a7c9b73b46e83e7f6"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:aca6f27f6124477919835a233e9760a6f"><td class="memItemLeft" align="right" valign="top">std::vector&lt; <a class="el" href="class_gene.html">Gene</a> &gt;&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_host.html#aca6f27f6124477919835a233e9760a6f">ChromosomeTwo</a></td></tr>
<tr class="separator:aca6f27f6124477919835a233e9760a6f"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a0a2a96cbc5292a5f5a7747e4125cd568"><td class="memItemLeft" align="right" valign="top">std::vector&lt; <a class="el" href="class_gene.html">Gene</a> &gt;&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_host.html#a0a2a96cbc5292a5f5a7747e4125cd568">UniqueAlleles</a></td></tr>
<tr class="separator:a0a2a96cbc5292a5f5a7747e4125cd568"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a141867e2ecdd009c016b41796cf80d8c"><td class="memItemLeft" align="right" valign="top">double&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_host.html#a141867e2ecdd009c016b41796cf80d8c">Fitness</a></td></tr>
<tr class="separator:a141867e2ecdd009c016b41796cf80d8c"><td class="memSeparator" colspan="2">&#160;</td></tr>
</table>
<a name="details" id="details"></a><h2 class="groupheader">Detailed Description</h2>
<div class="textblock"><p>Core class. Stores and handles a single host object. Each host has multiple instances of <a class="el" href="class_gene.html" title="Core class. Stores and handles a single gene object. Has methods to access and to mutate a gene...">Gene</a> class objects and stores them in 2 vectors called ChromosomeOne and ChromosomeTwo which simulate a diploid genome. </p>

<p>Definition at line <a class="el" href="_host_8h_source.html#l00040">40</a> of file <a class="el" href="_host_8h_source.html">Host.h</a>.</p>
</div><h2 class="groupheader">Constructor &amp; Destructor Documentation</h2>
<a class="anchor" id="a369c12e7425c030528bab0784da40e4b"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">Host::Host </td>
          <td>(</td>
          <td class="paramname"></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Definition at line <a class="el" href="_host_8cpp_source.html#l00035">35</a> of file <a class="el" href="_host_8cpp_source.html">Host.cpp</a>.</p>

</div>
</div>
<a class="anchor" id="a6c14f16c39ae6bc5db2672e0141adccd"></a>
<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">Host::~Host </td>
          <td>(</td>
          <td class="paramname"></td><td>)</td>
          <td></td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">virtual</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">

<p>Definition at line <a class="el" href="_host_8cpp_source.html#l00041">41</a> of file <a class="el" href="_host_8cpp_source.html">Host.cpp</a>.</p>

</div>
</div>
<h2 class="groupheader">Member Function Documentation</h2>
<a class="anchor" id="ac6ecd452f531bbcaf469d5d9d42b8028"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void Host::assignChromOne </td>
          <td>(</td>
          <td class="paramtype"><a class="el" href="_h2_pinteraction_8cpp.html#adcfab8f2ed9ed0c694cb95d35c91aa78">chromovector</a>&#160;</td>
          <td class="paramname"><em>One</em></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Core method. Assigns a new chromosome to host's Chromosome ONE. </p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">One</td><td>- a STL vector of genes </td></tr>
  </table>
  </dd>
</dl>

<p>Definition at line <a class="el" href="_host_8cpp_source.html#l00518">518</a> of file <a class="el" href="_host_8cpp_source.html">Host.cpp</a>.</p>

</div>
</div>
<a class="anchor" id="ab3e04fdf569863ddeee00d6d2f5c990d"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void Host::assignChromTwo </td>
          <td>(</td>
          <td class="paramtype"><a class="el" href="_h2_pinteraction_8cpp.html#adcfab8f2ed9ed0c694cb95d35c91aa78">chromovector</a>&#160;</td>
          <td class="paramname"><em>Two</em></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Core method. Assigns a new chromosome to host's Chromosome TWO. </p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">One</td><td>- a STL vector of genes </td></tr>
  </table>
  </dd>
</dl>

<p>Definition at line <a class="el" href="_host_8cpp_source.html#l00527">527</a> of file <a class="el" href="_host_8cpp_source.html">Host.cpp</a>.</p>

</div>
</div>
<a class="anchor" id="a6079285b3b8bb5280d52500e6b7fc7ef"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void Host::calculateFitnessAccChromSize </td>
          <td>(</td>
          <td class="paramname"></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Core method. Calculates host individual fitness as the number of exposed pathogens divided by the genome size. </p>

<p>Definition at line <a class="el" href="_host_8cpp_source.html#l00576">576</a> of file <a class="el" href="_host_8cpp_source.html">Host.cpp</a>.</p>

</div>
</div>
<a class="anchor" id="aca12c9829c01995669e842876478c7fb"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void Host::calculateFitnessAlphaXSqr </td>
          <td>(</td>
          <td class="paramtype">double&#160;</td>
          <td class="paramname"><em>alpha</em></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Core method. Calculates host individual fitness in proportion to one over the square of the number of genes scaled to by factor <img class="formulaInl" alt="$ \alpha $" src="form_12.png"/>: </p>
<p><img class="formulaInl" alt="$ F = \frac{P}{(\alpha \cdot N)^{2}} $" src="form_8.png"/></p>
<p>where <img class="formulaInl" alt="$0 &lt; \alpha &lt; 1 $" src="form_9.png"/>, <img class="formulaInl" alt="$ P $" src="form_10.png"/> is the number of pathogens exposed and <img class="formulaInl" alt="$ N $" src="form_11.png"/> in the sum of number of genes in both chromosomes.</p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">alpha</td><td>- scaling parameter for F() shape </td></tr>
  </table>
  </dd>
</dl>

<p>Definition at line <a class="el" href="_host_8cpp_source.html#l00613">613</a> of file <a class="el" href="_host_8cpp_source.html">Host.cpp</a>.</p>

</div>
</div>
<a class="anchor" id="a9134661842de046453acc0342ef74eb6"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void Host::calculateFitnessExpFunc </td>
          <td>(</td>
          <td class="paramtype">double&#160;</td>
          <td class="paramname"><em>alpha</em></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Core method. Calculates host individual fitness in proportion to the Gaussian function of the number of genes scaled to by factor <img class="formulaInl" alt="$ \alpha $" src="form_12.png"/>: </p>
<p><img class="formulaInl" alt="$ F = P \cdot \exp \left[-(\alpha N)^{2}\right] $" src="form_15.png"/></p>
<p>where <img class="formulaInl" alt="$0 &lt; \alpha &lt; 1 $" src="form_9.png"/> and scales the slope of the Gaussian function, <img class="formulaInl" alt="$ P $" src="form_10.png"/> is the number of pathogens exposed and <img class="formulaInl" alt="$ N $" src="form_11.png"/> in the sum of number of genes in both chromosomes.</p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">alpha</td><td>- scaling parameter for F() shape </td></tr>
  </table>
  </dd>
</dl>

<p>Definition at line <a class="el" href="_host_8cpp_source.html#l00634">634</a> of file <a class="el" href="_host_8cpp_source.html">Host.cpp</a>.</p>

</div>
</div>
<a class="anchor" id="ad185a5e02c8ed0930e4d400e9c0e6a9a"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void Host::calculateFitnessExpFuncUniqAlleles </td>
          <td>(</td>
          <td class="paramtype">double&#160;</td>
          <td class="paramname"><em>alpha</em></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Core method. Calculates host individual fitness in proportion to the Gaussian function of the number of unique MHC alleles in the chromosome scaled to by factor <img class="formulaInl" alt="$ \alpha $" src="form_12.png"/>: </p>
<p><img class="formulaInl" alt="$ F = P \cdot \exp \left[-(\alpha N)^{2}\right] $" src="form_15.png"/></p>
<p>where <img class="formulaInl" alt="$0 &lt; \alpha &lt; 1 $" src="form_9.png"/> and scales the slope of the Gaussian function, <img class="formulaInl" alt="$ P $" src="form_10.png"/> is the number of pathogens exposed and <img class="formulaInl" alt="$ N $" src="form_11.png"/> in the sum of number of unique MHC alleles in both chromosomes.</p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">alpha</td><td>- scaling parameter for F() shape </td></tr>
  </table>
  </dd>
</dl>

<p>Definition at line <a class="el" href="_host_8cpp_source.html#l00657">657</a> of file <a class="el" href="_host_8cpp_source.html">Host.cpp</a>.</p>

</div>
</div>
<a class="anchor" id="a614430aea49534f29b63232414cf4fcd"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void Host::calculateFitnessForDrift </td>
          <td>(</td>
          <td class="paramname"></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Core method. Sets fitness as 1, a fixed value to simulate genetic drift. </p>

<p>Definition at line <a class="el" href="_host_8cpp_source.html#l00598">598</a> of file <a class="el" href="_host_8cpp_source.html">Host.cpp</a>.</p>

</div>
</div>
<a class="anchor" id="aeb8564ae54f8bbf968a62f13c61083df"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void Host::calculateFitnessJustInfection </td>
          <td>(</td>
          <td class="paramname"></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Core method. Calculates host individual fitness simply as the number of exposed pathogens. </p>

<p>Definition at line <a class="el" href="_host_8cpp_source.html#l00590">590</a> of file <a class="el" href="_host_8cpp_source.html">Host.cpp</a>.</p>

</div>
</div>
<a class="anchor" id="a5911792258205cae25a2a4796061fe49"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void Host::chromoMutProcess </td>
          <td>(</td>
          <td class="paramtype">double&#160;</td>
          <td class="paramname"><em>mut_probabl</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">int&#160;</td>
          <td class="paramname"><em>timeStamp</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Core method. Decides (on a random basis) if there will be any mutations in the genome. </p>
<p>Iterates through the both chromosomes and calls gene mutation function. Genes mutate at random depending on the probability which was user-defined.</p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">mut_probabl</td><td>- mutation probability, a probability a gene will be replaced by a new one </td></tr>
    <tr><td class="paramname">timeStamp</td><td>- current time (number of the model iteration) </td></tr>
  </table>
  </dd>
</dl>

<p>Definition at line <a class="el" href="_host_8cpp_source.html#l00108">108</a> of file <a class="el" href="_host_8cpp_source.html">Host.cpp</a>.</p>

<p><div class="dynheader">
Here is the call graph for this function:</div>
<div class="dyncontent">
<div class="center"><img src="class_host_a5911792258205cae25a2a4796061fe49_cgraph.png" border="0" usemap="#class_host_a5911792258205cae25a2a4796061fe49_cgraph" alt=""/></div>
<map name="class_host_a5911792258205cae25a2a4796061fe49_cgraph" id="class_host_a5911792258205cae25a2a4796061fe49_cgraph">
<area shape="rect" id="node2" href="class_host.html#a3ae32e9601a5c79da23b0ba4970ff0a6" title="Core method. Creates a chromosome&#45;like vector containing only the unique MHC genes from two chromosom..." alt="" coords="224,5,385,32"/>
</map>
</div>
</p>

</div>
</div>
<a class="anchor" id="a03a4d86562ede11801244ee1242105e2"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void Host::chromoMutProcessWithDelDupl </td>
          <td>(</td>
          <td class="paramtype">double&#160;</td>
          <td class="paramname"><em>mut_probabl</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">double&#160;</td>
          <td class="paramname"><em>del</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">double&#160;</td>
          <td class="paramname"><em>dupli</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">unsigned long&#160;</td>
          <td class="paramname"><em>maxGene</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">int&#160;</td>
          <td class="paramname"><em>timeStamp</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Core method. Decides (on a random basis) if there will be any mutations in the genome: deletion, duplication or change of a gene. </p>
<p>Iterates through the both chromosomes and calls gene mutation function. Genes mutate at random depending on the probability which was user-defined. Also at random a gene can be deleted or duplicated. If there's only one single gene in the chromosome then no deletion will be carried on.</p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">mut_probabl</td><td>- mutation probability, a probability a gene will be replaced by a new one </td></tr>
    <tr><td class="paramname">del</td><td>- mutation probability, probability a gene will be deleted </td></tr>
    <tr><td class="paramname">dupli</td><td>- mutation probability, probability a gene will be duplicated (and added at the end of the Chromosome vector) </td></tr>
    <tr><td class="paramname">maxGene</td><td>- maximal allowed number of genes in a chromosome (user defined parameter). </td></tr>
    <tr><td class="paramname">timeStamp</td><td>- current time (current number of the model iteration) </td></tr>
  </table>
  </dd>
</dl>

<p>Definition at line <a class="el" href="_host_8cpp_source.html#l00138">138</a> of file <a class="el" href="_host_8cpp_source.html">Host.cpp</a>.</p>

<p><div class="dynheader">
Here is the call graph for this function:</div>
<div class="dyncontent">
<div class="center"><img src="class_host_a03a4d86562ede11801244ee1242105e2_cgraph.png" border="0" usemap="#class_host_a03a4d86562ede11801244ee1242105e2_cgraph" alt=""/></div>
<map name="class_host_a03a4d86562ede11801244ee1242105e2_cgraph" id="class_host_a03a4d86562ede11801244ee1242105e2_cgraph">
<area shape="rect" id="node2" href="class_random_numbs.html#a364c415e5105061274013a1e68576a64" title="Core method. Gets the instance of a singleton class. " alt="" coords="252,5,441,32"/>
<area shape="rect" id="node3" href="class_random_numbs.html#aaae77654579060738e55b46c8caba9a6" title="Core method. Generates one real random number from a given interval with an uniform distribution..." alt="" coords="259,56,435,83"/>
<area shape="rect" id="node4" href="class_host.html#a3ae32e9601a5c79da23b0ba4970ff0a6" title="Core method. Creates a chromosome&#45;like vector containing only the unique MHC genes from two chromosom..." alt="" coords="266,107,427,133"/>
</map>
</div>
</p>

</div>
</div>
<a class="anchor" id="ad39c5fbaae3af86c44a312ebe9fb0baa"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void Host::chromoMutProcessWithDelDuplPointMuts </td>
          <td>(</td>
          <td class="paramtype">double&#160;</td>
          <td class="paramname"><em>pm_mut_probabl</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">double&#160;</td>
          <td class="paramname"><em>del</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">double&#160;</td>
          <td class="paramname"><em>dupli</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">unsigned long&#160;</td>
          <td class="paramname"><em>maxGene</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">int&#160;</td>
          <td class="paramname"><em>timeStamp</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Core method. Decides (on a random basis) if there will be any mutations in the genome. </p>
<p>Iterates through the both chromosomes and calls gene mutation function. Point sites in genes are mutated at random depending on the probability which was user-defined. Also at random a gene can be deleted or duplicated. If there's only one single gene in the chromosome then no deletion will be carried on.</p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">pm_mut_probabl</td><td>- point mutation probability, a probability a gene will be replaced by a new one </td></tr>
    <tr><td class="paramname">del</td><td>- mutation probability, probability a gene will be deleted </td></tr>
    <tr><td class="paramname">dupli</td><td>- mutation probability, probability a gene will be duplicated (and added at the end of the Chromosome vector) </td></tr>
    <tr><td class="paramname">maxGene</td><td>- maximal allowed number of genes in a chromosome (user defined parameter). </td></tr>
    <tr><td class="paramname">timeStamp</td><td>- current time (current number of the model iteration) </td></tr>
  </table>
  </dd>
</dl>

<p>Definition at line <a class="el" href="_host_8cpp_source.html#l00186">186</a> of file <a class="el" href="_host_8cpp_source.html">Host.cpp</a>.</p>

<p><div class="dynheader">
Here is the call graph for this function:</div>
<div class="dyncontent">
<div class="center"><img src="class_host_ad39c5fbaae3af86c44a312ebe9fb0baa_cgraph.png" border="0" usemap="#class_host_ad39c5fbaae3af86c44a312ebe9fb0baa_cgraph" alt=""/></div>
<map name="class_host_ad39c5fbaae3af86c44a312ebe9fb0baa_cgraph" id="class_host_ad39c5fbaae3af86c44a312ebe9fb0baa_cgraph">
<area shape="rect" id="node2" href="class_random_numbs.html#a364c415e5105061274013a1e68576a64" title="Core method. Gets the instance of a singleton class. " alt="" coords="252,5,441,32"/>
<area shape="rect" id="node3" href="class_random_numbs.html#aaae77654579060738e55b46c8caba9a6" title="Core method. Generates one real random number from a given interval with an uniform distribution..." alt="" coords="259,56,435,83"/>
<area shape="rect" id="node4" href="class_host.html#a3ae32e9601a5c79da23b0ba4970ff0a6" title="Core method. Creates a chromosome&#45;like vector containing only the unique MHC genes from two chromosom..." alt="" coords="266,107,427,133"/>
</map>
</div>
</p>

</div>
</div>
<a class="anchor" id="a5da864a08370512a065cf5779795be5a"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void Host::chromoRecombination </td>
          <td>(</td>
          <td class="paramtype">double&#160;</td>
          <td class="paramname"><em>recomb_prob</em></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Core method. Performs recombination between host's chromosomes by inserting a brake in randomly selected genes on Chromosome One and Two and swapping the resulting parts of the chromosomes. </p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">recomb_prob</td><td>- probability of the recombination event in host's genome </td></tr>
    <tr><td class="paramname">timeStamp</td><td>- current time (current number of the model iteration) </td></tr>
  </table>
  </dd>
</dl>

<p>Definition at line <a class="el" href="_host_8cpp_source.html#l00224">224</a> of file <a class="el" href="_host_8cpp_source.html">Host.cpp</a>.</p>

<p><div class="dynheader">
Here is the call graph for this function:</div>
<div class="dyncontent">
<div class="center"><img src="class_host_a5da864a08370512a065cf5779795be5a_cgraph.png" border="0" usemap="#class_host_a5da864a08370512a065cf5779795be5a_cgraph" alt=""/></div>
<map name="class_host_a5da864a08370512a065cf5779795be5a_cgraph" id="class_host_a5da864a08370512a065cf5779795be5a_cgraph">
<area shape="rect" id="node2" href="class_random_numbs.html#a364c415e5105061274013a1e68576a64" title="Core method. Gets the instance of a singleton class. " alt="" coords="241,5,431,32"/>
<area shape="rect" id="node3" href="class_random_numbs.html#aaae77654579060738e55b46c8caba9a6" title="Core method. Generates one real random number from a given interval with an uniform distribution..." alt="" coords="248,56,424,83"/>
</map>
</div>
</p>

</div>
</div>
<a class="anchor" id="a7c21e1938c61966dc2657a437aeec9b9"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void Host::clearInfections </td>
          <td>(</td>
          <td class="paramname"></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Core method. Zeroes all data regarding infections and fitness. </p>

<p>Definition at line <a class="el" href="_host_8cpp_source.html#l00690">690</a> of file <a class="el" href="_host_8cpp_source.html">Host.cpp</a>.</p>

</div>
</div>
<a class="anchor" id="a505527db337d87670e5ab37eb9818fa3"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname"><a class="el" href="_h2_pinteraction_8cpp.html#adcfab8f2ed9ed0c694cb95d35c91aa78">chromovector</a> Host::doCrossAndMeiosis </td>
          <td>(</td>
          <td class="paramtype">double&#160;</td>
          <td class="paramname"><em>corssing_prob</em></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Core method. Performing crossing over and chromosome selection before mating. </p>
<p>It mixes up genes from two host chromosomes to form one child chromosome later pushed for mating. One can regulate which chromosome is preferred as a source of genes by changing crossing-over probability parameter. If the random selection procedure will fail to form a new chromosome, then new chromosome will have one gene from randomly selected parental chromosome.</p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">corssing_prob</td><td>-probability of a gene crossing over; p=0.5 means that each chromosome gives half of its genes to the resulting meiotic chromosome. p lesser then 0.5 means ChromosomeOne will give majority of genes; p larger then 0.5 will favor ChromosomeTwo as a donor.</td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>a Chromosome vector. </dd></dl>

<p>Definition at line <a class="el" href="_host_8cpp_source.html#l00248">248</a> of file <a class="el" href="_host_8cpp_source.html">Host.cpp</a>.</p>

<p><div class="dynheader">
Here is the call graph for this function:</div>
<div class="dyncontent">
<div class="center"><img src="class_host_a505527db337d87670e5ab37eb9818fa3_cgraph.png" border="0" usemap="#class_host_a505527db337d87670e5ab37eb9818fa3_cgraph" alt=""/></div>
<map name="class_host_a505527db337d87670e5ab37eb9818fa3_cgraph" id="class_host_a505527db337d87670e5ab37eb9818fa3_cgraph">
<area shape="rect" id="node2" href="class_random_numbs.html#a364c415e5105061274013a1e68576a64" title="Core method. Gets the instance of a singleton class. " alt="" coords="228,5,417,32"/>
<area shape="rect" id="node3" href="class_random_numbs.html#aaae77654579060738e55b46c8caba9a6" title="Core method. Generates one real random number from a given interval with an uniform distribution..." alt="" coords="235,56,411,83"/>
</map>
</div>
</p>

</div>
</div>
<a class="anchor" id="a3ae32e9601a5c79da23b0ba4970ff0a6"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void Host::evalUniqueMHCs </td>
          <td>(</td>
          <td class="paramname"></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Core method. Creates a chromosome-like vector containing only the unique MHC genes from two chromosomes. This vector is useful when interacting with pathogens or evaluating fitness. </p>

<p>Definition at line <a class="el" href="_host_8cpp_source.html#l00550">550</a> of file <a class="el" href="_host_8cpp_source.html">Host.cpp</a>.</p>

<p><div class="dynheader">
Here is the caller graph for this function:</div>
<div class="dyncontent">
<div class="center"><img src="class_host_a3ae32e9601a5c79da23b0ba4970ff0a6_icgraph.png" border="0" usemap="#class_host_a3ae32e9601a5c79da23b0ba4970ff0a6_icgraph" alt=""/></div>
<map name="class_host_a3ae32e9601a5c79da23b0ba4970ff0a6_icgraph" id="class_host_a3ae32e9601a5c79da23b0ba4970ff0a6_icgraph">
<area shape="rect" id="node2" href="class_host.html#a4c017e042f00858c8ae173786e077bb4" title="Core method. Sets a new Host object and assigns a user&#45;defined number of genes of user&#45;defined number..." alt="" coords="249,5,379,32"/>
<area shape="rect" id="node3" href="class_host.html#afdfa6fd4262695491d83d3bea9cd29e7" title="Core method. Sets a new Host object and assigns a user&#45;defined number of genes of user&#45;defined number..." alt="" coords="221,56,407,83"/>
<area shape="rect" id="node4" href="class_host.html#a5911792258205cae25a2a4796061fe49" title="Core method. Decides (on a random basis) if there will be any mutations in the genome. " alt="" coords="229,107,399,133"/>
<area shape="rect" id="node5" href="class_host.html#a03a4d86562ede11801244ee1242105e2" title="Core method. Decides (on a random basis) if there will be any mutations in the genome: deletion..." alt="" coords="215,158,413,199"/>
<area shape="rect" id="node6" href="class_host.html#ad39c5fbaae3af86c44a312ebe9fb0baa" title="Core method. Decides (on a random basis) if there will be any mutations in the genome. " alt="" coords="215,223,413,265"/>
</map>
</div>
</p>

</div>
</div>
<a class="anchor" id="a51d5609864393846c577672908f824b0"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">unsigned long Host::getChromoOneSize </td>
          <td>(</td>
          <td class="paramname"></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Returns the number of genes in the Chromosome One. </p>
<dl class="section return"><dt>Returns</dt><dd>number of MHC genes in the Chromosome One. </dd></dl>

<p>Definition at line <a class="el" href="_host_8cpp_source.html#l00360">360</a> of file <a class="el" href="_host_8cpp_source.html">Host.cpp</a>.</p>

<p><div class="dynheader">
Here is the caller graph for this function:</div>
<div class="dyncontent">
<div class="center"><img src="class_host_a51d5609864393846c577672908f824b0_icgraph.png" border="0" usemap="#class_host_a51d5609864393846c577672908f824b0_icgraph" alt=""/></div>
<map name="class_host_a51d5609864393846c577672908f824b0_icgraph" id="class_host_a51d5609864393846c577672908f824b0_icgraph">
<area shape="rect" id="node2" href="class_h2_pinteraction.html#a045058093b06eff43c33ab6c6729176e" title="Core method. Checks if a host gets infected with a pathogen. Heterozygote has an advantage here over ..." alt="" coords="225,5,439,47"/>
</map>
</div>
</p>

</div>
</div>
<a class="anchor" id="ac387176ba203f6aa1f20610f5679385a"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname"><a class="el" href="_h2_pinteraction_8cpp.html#adcfab8f2ed9ed0c694cb95d35c91aa78">chromovector</a> Host::getChromosomeOne </td>
          <td>(</td>
          <td class="paramname"></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Core method. Returns the first chromosome. </p>
<dl class="section return"><dt>Returns</dt><dd>Chromosome One </dd></dl>

<p>Definition at line <a class="el" href="_host_8cpp_source.html#l00477">477</a> of file <a class="el" href="_host_8cpp_source.html">Host.cpp</a>.</p>

</div>
</div>
<a class="anchor" id="aa5afa38408ebba644f56667364213f0d"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname"><a class="el" href="_h2_pinteraction_8cpp.html#adcfab8f2ed9ed0c694cb95d35c91aa78">chromovector</a> Host::getChromosomeTwo </td>
          <td>(</td>
          <td class="paramname"></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Core method. Returns the first chromosome. </p>
<dl class="section return"><dt>Returns</dt><dd>Chromosome Two </dd></dl>

<p>Definition at line <a class="el" href="_host_8cpp_source.html#l00486">486</a> of file <a class="el" href="_host_8cpp_source.html">Host.cpp</a>.</p>

</div>
</div>
<a class="anchor" id="a5158ada636bdc6c9985abfeb51cc5642"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">unsigned long Host::getChromoTwoSize </td>
          <td>(</td>
          <td class="paramname"></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Returns the number of genes in the Chromosome Two. </p>
<dl class="section return"><dt>Returns</dt><dd>number of MHC genes in the Chromosome Two. </dd></dl>

<p>Definition at line <a class="el" href="_host_8cpp_source.html#l00369">369</a> of file <a class="el" href="_host_8cpp_source.html">Host.cpp</a>.</p>

<p><div class="dynheader">
Here is the caller graph for this function:</div>
<div class="dyncontent">
<div class="center"><img src="class_host_a5158ada636bdc6c9985abfeb51cc5642_icgraph.png" border="0" usemap="#class_host_a5158ada636bdc6c9985abfeb51cc5642_icgraph" alt=""/></div>
<map name="class_host_a5158ada636bdc6c9985abfeb51cc5642_icgraph" id="class_host_a5158ada636bdc6c9985abfeb51cc5642_icgraph">
<area shape="rect" id="node2" href="class_h2_pinteraction.html#a045058093b06eff43c33ab6c6729176e" title="Core method. Checks if a host gets infected with a pathogen. Heterozygote has an advantage here over ..." alt="" coords="227,5,440,47"/>
</map>
</div>
</p>

</div>
</div>
<a class="anchor" id="a87d357c93783916037446bf8096ed45d"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">double Host::getFitness </td>
          <td>(</td>
          <td class="paramname"></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Core method. Returns current value of host's fitness. </p>
<dl class="section return"><dt>Returns</dt><dd>current value of host's fitness </dd></dl>

<p>Definition at line <a class="el" href="_host_8cpp_source.html#l00683">683</a> of file <a class="el" href="_host_8cpp_source.html">Host.cpp</a>.</p>

</div>
</div>
<a class="anchor" id="a8fd67f5e888387ddd188e1ba66d52727"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">unsigned long Host::getGenomeSize </td>
          <td>(</td>
          <td class="paramname"></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Returns the total number of genes in the genome. </p>
<dl class="section return"><dt>Returns</dt><dd>number of all MHC genes in the host </dd></dl>

<p>Definition at line <a class="el" href="_host_8cpp_source.html#l00351">351</a> of file <a class="el" href="_host_8cpp_source.html">Host.cpp</a>.</p>

</div>
</div>
<a class="anchor" id="a21a7fead03f3fb31005028f6575ea867"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">unsigned Host::getNumberOfPresentedPatho </td>
          <td>(</td>
          <td class="paramname"></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Core method. Returns the number of pathogen species a host has presented. </p>
<dl class="section return"><dt>Returns</dt><dd>number of presented pathogen species </dd></dl>

<p>Definition at line <a class="el" href="_host_8cpp_source.html#l00673">673</a> of file <a class="el" href="_host_8cpp_source.html">Host.cpp</a>.</p>

</div>
</div>
<a class="anchor" id="a9757c15b19a45468e3ce5921b5d2deab"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">double Host::getNumbOfChromoOneUniqAlleles </td>
          <td>(</td>
          <td class="paramname"></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Returns the number of unique MHC alleles in the Chromosome One. </p>
<dl class="section return"><dt>Returns</dt><dd>number of unique MHC alleles in the Chromosome One. </dd></dl>

<p>Definition at line <a class="el" href="_host_8cpp_source.html#l00380">380</a> of file <a class="el" href="_host_8cpp_source.html">Host.cpp</a>.</p>

</div>
</div>
<a class="anchor" id="a533fcc0f3f715fe36abdb878b0491aaf"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">double Host::getNumbOfChromoTwoUniqAlleles </td>
          <td>(</td>
          <td class="paramname"></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Returns the number of unique MHC alleles in the Chromosome Two. </p>
<dl class="section return"><dt>Returns</dt><dd>number of unique MHC alleles in the Chromosome Two. </dd></dl>

<p>Definition at line <a class="el" href="_host_8cpp_source.html#l00412">412</a> of file <a class="el" href="_host_8cpp_source.html">Host.cpp</a>.</p>

</div>
</div>
<a class="anchor" id="a87feca6e68d17438266fe8e92468382f"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">double Host::getNumbOfUniqMHCgenes </td>
          <td>(</td>
          <td class="paramname"></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Data harvesting method. Returns the number (double) of unique MHC in genome. </p>
<dl class="section return"><dt>Returns</dt><dd>number of unique MHC genes in individual host's genome. </dd></dl>

<p>Definition at line <a class="el" href="_host_8cpp_source.html#l00767">767</a> of file <a class="el" href="_host_8cpp_source.html">Host.cpp</a>.</p>

</div>
</div>
<a class="anchor" id="a48ffcaf86f00239f8dd8c56683478e2a"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">unsigned long int Host::getOneGeneFromOne </td>
          <td>(</td>
          <td class="paramtype">unsigned long&#160;</td>
          <td class="paramname"><em>indx</em></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Core method. Fetches a gene (as integer) with a given position on the Chromosome One. </p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">indx</td><td>- index of the gene you wanna get. </td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>an integer representation of a gene with a given index. </dd></dl>

<p>Definition at line <a class="el" href="_host_8cpp_source.html#l00279">279</a> of file <a class="el" href="_host_8cpp_source.html">Host.cpp</a>.</p>

</div>
</div>
<a class="anchor" id="a2540e6d2f6d139292a964b61f04449f0"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">unsigned long int Host::getOneGeneFromTwo </td>
          <td>(</td>
          <td class="paramtype">unsigned long&#160;</td>
          <td class="paramname"><em>indx</em></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Core method. Fetches a gene (as integer) with a given position on the Chromosome Two. </p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">indx</td><td>- index of the gene you wanna get. </td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>an integer representation of a gene with a given index. </dd></dl>

<p>Definition at line <a class="el" href="_host_8cpp_source.html#l00303">303</a> of file <a class="el" href="_host_8cpp_source.html">Host.cpp</a>.</p>

</div>
</div>
<a class="anchor" id="ab51c69cfe9d0f1b2bf4f3ca4225ae3d9"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">unsigned long int Host::getOneGeneFromUniqVect </td>
          <td>(</td>
          <td class="paramtype">unsigned long&#160;</td>
          <td class="paramname"><em>indx</em></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Core method. Fetches a gene (as integer) with a given position on the vector of unique MHCs. </p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">indx</td><td>- index of the gene you wanna get. </td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>an integer representation of a gene with a given index. </dd></dl>

<p>Definition at line <a class="el" href="_host_8cpp_source.html#l00328">328</a> of file <a class="el" href="_host_8cpp_source.html">Host.cpp</a>.</p>

</div>
</div>
<a class="anchor" id="a093534dcf4dc55de1bd817115676adb0"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname"><a class="el" href="_gene_8cpp.html#a108efa7670466d15a5ecafb60bbc9c37">genestring</a> Host::getSingleGeneFromOne </td>
          <td>(</td>
          <td class="paramtype">unsigned long&#160;</td>
          <td class="paramname"><em>indx</em></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Core method. Fetches a gene (bit string) with a given position on the chromosome One. </p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">indx</td><td>- index of the gene you wanna get. </td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>a bit-string representation of a gene with a given index. </dd></dl>

<p>Definition at line <a class="el" href="_host_8cpp_source.html#l00445">445</a> of file <a class="el" href="_host_8cpp_source.html">Host.cpp</a>.</p>

</div>
</div>
<a class="anchor" id="a25421bd536db26a7e83cf3a67e7b4cbe"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname"><a class="el" href="_gene_8cpp.html#a108efa7670466d15a5ecafb60bbc9c37">genestring</a> Host::getSingleGeneFromTwo </td>
          <td>(</td>
          <td class="paramtype">unsigned long&#160;</td>
          <td class="paramname"><em>indx</em></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Core method. Fetches a gene (bit string) with a given position on the chromosome Two. </p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">indx</td><td>- index of the gene you wanna get. </td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>a bitstring representation of a gene with a given index. </dd></dl>

<p>Definition at line <a class="el" href="_host_8cpp_source.html#l00462">462</a> of file <a class="el" href="_host_8cpp_source.html">Host.cpp</a>.</p>

</div>
</div>
<a class="anchor" id="a0d36e2b60e48e985e6e561fc46213564"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname"><a class="el" href="_h2_pinteraction_8cpp.html#adcfab8f2ed9ed0c694cb95d35c91aa78">chromovector</a> Host::getUniqueMHCs </td>
          <td>(</td>
          <td class="paramname"></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Core method. Returns unique MHC alleles from the host. </p>
<dl class="section return"><dt>Returns</dt><dd>chromosome-like vector if unique MHC alleles. </dd></dl>

<p>Definition at line <a class="el" href="_host_8cpp_source.html#l00509">509</a> of file <a class="el" href="_host_8cpp_source.html">Host.cpp</a>.</p>

</div>
</div>
<a class="anchor" id="a6c8640b0cd873b9669983330b2c4c59e"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">unsigned long Host::getUniqueMhcSize </td>
          <td>(</td>
          <td class="paramname"></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Core method. Returns the size of the vector containing unique MHCs. </p>
<dl class="section return"><dt>Returns</dt><dd>number of unique MHCs in unsigned long format </dd></dl>

<p>Definition at line <a class="el" href="_host_8cpp_source.html#l00777">777</a> of file <a class="el" href="_host_8cpp_source.html">Host.cpp</a>.</p>

<p><div class="dynheader">
Here is the caller graph for this function:</div>
<div class="dyncontent">
<div class="center"><img src="class_host_a6c8640b0cd873b9669983330b2c4c59e_icgraph.png" border="0" usemap="#class_host_a6c8640b0cd873b9669983330b2c4c59e_icgraph" alt=""/></div>
<map name="class_host_a6c8640b0cd873b9669983330b2c4c59e_icgraph" id="class_host_a6c8640b0cd873b9669983330b2c4c59e_icgraph">
<area shape="rect" id="node2" href="class_h2_pinteraction.html#a719ea8b86e86701f86d89cddfddc47db" title="Core method. Checks if a host gets infected with a pathogen. Heterozygote has an advantage here over ..." alt="" coords="221,5,412,47"/>
<area shape="rect" id="node3" href="class_environment.html#aecd592b734359557235bb0a6980f77f9" title="Core method. Iterates through the host population and the parasite population to &quot;infect&quot; the hosts w..." alt="" coords="460,5,649,47"/>
<area shape="rect" id="node4" href="main_8cpp.html#a3c04138a5bfe5d72780bb7e82a18e627" title="The main function. Things are happening here. " alt="" coords="697,13,748,39"/>
</map>
</div>
</p>

</div>
</div>
<a class="anchor" id="a55edfefc42f7306c85983474500ef10b"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname"><a class="el" href="_h2_pinteraction_8cpp.html#adcfab8f2ed9ed0c694cb95d35c91aa78">chromovector</a> Host::mergeChromosomes </td>
          <td>(</td>
          <td class="paramname"></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Data harvesting method. Merges both chromosomes to create an object easier to handle is some situations. </p>
<dl class="section return"><dt>Returns</dt><dd>Chromosome vector containing genes from both, Chromosome One and Two </dd></dl>

<p>Definition at line <a class="el" href="_host_8cpp_source.html#l00496">496</a> of file <a class="el" href="_host_8cpp_source.html">Host.cpp</a>.</p>

</div>
</div>
<a class="anchor" id="afdfa6fd4262695491d83d3bea9cd29e7"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void Host::setNewHomozygHost </td>
          <td>(</td>
          <td class="paramtype">unsigned long&#160;</td>
          <td class="paramname"><em>num_of_loci</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">unsigned long&#160;</td>
          <td class="paramname"><em>gene_size</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">int&#160;</td>
          <td class="paramname"><em>timeStamp</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Core method. Sets a new <a class="el" href="class_host.html" title="Core class. Stores and handles a single host object. Each host has multiple instances of Gene class o...">Host</a> object and assigns a user-defined number of genes of user-defined number of bits (as genes are bit-strings). </p>
<p>A host has two chromosome of equal length. Each corresponding pair of loci is homozygous.</p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">num_of_loci</td><td>- number of gene loci in a chromosome </td></tr>
    <tr><td class="paramname">gene_size</td><td>- the length of the bit-string representing a gene </td></tr>
    <tr><td class="paramname">timeStamp</td><td>- current time (number of the model iteration) </td></tr>
  </table>
  </dd>
</dl>

<p>Definition at line <a class="el" href="_host_8cpp_source.html#l00080">80</a> of file <a class="el" href="_host_8cpp_source.html">Host.cpp</a>.</p>

<p><div class="dynheader">
Here is the call graph for this function:</div>
<div class="dyncontent">
<div class="center"><img src="class_host_afdfa6fd4262695491d83d3bea9cd29e7_cgraph.png" border="0" usemap="#class_host_afdfa6fd4262695491d83d3bea9cd29e7_cgraph" alt=""/></div>
<map name="class_host_afdfa6fd4262695491d83d3bea9cd29e7_cgraph" id="class_host_afdfa6fd4262695491d83d3bea9cd29e7_cgraph">
<area shape="rect" id="node2" href="class_host.html#a3ae32e9601a5c79da23b0ba4970ff0a6" title="Core method. Creates a chromosome&#45;like vector containing only the unique MHC genes from two chromosom..." alt="" coords="240,5,401,32"/>
</map>
</div>
</p>

</div>
</div>
<a class="anchor" id="a4c017e042f00858c8ae173786e077bb4"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void Host::setNewHost </td>
          <td>(</td>
          <td class="paramtype">unsigned long&#160;</td>
          <td class="paramname"><em>num_of_loci</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">unsigned long&#160;</td>
          <td class="paramname"><em>gene_size</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">int&#160;</td>
          <td class="paramname"><em>timeStamp</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Core method. Sets a new <a class="el" href="class_host.html" title="Core class. Stores and handles a single host object. Each host has multiple instances of Gene class o...">Host</a> object and assigns a user-defined number of genes of user-defined number of bits (as genes are bit-strings). </p>
<p>A host has two chromosome of equal length. Each gene is set at random and loci are most likely to be heterozygous.</p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">num_of_loci</td><td>- number of gene loci in a chromosome </td></tr>
    <tr><td class="paramname">gene_size</td><td>- the length of the bit-string representing a gene </td></tr>
    <tr><td class="paramname">timeStamp</td><td>- current time (number of the model iteration) </td></tr>
  </table>
  </dd>
</dl>

<p>Definition at line <a class="el" href="_host_8cpp_source.html#l00055">55</a> of file <a class="el" href="_host_8cpp_source.html">Host.cpp</a>.</p>

<p><div class="dynheader">
Here is the call graph for this function:</div>
<div class="dyncontent">
<div class="center"><img src="class_host_a4c017e042f00858c8ae173786e077bb4_cgraph.png" border="0" usemap="#class_host_a4c017e042f00858c8ae173786e077bb4_cgraph" alt=""/></div>
<map name="class_host_a4c017e042f00858c8ae173786e077bb4_cgraph" id="class_host_a4c017e042f00858c8ae173786e077bb4_cgraph">
<area shape="rect" id="node2" href="class_host.html#a3ae32e9601a5c79da23b0ba4970ff0a6" title="Core method. Creates a chromosome&#45;like vector containing only the unique MHC genes from two chromosom..." alt="" coords="184,5,345,32"/>
</map>
</div>
</p>

</div>
</div>
<a class="anchor" id="a481373a4981aad450aa666962f0c8252"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">std::string Host::stringChromosomes </td>
          <td>(</td>
          <td class="paramname"></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Data harvesting method. Gives a host's genome in a human-readable format. With all the gene specs. </p>
<dl class="section return"><dt>Returns</dt><dd>a STL string containing the host's genome and its annotations in a human-readable format. </dd></dl>

<p>Definition at line <a class="el" href="_host_8cpp_source.html#l00714">714</a> of file <a class="el" href="_host_8cpp_source.html">Host.cpp</a>.</p>

</div>
</div>
<a class="anchor" id="a118fc07706467cd5070dd7b8dd53a3a6"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">std::string Host::stringUniqMHCs </td>
          <td>(</td>
          <td class="paramname"></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Data harvesting method. Gives a host's unique MHC genes in a human-readable format. With all the gene specs. </p>
<dl class="section return"><dt>Returns</dt><dd>a STL string containing the host's unique MHC genes and its annotations in a human-readable format. </dd></dl>

<p>Definition at line <a class="el" href="_host_8cpp_source.html#l00789">789</a> of file <a class="el" href="_host_8cpp_source.html">Host.cpp</a>.</p>

</div>
</div>
<a class="anchor" id="ab30250ea6fdd47221d65914b3e6de4e0"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void Host::swapChromosomes </td>
          <td>(</td>
          <td class="paramname"></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Core method. Randomly swaps places of Chromosome One and Chromosome Two to avoid situation when they effectively become two separate populations. </p>

<p>Definition at line <a class="el" href="_host_8cpp_source.html#l00535">535</a> of file <a class="el" href="_host_8cpp_source.html">Host.cpp</a>.</p>

<p><div class="dynheader">
Here is the call graph for this function:</div>
<div class="dyncontent">
<div class="center"><img src="class_host_ab30250ea6fdd47221d65914b3e6de4e0_cgraph.png" border="0" usemap="#class_host_ab30250ea6fdd47221d65914b3e6de4e0_cgraph" alt=""/></div>
<map name="class_host_ab30250ea6fdd47221d65914b3e6de4e0_cgraph" id="class_host_ab30250ea6fdd47221d65914b3e6de4e0_cgraph">
<area shape="rect" id="node2" href="class_random_numbs.html#a364c415e5105061274013a1e68576a64" title="Core method. Gets the instance of a singleton class. " alt="" coords="228,5,417,32"/>
<area shape="rect" id="node3" href="class_random_numbs.html#aaae77654579060738e55b46c8caba9a6" title="Core method. Generates one real random number from a given interval with an uniform distribution..." alt="" coords="235,56,411,83"/>
</map>
</div>
</p>

</div>
</div>
<h2 class="groupheader">Member Data Documentation</h2>
<a class="anchor" id="af4c460c65f085a6a7c9b73b46e83e7f6"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">std::vector&lt;<a class="el" href="class_gene.html">Gene</a>&gt; Host::ChromosomeOne</td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Definition at line <a class="el" href="_host_8h_source.html#l00098">98</a> of file <a class="el" href="_host_8h_source.html">Host.h</a>.</p>

</div>
</div>
<a class="anchor" id="aca6f27f6124477919835a233e9760a6f"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">std::vector&lt;<a class="el" href="class_gene.html">Gene</a>&gt; Host::ChromosomeTwo</td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Definition at line <a class="el" href="_host_8h_source.html#l00099">99</a> of file <a class="el" href="_host_8h_source.html">Host.h</a>.</p>

</div>
</div>
<a class="anchor" id="a141867e2ecdd009c016b41796cf80d8c"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">double Host::Fitness</td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Definition at line <a class="el" href="_host_8h_source.html#l00101">101</a> of file <a class="el" href="_host_8h_source.html">Host.h</a>.</p>

</div>
</div>
<a class="anchor" id="ac173100d13da6dfeb93387f0d689c349"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">unsigned Host::NumOfPathogesInfecting</td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Definition at line <a class="el" href="_host_8h_source.html#l00048">48</a> of file <a class="el" href="_host_8h_source.html">Host.h</a>.</p>

</div>
</div>
<a class="anchor" id="ad0411c059e63ed72507b0c370880b5ab"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">unsigned Host::NumOfPathogesPresented</td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Definition at line <a class="el" href="_host_8h_source.html#l00049">49</a> of file <a class="el" href="_host_8h_source.html">Host.h</a>.</p>

</div>
</div>
<a class="anchor" id="ac8b084352039979629a7896e2aaefb45"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">std::vector&lt;int&gt; Host::PathogesPresented</td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Definition at line <a class="el" href="_host_8h_source.html#l00047">47</a> of file <a class="el" href="_host_8h_source.html">Host.h</a>.</p>

</div>
</div>
<a class="anchor" id="a0d79bd02d6147f8e94ae975dc42ae026"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">std::vector&lt;int&gt; Host::PathoSpecInfecting</td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Definition at line <a class="el" href="_host_8h_source.html#l00046">46</a> of file <a class="el" href="_host_8h_source.html">Host.h</a>.</p>

</div>
</div>
<a class="anchor" id="a50bc219031501dd2c19a69d4e3f19533"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">int Host::SelectedForReproduction</td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Definition at line <a class="el" href="_host_8h_source.html#l00050">50</a> of file <a class="el" href="_host_8h_source.html">Host.h</a>.</p>

</div>
</div>
<a class="anchor" id="a0a2a96cbc5292a5f5a7747e4125cd568"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">std::vector&lt;<a class="el" href="class_gene.html">Gene</a>&gt; Host::UniqueAlleles</td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Definition at line <a class="el" href="_host_8h_source.html#l00100">100</a> of file <a class="el" href="_host_8h_source.html">Host.h</a>.</p>

</div>
</div>
<hr/>The documentation for this class was generated from the following files:<ul>
<li><a class="el" href="_host_8h_source.html">Host.h</a></li>
<li><a class="el" href="_host_8cpp_source.html">Host.cpp</a></li>
</ul>
</div><!-- contents -->
</div><!-- doc-content -->
<!-- start footer part -->
<div id="nav-path" class="navpath"><!-- id is needed for treeview function! -->
  <ul>
    <li class="navelem"><a class="el" href="class_host.html">Host</a></li>
    <li class="footer">Generated on Fri Dec 8 2017 18:23:29 for Modelling MHC Evolutionary Dynamics by
    <a href="http://www.doxygen.org/index.html">
    <img class="footer" src="doxygen.png" alt="doxygen"/></a> 1.8.11 </li>
  </ul>
</div>
</body>
</html>
