<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml">
<head>
<meta http-equiv="Content-Type" content="text/xhtml;charset=UTF-8"/>
<meta http-equiv="X-UA-Compatible" content="IE=9"/>
<meta name="generator" content="Doxygen 1.8.11"/>
<title>Modelling MHC Evolutionary Dynamics: Environment Class Reference</title>
<link href="tabs.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="jquery.js"></script>
<script type="text/javascript" src="dynsections.js"></script>
<link href="navtree.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="resize.js"></script>
<script type="text/javascript" src="navtreedata.js"></script>
<script type="text/javascript" src="navtree.js"></script>
<script type="text/javascript">
  $(document).ready(initResizable);
  $(window).load(resizeHeight);
</script>
<link href="search/search.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="search/searchdata.js"></script>
<script type="text/javascript" src="search/search.js"></script>
<script type="text/javascript">
  $(document).ready(function() { init_search(); });
</script>
<link href="doxygen.css" rel="stylesheet" type="text/css" />
</head>
<body>
<div id="top"><!-- do not remove this div, it is closed by doxygen! -->
<div id="titlearea">
<table cellspacing="0" cellpadding="0">
 <tbody>
 <tr style="height: 56px;">
  <td id="projectalign" style="padding-left: 0.5em;">
   <div id="projectname">Modelling MHC Evolutionary Dynamics
   &#160;<span id="projectnumber">0.9.1</span>
   </div>
  </td>
 </tr>
 </tbody>
</table>
</div>
<!-- end header part -->
<!-- Generated by Doxygen 1.8.11 -->
<script type="text/javascript">
var searchBox = new SearchBox("searchBox", "search",false,'Search');
</script>
  <div id="navrow1" class="tabs">
    <ul class="tablist">
      <li><a href="index.html"><span>Main&#160;Page</span></a></li>
      <li class="current"><a href="annotated.html"><span>Classes</span></a></li>
      <li><a href="files.html"><span>Files</span></a></li>
      <li>
        <div id="MSearchBox" class="MSearchBoxInactive">
        <span class="left">
          <img id="MSearchSelect" src="search/mag_sel.png"
               onmouseover="return searchBox.OnSearchSelectShow()"
               onmouseout="return searchBox.OnSearchSelectHide()"
               alt=""/>
          <input type="text" id="MSearchField" value="Search" accesskey="S"
               onfocus="searchBox.OnSearchFieldFocus(true)" 
               onblur="searchBox.OnSearchFieldFocus(false)" 
               onkeyup="searchBox.OnSearchFieldChange(event)"/>
          </span><span class="right">
            <a id="MSearchClose" href="javascript:searchBox.CloseResultsWindow()"><img id="MSearchCloseImg" border="0" src="search/close.png" alt=""/></a>
          </span>
        </div>
      </li>
    </ul>
  </div>
  <div id="navrow2" class="tabs2">
    <ul class="tablist">
      <li><a href="annotated.html"><span>Class&#160;List</span></a></li>
      <li><a href="classes.html"><span>Class&#160;Index</span></a></li>
      <li><a href="functions.html"><span>Class&#160;Members</span></a></li>
    </ul>
  </div>
</div><!-- top -->
<div id="side-nav" class="ui-resizable side-nav-resizable">
  <div id="nav-tree">
    <div id="nav-tree-contents">
      <div id="nav-sync" class="sync"></div>
    </div>
  </div>
  <div id="splitbar" style="-moz-user-select:none;" 
       class="ui-resizable-handle">
  </div>
</div>
<script type="text/javascript">
$(document).ready(function(){initNavTree('class_environment.html','');});
</script>
<div id="doc-content">
<!-- window showing the filter options -->
<div id="MSearchSelectWindow"
     onmouseover="return searchBox.OnSearchSelectShow()"
     onmouseout="return searchBox.OnSearchSelectHide()"
     onkeydown="return searchBox.OnSearchSelectKey(event)">
</div>

<!-- iframe showing the search results (closed by default) -->
<div id="MSearchResultsWindow">
<iframe src="javascript:void(0)" frameborder="0" 
        name="MSearchResults" id="MSearchResults">
</iframe>
</div>

<div class="header">
  <div class="summary">
<a href="#pub-methods">Public Member Functions</a> &#124;
<a href="class_environment-members.html">List of all members</a>  </div>
  <div class="headertitle">
<div class="title">Environment Class Reference</div>  </div>
</div><!--header-->
<div class="contents">

<p>Core class. Stores and handles the environment object that is the place where all the things are happening. It contains host and pathogen populations stored in separate STL Vectors. Handles all the population-wide things (setting populations up, calculating fitness in all population, matching pathogens and hosts for infections, mating hosts etc.). Also has methods for fetching some of the stats.  
 <a href="class_environment.html#details">More...</a></p>

<p><code>#include &lt;<a class="el" href="_environment_8h_source.html">Environment.h</a>&gt;</code></p>
<table class="memberdecls">
<tr class="heading"><td colspan="2"><h2 class="groupheader"><a name="pub-methods"></a>
Public Member Functions</h2></td></tr>
<tr class="memitem:aeff4f165bb55ef7abe27662598bb82f2"><td class="memItemLeft" align="right" valign="top">&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_environment.html#aeff4f165bb55ef7abe27662598bb82f2">Environment</a> (unsigned int numberOfThreads)</td></tr>
<tr class="memdesc:aeff4f165bb55ef7abe27662598bb82f2"><td class="mdescLeft">&#160;</td><td class="mdescRight">Core method. Sets the number of threads used later for computing the model.  <a href="#aeff4f165bb55ef7abe27662598bb82f2">More...</a><br /></td></tr>
<tr class="separator:aeff4f165bb55ef7abe27662598bb82f2"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a8e294735187880dd3d59be10c425b29d"><td class="memItemLeft" align="right" valign="top">virtual&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_environment.html#a8e294735187880dd3d59be10c425b29d">~Environment</a> ()</td></tr>
<tr class="separator:a8e294735187880dd3d59be10c425b29d"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a4003e6947de2059bc6ba76af273cf5f7"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_environment.html#a4003e6947de2059bc6ba76af273cf5f7">seedEnvsRNG</a> ()</td></tr>
<tr class="separator:a4003e6947de2059bc6ba76af273cf5f7"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a54b80db1bda904983fda2c38ae1c5dab"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_environment.html#a54b80db1bda904983fda2c38ae1c5dab">setNoMutsVector</a> (int numb_of_species, unsigned long antigen_size, double fixedAntigenFrac)</td></tr>
<tr class="memdesc:a54b80db1bda904983fda2c38ae1c5dab"><td class="mdescLeft">&#160;</td><td class="mdescRight">Core method. It defines "no mutation sites" of the antigen for all individual pathogen species in the simulation. It should be run only ones per simulation.  <a href="#a54b80db1bda904983fda2c38ae1c5dab">More...</a><br /></td></tr>
<tr class="separator:a54b80db1bda904983fda2c38ae1c5dab"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:aa4bcb97295aa866eebaf01d6de58b471"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_environment.html#aa4bcb97295aa866eebaf01d6de58b471">setHostRandomPopulation</a> (int pop_size, unsigned long gene_size, unsigned long chrom_size, int timeStamp, <a class="el" href="class_tagging__system.html">Tagging_system</a> &amp;tag)</td></tr>
<tr class="memdesc:aa4bcb97295aa866eebaf01d6de58b471"><td class="mdescLeft">&#160;</td><td class="mdescRight">Core method. Initializes a vector containing host population.  <a href="#aa4bcb97295aa866eebaf01d6de58b471">More...</a><br /></td></tr>
<tr class="separator:aa4bcb97295aa866eebaf01d6de58b471"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ada465b8ad37b2688ab411fc8ab07f3d7"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_environment.html#ada465b8ad37b2688ab411fc8ab07f3d7">setHostRandomPopulation</a> (int pop_size, unsigned long gene_size, unsigned long chrom_size_lower, unsigned long chrom_size_uper, int timeStamp, <a class="el" href="class_tagging__system.html">Tagging_system</a> &amp;tag)</td></tr>
<tr class="memdesc:ada465b8ad37b2688ab411fc8ab07f3d7"><td class="mdescLeft">&#160;</td><td class="mdescRight">Core method. Initializes a vector containing host population.  <a href="#ada465b8ad37b2688ab411fc8ab07f3d7">More...</a><br /></td></tr>
<tr class="separator:ada465b8ad37b2688ab411fc8ab07f3d7"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a28cc07198ee462d0708ac4e6d43b062f"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_environment.html#a28cc07198ee462d0708ac4e6d43b062f">setHostClonalPopulation</a> (int pop_size, unsigned long gene_size, unsigned long chrom_size, int timeStamp, <a class="el" href="class_tagging__system.html">Tagging_system</a> &amp;tag)</td></tr>
<tr class="memdesc:a28cc07198ee462d0708ac4e6d43b062f"><td class="mdescLeft">&#160;</td><td class="mdescRight">Core method. Initializes a vector containing clonal host population.  <a href="#a28cc07198ee462d0708ac4e6d43b062f">More...</a><br /></td></tr>
<tr class="separator:a28cc07198ee462d0708ac4e6d43b062f"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a5e09ffa9aa8d501096f19131571bace1"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_environment.html#a5e09ffa9aa8d501096f19131571bace1">setPathoPopulatioUniformGenome</a> (int pop_size, unsigned long gene_size, int chrom_size, int numb_of_species, unsigned long mhcSize, int timeStamp, double fixedAntigenFrac, <a class="el" href="class_tagging__system.html">Tagging_system</a> &amp;tag)</td></tr>
<tr class="separator:a5e09ffa9aa8d501096f19131571bace1"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:adcd7c4b6fb8b8181be1f2099a1c7559f"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_environment.html#adcd7c4b6fb8b8181be1f2099a1c7559f">setPathoPopulatioDivSpecies</a> (int pop_size, unsigned long gene_size, int numb_of_species, unsigned long mhcSize, int timeStamp, double fixedAntigenFrac, <a class="el" href="class_tagging__system.html">Tagging_system</a> &amp;tag)</td></tr>
<tr class="memdesc:adcd7c4b6fb8b8181be1f2099a1c7559f"><td class="mdescLeft">&#160;</td><td class="mdescRight">Core method. Initializes the pathogen population.  <a href="#adcd7c4b6fb8b8181be1f2099a1c7559f">More...</a><br /></td></tr>
<tr class="separator:adcd7c4b6fb8b8181be1f2099a1c7559f"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:aecd592b734359557235bb0a6980f77f9"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_environment.html#aecd592b734359557235bb0a6980f77f9">infectOneFromOneSpecHetero</a> ()</td></tr>
<tr class="memdesc:aecd592b734359557235bb0a6980f77f9"><td class="mdescLeft">&#160;</td><td class="mdescRight">Core method. Iterates through the host population and the parasite population to "infect" the hosts with parasites. With heterozygote advantage and it permits a pathogen species to infect a host only ONES.  <a href="#aecd592b734359557235bb0a6980f77f9">More...</a><br /></td></tr>
<tr class="separator:aecd592b734359557235bb0a6980f77f9"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ae63fad9a51f8700cf96cd3019f5a3bec"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_environment.html#ae63fad9a51f8700cf96cd3019f5a3bec">calculateHostsFitnessPerGene</a> ()</td></tr>
<tr class="memdesc:ae63fad9a51f8700cf96cd3019f5a3bec"><td class="mdescLeft">&#160;</td><td class="mdescRight">Core method. Iterates through the host population and calculates the Fitness for each single individual by calling <a class="el" href="class_host.html#a6079285b3b8bb5280d52500e6b7fc7ef" title="Core method. Calculates host individual fitness as the number of exposed pathogens divided by the gen...">Host::calculateFitnessAccChromSize()</a>, which takes the number of MHC genes under account.  <a href="#ae63fad9a51f8700cf96cd3019f5a3bec">More...</a><br /></td></tr>
<tr class="separator:ae63fad9a51f8700cf96cd3019f5a3bec"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a2d11b6356c2d62e7ebf4c97115118c5c"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_environment.html#a2d11b6356c2d62e7ebf4c97115118c5c">calculateHostsFitnessPlainPresent</a> ()</td></tr>
<tr class="memdesc:a2d11b6356c2d62e7ebf4c97115118c5c"><td class="mdescLeft">&#160;</td><td class="mdescRight">Core method. Iterates through the host population and calculates the Fitness for each single individual by calling Host::calculateHostsFitnessPlainInfect(), which is the plain-and-lame sum of presented pathogens.  <a href="#a2d11b6356c2d62e7ebf4c97115118c5c">More...</a><br /></td></tr>
<tr class="separator:a2d11b6356c2d62e7ebf4c97115118c5c"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a596cf4db11a6e6356211a45540e04c7d"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_environment.html#a596cf4db11a6e6356211a45540e04c7d">calculateHostsFitnessForDrift</a> ()</td></tr>
<tr class="memdesc:a596cf4db11a6e6356211a45540e04c7d"><td class="mdescLeft">&#160;</td><td class="mdescRight">Core method. Iterates through the host population and calculates the Fitness for each single individual by calling <a class="el" href="class_host.html#a614430aea49534f29b63232414cf4fcd" title="Core method. Sets fitness as 1, a fixed value to simulate genetic drift. ">Host::calculateFitnessForDrift()</a>, which assigns "1" for each cell to make the genetic driff work.  <a href="#a596cf4db11a6e6356211a45540e04c7d">More...</a><br /></td></tr>
<tr class="separator:a596cf4db11a6e6356211a45540e04c7d"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a0994a6e48ab6a97cbff9feb68ce9dcd3"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_environment.html#a0994a6e48ab6a97cbff9feb68ce9dcd3">calculateHostsFitnessAlphaXsqr</a> (double alpha)</td></tr>
<tr class="memdesc:a0994a6e48ab6a97cbff9feb68ce9dcd3"><td class="mdescLeft">&#160;</td><td class="mdescRight">Core method. Iterates through the host population and calculates the Fitness for each single individual by calling Host::calculateHostsFitnessAlphaXsqr(), which uses one over the square on number of genes as a fitness cost.  <a href="#a0994a6e48ab6a97cbff9feb68ce9dcd3">More...</a><br /></td></tr>
<tr class="separator:a0994a6e48ab6a97cbff9feb68ce9dcd3"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a5f156d009868bce000e653b25fe53967"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_environment.html#a5f156d009868bce000e653b25fe53967">calculateHostsFitnessExpScaling</a> (double alpha)</td></tr>
<tr class="memdesc:a5f156d009868bce000e653b25fe53967"><td class="mdescLeft">&#160;</td><td class="mdescRight">Core method. Iterates through the host population and calculates the Fitness for each single individual by calling <a class="el" href="class_host.html#a9134661842de046453acc0342ef74eb6" title="Core method. Calculates host individual fitness in proportion to the Gaussian function of the number ...">Host::calculateFitnessExpFunc()</a>, which uses a Gaussian function to accommodate the costs of having lots of genes.  <a href="#a5f156d009868bce000e653b25fe53967">More...</a><br /></td></tr>
<tr class="separator:a5f156d009868bce000e653b25fe53967"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:aad5bb2197b43d307428356c123e34f43"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_environment.html#aad5bb2197b43d307428356c123e34f43">calculateHostsFitnessExpScalingUniqAlleles</a> (double alpha)</td></tr>
<tr class="memdesc:aad5bb2197b43d307428356c123e34f43"><td class="mdescLeft">&#160;</td><td class="mdescRight">Core method. Iterates through the host population and calculates the Fitness for each single individual by calling <a class="el" href="class_host.html#ad185a5e02c8ed0930e4d400e9c0e6a9a" title="Core method. Calculates host individual fitness in proportion to the Gaussian function of the number ...">Host::calculateFitnessExpFuncUniqAlleles()</a>, which uses a Gaussian function to accommodate the costs of having lots of unique MHC alleles in chromosomes.  <a href="#aad5bb2197b43d307428356c123e34f43">More...</a><br /></td></tr>
<tr class="separator:aad5bb2197b43d307428356c123e34f43"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a09af2890db9eb1a02069adfb8836aa47"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_environment.html#a09af2890db9eb1a02069adfb8836aa47">selectAndReprodHostsReplace</a> ()</td></tr>
<tr class="memdesc:a09af2890db9eb1a02069adfb8836aa47"><td class="mdescLeft">&#160;</td><td class="mdescRight">Core method. Forms the next generation of hosts using the fitness proportionate selection method. Replaces the old population with a new one.  <a href="#a09af2890db9eb1a02069adfb8836aa47">More...</a><br /></td></tr>
<tr class="separator:a09af2890db9eb1a02069adfb8836aa47"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:aa3e7b74e17329b650089507642e76e5f"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_environment.html#aa3e7b74e17329b650089507642e76e5f">selectAndReprodHostsNoMating</a> ()</td></tr>
<tr class="memdesc:aa3e7b74e17329b650089507642e76e5f"><td class="mdescLeft">&#160;</td><td class="mdescRight">Core method. Forms the next generation of hosts using the fitness proportionate selection method. Replaces the old population with a new one.  <a href="#aa3e7b74e17329b650089507642e76e5f">More...</a><br /></td></tr>
<tr class="separator:aa3e7b74e17329b650089507642e76e5f"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a90177792fb03087e1fc64dca78f7e312"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_environment.html#a90177792fb03087e1fc64dca78f7e312">selectAndReproducePathoFixedPopSizes</a> ()</td></tr>
<tr class="memdesc:a90177792fb03087e1fc64dca78f7e312"><td class="mdescLeft">&#160;</td><td class="mdescRight">Core method. Forms the next generation of hosts using the fitness proportionate selection method. It keeps population sizes of different pathogens species at a fixed number.  <a href="#a90177792fb03087e1fc64dca78f7e312">More...</a><br /></td></tr>
<tr class="separator:a90177792fb03087e1fc64dca78f7e312"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:adff3e1c9799ef7a3bc8d7b988bda3381"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_environment.html#adff3e1c9799ef7a3bc8d7b988bda3381">clearHostInfectionsData</a> ()</td></tr>
<tr class="memdesc:adff3e1c9799ef7a3bc8d7b988bda3381"><td class="mdescLeft">&#160;</td><td class="mdescRight">. Core method. Clears information about infection and fitness in the whole host population.  <a href="#adff3e1c9799ef7a3bc8d7b988bda3381">More...</a><br /></td></tr>
<tr class="separator:adff3e1c9799ef7a3bc8d7b988bda3381"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:af43ff3d50276e23f42895c6c7ddac69c"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_environment.html#af43ff3d50276e23f42895c6c7ddac69c">clearPathoInfectionData</a> ()</td></tr>
<tr class="memdesc:af43ff3d50276e23f42895c6c7ddac69c"><td class="mdescLeft">&#160;</td><td class="mdescRight">. Core method. Clears information about infection and fitness in the whole pathogen population.  <a href="#af43ff3d50276e23f42895c6c7ddac69c">More...</a><br /></td></tr>
<tr class="separator:af43ff3d50276e23f42895c6c7ddac69c"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a2e2e96c34f39413e6c77b8608c3a0822"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_environment.html#a2e2e96c34f39413e6c77b8608c3a0822">mutatePathogens</a> (double mut_probabl, unsigned long mhcSize, int timeStamp)</td></tr>
<tr class="separator:a2e2e96c34f39413e6c77b8608c3a0822"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:abc688aaa01ad739484af2b4d47bf1081"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_environment.html#abc688aaa01ad739484af2b4d47bf1081">mutatePathogensWithRestric</a> (double mut_probabl, unsigned long mhcSize, int timeStamp, <a class="el" href="class_tagging__system.html">Tagging_system</a> &amp;tag)</td></tr>
<tr class="memdesc:abc688aaa01ad739484af2b4d47bf1081"><td class="mdescLeft">&#160;</td><td class="mdescRight">Core method. Iterates through the all genes of the pathogen population and performs mutations in genes with a given probability.  <a href="#abc688aaa01ad739484af2b4d47bf1081">More...</a><br /></td></tr>
<tr class="separator:abc688aaa01ad739484af2b4d47bf1081"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ab16a6139a196350da4e078bb46d4c8ac"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_environment.html#ab16a6139a196350da4e078bb46d4c8ac">mutateHostsWithDelDuplPointMuts</a> (double pm_mut_probabl, double del, double dupl, unsigned long maxGene, int timeStamp, <a class="el" href="class_tagging__system.html">Tagging_system</a> &amp;tag)</td></tr>
<tr class="memdesc:ab16a6139a196350da4e078bb46d4c8ac"><td class="mdescLeft">&#160;</td><td class="mdescRight">Core method. Iterates through the all genes of the host population and performs POINT mutations in genes with a given probability. Also deletions and duplications of genes.  <a href="#ab16a6139a196350da4e078bb46d4c8ac">More...</a><br /></td></tr>
<tr class="separator:ab16a6139a196350da4e078bb46d4c8ac"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ad4004678390faa9ee17889a91344c6db"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_environment.html#ad4004678390faa9ee17889a91344c6db">mutateHostWithDelDuplAllMHCchange</a> (double mut_probabl, double del, double dupl, unsigned long maxGene, int timeStamp, <a class="el" href="class_tagging__system.html">Tagging_system</a> &amp;tag)</td></tr>
<tr class="memdesc:ad4004678390faa9ee17889a91344c6db"><td class="mdescLeft">&#160;</td><td class="mdescRight">Core method. Iterates through the all genes of the host population and performs all-MHC-mutation (writing a whole new bit string over an existing MHC) of genes with a given probability. Also deletions and duplications of genes.  <a href="#ad4004678390faa9ee17889a91344c6db">More...</a><br /></td></tr>
<tr class="separator:ad4004678390faa9ee17889a91344c6db"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:afa7335b658f28dabe2e81ac92aa556bb"><td class="memItemLeft" align="right" valign="top">double&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_environment.html#afa7335b658f28dabe2e81ac92aa556bb">MMtoPMscaling</a> (double MM_prob_mut, unsigned long geneLength)</td></tr>
<tr class="memdesc:afa7335b658f28dabe2e81ac92aa556bb"><td class="mdescLeft">&#160;</td><td class="mdescRight">Core method. When given micro-recombination mutation probability it returns point-mutation probability calculated the way that it the average number of new MHC match in both mutation scenarios.  <a href="#afa7335b658f28dabe2e81ac92aa556bb">More...</a><br /></td></tr>
<tr class="separator:afa7335b658f28dabe2e81ac92aa556bb"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a9133b29bb593bf59549903b6895c9c02"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_environment.html#a9133b29bb593bf59549903b6895c9c02">matingWithNoCommonMHCsmallSubset</a> (unsigned long matingPartnerNumber)</td></tr>
<tr class="memdesc:a9133b29bb593bf59549903b6895c9c02"><td class="mdescLeft">&#160;</td><td class="mdescRight">Core method. Creates a new generation of hosts by sexual reproduction with strong negative preference towards MHC similarity between mates. Takes given number of possible sexual partners.  <a href="#a9133b29bb593bf59549903b6895c9c02">More...</a><br /></td></tr>
<tr class="separator:a9133b29bb593bf59549903b6895c9c02"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a20274efa19fd5d3f3d017cdf589572f3"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_environment.html#a20274efa19fd5d3f3d017cdf589572f3">matingWithOneDifferentMHCsmallSubset</a> (int matingPartnerNumber)</td></tr>
<tr class="memdesc:a20274efa19fd5d3f3d017cdf589572f3"><td class="mdescLeft">&#160;</td><td class="mdescRight">Core method. Creates a new generation of hosts by sexual reproduction with weak negative preference towards MHC similarity between mates.  <a href="#a20274efa19fd5d3f3d017cdf589572f3">More...</a><br /></td></tr>
<tr class="separator:a20274efa19fd5d3f3d017cdf589572f3"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a9111a354d21ef5e3b967e42feda93a94"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_environment.html#a9111a354d21ef5e3b967e42feda93a94">matingMeanOptimalNumberMHCsmallSubset</a> (int matingPartnerNumber)</td></tr>
<tr class="memdesc:a9111a354d21ef5e3b967e42feda93a94"><td class="mdescLeft">&#160;</td><td class="mdescRight">Core method. Creates a new generation of hosts by sexual reproduction picking a mate which ensures that the offspring will have the number of unique MHC greater then their parents. Takes given number of possible sexual partners.  <a href="#a9111a354d21ef5e3b967e42feda93a94">More...</a><br /></td></tr>
<tr class="separator:a9111a354d21ef5e3b967e42feda93a94"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:aff0ee157919d3ee6e97cdbe7423761a6"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_environment.html#aff0ee157919d3ee6e97cdbe7423761a6">matingMaxDifferentMHCs</a> (int matingPartnerNumber)</td></tr>
<tr class="memdesc:aff0ee157919d3ee6e97cdbe7423761a6"><td class="mdescLeft">&#160;</td><td class="mdescRight">Core method. Creates a new generation of hosts by sexual reproduction picking a mate which has the most different MHCs then the selecting partner. Takes a given number of possible sexual partners that are selected. It rather should be called 'minimal overlap'.  <a href="#aff0ee157919d3ee6e97cdbe7423761a6">More...</a><br /></td></tr>
<tr class="separator:aff0ee157919d3ee6e97cdbe7423761a6"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a1331c6f789c4d8a5a47138175e1ce6d0"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_environment.html#a1331c6f789c4d8a5a47138175e1ce6d0">matingRandom</a> ()</td></tr>
<tr class="memdesc:a1331c6f789c4d8a5a47138175e1ce6d0"><td class="mdescLeft">&#160;</td><td class="mdescRight">Core method. Creates a new generation of hosts by sexual reproduction picking mates at random.  <a href="#a1331c6f789c4d8a5a47138175e1ce6d0">More...</a><br /></td></tr>
<tr class="separator:a1331c6f789c4d8a5a47138175e1ce6d0"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a9bbe18137bc9b931826ec18a036f5697"><td class="memItemLeft" align="right" valign="top">unsigned long&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_environment.html#a9bbe18137bc9b931826ec18a036f5697">getPathoNumOfSpecies</a> ()</td></tr>
<tr class="memdesc:a9bbe18137bc9b931826ec18a036f5697"><td class="mdescLeft">&#160;</td><td class="mdescRight">Core method. Gets the number of species of pathogens.  <a href="#a9bbe18137bc9b931826ec18a036f5697">More...</a><br /></td></tr>
<tr class="separator:a9bbe18137bc9b931826ec18a036f5697"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:af8726165c8bdeb45dcc0e0f91e894c38"><td class="memItemLeft" align="right" valign="top">unsigned long&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_environment.html#af8726165c8bdeb45dcc0e0f91e894c38">getPathoSpeciesPopSize</a> (unsigned long spec_numb)</td></tr>
<tr class="memdesc:af8726165c8bdeb45dcc0e0f91e894c38"><td class="mdescLeft">&#160;</td><td class="mdescRight">Core method. Gets a number of individuals in a selected species of pathogen.  <a href="#af8726165c8bdeb45dcc0e0f91e894c38">More...</a><br /></td></tr>
<tr class="separator:af8726165c8bdeb45dcc0e0f91e894c38"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a45aeebf99b44c9b1e66ca16b923605ae"><td class="memItemLeft" align="right" valign="top">unsigned long&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_environment.html#a45aeebf99b44c9b1e66ca16b923605ae">getHostsPopSize</a> ()</td></tr>
<tr class="memdesc:a45aeebf99b44c9b1e66ca16b923605ae"><td class="mdescLeft">&#160;</td><td class="mdescRight">Core method. Gets the host population size.  <a href="#a45aeebf99b44c9b1e66ca16b923605ae">More...</a><br /></td></tr>
<tr class="separator:a45aeebf99b44c9b1e66ca16b923605ae"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a93a98eb73207ed37097a5b398579b44a"><td class="memItemLeft" align="right" valign="top">std::string&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_environment.html#a93a98eb73207ed37097a5b398579b44a">getHostsTags</a> ()</td></tr>
<tr class="separator:a93a98eb73207ed37097a5b398579b44a"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:aa0b7fddb5c105c65ef4817531593e64f"><td class="memItemLeft" align="right" valign="top">std::string&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_environment.html#aa0b7fddb5c105c65ef4817531593e64f">getPathoGenesToString</a> (unsigned long i, unsigned long j)</td></tr>
<tr class="memdesc:aa0b7fddb5c105c65ef4817531593e64f"><td class="mdescLeft">&#160;</td><td class="mdescRight">Data harvesting method. Gets the genome of selected pathogen in human-readable format.  <a href="#aa0b7fddb5c105c65ef4817531593e64f">More...</a><br /></td></tr>
<tr class="separator:aa0b7fddb5c105c65ef4817531593e64f"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a71dc8a1828a2830ffefa901d0974b2c1"><td class="memItemLeft" align="right" valign="top">std::string&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_environment.html#a71dc8a1828a2830ffefa901d0974b2c1">getHostGenesToString</a> (int i)</td></tr>
<tr class="memdesc:a71dc8a1828a2830ffefa901d0974b2c1"><td class="mdescLeft">&#160;</td><td class="mdescRight">Data harvesting method. Gets the genome of selected host in human-readable format.  <a href="#a71dc8a1828a2830ffefa901d0974b2c1">More...</a><br /></td></tr>
<tr class="separator:a71dc8a1828a2830ffefa901d0974b2c1"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:aac0eb33e3bce3534c5513c2867c4a298"><td class="memItemLeft" align="right" valign="top">std::string&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_environment.html#aac0eb33e3bce3534c5513c2867c4a298">getHostUniqMHCtoString</a> (int i)</td></tr>
<tr class="separator:aac0eb33e3bce3534c5513c2867c4a298"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a7ca87db3a1ceb9e27670538fce828675"><td class="memItemLeft" align="right" valign="top">std::string&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_environment.html#a7ca87db3a1ceb9e27670538fce828675">getFixedBitsInAntigens</a> ()</td></tr>
<tr class="memdesc:a7ca87db3a1ceb9e27670538fce828675"><td class="mdescLeft">&#160;</td><td class="mdescRight">Data harvesting method. Prepares a string with the list of fixed "no mutation sites" in antigens used by the DataHarvester to write these data to file.  <a href="#a7ca87db3a1ceb9e27670538fce828675">More...</a><br /></td></tr>
<tr class="separator:a7ca87db3a1ceb9e27670538fce828675"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a12173f6c8c70376c82bace9cc578a738"><td class="memItemLeft" align="right" valign="top">std::string&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_environment.html#a12173f6c8c70376c82bace9cc578a738">getNumbersOfPathogensPresented</a> ()</td></tr>
<tr class="memdesc:a12173f6c8c70376c82bace9cc578a738"><td class="mdescLeft">&#160;</td><td class="mdescRight">Data harvesting method. Prepares a string with the list of how many pathogen species each host had presented.  <a href="#a12173f6c8c70376c82bace9cc578a738">More...</a><br /></td></tr>
<tr class="separator:a12173f6c8c70376c82bace9cc578a738"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:aaaccca98aea69e0eede4010fec439e56"><td class="memItemLeft" align="right" valign="top">std::string&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_environment.html#aaaccca98aea69e0eede4010fec439e56">getNumbersOfMhcInMother</a> ()</td></tr>
<tr class="memdesc:aaaccca98aea69e0eede4010fec439e56"><td class="mdescLeft">&#160;</td><td class="mdescRight">Data harvesting method. Prepares a string with the list of how many unique MHC a "mother" had (a selecting party in mating procedures)  <a href="#aaaccca98aea69e0eede4010fec439e56">More...</a><br /></td></tr>
<tr class="separator:aaaccca98aea69e0eede4010fec439e56"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:abc923f56f8f796f81f05a41bea0b4c82"><td class="memItemLeft" align="right" valign="top">std::string&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_environment.html#abc923f56f8f796f81f05a41bea0b4c82">getNumbersOfMhcInFather</a> ()</td></tr>
<tr class="memdesc:abc923f56f8f796f81f05a41bea0b4c82"><td class="mdescLeft">&#160;</td><td class="mdescRight">Data harvesting method. Prepares a string with the list of how many unique MHC a "father" had (the party that is selected in mating procedures)  <a href="#abc923f56f8f796f81f05a41bea0b4c82">More...</a><br /></td></tr>
<tr class="separator:abc923f56f8f796f81f05a41bea0b4c82"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ab5cccffb4eaae7d47d8dc9078d7dc601"><td class="memItemLeft" align="right" valign="top">std::string&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_environment.html#ab5cccffb4eaae7d47d8dc9078d7dc601">getNumbersOfUniqueMHCs</a> ()</td></tr>
<tr class="separator:ab5cccffb4eaae7d47d8dc9078d7dc601"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a8a77248df68fa9c5daa0924e5c9ee585"><td class="memItemLeft" align="right" valign="top">unsigned long&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_environment.html#a8a77248df68fa9c5daa0924e5c9ee585">getSingleHostGenomeSize</a> (unsigned long indx)</td></tr>
<tr class="separator:a8a77248df68fa9c5daa0924e5c9ee585"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:af7fee696722912338dd1b540adf3dc02"><td class="memItemLeft" align="right" valign="top">unsigned long&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_environment.html#af7fee696722912338dd1b540adf3dc02">getSingleHostChromoOneSize</a> (unsigned long indx)</td></tr>
<tr class="separator:af7fee696722912338dd1b540adf3dc02"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a15990da9399f9c49b7e4bcb84ccace45"><td class="memItemLeft" align="right" valign="top">unsigned long&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_environment.html#a15990da9399f9c49b7e4bcb84ccace45">getSingleHostChromoTwoSize</a> (unsigned long indx)</td></tr>
<tr class="separator:a15990da9399f9c49b7e4bcb84ccace45"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ad5a2510424df27c93774fe0fde37e08b"><td class="memItemLeft" align="right" valign="top">unsigned long&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_environment.html#ad5a2510424df27c93774fe0fde37e08b">getSingleHostRealGeneOne</a> (unsigned long i, unsigned long j)</td></tr>
<tr class="separator:ad5a2510424df27c93774fe0fde37e08b"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:af838d055fa650935798fe466e19809e4"><td class="memItemLeft" align="right" valign="top">unsigned long&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_environment.html#af838d055fa650935798fe466e19809e4">getSingleHostRealGeneTwo</a> (unsigned long i, unsigned long j)</td></tr>
<tr class="separator:af838d055fa650935798fe466e19809e4"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ab34c398994fe34c54bdb4239ac624ac0"><td class="memItemLeft" align="right" valign="top">double&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_environment.html#ab34c398994fe34c54bdb4239ac624ac0">getHostFitness</a> (unsigned long indx)</td></tr>
<tr class="separator:ab34c398994fe34c54bdb4239ac624ac0"><td class="memSeparator" colspan="2">&#160;</td></tr>
</table>
<a name="details" id="details"></a><h2 class="groupheader">Detailed Description</h2>
<div class="textblock"><p>Core class. Stores and handles the environment object that is the place where all the things are happening. It contains host and pathogen populations stored in separate STL Vectors. Handles all the population-wide things (setting populations up, calculating fitness in all population, matching pathogens and hosts for infections, mating hosts etc.). Also has methods for fetching some of the stats. </p>

<p>Definition at line <a class="el" href="_environment_8h_source.html#l00042">42</a> of file <a class="el" href="_environment_8h_source.html">Environment.h</a>.</p>
</div><h2 class="groupheader">Constructor &amp; Destructor Documentation</h2>
<a class="anchor" id="aeff4f165bb55ef7abe27662598bb82f2"></a>
<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">Environment::Environment </td>
          <td>(</td>
          <td class="paramtype">unsigned int&#160;</td>
          <td class="paramname"><em>numberOfThreads</em></td><td>)</td>
          <td></td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">explicit</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">

<p>Core method. Sets the number of threads used later for computing the model. </p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">numberOfThreads</td><td>- number of threads you want to use for computing this mode. </td></tr>
  </table>
  </dd>
</dl>

<p>Definition at line <a class="el" href="_environment_8cpp_source.html#l00042">42</a> of file <a class="el" href="_environment_8cpp_source.html">Environment.cpp</a>.</p>

<p><div class="dynheader">
Here is the call graph for this function:</div>
<div class="dyncontent">
<div class="center"><img src="class_environment_aeff4f165bb55ef7abe27662598bb82f2_cgraph.png" border="0" usemap="#class_environment_aeff4f165bb55ef7abe27662598bb82f2_cgraph" alt=""/></div>
<map name="class_environment_aeff4f165bb55ef7abe27662598bb82f2_cgraph" id="class_environment_aeff4f165bb55ef7abe27662598bb82f2_cgraph">
<area shape="rect" id="node2" href="class_environment.html#a4003e6947de2059bc6ba76af273cf5f7" title="Environment::seedEnvsRNG" alt="" coords="229,5,420,32"/>
<area shape="rect" id="node3" href="class_environment.html#a8e294735187880dd3d59be10c425b29d" title="Environment::~Environment" alt="" coords="233,56,417,83"/>
</map>
</div>
</p>

</div>
</div>
<a class="anchor" id="a8e294735187880dd3d59be10c425b29d"></a>
<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">Environment::~Environment </td>
          <td>(</td>
          <td class="paramname"></td><td>)</td>
          <td></td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">virtual</span><span class="mlabel">default</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">

<p><div class="dynheader">
Here is the caller graph for this function:</div>
<div class="dyncontent">
<div class="center"><img src="class_environment_a8e294735187880dd3d59be10c425b29d_icgraph.png" border="0" usemap="#class_environment_a8e294735187880dd3d59be10c425b29d_icgraph" alt=""/></div>
<map name="class_environment_a8e294735187880dd3d59be10c425b29d_icgraph" id="class_environment_a8e294735187880dd3d59be10c425b29d_icgraph">
<area shape="rect" id="node2" href="class_environment.html#aeff4f165bb55ef7abe27662598bb82f2" title="Core method. Sets the number of threads used later for computing the model. " alt="" coords="237,5,413,32"/>
</map>
</div>
</p>

</div>
</div>
<h2 class="groupheader">Member Function Documentation</h2>
<a class="anchor" id="a0994a6e48ab6a97cbff9feb68ce9dcd3"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void Environment::calculateHostsFitnessAlphaXsqr </td>
          <td>(</td>
          <td class="paramtype">double&#160;</td>
          <td class="paramname"><em>alpha</em></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Core method. Iterates through the host population and calculates the Fitness for each single individual by calling Host::calculateHostsFitnessAlphaXsqr(), which uses one over the square on number of genes as a fitness cost. </p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">alpha</td><td>- penalty for having too many MHC types factor for the host fitness function </td></tr>
  </table>
  </dd>
</dl>

<p>Definition at line <a class="el" href="_environment_8cpp_source.html#l00372">372</a> of file <a class="el" href="_environment_8cpp_source.html">Environment.cpp</a>.</p>

</div>
</div>
<a class="anchor" id="a5f156d009868bce000e653b25fe53967"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void Environment::calculateHostsFitnessExpScaling </td>
          <td>(</td>
          <td class="paramtype">double&#160;</td>
          <td class="paramname"><em>alpha</em></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Core method. Iterates through the host population and calculates the Fitness for each single individual by calling <a class="el" href="class_host.html#a9134661842de046453acc0342ef74eb6" title="Core method. Calculates host individual fitness in proportion to the Gaussian function of the number ...">Host::calculateFitnessExpFunc()</a>, which uses a Gaussian function to accommodate the costs of having lots of genes. </p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">alpha</td><td>- penalty for having too many MHC types factor for the host fitness function </td></tr>
  </table>
  </dd>
</dl>

<p>Definition at line <a class="el" href="_environment_8cpp_source.html#l00388">388</a> of file <a class="el" href="_environment_8cpp_source.html">Environment.cpp</a>.</p>

</div>
</div>
<a class="anchor" id="aad5bb2197b43d307428356c123e34f43"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void Environment::calculateHostsFitnessExpScalingUniqAlleles </td>
          <td>(</td>
          <td class="paramtype">double&#160;</td>
          <td class="paramname"><em>alpha</em></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Core method. Iterates through the host population and calculates the Fitness for each single individual by calling <a class="el" href="class_host.html#ad185a5e02c8ed0930e4d400e9c0e6a9a" title="Core method. Calculates host individual fitness in proportion to the Gaussian function of the number ...">Host::calculateFitnessExpFuncUniqAlleles()</a>, which uses a Gaussian function to accommodate the costs of having lots of unique MHC alleles in chromosomes. </p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">alpha</td><td>- penalty for having too many MHC types factor for the host fitness function </td></tr>
  </table>
  </dd>
</dl>

<p>Definition at line <a class="el" href="_environment_8cpp_source.html#l00404">404</a> of file <a class="el" href="_environment_8cpp_source.html">Environment.cpp</a>.</p>

</div>
</div>
<a class="anchor" id="a596cf4db11a6e6356211a45540e04c7d"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void Environment::calculateHostsFitnessForDrift </td>
          <td>(</td>
          <td class="paramname"></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Core method. Iterates through the host population and calculates the Fitness for each single individual by calling <a class="el" href="class_host.html#a614430aea49534f29b63232414cf4fcd" title="Core method. Sets fitness as 1, a fixed value to simulate genetic drift. ">Host::calculateFitnessForDrift()</a>, which assigns "1" for each cell to make the genetic driff work. </p>

<p>Definition at line <a class="el" href="_environment_8cpp_source.html#l00356">356</a> of file <a class="el" href="_environment_8cpp_source.html">Environment.cpp</a>.</p>

</div>
</div>
<a class="anchor" id="ae63fad9a51f8700cf96cd3019f5a3bec"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void Environment::calculateHostsFitnessPerGene </td>
          <td>(</td>
          <td class="paramname"></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Core method. Iterates through the host population and calculates the Fitness for each single individual by calling <a class="el" href="class_host.html#a6079285b3b8bb5280d52500e6b7fc7ef" title="Core method. Calculates host individual fitness as the number of exposed pathogens divided by the gen...">Host::calculateFitnessAccChromSize()</a>, which takes the number of MHC genes under account. </p>

<p>Definition at line <a class="el" href="_environment_8cpp_source.html#l00328">328</a> of file <a class="el" href="_environment_8cpp_source.html">Environment.cpp</a>.</p>

</div>
</div>
<a class="anchor" id="a2d11b6356c2d62e7ebf4c97115118c5c"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void Environment::calculateHostsFitnessPlainPresent </td>
          <td>(</td>
          <td class="paramname"></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Core method. Iterates through the host population and calculates the Fitness for each single individual by calling Host::calculateHostsFitnessPlainInfect(), which is the plain-and-lame sum of presented pathogens. </p>

<p>Definition at line <a class="el" href="_environment_8cpp_source.html#l00342">342</a> of file <a class="el" href="_environment_8cpp_source.html">Environment.cpp</a>.</p>

</div>
</div>
<a class="anchor" id="adff3e1c9799ef7a3bc8d7b988bda3381"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void Environment::clearHostInfectionsData </td>
          <td>(</td>
          <td class="paramname"></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>. Core method. Clears information about infection and fitness in the whole host population. </p>

<p>Definition at line <a class="el" href="_environment_8cpp_source.html#l00623">623</a> of file <a class="el" href="_environment_8cpp_source.html">Environment.cpp</a>.</p>

</div>
</div>
<a class="anchor" id="af43ff3d50276e23f42895c6c7ddac69c"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void Environment::clearPathoInfectionData </td>
          <td>(</td>
          <td class="paramname"></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>. Core method. Clears information about infection and fitness in the whole pathogen population. </p>

<p>Definition at line <a class="el" href="_environment_8cpp_source.html#l00607">607</a> of file <a class="el" href="_environment_8cpp_source.html">Environment.cpp</a>.</p>

</div>
</div>
<a class="anchor" id="a7ca87db3a1ceb9e27670538fce828675"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">std::string Environment::getFixedBitsInAntigens </td>
          <td>(</td>
          <td class="paramname"></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Data harvesting method. Prepares a string with the list of fixed "no mutation sites" in antigens used by the DataHarvester to write these data to file. </p>
<dl class="section return"><dt>Returns</dt><dd>a string listing all the fixed sites in all pathogen species </dd></dl>

<p>Definition at line <a class="el" href="_environment_8cpp_source.html#l01203">1203</a> of file <a class="el" href="_environment_8cpp_source.html">Environment.cpp</a>.</p>

<p><div class="dynheader">
Here is the caller graph for this function:</div>
<div class="dyncontent">
<div class="center"><img src="class_environment_a7ca87db3a1ceb9e27670538fce828675_icgraph.png" border="0" usemap="#class_environment_a7ca87db3a1ceb9e27670538fce828675_icgraph" alt=""/></div>
<map name="class_environment_a7ca87db3a1ceb9e27670538fce828675_icgraph" id="class_environment_a7ca87db3a1ceb9e27670538fce828675_icgraph">
<area shape="rect" id="node2" href="class_data_handler.html#a6378c36aa8e4daf8f525c0fee446897e" title="Data harvesting method. Record the indices of fixed bits that are not allow to mutate in all pathogen..." alt="" coords="229,5,412,47"/>
</map>
</div>
</p>

</div>
</div>
<a class="anchor" id="ab34c398994fe34c54bdb4239ac624ac0"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">double Environment::getHostFitness </td>
          <td>(</td>
          <td class="paramtype">unsigned long&#160;</td>
          <td class="paramname"><em>indx</em></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Definition at line <a class="el" href="_environment_8cpp_source.html#l01292">1292</a> of file <a class="el" href="_environment_8cpp_source.html">Environment.cpp</a>.</p>

<p><div class="dynheader">
Here is the caller graph for this function:</div>
<div class="dyncontent">
<div class="center"><img src="class_environment_ab34c398994fe34c54bdb4239ac624ac0_icgraph.png" border="0" usemap="#class_environment_ab34c398994fe34c54bdb4239ac624ac0_icgraph" alt=""/></div>
<map name="class_environment_ab34c398994fe34c54bdb4239ac624ac0_icgraph" id="class_environment_ab34c398994fe34c54bdb4239ac624ac0_icgraph">
<area shape="rect" id="node2" href="class_data_handler.html#a23ad0de50b3da065674e92b150e12d9f" title="Data harvesting method. Calculates and writes to a file some stats about the hosts population genetic..." alt="" coords="245,5,449,47"/>
</map>
</div>
</p>

</div>
</div>
<a class="anchor" id="a71dc8a1828a2830ffefa901d0974b2c1"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">std::string Environment::getHostGenesToString </td>
          <td>(</td>
          <td class="paramtype">int&#160;</td>
          <td class="paramname"><em>i</em></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Data harvesting method. Gets the genome of selected host in human-readable format. </p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">i</td><td>individual's index within the host vector </td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>a string of the host's chromosome in a human-readable format. </dd></dl>

<p>Definition at line <a class="el" href="_environment_8cpp_source.html#l01187">1187</a> of file <a class="el" href="_environment_8cpp_source.html">Environment.cpp</a>.</p>

<p><div class="dynheader">
Here is the caller graph for this function:</div>
<div class="dyncontent">
<div class="center"><img src="class_environment_a71dc8a1828a2830ffefa901d0974b2c1_icgraph.png" border="0" usemap="#class_environment_a71dc8a1828a2830ffefa901d0974b2c1_icgraph" alt=""/></div>
<map name="class_environment_a71dc8a1828a2830ffefa901d0974b2c1_icgraph" id="class_environment_a71dc8a1828a2830ffefa901d0974b2c1_icgraph">
<area shape="rect" id="node2" href="class_data_handler.html#a46129d74eac82af0caae37a9e20e3e2e" title="Data harvesting method. Writes to a file all hosts with their genomes in a human&#45;readable format..." alt="" coords="240,5,432,47"/>
</map>
</div>
</p>

</div>
</div>
<a class="anchor" id="a45aeebf99b44c9b1e66ca16b923605ae"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">unsigned long Environment::getHostsPopSize </td>
          <td>(</td>
          <td class="paramname"></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Core method. Gets the host population size. </p>
<dl class="section return"><dt>Returns</dt><dd>host population size </dd></dl>

<p>Definition at line <a class="el" href="_environment_8cpp_source.html#l00792">792</a> of file <a class="el" href="_environment_8cpp_source.html">Environment.cpp</a>.</p>

<p><div class="dynheader">
Here is the caller graph for this function:</div>
<div class="dyncontent">
<div class="center"><img src="class_environment_a45aeebf99b44c9b1e66ca16b923605ae_icgraph.png" border="0" usemap="#class_environment_a45aeebf99b44c9b1e66ca16b923605ae_icgraph" alt=""/></div>
<map name="class_environment_a45aeebf99b44c9b1e66ca16b923605ae_icgraph" id="class_environment_a45aeebf99b44c9b1e66ca16b923605ae_icgraph">
<area shape="rect" id="node2" href="class_data_handler.html#a46129d74eac82af0caae37a9e20e3e2e" title="Data harvesting method. Writes to a file all hosts with their genomes in a human&#45;readable format..." alt="" coords="265,5,457,47"/>
<area shape="rect" id="node3" href="class_data_handler.html#a23ad0de50b3da065674e92b150e12d9f" title="Data harvesting method. Calculates and writes to a file some stats about the hosts population genetic..." alt="" coords="259,71,463,112"/>
<area shape="rect" id="node4" href="class_data_handler.html#a5661a45f0a9965bcd041c5617f1a1e8b" title="Data harvesting method. Saves the number of genes in Chromosome ONE and the number of unique MHC type..." alt="" coords="265,136,456,177"/>
</map>
</div>
</p>

</div>
</div>
<a class="anchor" id="a93a98eb73207ed37097a5b398579b44a"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">std::string Environment::getHostsTags </td>
          <td>(</td>
          <td class="paramname"></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">

</div>
</div>
<a class="anchor" id="aac0eb33e3bce3534c5513c2867c4a298"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">std::string Environment::getHostUniqMHCtoString </td>
          <td>(</td>
          <td class="paramtype">int&#160;</td>
          <td class="paramname"><em>i</em></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Definition at line <a class="el" href="_environment_8cpp_source.html#l01192">1192</a> of file <a class="el" href="_environment_8cpp_source.html">Environment.cpp</a>.</p>

</div>
</div>
<a class="anchor" id="abc923f56f8f796f81f05a41bea0b4c82"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">std::string Environment::getNumbersOfMhcInFather </td>
          <td>(</td>
          <td class="paramname"></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Data harvesting method. Prepares a string with the list of how many unique MHC a "father" had (the party that is selected in mating procedures) </p>
<dl class="section return"><dt>Returns</dt><dd>a string listing how many MHCs a "father" had </dd></dl>

<p>Definition at line <a class="el" href="_environment_8cpp_source.html#l01250">1250</a> of file <a class="el" href="_environment_8cpp_source.html">Environment.cpp</a>.</p>

<p><div class="dynheader">
Here is the caller graph for this function:</div>
<div class="dyncontent">
<div class="center"><img src="class_environment_abc923f56f8f796f81f05a41bea0b4c82_icgraph.png" border="0" usemap="#class_environment_abc923f56f8f796f81f05a41bea0b4c82_icgraph" alt=""/></div>
<map name="class_environment_abc923f56f8f796f81f05a41bea0b4c82_icgraph" id="class_environment_abc923f56f8f796f81f05a41bea0b4c82_icgraph">
<area shape="rect" id="node2" href="class_data_handler.html#ab7b38c92d0d564999a1159ceb56494e4" title="Data harvesting method. Records the number of unique MHCs a &quot;mother&quot; and a &quot;farther&quot; had..." alt="" coords="241,5,451,47"/>
</map>
</div>
</p>

</div>
</div>
<a class="anchor" id="aaaccca98aea69e0eede4010fec439e56"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">std::string Environment::getNumbersOfMhcInMother </td>
          <td>(</td>
          <td class="paramname"></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Data harvesting method. Prepares a string with the list of how many unique MHC a "mother" had (a selecting party in mating procedures) </p>
<dl class="section return"><dt>Returns</dt><dd>a string listing how many MHCs a "mother" had </dd></dl>

<p>Definition at line <a class="el" href="_environment_8cpp_source.html#l01235">1235</a> of file <a class="el" href="_environment_8cpp_source.html">Environment.cpp</a>.</p>

<p><div class="dynheader">
Here is the caller graph for this function:</div>
<div class="dyncontent">
<div class="center"><img src="class_environment_aaaccca98aea69e0eede4010fec439e56_icgraph.png" border="0" usemap="#class_environment_aaaccca98aea69e0eede4010fec439e56_icgraph" alt=""/></div>
<map name="class_environment_aaaccca98aea69e0eede4010fec439e56_icgraph" id="class_environment_aaaccca98aea69e0eede4010fec439e56_icgraph">
<area shape="rect" id="node2" href="class_data_handler.html#ab7b38c92d0d564999a1159ceb56494e4" title="Data harvesting method. Records the number of unique MHCs a &quot;mother&quot; and a &quot;farther&quot; had..." alt="" coords="241,5,451,47"/>
</map>
</div>
</p>

</div>
</div>
<a class="anchor" id="a12173f6c8c70376c82bace9cc578a738"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">std::string Environment::getNumbersOfPathogensPresented </td>
          <td>(</td>
          <td class="paramname"></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Data harvesting method. Prepares a string with the list of how many pathogen species each host had presented. </p>
<dl class="section return"><dt>Returns</dt><dd>a string listing how many pathogens hosts presented </dd></dl>

<p>Definition at line <a class="el" href="_environment_8cpp_source.html#l01220">1220</a> of file <a class="el" href="_environment_8cpp_source.html">Environment.cpp</a>.</p>

<p><div class="dynheader">
Here is the caller graph for this function:</div>
<div class="dyncontent">
<div class="center"><img src="class_environment_a12173f6c8c70376c82bace9cc578a738_icgraph.png" border="0" usemap="#class_environment_a12173f6c8c70376c82bace9cc578a738_icgraph" alt=""/></div>
<map name="class_environment_a12173f6c8c70376c82bace9cc578a738_icgraph" id="class_environment_a12173f6c8c70376c82bace9cc578a738_icgraph">
<area shape="rect" id="node2" href="class_data_handler.html#ac0bd3c1ba0fa18e132ea9095db9747cd" title="Data harvesting method. Saves the number of presented pathogens by each indrividual host in each time..." alt="" coords="241,5,432,47"/>
</map>
</div>
</p>

</div>
</div>
<a class="anchor" id="ab5cccffb4eaae7d47d8dc9078d7dc601"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">std::string Environment::getNumbersOfUniqueMHCs </td>
          <td>(</td>
          <td class="paramname"></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Definition at line <a class="el" href="_environment_8cpp_source.html#l01260">1260</a> of file <a class="el" href="_environment_8cpp_source.html">Environment.cpp</a>.</p>

<p><div class="dynheader">
Here is the caller graph for this function:</div>
<div class="dyncontent">
<div class="center"><img src="class_environment_ab5cccffb4eaae7d47d8dc9078d7dc601_icgraph.png" border="0" usemap="#class_environment_ab5cccffb4eaae7d47d8dc9078d7dc601_icgraph" alt=""/></div>
<map name="class_environment_ab5cccffb4eaae7d47d8dc9078d7dc601_icgraph" id="class_environment_ab5cccffb4eaae7d47d8dc9078d7dc601_icgraph">
<area shape="rect" id="node2" href="class_data_handler.html#a791581a6ac572d0daabd2c5bf8a53033" title="Saves the number of unique MHC types of each individual in each time step. Should be used BEFORE mati..." alt="" coords="241,5,451,47"/>
<area shape="rect" id="node3" href="class_data_handler.html#a95cd188763d53b65e7e2e658bb76760b" title="Saves the number of unique MHC types of each individual in each time step. Should be used AFTER matin..." alt="" coords="241,71,451,112"/>
</map>
</div>
</p>

</div>
</div>
<a class="anchor" id="aa0b7fddb5c105c65ef4817531593e64f"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">std::string Environment::getPathoGenesToString </td>
          <td>(</td>
          <td class="paramtype">unsigned long&#160;</td>
          <td class="paramname"><em>i</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">unsigned long&#160;</td>
          <td class="paramname"><em>j</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Data harvesting method. Gets the genome of selected pathogen in human-readable format. </p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">i</td><td>pathogen species number </td></tr>
    <tr><td class="paramname">j</td><td>individual's index within the pathogen species vector </td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>a string of the pathogen's chromosome in a human-readable format. </dd></dl>

<p>Definition at line <a class="el" href="_environment_8cpp_source.html#l01176">1176</a> of file <a class="el" href="_environment_8cpp_source.html">Environment.cpp</a>.</p>

<p><div class="dynheader">
Here is the caller graph for this function:</div>
<div class="dyncontent">
<div class="center"><img src="class_environment_aa0b7fddb5c105c65ef4817531593e64f_icgraph.png" border="0" usemap="#class_environment_aa0b7fddb5c105c65ef4817531593e64f_icgraph" alt=""/></div>
<map name="class_environment_aa0b7fddb5c105c65ef4817531593e64f_icgraph" id="class_environment_aa0b7fddb5c105c65ef4817531593e64f_icgraph">
<area shape="rect" id="node2" href="class_data_handler.html#a877ed5bfd00f0cbeef0413bf1d47f865" title="Data harvesting method. Writes to a file all pathogens with their genomes in a human&#45;readable format..." alt="" coords="247,5,445,47"/>
</map>
</div>
</p>

</div>
</div>
<a class="anchor" id="a9bbe18137bc9b931826ec18a036f5697"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">unsigned long Environment::getPathoNumOfSpecies </td>
          <td>(</td>
          <td class="paramname"></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Core method. Gets the number of species of pathogens. </p>
<dl class="section return"><dt>Returns</dt><dd>number of pathogen species </dd></dl>

<p>Definition at line <a class="el" href="_environment_8cpp_source.html#l00772">772</a> of file <a class="el" href="_environment_8cpp_source.html">Environment.cpp</a>.</p>

<p><div class="dynheader">
Here is the caller graph for this function:</div>
<div class="dyncontent">
<div class="center"><img src="class_environment_a9bbe18137bc9b931826ec18a036f5697_icgraph.png" border="0" usemap="#class_environment_a9bbe18137bc9b931826ec18a036f5697_icgraph" alt=""/></div>
<map name="class_environment_a9bbe18137bc9b931826ec18a036f5697_icgraph" id="class_environment_a9bbe18137bc9b931826ec18a036f5697_icgraph">
<area shape="rect" id="node2" href="class_data_handler.html#a89f86a67a11f719ec282f37740663f95" title="Data harvesting method. Writes to a file population sizes of all pathogen species in a given time..." alt="" coords="236,5,443,47"/>
<area shape="rect" id="node3" href="class_data_handler.html#a877ed5bfd00f0cbeef0413bf1d47f865" title="Data harvesting method. Writes to a file all pathogens with their genomes in a human&#45;readable format..." alt="" coords="240,71,439,112"/>
</map>
</div>
</p>

</div>
</div>
<a class="anchor" id="af8726165c8bdeb45dcc0e0f91e894c38"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">unsigned long Environment::getPathoSpeciesPopSize </td>
          <td>(</td>
          <td class="paramtype">unsigned long&#160;</td>
          <td class="paramname"><em>spec_numb</em></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Core method. Gets a number of individuals in a selected species of pathogen. </p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">spec_numb</td><td>- number of selected species. </td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>number of individuals of a selected species. </dd></dl>

<p>Definition at line <a class="el" href="_environment_8cpp_source.html#l00783">783</a> of file <a class="el" href="_environment_8cpp_source.html">Environment.cpp</a>.</p>

<p><div class="dynheader">
Here is the caller graph for this function:</div>
<div class="dyncontent">
<div class="center"><img src="class_environment_af8726165c8bdeb45dcc0e0f91e894c38_icgraph.png" border="0" usemap="#class_environment_af8726165c8bdeb45dcc0e0f91e894c38_icgraph" alt=""/></div>
<map name="class_environment_af8726165c8bdeb45dcc0e0f91e894c38_icgraph" id="class_environment_af8726165c8bdeb45dcc0e0f91e894c38_icgraph">
<area shape="rect" id="node2" href="class_data_handler.html#a89f86a67a11f719ec282f37740663f95" title="Data harvesting method. Writes to a file population sizes of all pathogen species in a given time..." alt="" coords="256,5,463,47"/>
<area shape="rect" id="node3" href="class_data_handler.html#a877ed5bfd00f0cbeef0413bf1d47f865" title="Data harvesting method. Writes to a file all pathogens with their genomes in a human&#45;readable format..." alt="" coords="260,71,459,112"/>
</map>
</div>
</p>

</div>
</div>
<a class="anchor" id="af7fee696722912338dd1b540adf3dc02"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">unsigned long Environment::getSingleHostChromoOneSize </td>
          <td>(</td>
          <td class="paramtype">unsigned long&#160;</td>
          <td class="paramname"><em>indx</em></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Definition at line <a class="el" href="_environment_8cpp_source.html#l01274">1274</a> of file <a class="el" href="_environment_8cpp_source.html">Environment.cpp</a>.</p>

<p><div class="dynheader">
Here is the caller graph for this function:</div>
<div class="dyncontent">
<div class="center"><img src="class_environment_af7fee696722912338dd1b540adf3dc02_icgraph.png" border="0" usemap="#class_environment_af7fee696722912338dd1b540adf3dc02_icgraph" alt=""/></div>
<map name="class_environment_af7fee696722912338dd1b540adf3dc02_icgraph" id="class_environment_af7fee696722912338dd1b540adf3dc02_icgraph">
<area shape="rect" id="node2" href="class_data_handler.html#a23ad0de50b3da065674e92b150e12d9f" title="Data harvesting method. Calculates and writes to a file some stats about the hosts population genetic..." alt="" coords="239,5,443,47"/>
<area shape="rect" id="node3" href="class_data_handler.html#a5661a45f0a9965bcd041c5617f1a1e8b" title="Data harvesting method. Saves the number of genes in Chromosome ONE and the number of unique MHC type..." alt="" coords="245,71,436,112"/>
</map>
</div>
</p>

</div>
</div>
<a class="anchor" id="a15990da9399f9c49b7e4bcb84ccace45"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">unsigned long Environment::getSingleHostChromoTwoSize </td>
          <td>(</td>
          <td class="paramtype">unsigned long&#160;</td>
          <td class="paramname"><em>indx</em></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Definition at line <a class="el" href="_environment_8cpp_source.html#l01279">1279</a> of file <a class="el" href="_environment_8cpp_source.html">Environment.cpp</a>.</p>

<p><div class="dynheader">
Here is the caller graph for this function:</div>
<div class="dyncontent">
<div class="center"><img src="class_environment_a15990da9399f9c49b7e4bcb84ccace45_icgraph.png" border="0" usemap="#class_environment_a15990da9399f9c49b7e4bcb84ccace45_icgraph" alt=""/></div>
<map name="class_environment_a15990da9399f9c49b7e4bcb84ccace45_icgraph" id="class_environment_a15990da9399f9c49b7e4bcb84ccace45_icgraph">
<area shape="rect" id="node2" href="class_data_handler.html#a23ad0de50b3da065674e92b150e12d9f" title="Data harvesting method. Calculates and writes to a file some stats about the hosts population genetic..." alt="" coords="239,5,443,47"/>
</map>
</div>
</p>

</div>
</div>
<a class="anchor" id="a8a77248df68fa9c5daa0924e5c9ee585"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">unsigned long Environment::getSingleHostGenomeSize </td>
          <td>(</td>
          <td class="paramtype">unsigned long&#160;</td>
          <td class="paramname"><em>indx</em></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Definition at line <a class="el" href="_environment_8cpp_source.html#l01269">1269</a> of file <a class="el" href="_environment_8cpp_source.html">Environment.cpp</a>.</p>

</div>
</div>
<a class="anchor" id="ad5a2510424df27c93774fe0fde37e08b"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">unsigned long Environment::getSingleHostRealGeneOne </td>
          <td>(</td>
          <td class="paramtype">unsigned long&#160;</td>
          <td class="paramname"><em>i</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">unsigned long&#160;</td>
          <td class="paramname"><em>j</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Definition at line <a class="el" href="_environment_8cpp_source.html#l01284">1284</a> of file <a class="el" href="_environment_8cpp_source.html">Environment.cpp</a>.</p>

<p><div class="dynheader">
Here is the caller graph for this function:</div>
<div class="dyncontent">
<div class="center"><img src="class_environment_ad5a2510424df27c93774fe0fde37e08b_icgraph.png" border="0" usemap="#class_environment_ad5a2510424df27c93774fe0fde37e08b_icgraph" alt=""/></div>
<map name="class_environment_ad5a2510424df27c93774fe0fde37e08b_icgraph" id="class_environment_ad5a2510424df27c93774fe0fde37e08b_icgraph">
<area shape="rect" id="node2" href="class_data_handler.html#a23ad0de50b3da065674e92b150e12d9f" title="Data harvesting method. Calculates and writes to a file some stats about the hosts population genetic..." alt="" coords="239,5,443,47"/>
<area shape="rect" id="node3" href="class_data_handler.html#a5661a45f0a9965bcd041c5617f1a1e8b" title="Data harvesting method. Saves the number of genes in Chromosome ONE and the number of unique MHC type..." alt="" coords="245,71,436,112"/>
</map>
</div>
</p>

</div>
</div>
<a class="anchor" id="af838d055fa650935798fe466e19809e4"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">unsigned long Environment::getSingleHostRealGeneTwo </td>
          <td>(</td>
          <td class="paramtype">unsigned long&#160;</td>
          <td class="paramname"><em>i</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">unsigned long&#160;</td>
          <td class="paramname"><em>j</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Definition at line <a class="el" href="_environment_8cpp_source.html#l01288">1288</a> of file <a class="el" href="_environment_8cpp_source.html">Environment.cpp</a>.</p>

<p><div class="dynheader">
Here is the caller graph for this function:</div>
<div class="dyncontent">
<div class="center"><img src="class_environment_af838d055fa650935798fe466e19809e4_icgraph.png" border="0" usemap="#class_environment_af838d055fa650935798fe466e19809e4_icgraph" alt=""/></div>
<map name="class_environment_af838d055fa650935798fe466e19809e4_icgraph" id="class_environment_af838d055fa650935798fe466e19809e4_icgraph">
<area shape="rect" id="node2" href="class_data_handler.html#a23ad0de50b3da065674e92b150e12d9f" title="Data harvesting method. Calculates and writes to a file some stats about the hosts population genetic..." alt="" coords="239,5,443,47"/>
</map>
</div>
</p>

</div>
</div>
<a class="anchor" id="aecd592b734359557235bb0a6980f77f9"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void Environment::infectOneFromOneSpecHetero </td>
          <td>(</td>
          <td class="paramname"></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Core method. Iterates through the host population and the parasite population to "infect" the hosts with parasites. With heterozygote advantage and it permits a pathogen species to infect a host only ONES. </p>
<p>Each host is exposed to one, randomly selected individual from a pathogen species. Exposition procedure for a single host is repeated for all pathogen species. Maximum number of pathogens a host can contract in one go equals to the number of pathogen species. Fitness is calculated with heterozygote advantage added (antigen recognition just by one allele gives a full advantage). One species can infect a host only ONES. </p>

<p>Definition at line <a class="el" href="_environment_8cpp_source.html#l00305">305</a> of file <a class="el" href="_environment_8cpp_source.html">Environment.cpp</a>.</p>

<p><div class="dynheader">
Here is the call graph for this function:</div>
<div class="dyncontent">
<div class="center"><img src="class_environment_aecd592b734359557235bb0a6980f77f9_cgraph.png" border="0" usemap="#class_environment_aecd592b734359557235bb0a6980f77f9_cgraph" alt=""/></div>
<map name="class_environment_aecd592b734359557235bb0a6980f77f9_cgraph" id="class_environment_aecd592b734359557235bb0a6980f77f9_cgraph">
<area shape="rect" id="node2" href="class_random.html#aff6c9a321cc83da8ae04f003fdfdd169" title="Returns a random unsigned int value in [min, max] from uniform distribution. " alt="" coords="243,48,465,75"/>
<area shape="rect" id="node3" href="class_h2_pinteraction.html#a46703648dcef05c6478d6249aeadd2c8" title="Core method. Checks if a host gets infected with a pathogen. Heterozygote has an advantage here over ..." alt="" coords="259,99,449,141"/>
<area shape="rect" id="node4" href="class_host.html#a0d36e2b60e48e985e6e561fc46213564" title="Core method. Returns unique MHC alleles from the host. " alt="" coords="537,5,692,32"/>
<area shape="rect" id="node5" href="class_pathogen.html#a751c89dc80a7a21207ce7d16beaf266c" title="Core method. Fetches the pathogene&#39;s antigens. " alt="" coords="529,56,700,83"/>
<area shape="rect" id="node6" href="class_antigen.html#aeebf005aeee7ea02f135550d127603dd" title="Core method. Returns the epitopes that can be generated from one antigen. " alt="" coords="543,107,686,133"/>
<area shape="rect" id="node7" href="class_pathogen.html#ada9e5929d6736f63377f0625153b224b" title="Core method. Fetches a species tag. " alt="" coords="529,157,701,184"/>
<area shape="rect" id="node8" href="class_h2_pinteraction.html#ae63bc3afa414686c4c7d9a500f3d4a53" title="Core method. Checks if the antigen is presented by a given MHC. " alt="" coords="513,208,716,235"/>
</map>
</div>
</p>

</div>
</div>
<a class="anchor" id="aff0ee157919d3ee6e97cdbe7423761a6"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void Environment::matingMaxDifferentMHCs </td>
          <td>(</td>
          <td class="paramtype">int&#160;</td>
          <td class="paramname"><em>matingPartnerNumber</em></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Core method. Creates a new generation of hosts by sexual reproduction picking a mate which has the most different MHCs then the selecting partner. Takes a given number of possible sexual partners that are selected. It rather should be called 'minimal overlap'. </p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">matingPartnerNumber</td><td>- number of randomly selected partners an individual will checks out eventually selecting one that is the best to mate with. </td></tr>
  </table>
  </dd>
</dl>

<p>Definition at line <a class="el" href="_environment_8cpp_source.html#l01055">1055</a> of file <a class="el" href="_environment_8cpp_source.html">Environment.cpp</a>.</p>

<p><div class="dynheader">
Here is the call graph for this function:</div>
<div class="dyncontent">
<div class="center"><img src="class_environment_aff0ee157919d3ee6e97cdbe7423761a6_cgraph.png" border="0" usemap="#class_environment_aff0ee157919d3ee6e97cdbe7423761a6_cgraph" alt=""/></div>
<map name="class_environment_aff0ee157919d3ee6e97cdbe7423761a6_cgraph" id="class_environment_aff0ee157919d3ee6e97cdbe7423761a6_cgraph">
<area shape="rect" id="node2" href="class_random.html#aff6c9a321cc83da8ae04f003fdfdd169" title="Returns a random unsigned int value in [min, max] from uniform distribution. " alt="" coords="308,5,531,32"/>
<area shape="rect" id="node3" href="class_random.html#a549d2fd18d531d9b8973eb49824073b2" title="Random::returnEngene" alt="" coords="341,56,497,83"/>
</map>
</div>
</p>

</div>
</div>
<a class="anchor" id="a9111a354d21ef5e3b967e42feda93a94"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void Environment::matingMeanOptimalNumberMHCsmallSubset </td>
          <td>(</td>
          <td class="paramtype">int&#160;</td>
          <td class="paramname"><em>matingPartnerNumber</em></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Core method. Creates a new generation of hosts by sexual reproduction picking a mate which ensures that the offspring will have the number of unique MHC greater then their parents. Takes given number of possible sexual partners. </p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">matingPartnerNumber</td><td>- number of randomly selected partners an individual will checks out eventually selecting one best to mate with. </td></tr>
  </table>
  </dd>
</dl>

<p>Definition at line <a class="el" href="_environment_8cpp_source.html#l00978">978</a> of file <a class="el" href="_environment_8cpp_source.html">Environment.cpp</a>.</p>

<p><div class="dynheader">
Here is the call graph for this function:</div>
<div class="dyncontent">
<div class="center"><img src="class_environment_a9111a354d21ef5e3b967e42feda93a94_cgraph.png" border="0" usemap="#class_environment_a9111a354d21ef5e3b967e42feda93a94_cgraph" alt=""/></div>
<map name="class_environment_a9111a354d21ef5e3b967e42feda93a94_cgraph" id="class_environment_a9111a354d21ef5e3b967e42feda93a94_cgraph">
<area shape="rect" id="node2" href="class_random.html#aff6c9a321cc83da8ae04f003fdfdd169" title="Returns a random unsigned int value in [min, max] from uniform distribution. " alt="" coords="272,5,495,32"/>
<area shape="rect" id="node3" href="class_random.html#a549d2fd18d531d9b8973eb49824073b2" title="Random::returnEngene" alt="" coords="305,56,461,83"/>
</map>
</div>
</p>

</div>
</div>
<a class="anchor" id="a1331c6f789c4d8a5a47138175e1ce6d0"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void Environment::matingRandom </td>
          <td>(</td>
          <td class="paramname"></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Core method. Creates a new generation of hosts by sexual reproduction picking mates at random. </p>
<p>Mates are randomly selected. It only makes sure that an individual does not mate with itself. </p>

<p>Definition at line <a class="el" href="_environment_8cpp_source.html#l01128">1128</a> of file <a class="el" href="_environment_8cpp_source.html">Environment.cpp</a>.</p>

<p><div class="dynheader">
Here is the call graph for this function:</div>
<div class="dyncontent">
<div class="center"><img src="class_environment_a1331c6f789c4d8a5a47138175e1ce6d0_cgraph.png" border="0" usemap="#class_environment_a1331c6f789c4d8a5a47138175e1ce6d0_cgraph" alt=""/></div>
<map name="class_environment_a1331c6f789c4d8a5a47138175e1ce6d0_cgraph" id="class_environment_a1331c6f789c4d8a5a47138175e1ce6d0_cgraph">
<area shape="rect" id="node2" href="class_random.html#aff6c9a321cc83da8ae04f003fdfdd169" title="Returns a random unsigned int value in [min, max] from uniform distribution. " alt="" coords="244,5,467,32"/>
</map>
</div>
</p>

</div>
</div>
<a class="anchor" id="a9133b29bb593bf59549903b6895c9c02"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void Environment::matingWithNoCommonMHCsmallSubset </td>
          <td>(</td>
          <td class="paramtype">unsigned long&#160;</td>
          <td class="paramname"><em>matingPartnerNumber</em></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Core method. Creates a new generation of hosts by sexual reproduction with strong negative preference towards MHC similarity between mates. Takes given number of possible sexual partners. </p>
<p>There are no sexes as the host species is assumed a hermaphrodite. Each individual checks out a user defined N number of random individuals from the population and mates with the individual that has the lowest number of similar MHC genes. The process of random mating and selection is repeated until the algorithm will recreate a population of the same size as the original one.</p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">matingPartnerNumber</td><td>- number of randomly selected partners an individual will checks out eventually selecting one best to mate with. </td></tr>
  </table>
  </dd>
</dl>

<p>Definition at line <a class="el" href="_environment_8cpp_source.html#l00811">811</a> of file <a class="el" href="_environment_8cpp_source.html">Environment.cpp</a>.</p>

<p><div class="dynheader">
Here is the call graph for this function:</div>
<div class="dyncontent">
<div class="center"><img src="class_environment_a9133b29bb593bf59549903b6895c9c02_cgraph.png" border="0" usemap="#class_environment_a9133b29bb593bf59549903b6895c9c02_cgraph" alt=""/></div>
<map name="class_environment_a9133b29bb593bf59549903b6895c9c02_cgraph" id="class_environment_a9133b29bb593bf59549903b6895c9c02_cgraph">
<area shape="rect" id="node2" href="class_random.html#aff6c9a321cc83da8ae04f003fdfdd169" title="Returns a random unsigned int value in [min, max] from uniform distribution. " alt="" coords="240,5,463,32"/>
<area shape="rect" id="node3" href="class_random.html#a549d2fd18d531d9b8973eb49824073b2" title="Random::returnEngene" alt="" coords="273,56,429,83"/>
</map>
</div>
</p>

</div>
</div>
<a class="anchor" id="a20274efa19fd5d3f3d017cdf589572f3"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void Environment::matingWithOneDifferentMHCsmallSubset </td>
          <td>(</td>
          <td class="paramtype">int&#160;</td>
          <td class="paramname"><em>matingPartnerNumber</em></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Core method. Creates a new generation of hosts by sexual reproduction with weak negative preference towards MHC similarity between mates. </p>
<p>There are no sexes as the host species is assumed a hermaphrodite. Each individual checks out a user defined N number of random individuals from the population and mates with the first of the N individual that has at least one different MHC gene not present in the choosing party genome. The process of random mating and selection is repeated until the algorithm will recreate a population of the same size as the original one.</p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">matingPartnerNumber</td><td>- number of randomly selected partners an individual will checks out eventually selecting one best to mate with. </td></tr>
  </table>
  </dd>
</dl>

<p>Definition at line <a class="el" href="_environment_8cpp_source.html#l00901">901</a> of file <a class="el" href="_environment_8cpp_source.html">Environment.cpp</a>.</p>

<p><div class="dynheader">
Here is the call graph for this function:</div>
<div class="dyncontent">
<div class="center"><img src="class_environment_a20274efa19fd5d3f3d017cdf589572f3_cgraph.png" border="0" usemap="#class_environment_a20274efa19fd5d3f3d017cdf589572f3_cgraph" alt=""/></div>
<map name="class_environment_a20274efa19fd5d3f3d017cdf589572f3_cgraph" id="class_environment_a20274efa19fd5d3f3d017cdf589572f3_cgraph">
<area shape="rect" id="node2" href="class_random.html#aff6c9a321cc83da8ae04f003fdfdd169" title="Returns a random unsigned int value in [min, max] from uniform distribution. " alt="" coords="247,5,469,32"/>
<area shape="rect" id="node3" href="class_random.html#a549d2fd18d531d9b8973eb49824073b2" title="Random::returnEngene" alt="" coords="280,56,436,83"/>
</map>
</div>
</p>

</div>
</div>
<a class="anchor" id="afa7335b658f28dabe2e81ac92aa556bb"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">double Environment::MMtoPMscaling </td>
          <td>(</td>
          <td class="paramtype">double&#160;</td>
          <td class="paramname"><em>MM_prob_mut</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">unsigned long&#160;</td>
          <td class="paramname"><em>geneLength</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Core method. When given micro-recombination mutation probability it returns point-mutation probability calculated the way that it the average number of new MHC match in both mutation scenarios. </p>
<p>It implements the way in Ejsmond MJ, Radwan J. 2015. <em>Red Queen drives positive selection on Major Histocompatibility Complex genes (MHC)</em> micro-recombination mutation is transformed to point-mutation probability:</p>
<p><img class="formulaInl" alt="$ p_{t} = 1 - \left[1 - p_{m} (1 - 0.5^{b})\right]^{1/b} $" src="form_0.png"/></p>
<p>where <img class="formulaInl" alt="$b$" src="form_1.png"/> is the number of bits, <img class="formulaInl" alt="$p_{m}$" src="form_2.png"/> is the old-style mutation probability. Only to avoid problems with numerical stability when calculating the power of <img class="formulaInl" alt="$1/b$" src="form_3.png"/> when <img class="formulaInl" alt="$b$" src="form_1.png"/> is large and the base is small we re-phrased it into:</p>
<p><img class="formulaInl" alt="$ p_{t} = 1 - \exp\left( 1/b \cdot \ln [1 - p_{m} (1 - 0.5^{b})]\right) $" src="form_4.png"/></p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">MM_prob_mut</td><td>- micro-recombination mutation ('old-style' mutation) </td></tr>
    <tr><td class="paramname">geneLength</td><td>- number of bits in a gene representation </td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>- calculated corresponding point mutation </dd></dl>

<p>Definition at line <a class="el" href="_environment_8cpp_source.html#l00709">709</a> of file <a class="el" href="_environment_8cpp_source.html">Environment.cpp</a>.</p>

</div>
</div>
<a class="anchor" id="ab16a6139a196350da4e078bb46d4c8ac"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void Environment::mutateHostsWithDelDuplPointMuts </td>
          <td>(</td>
          <td class="paramtype">double&#160;</td>
          <td class="paramname"><em>pm_mut_probabl</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">double&#160;</td>
          <td class="paramname"><em>del</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">double&#160;</td>
          <td class="paramname"><em>dupl</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">unsigned long&#160;</td>
          <td class="paramname"><em>maxGene</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">int&#160;</td>
          <td class="paramname"><em>timeStamp</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype"><a class="el" href="class_tagging__system.html">Tagging_system</a> &amp;&#160;</td>
          <td class="paramname"><em>tag</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Core method. Iterates through the all genes of the host population and performs POINT mutations in genes with a given probability. Also deletions and duplications of genes. </p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">pm_mut_probabl</td><td>- probability of a point mutation in a single gene. </td></tr>
    <tr><td class="paramname">del</td><td>- mutation probability, probability a gene will be deleted </td></tr>
    <tr><td class="paramname">dupli</td><td>- mutation probability, probability a gene will be duplicated (and added at the end of the Chromosome vector) </td></tr>
    <tr><td class="paramname">timeStamp</td><td>- current time (number of the model iteration) </td></tr>
    <tr><td class="paramname">maxGene</td><td>- maximal allowed number of genes in a chromosome (user defined parameter). </td></tr>
    <tr><td class="paramname">tag</td><td>- pointer to the tagging system marking each gene variant </td></tr>
  </table>
  </dd>
</dl>

<p>Definition at line <a class="el" href="_environment_8cpp_source.html#l00647">647</a> of file <a class="el" href="_environment_8cpp_source.html">Environment.cpp</a>.</p>

</div>
</div>
<a class="anchor" id="ad4004678390faa9ee17889a91344c6db"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void Environment::mutateHostWithDelDuplAllMHCchange </td>
          <td>(</td>
          <td class="paramtype">double&#160;</td>
          <td class="paramname"><em>mut_probabl</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">double&#160;</td>
          <td class="paramname"><em>del</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">double&#160;</td>
          <td class="paramname"><em>dupl</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">unsigned long&#160;</td>
          <td class="paramname"><em>maxGene</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">int&#160;</td>
          <td class="paramname"><em>timeStamp</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype"><a class="el" href="class_tagging__system.html">Tagging_system</a> &amp;&#160;</td>
          <td class="paramname"><em>tag</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Core method. Iterates through the all genes of the host population and performs all-MHC-mutation (writing a whole new bit string over an existing MHC) of genes with a given probability. Also deletions and duplications of genes. </p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">mut_probabl</td><td>- probability of mutation of the whole MHC flip. </td></tr>
    <tr><td class="paramname">del</td><td>- mutation probability, probability a gene will be deleted </td></tr>
    <tr><td class="paramname">dupli</td><td>- mutation probability, probability a gene will be duplicated (and added at the end of the Chromosome vector) </td></tr>
    <tr><td class="paramname">timeStamp</td><td>- current time (number of the model iteration) </td></tr>
    <tr><td class="paramname">maxGene</td><td>- maximal allowed number of genes in a chromosome (user defined parameter). </td></tr>
    <tr><td class="paramname">tag</td><td>- pointer to the tagging system marking each gene variant </td></tr>
  </table>
  </dd>
</dl>

<p>Definition at line <a class="el" href="_environment_8cpp_source.html#l00674">674</a> of file <a class="el" href="_environment_8cpp_source.html">Environment.cpp</a>.</p>

</div>
</div>
<a class="anchor" id="a2e2e96c34f39413e6c77b8608c3a0822"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void Environment::mutatePathogens </td>
          <td>(</td>
          <td class="paramtype">double&#160;</td>
          <td class="paramname"><em>mut_probabl</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">unsigned long&#160;</td>
          <td class="paramname"><em>mhcSize</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">int&#160;</td>
          <td class="paramname"><em>timeStamp</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

</div>
</div>
<a class="anchor" id="abc688aaa01ad739484af2b4d47bf1081"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void Environment::mutatePathogensWithRestric </td>
          <td>(</td>
          <td class="paramtype">double&#160;</td>
          <td class="paramname"><em>mut_probabl</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">unsigned long&#160;</td>
          <td class="paramname"><em>mhcSize</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">int&#160;</td>
          <td class="paramname"><em>timeStamp</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype"><a class="el" href="class_tagging__system.html">Tagging_system</a> &amp;&#160;</td>
          <td class="paramname"><em>tag</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Core method. Iterates through the all genes of the pathogen population and performs mutations in genes with a given probability. </p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">mut_probabl</td><td>- probability of a mutation in a single gene. </td></tr>
    <tr><td class="paramname">mhcSize</td><td>- number of bits in MHC protein </td></tr>
    <tr><td class="paramname">timeStamp</td><td>- current time (number of the model iteration)</td></tr>
  </table>
  </dd>
</dl>
<p>void <a class="el" href="class_environment.html#a2e2e96c34f39413e6c77b8608c3a0822">Environment::mutatePathogens(double mut_probabl, unsigned long mhcSize, int timeStamp)</a>{ unsigned long PathPopulationSize = PathPopulation.size(); for (int i = 0; i &lt; PathPopulationSize; ++i){ unsigned long PathPopulationIthSize = PathPopulation[i].size(); for (int j = 0; j &lt; PathPopulationIthSize; ++j){ PathPopulation[i][j].chromoMutProcess(mut_probabl, mhcSize, timeStamp); } } } Core method. Iterates through the all genes of the pathogen population and performs mutations in genes with a given probability. But some positions in the bit-string (gene) are not allowed to change.</p>
<p>The "No-Mutation Vector" is defined within the <a class="el" href="class_environment.html" title="Core class. Stores and handles the environment object that is the place where all the things are happ...">Environment</a> object.</p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">mut_probabl</td><td>- probability of a mutation in a single gene. </td></tr>
    <tr><td class="paramname">mhcSize</td><td>- number of bits in MHC protein </td></tr>
    <tr><td class="paramname">timeStamp</td><td>- current time (number of the model iteration) </td></tr>
    <tr><td class="paramname">tag</td><td>- pointer to the tagging system marking each gene variant </td></tr>
  </table>
  </dd>
</dl>

<p>Definition at line <a class="el" href="_environment_8cpp_source.html#l00746">746</a> of file <a class="el" href="_environment_8cpp_source.html">Environment.cpp</a>.</p>

</div>
</div>
<a class="anchor" id="a4003e6947de2059bc6ba76af273cf5f7"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void Environment::seedEnvsRNG </td>
          <td>(</td>
          <td class="paramname"></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Definition at line <a class="el" href="_environment_8cpp_source.html#l00060">60</a> of file <a class="el" href="_environment_8cpp_source.html">Environment.cpp</a>.</p>

<p><div class="dynheader">
Here is the caller graph for this function:</div>
<div class="dyncontent">
<div class="center"><img src="class_environment_a4003e6947de2059bc6ba76af273cf5f7_icgraph.png" border="0" usemap="#class_environment_a4003e6947de2059bc6ba76af273cf5f7_icgraph" alt=""/></div>
<map name="class_environment_a4003e6947de2059bc6ba76af273cf5f7_icgraph" id="class_environment_a4003e6947de2059bc6ba76af273cf5f7_icgraph">
<area shape="rect" id="node2" href="class_environment.html#aeff4f165bb55ef7abe27662598bb82f2" title="Core method. Sets the number of threads used later for computing the model. " alt="" coords="244,5,420,32"/>
</map>
</div>
</p>

</div>
</div>
<a class="anchor" id="aa3e7b74e17329b650089507642e76e5f"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void Environment::selectAndReprodHostsNoMating </td>
          <td>(</td>
          <td class="paramname"></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Core method. Forms the next generation of hosts using the fitness proportionate selection method. Replaces the old population with a new one. </p>
<p>Iterates through the host population selecting the next generation for reproduction using <a href="http://en.wikipedia.org/wiki/Fitness_proportionate_selection">fitness proportionate selection method</a> (also known as the roulette wheel selection). Successful individuals are simply cloned replacing the weak ones. </p>

<p>Definition at line <a class="el" href="_environment_8cpp_source.html#l00497">497</a> of file <a class="el" href="_environment_8cpp_source.html">Environment.cpp</a>.</p>

<p><div class="dynheader">
Here is the call graph for this function:</div>
<div class="dyncontent">
<div class="center"><img src="class_environment_aa3e7b74e17329b650089507642e76e5f_cgraph.png" border="0" usemap="#class_environment_aa3e7b74e17329b650089507642e76e5f_cgraph" alt=""/></div>
<map name="class_environment_aa3e7b74e17329b650089507642e76e5f_cgraph" id="class_environment_aa3e7b74e17329b650089507642e76e5f_cgraph">
<area shape="rect" id="node2" href="class_random.html#aad6bba324bd040a1b425026e3dd7b019" title="Random::getRealDouble" alt="" coords="256,13,421,39"/>
</map>
</div>
</p>

</div>
</div>
<a class="anchor" id="a09af2890db9eb1a02069adfb8836aa47"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void Environment::selectAndReprodHostsReplace </td>
          <td>(</td>
          <td class="paramname"></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Core method. Forms the next generation of hosts using the fitness proportionate selection method. Replaces the old population with a new one. </p>
<p>Iterates through the host population selecting the next generation for reproduction using <a href="http://en.wikipedia.org/wiki/Fitness_proportionate_selection">fitness proportionate selection method</a> (also known as the roulette wheel selection). Simulates random mating of hermaphrodites with no difference between sexes. </p>

<p>Definition at line <a class="el" href="_environment_8cpp_source.html#l00424">424</a> of file <a class="el" href="_environment_8cpp_source.html">Environment.cpp</a>.</p>

<p><div class="dynheader">
Here is the call graph for this function:</div>
<div class="dyncontent">
<div class="center"><img src="class_environment_a09af2890db9eb1a02069adfb8836aa47_cgraph.png" border="0" usemap="#class_environment_a09af2890db9eb1a02069adfb8836aa47_cgraph" alt=""/></div>
<map name="class_environment_a09af2890db9eb1a02069adfb8836aa47_cgraph" id="class_environment_a09af2890db9eb1a02069adfb8836aa47_cgraph">
<area shape="rect" id="node2" href="class_random.html#aad6bba324bd040a1b425026e3dd7b019" title="Random::getRealDouble" alt="" coords="256,13,421,39"/>
</map>
</div>
</p>

</div>
</div>
<a class="anchor" id="a90177792fb03087e1fc64dca78f7e312"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void Environment::selectAndReproducePathoFixedPopSizes </td>
          <td>(</td>
          <td class="paramname"></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Core method. Forms the next generation of hosts using the fitness proportionate selection method. It keeps population sizes of different pathogens species at a fixed number. </p>
<p>Iterates through the pathogen population selecting the next generation for reproduction using fitness proportionate selection method (roulette wheel selection). </p>

<p>Definition at line <a class="el" href="_environment_8cpp_source.html#l00549">549</a> of file <a class="el" href="_environment_8cpp_source.html">Environment.cpp</a>.</p>

<p><div class="dynheader">
Here is the call graph for this function:</div>
<div class="dyncontent">
<div class="center"><img src="class_environment_a90177792fb03087e1fc64dca78f7e312_cgraph.png" border="0" usemap="#class_environment_a90177792fb03087e1fc64dca78f7e312_cgraph" alt=""/></div>
<map name="class_environment_a90177792fb03087e1fc64dca78f7e312_cgraph" id="class_environment_a90177792fb03087e1fc64dca78f7e312_cgraph">
<area shape="rect" id="node2" href="class_random.html#aff6c9a321cc83da8ae04f003fdfdd169" title="Returns a random unsigned int value in [min, max] from uniform distribution. " alt="" coords="277,13,500,39"/>
</map>
</div>
</p>

</div>
</div>
<a class="anchor" id="a28cc07198ee462d0708ac4e6d43b062f"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void Environment::setHostClonalPopulation </td>
          <td>(</td>
          <td class="paramtype">int&#160;</td>
          <td class="paramname"><em>pop_size</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">unsigned long&#160;</td>
          <td class="paramname"><em>gene_size</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">unsigned long&#160;</td>
          <td class="paramname"><em>chrom_size</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">int&#160;</td>
          <td class="paramname"><em>timeStamp</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype"><a class="el" href="class_tagging__system.html">Tagging_system</a> &amp;&#160;</td>
          <td class="paramname"><em>tag</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Core method. Initializes a vector containing clonal host population. </p>
<p>Sets and fills a vector containing the host population consisting of a single homozygous clone. Parameters like size of the population, length of genes represented by bit-strings, number of genes in a chromosome (remember that there are two chromosomes) are user-defined, the rest is set at random (e.g. actual gene values).</p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">pop_size</td><td>- number of host individuals in a simulation </td></tr>
    <tr><td class="paramname">gene_size</td><td>- number of bits in bit-represented genes </td></tr>
    <tr><td class="paramname">chrom_size</td><td>- number of genes in a chromosome </td></tr>
    <tr><td class="paramname">timeStamp</td><td>- current time (number of the model iteration) </td></tr>
    <tr><td class="paramname">tag</td><td>- pointer to the tagging system marking each gene variant </td></tr>
  </table>
  </dd>
</dl>

<p>Definition at line <a class="el" href="_environment_8cpp_source.html#l00176">176</a> of file <a class="el" href="_environment_8cpp_source.html">Environment.cpp</a>.</p>

</div>
</div>
<a class="anchor" id="aa4bcb97295aa866eebaf01d6de58b471"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void Environment::setHostRandomPopulation </td>
          <td>(</td>
          <td class="paramtype">int&#160;</td>
          <td class="paramname"><em>pop_size</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">unsigned long&#160;</td>
          <td class="paramname"><em>gene_size</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">unsigned long&#160;</td>
          <td class="paramname"><em>chrom_size</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">int&#160;</td>
          <td class="paramname"><em>timeStamp</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype"><a class="el" href="class_tagging__system.html">Tagging_system</a> &amp;&#160;</td>
          <td class="paramname"><em>tag</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Core method. Initializes a vector containing host population. </p>
<p>Sets and fills a vector containing the host population. Parameters like size of the population, length of genes represented by bit-strings, number of genes in a chromosome (remember that there are two chromosomes) are user-defined, the rest is set at random (e.g. actual gene values).</p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">pop_size</td><td>- number of host individuals in a simulation </td></tr>
    <tr><td class="paramname">gene_size</td><td>- number of bits in bit-represented genes </td></tr>
    <tr><td class="paramname">chrom_size</td><td>- number of genes in a chromosome </td></tr>
    <tr><td class="paramname">timeStamp</td><td>- current time (number of the model iteration) </td></tr>
    <tr><td class="paramname">tag</td><td>- pointer to the tagging system marking each gene variant </td></tr>
  </table>
  </dd>
</dl>

<p>Definition at line <a class="el" href="_environment_8cpp_source.html#l00111">111</a> of file <a class="el" href="_environment_8cpp_source.html">Environment.cpp</a>.</p>

</div>
</div>
<a class="anchor" id="ada465b8ad37b2688ab411fc8ab07f3d7"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void Environment::setHostRandomPopulation </td>
          <td>(</td>
          <td class="paramtype">int&#160;</td>
          <td class="paramname"><em>pop_size</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">unsigned long&#160;</td>
          <td class="paramname"><em>gene_size</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">unsigned long&#160;</td>
          <td class="paramname"><em>chrom_size_lower</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">unsigned long&#160;</td>
          <td class="paramname"><em>chrom_size_uper</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">int&#160;</td>
          <td class="paramname"><em>timeStamp</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype"><a class="el" href="class_tagging__system.html">Tagging_system</a> &amp;&#160;</td>
          <td class="paramname"><em>tag</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Core method. Initializes a vector containing host population. </p>
<p>Sets and fills a vector containing the host population. Parameters like size of the population and the length of genes represented by bit-strings are user-defined. Whereas the range of the number of genes in a chromosome (remember that there are two chromosomes) is selected randomly from a set range of possible values, the rest of parameters is set totally at random (e.g. actual gene values).</p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">pop_size</td><td>- number of host individuals in a simulation </td></tr>
    <tr><td class="paramname">gene_size</td><td>- number of bits in bit-represented genes </td></tr>
    <tr><td class="paramname">chrom_size_lower</td><td>- lower limit of the number of genes in a chromosome </td></tr>
    <tr><td class="paramname">chrom_size_uper</td><td>- upper limit of the number of genes in a chromosome </td></tr>
    <tr><td class="paramname">timeStamp</td><td>- current time (number of the model iteration) </td></tr>
    <tr><td class="paramname">tag</td><td>- pointer to the tagging system marking each gene variant </td></tr>
  </table>
  </dd>
</dl>

<p>Definition at line <a class="el" href="_environment_8cpp_source.html#l00141">141</a> of file <a class="el" href="_environment_8cpp_source.html">Environment.cpp</a>.</p>

</div>
</div>
<a class="anchor" id="a54b80db1bda904983fda2c38ae1c5dab"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void Environment::setNoMutsVector </td>
          <td>(</td>
          <td class="paramtype">int&#160;</td>
          <td class="paramname"><em>numb_of_species</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">unsigned long&#160;</td>
          <td class="paramname"><em>antigen_size</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">double&#160;</td>
          <td class="paramname"><em>fixedAntigenFrac</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Core method. It defines "no mutation sites" of the antigen for all individual pathogen species in the simulation. It should be run only ones per simulation. </p>
<p>Creates a vector of <a href="http://www.cplusplus.com/reference/set/set/">STD sets</a> containing indices of antigen bits in which mutations are not allowed. The length of the vector equals the number of pathogen species. Each species has its own vector. Number of fixed sites is a user-defined parameter.</p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">numb_of_species</td><td>- number of pathogen species </td></tr>
    <tr><td class="paramname">antigen_size</td><td>- number of bits per antigen </td></tr>
    <tr><td class="paramname">fixedAntigenFrac</td><td>- fraction of bits in antigens which need to be fixed </td></tr>
  </table>
  </dd>
</dl>

<p>Definition at line <a class="el" href="_environment_8cpp_source.html#l00079">79</a> of file <a class="el" href="_environment_8cpp_source.html">Environment.cpp</a>.</p>

<p><div class="dynheader">
Here is the caller graph for this function:</div>
<div class="dyncontent">
<div class="center"><img src="class_environment_a54b80db1bda904983fda2c38ae1c5dab_icgraph.png" border="0" usemap="#class_environment_a54b80db1bda904983fda2c38ae1c5dab_icgraph" alt=""/></div>
<map name="class_environment_a54b80db1bda904983fda2c38ae1c5dab_icgraph" id="class_environment_a54b80db1bda904983fda2c38ae1c5dab_icgraph">
<area shape="rect" id="node2" href="class_environment.html#adcd7c4b6fb8b8181be1f2099a1c7559f" title="Core method. Initializes the pathogen population. " alt="" coords="259,5,468,47"/>
</map>
</div>
</p>

</div>
</div>
<a class="anchor" id="adcd7c4b6fb8b8181be1f2099a1c7559f"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void Environment::setPathoPopulatioDivSpecies </td>
          <td>(</td>
          <td class="paramtype">int&#160;</td>
          <td class="paramname"><em>pop_size</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">unsigned long&#160;</td>
          <td class="paramname"><em>antigenSize</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">int&#160;</td>
          <td class="paramname"><em>numb_of_species</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">unsigned long&#160;</td>
          <td class="paramname"><em>mhcSize</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">int&#160;</td>
          <td class="paramname"><em>timeStamp</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">double&#160;</td>
          <td class="paramname"><em>fixedAntigenFrac</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype"><a class="el" href="class_tagging__system.html">Tagging_system</a> &amp;&#160;</td>
          <td class="paramname"><em>tag</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Core method. Initializes the pathogen population. </p>
<p>Given the number of individuals, number of bit per gene, desired number of genes in a genome and desired number of pathogen species it generates random population of pathogens. Number of individuals will be evenly distributed between species and each species draws its genes from the same pool of possible bit strings.</p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">pop_size</td><td>- total number of individuals </td></tr>
    <tr><td class="paramname">antigenSize</td><td>- number of bits per gene </td></tr>
    <tr><td class="paramname">chrom_size</td><td>- number of genes per genome </td></tr>
    <tr><td class="paramname">numb_of_species</td><td>- number of pathogen species </td></tr>
    <tr><td class="paramname">mhcSize</td><td>- number of bits in MHC protein </td></tr>
    <tr><td class="paramname">timeStamp</td><td>- current time (number of the model iteration) </td></tr>
    <tr><td class="paramname">fixedAntigenFrac</td><td>- fraction of bits in antigens which need to be fixed </td></tr>
    <tr><td class="paramname">tag</td><td>- pointer to the tagging system marking each gene variant</td></tr>
  </table>
  </dd>
</dl>
<p>void <a class="el" href="class_environment.html#a5e09ffa9aa8d501096f19131571bace1">Environment::setPathoPopulatioUniformGenome</a>(int pop_size, unsigned long antigenSize, int chrom_size, int numb_of_species, unsigned long mhcSize, int timeStamp, double fixedAntigenFrac, <a class="el" href="class_tagging__system.html" title="A little handy machine which generates unique tags for genes. These tags are used when keeping the tr...">Tagging_system</a> &amp;tag){ Environment::setNoMutsVector(numb_of_species, antigenSize, fixedAntigenFrac, rngGenPtr[omp_get_thread_num()]); if (numb_of_species &gt; pop_size) numb_of_species = pop_size; int indiv_per_species = pop_size / numb_of_species; int indiv_left = pop_size % numb_of_species; std::vector&lt;Pathogen&gt; OneSpeciesVector; for (int i = 0; i &lt; numb_of_species; ++i){ for(int j = 0; j &lt; indiv_per_species; ++j){ OneSpeciesVector.push_back(Pathogen()); if(i+1 != indiv_per_species){ OneSpeciesVector.back().setNewPathogen(chrom_size, antigenSize, mhcSize, i, timeStamp, randGen, tag); } else { OneSpeciesVector.back().setNewPathogen(chrom_size, antigenSize, mhcSize, i, timeStamp, randGen, tag); } if(indiv_left){ OneSpeciesVector.push_back(Pathogen()); OneSpeciesVector.back().setNewPathogen(chrom_size, antigenSize, mhcSize, i, timeStamp, randGen, tag); indiv_left&ndash;; } } PathPopulation.push_back(OneSpeciesVector); OneSpeciesVector.clear(); } } Core method. Initializes the pathogen population.</p>
<p>Given the number of individuals, number of bit per gene, desired number of genes in a genome and desired number of pathogen species it generates random population of pathogens. Number of individuals will be evenly distributed between species and each species consists of identical clones.</p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">pop_size</td><td>- total number of individuals </td></tr>
    <tr><td class="paramname">antigenSize</td><td>- number of bits per gene </td></tr>
    <tr><td class="paramname">chrom_size</td><td>- number of genes per genome </td></tr>
    <tr><td class="paramname">numb_of_species</td><td>- number of pathogen species </td></tr>
    <tr><td class="paramname">mhcSize</td><td>- number of bits in MHC protein </td></tr>
    <tr><td class="paramname">timeStamp</td><td>- current time (number of the model iteration) </td></tr>
    <tr><td class="paramname">fixedAntigenFrac</td><td>- fraction of bits in antigens which need to be fixed </td></tr>
    <tr><td class="paramname">tag</td><td>- pointer to the tagging system marking each gene variant </td></tr>
  </table>
  </dd>
</dl>

<p>Definition at line <a class="el" href="_environment_8cpp_source.html#l00258">258</a> of file <a class="el" href="_environment_8cpp_source.html">Environment.cpp</a>.</p>

<p><div class="dynheader">
Here is the call graph for this function:</div>
<div class="dyncontent">
<div class="center"><img src="class_environment_adcd7c4b6fb8b8181be1f2099a1c7559f_cgraph.png" border="0" usemap="#class_environment_adcd7c4b6fb8b8181be1f2099a1c7559f_cgraph" alt=""/></div>
<map name="class_environment_adcd7c4b6fb8b8181be1f2099a1c7559f_cgraph" id="class_environment_adcd7c4b6fb8b8181be1f2099a1c7559f_cgraph">
<area shape="rect" id="node2" href="class_environment.html#a54b80db1bda904983fda2c38ae1c5dab" title="Core method. It defines &quot;no mutation sites&quot; of the antigen for all individual pathogen species in the..." alt="" coords="263,13,468,39"/>
</map>
</div>
</p>

</div>
</div>
<a class="anchor" id="a5e09ffa9aa8d501096f19131571bace1"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void Environment::setPathoPopulatioUniformGenome </td>
          <td>(</td>
          <td class="paramtype">int&#160;</td>
          <td class="paramname"><em>pop_size</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">unsigned long&#160;</td>
          <td class="paramname"><em>gene_size</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">int&#160;</td>
          <td class="paramname"><em>chrom_size</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">int&#160;</td>
          <td class="paramname"><em>numb_of_species</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">unsigned long&#160;</td>
          <td class="paramname"><em>mhcSize</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">int&#160;</td>
          <td class="paramname"><em>timeStamp</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">double&#160;</td>
          <td class="paramname"><em>fixedAntigenFrac</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype"><a class="el" href="class_tagging__system.html">Tagging_system</a> &amp;&#160;</td>
          <td class="paramname"><em>tag</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

</div>
</div>
<hr/>The documentation for this class was generated from the following files:<ul>
<li>src/<a class="el" href="_environment_8h_source.html">Environment.h</a></li>
<li>src/<a class="el" href="_environment_8cpp_source.html">Environment.cpp</a></li>
</ul>
</div><!-- contents -->
</div><!-- doc-content -->
<!-- start footer part -->
<div id="nav-path" class="navpath"><!-- id is needed for treeview function! -->
  <ul>
    <li class="navelem"><a class="el" href="class_environment.html">Environment</a></li>
    <li class="footer">Generated on Mon Apr 1 2019 15:29:37 for Modelling MHC Evolutionary Dynamics by
    <a href="http://www.doxygen.org/index.html">
    <img class="footer" src="doxygen.png" alt="doxygen"/></a> 1.8.11 </li>
  </ul>
</div>
</body>
</html>
