<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml">
<head>
<meta http-equiv="Content-Type" content="text/xhtml;charset=UTF-8"/>
<meta http-equiv="X-UA-Compatible" content="IE=9"/>
<meta name="generator" content="Doxygen 1.8.7"/>
<title>Modelling MHC Evolutionary Dynamics: Host Class Reference</title>
<link href="tabs.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="jquery.js"></script>
<script type="text/javascript" src="dynsections.js"></script>
<link href="navtree.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="resize.js"></script>
<script type="text/javascript" src="navtree.js"></script>
<script type="text/javascript">
  $(document).ready(initResizable);
  $(window).load(resizeHeight);
</script>
<link href="search/search.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="search/search.js"></script>
<script type="text/javascript">
  $(document).ready(function() { searchBox.OnSelectItem(0); });
</script>
<link href="doxygen.css" rel="stylesheet" type="text/css" />
</head>
<body>
<div id="top"><!-- do not remove this div, it is closed by doxygen! -->
<div id="titlearea">
<table cellspacing="0" cellpadding="0">
 <tbody>
 <tr style="height: 56px;">
  <td style="padding-left: 0.5em;">
   <div id="projectname">Modelling MHC Evolutionary Dynamics
   &#160;<span id="projectnumber">0.9</span>
   </div>
  </td>
 </tr>
 </tbody>
</table>
</div>
<!-- end header part -->
<!-- Generated by Doxygen 1.8.7 -->
<script type="text/javascript">
var searchBox = new SearchBox("searchBox", "search",false,'Search');
</script>
  <div id="navrow1" class="tabs">
    <ul class="tablist">
      <li><a href="index.html"><span>Main&#160;Page</span></a></li>
      <li><a href="pages.html"><span>Related&#160;Pages</span></a></li>
      <li class="current"><a href="annotated.html"><span>Classes</span></a></li>
      <li><a href="files.html"><span>Files</span></a></li>
      <li>
        <div id="MSearchBox" class="MSearchBoxInactive">
        <span class="left">
          <img id="MSearchSelect" src="search/mag_sel.png"
               onmouseover="return searchBox.OnSearchSelectShow()"
               onmouseout="return searchBox.OnSearchSelectHide()"
               alt=""/>
          <input type="text" id="MSearchField" value="Search" accesskey="S"
               onfocus="searchBox.OnSearchFieldFocus(true)" 
               onblur="searchBox.OnSearchFieldFocus(false)" 
               onkeyup="searchBox.OnSearchFieldChange(event)"/>
          </span><span class="right">
            <a id="MSearchClose" href="javascript:searchBox.CloseResultsWindow()"><img id="MSearchCloseImg" border="0" src="search/close.png" alt=""/></a>
          </span>
        </div>
      </li>
    </ul>
  </div>
  <div id="navrow2" class="tabs2">
    <ul class="tablist">
      <li><a href="annotated.html"><span>Class&#160;List</span></a></li>
      <li><a href="classes.html"><span>Class&#160;Index</span></a></li>
      <li><a href="functions.html"><span>Class&#160;Members</span></a></li>
    </ul>
  </div>
</div><!-- top -->
<div id="side-nav" class="ui-resizable side-nav-resizable">
  <div id="nav-tree">
    <div id="nav-tree-contents">
      <div id="nav-sync" class="sync"></div>
    </div>
  </div>
  <div id="splitbar" style="-moz-user-select:none;" 
       class="ui-resizable-handle">
  </div>
</div>
<script type="text/javascript">
$(document).ready(function(){initNavTree('class_host.html','');});
</script>
<div id="doc-content">
<!-- window showing the filter options -->
<div id="MSearchSelectWindow"
     onmouseover="return searchBox.OnSearchSelectShow()"
     onmouseout="return searchBox.OnSearchSelectHide()"
     onkeydown="return searchBox.OnSearchSelectKey(event)">
<a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(0)"><span class="SelectionMark">&#160;</span>All</a><a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(1)"><span class="SelectionMark">&#160;</span>Classes</a><a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(2)"><span class="SelectionMark">&#160;</span>Files</a><a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(3)"><span class="SelectionMark">&#160;</span>Functions</a><a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(4)"><span class="SelectionMark">&#160;</span>Variables</a><a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(5)"><span class="SelectionMark">&#160;</span>Typedefs</a><a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(6)"><span class="SelectionMark">&#160;</span>Pages</a></div>

<!-- iframe showing the search results (closed by default) -->
<div id="MSearchResultsWindow">
<iframe src="javascript:void(0)" frameborder="0" 
        name="MSearchResults" id="MSearchResults">
</iframe>
</div>

<div class="header">
  <div class="summary">
<a href="#pub-methods">Public Member Functions</a> &#124;
<a href="#pub-attribs">Public Attributes</a> &#124;
<a href="class_host-members.html">List of all members</a>  </div>
  <div class="headertitle">
<div class="title">Host Class Reference</div>  </div>
</div><!--header-->
<div class="contents">

<p>Core class. Stores and handles a single host object. Each host has multiple instances of <a class="el" href="class_gene.html" title="Core class. Stores and handles a single gene object. Has methods to access and to mutate a gene...">Gene</a> class objects and stores them in 2 vectors called ChromosomeOne and ChromosomeTwo which simulate a diploid genome.  
 <a href="class_host.html#details">More...</a></p>

<p><code>#include &lt;<a class="el" href="_host_8h_source.html">Host.h</a>&gt;</code></p>
<table class="memberdecls">
<tr class="heading"><td colspan="2"><h2 class="groupheader"><a name="pub-methods"></a>
Public Member Functions</h2></td></tr>
<tr class="memitem:a369c12e7425c030528bab0784da40e4b"><td class="memItemLeft" align="right" valign="top">&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_host.html#a369c12e7425c030528bab0784da40e4b">Host</a> ()</td></tr>
<tr class="separator:a369c12e7425c030528bab0784da40e4b"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a6c14f16c39ae6bc5db2672e0141adccd"><td class="memItemLeft" align="right" valign="top">virtual&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_host.html#a6c14f16c39ae6bc5db2672e0141adccd">~Host</a> ()</td></tr>
<tr class="separator:a6c14f16c39ae6bc5db2672e0141adccd"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:aafa1539a8942e3e99dc6b49110a618c1"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_host.html#aafa1539a8942e3e99dc6b49110a618c1">setNewHost</a> (int num_of_loci, int gene_size, int timeStamp)</td></tr>
<tr class="memdesc:aafa1539a8942e3e99dc6b49110a618c1"><td class="mdescLeft">&#160;</td><td class="mdescRight">Core method. Sets a new <a class="el" href="class_host.html" title="Core class. Stores and handles a single host object. Each host has multiple instances of Gene class o...">Host</a> object and assigns a user-defined number of genes of user-defined number of bits (as genes are bit-strings).  <a href="#aafa1539a8942e3e99dc6b49110a618c1">More...</a><br /></td></tr>
<tr class="separator:aafa1539a8942e3e99dc6b49110a618c1"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:acea27ac91fb04a2d711d5af25200c557"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_host.html#acea27ac91fb04a2d711d5af25200c557">setNewHomozygHost</a> (int num_of_loci, int gene_size, int timeStamp)</td></tr>
<tr class="memdesc:acea27ac91fb04a2d711d5af25200c557"><td class="mdescLeft">&#160;</td><td class="mdescRight">Core method. Sets a new <a class="el" href="class_host.html" title="Core class. Stores and handles a single host object. Each host has multiple instances of Gene class o...">Host</a> object and assigns a user-defined number of genes of user-defined number of bits (as genes are bit-strings).  <a href="#acea27ac91fb04a2d711d5af25200c557">More...</a><br /></td></tr>
<tr class="separator:acea27ac91fb04a2d711d5af25200c557"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a5911792258205cae25a2a4796061fe49"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_host.html#a5911792258205cae25a2a4796061fe49">chromoMutProcess</a> (double mut_probabl, int timeStamp)</td></tr>
<tr class="memdesc:a5911792258205cae25a2a4796061fe49"><td class="mdescLeft">&#160;</td><td class="mdescRight">Core method. Decides (on a random basis) if there will be any mutations in the genome.  <a href="#a5911792258205cae25a2a4796061fe49">More...</a><br /></td></tr>
<tr class="separator:a5911792258205cae25a2a4796061fe49"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a48c01d4345818b873ceed7597588bd6d"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_host.html#a48c01d4345818b873ceed7597588bd6d">chromoMutProcessWithDelDupl</a> (double mut_probabl, double del, double dupli, int maxGene, int timeStamp)</td></tr>
<tr class="memdesc:a48c01d4345818b873ceed7597588bd6d"><td class="mdescLeft">&#160;</td><td class="mdescRight">Core method. Decides (on a random basis) if there will be any mutations in the genome: deletion, duplication or change of a gene.  <a href="#a48c01d4345818b873ceed7597588bd6d">More...</a><br /></td></tr>
<tr class="separator:a48c01d4345818b873ceed7597588bd6d"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a303c0d61ab48aaa709fccfcdd14a156e"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_host.html#a303c0d61ab48aaa709fccfcdd14a156e">chromoMutProcessWithDelDuplPointMuts</a> (double mut_probabl, double del, double dupli, int maxGene, int timeStamp)</td></tr>
<tr class="memdesc:a303c0d61ab48aaa709fccfcdd14a156e"><td class="mdescLeft">&#160;</td><td class="mdescRight">Core method. Decides (on a random basis) if there will be any mutations in the genome.  <a href="#a303c0d61ab48aaa709fccfcdd14a156e">More...</a><br /></td></tr>
<tr class="separator:a303c0d61ab48aaa709fccfcdd14a156e"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ada5564c9d57e18ce046e131a5968e08f"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_host.html#ada5564c9d57e18ce046e131a5968e08f">chromoRecombination</a> (double recomb_prob, int timeStamp)</td></tr>
<tr class="memdesc:ada5564c9d57e18ce046e131a5968e08f"><td class="mdescLeft">&#160;</td><td class="mdescRight">Core method. Performs recombination between host's chromosomes by inserting a brake in randomly selected genes on Chromosome One and Two and swapping the resulting parts of the chromosomes.  <a href="#ada5564c9d57e18ce046e131a5968e08f">More...</a><br /></td></tr>
<tr class="separator:ada5564c9d57e18ce046e131a5968e08f"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a7c21e1938c61966dc2657a437aeec9b9"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_host.html#a7c21e1938c61966dc2657a437aeec9b9">clearInfections</a> ()</td></tr>
<tr class="memdesc:a7c21e1938c61966dc2657a437aeec9b9"><td class="mdescLeft">&#160;</td><td class="mdescRight">Core method. Clears data regarding infections and fitness.  <a href="#a7c21e1938c61966dc2657a437aeec9b9">More...</a><br /></td></tr>
<tr class="separator:a7c21e1938c61966dc2657a437aeec9b9"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a505527db337d87670e5ab37eb9818fa3"><td class="memItemLeft" align="right" valign="top"><a class="el" href="_h2_pinteraction_8cpp.html#adcfab8f2ed9ed0c694cb95d35c91aa78">chromovector</a>&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_host.html#a505527db337d87670e5ab37eb9818fa3">doCrossAndMeiosis</a> (double corssing_prob)</td></tr>
<tr class="memdesc:a505527db337d87670e5ab37eb9818fa3"><td class="mdescLeft">&#160;</td><td class="mdescRight">Core method. Performing crossing over and chromosome selection before mating.  <a href="#a505527db337d87670e5ab37eb9818fa3">More...</a><br /></td></tr>
<tr class="separator:a505527db337d87670e5ab37eb9818fa3"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ac387176ba203f6aa1f20610f5679385a"><td class="memItemLeft" align="right" valign="top"><a class="el" href="_h2_pinteraction_8cpp.html#adcfab8f2ed9ed0c694cb95d35c91aa78">chromovector</a>&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_host.html#ac387176ba203f6aa1f20610f5679385a">getChromosomeOne</a> ()</td></tr>
<tr class="memdesc:ac387176ba203f6aa1f20610f5679385a"><td class="mdescLeft">&#160;</td><td class="mdescRight">Core method. Returns the first chromosome.  <a href="#ac387176ba203f6aa1f20610f5679385a">More...</a><br /></td></tr>
<tr class="separator:ac387176ba203f6aa1f20610f5679385a"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:aa5afa38408ebba644f56667364213f0d"><td class="memItemLeft" align="right" valign="top"><a class="el" href="_h2_pinteraction_8cpp.html#adcfab8f2ed9ed0c694cb95d35c91aa78">chromovector</a>&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_host.html#aa5afa38408ebba644f56667364213f0d">getChromosomeTwo</a> ()</td></tr>
<tr class="memdesc:aa5afa38408ebba644f56667364213f0d"><td class="mdescLeft">&#160;</td><td class="mdescRight">Core method. Returns the first chromosome.  <a href="#aa5afa38408ebba644f56667364213f0d">More...</a><br /></td></tr>
<tr class="separator:aa5afa38408ebba644f56667364213f0d"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a55edfefc42f7306c85983474500ef10b"><td class="memItemLeft" align="right" valign="top"><a class="el" href="_h2_pinteraction_8cpp.html#adcfab8f2ed9ed0c694cb95d35c91aa78">chromovector</a>&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_host.html#a55edfefc42f7306c85983474500ef10b">mergeChromosomes</a> ()</td></tr>
<tr class="memdesc:a55edfefc42f7306c85983474500ef10b"><td class="mdescLeft">&#160;</td><td class="mdescRight">Data harvesting method. Merges both chromosomes to create an object easier to handle is some situations.  <a href="#a55edfefc42f7306c85983474500ef10b">More...</a><br /></td></tr>
<tr class="separator:a55edfefc42f7306c85983474500ef10b"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a10ec4eb6494084cbc99a4c28164b0761"><td class="memItemLeft" align="right" valign="top">unsigned&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_host.html#a10ec4eb6494084cbc99a4c28164b0761">getGenomeSize</a> ()</td></tr>
<tr class="memdesc:a10ec4eb6494084cbc99a4c28164b0761"><td class="mdescLeft">&#160;</td><td class="mdescRight">Returns the total number of genes in the genome.  <a href="#a10ec4eb6494084cbc99a4c28164b0761">More...</a><br /></td></tr>
<tr class="separator:a10ec4eb6494084cbc99a4c28164b0761"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:aec728ae16d7c1186739730510a2c7e2e"><td class="memItemLeft" align="right" valign="top">unsigned&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_host.html#aec728ae16d7c1186739730510a2c7e2e">getChromoOneSize</a> ()</td></tr>
<tr class="memdesc:aec728ae16d7c1186739730510a2c7e2e"><td class="mdescLeft">&#160;</td><td class="mdescRight">Returns the number of genes in the Chromosome One.  <a href="#aec728ae16d7c1186739730510a2c7e2e">More...</a><br /></td></tr>
<tr class="separator:aec728ae16d7c1186739730510a2c7e2e"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a03e175acb665dc22d96bb3a038571af7"><td class="memItemLeft" align="right" valign="top">unsigned&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_host.html#a03e175acb665dc22d96bb3a038571af7">getChromoTwoSize</a> ()</td></tr>
<tr class="memdesc:a03e175acb665dc22d96bb3a038571af7"><td class="mdescLeft">&#160;</td><td class="mdescRight">Returns the number of genes in the Chromosome Two.  <a href="#a03e175acb665dc22d96bb3a038571af7">More...</a><br /></td></tr>
<tr class="separator:a03e175acb665dc22d96bb3a038571af7"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ad84a6fab9904782aa633deab0ad5e506"><td class="memItemLeft" align="right" valign="top">double&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_host.html#ad84a6fab9904782aa633deab0ad5e506">getChromoOneUniqAlleles</a> ()</td></tr>
<tr class="memdesc:ad84a6fab9904782aa633deab0ad5e506"><td class="mdescLeft">&#160;</td><td class="mdescRight">Returns the number of unique MHC alleles in the Chromosome One.  <a href="#ad84a6fab9904782aa633deab0ad5e506">More...</a><br /></td></tr>
<tr class="separator:ad84a6fab9904782aa633deab0ad5e506"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ad05b417630dd1095b967de411bdb350a"><td class="memItemLeft" align="right" valign="top">double&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_host.html#ad05b417630dd1095b967de411bdb350a">getChromoTwoUniqAlleles</a> ()</td></tr>
<tr class="memdesc:ad05b417630dd1095b967de411bdb350a"><td class="mdescLeft">&#160;</td><td class="mdescRight">Returns the number of unique MHC alleles in the Chromosome Two.  <a href="#ad05b417630dd1095b967de411bdb350a">More...</a><br /></td></tr>
<tr class="separator:ad05b417630dd1095b967de411bdb350a"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ac6ecd452f531bbcaf469d5d9d42b8028"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_host.html#ac6ecd452f531bbcaf469d5d9d42b8028">assignChromOne</a> (<a class="el" href="_h2_pinteraction_8cpp.html#adcfab8f2ed9ed0c694cb95d35c91aa78">chromovector</a> One)</td></tr>
<tr class="memdesc:ac6ecd452f531bbcaf469d5d9d42b8028"><td class="mdescLeft">&#160;</td><td class="mdescRight">Core method. Assigns a new chromosome to host's Chromosome ONE.  <a href="#ac6ecd452f531bbcaf469d5d9d42b8028">More...</a><br /></td></tr>
<tr class="separator:ac6ecd452f531bbcaf469d5d9d42b8028"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ab3e04fdf569863ddeee00d6d2f5c990d"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_host.html#ab3e04fdf569863ddeee00d6d2f5c990d">assignChromTwo</a> (<a class="el" href="_h2_pinteraction_8cpp.html#adcfab8f2ed9ed0c694cb95d35c91aa78">chromovector</a> Two)</td></tr>
<tr class="memdesc:ab3e04fdf569863ddeee00d6d2f5c990d"><td class="mdescLeft">&#160;</td><td class="mdescRight">Core method. Assigns a new chromosome to host's Chromosome TWO.  <a href="#ab3e04fdf569863ddeee00d6d2f5c990d">More...</a><br /></td></tr>
<tr class="separator:ab3e04fdf569863ddeee00d6d2f5c990d"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a8b8d1bb28ee49fcd337a0a5656595497"><td class="memItemLeft" align="right" valign="top"><a class="el" href="_gene_8cpp.html#a108efa7670466d15a5ecafb60bbc9c37">genestring</a>&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_host.html#a8b8d1bb28ee49fcd337a0a5656595497">getSingleGeneFromOne</a> (int indx)</td></tr>
<tr class="memdesc:a8b8d1bb28ee49fcd337a0a5656595497"><td class="mdescLeft">&#160;</td><td class="mdescRight">Core method. Fetches a gene (bit string) with a given position on the chromosome One.  <a href="#a8b8d1bb28ee49fcd337a0a5656595497">More...</a><br /></td></tr>
<tr class="separator:a8b8d1bb28ee49fcd337a0a5656595497"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a8e57071a6138a8a975dba81be1d39e44"><td class="memItemLeft" align="right" valign="top"><a class="el" href="_gene_8cpp.html#a108efa7670466d15a5ecafb60bbc9c37">genestring</a>&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_host.html#a8e57071a6138a8a975dba81be1d39e44">getSingleGeneFromTwo</a> (int indx)</td></tr>
<tr class="memdesc:a8e57071a6138a8a975dba81be1d39e44"><td class="mdescLeft">&#160;</td><td class="mdescRight">Core method. Fetches a gene (bit string) with a given position on the chromosome Two.  <a href="#a8e57071a6138a8a975dba81be1d39e44">More...</a><br /></td></tr>
<tr class="separator:a8e57071a6138a8a975dba81be1d39e44"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a88a454efb7a483985eea815e2b285809"><td class="memItemLeft" align="right" valign="top">unsigned long&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_host.html#a88a454efb7a483985eea815e2b285809">getHostIndvTag</a> ()</td></tr>
<tr class="separator:a88a454efb7a483985eea815e2b285809"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a44931bdccc9732456a67db776b5e69c4"><td class="memItemLeft" align="right" valign="top">unsigned long&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_host.html#a44931bdccc9732456a67db776b5e69c4">getHostMotherTag</a> ()</td></tr>
<tr class="separator:a44931bdccc9732456a67db776b5e69c4"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a9c86ac277ceed9299afccff49c954837"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_host.html#a9c86ac277ceed9299afccff49c954837">setHostIndvTag</a> (unsigned long theTag)</td></tr>
<tr class="separator:a9c86ac277ceed9299afccff49c954837"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a8829231a3659a4b206734b4fa14341f0"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_host.html#a8829231a3659a4b206734b4fa14341f0">setHostMotherTag</a> (unsigned long theTag)</td></tr>
<tr class="separator:a8829231a3659a4b206734b4fa14341f0"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ab30250ea6fdd47221d65914b3e6de4e0"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_host.html#ab30250ea6fdd47221d65914b3e6de4e0">swapChromosomes</a> ()</td></tr>
<tr class="memdesc:ab30250ea6fdd47221d65914b3e6de4e0"><td class="mdescLeft">&#160;</td><td class="mdescRight">Core method. Randomly swaps places of Chromosome One and Chromosome Two to avoid situation when they effectively become two separate populations.  <a href="#ab30250ea6fdd47221d65914b3e6de4e0">More...</a><br /></td></tr>
<tr class="separator:ab30250ea6fdd47221d65914b3e6de4e0"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:aeb8564ae54f8bbf968a62f13c61083df"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_host.html#aeb8564ae54f8bbf968a62f13c61083df">calculateFitnessJustInfection</a> ()</td></tr>
<tr class="memdesc:aeb8564ae54f8bbf968a62f13c61083df"><td class="mdescLeft">&#160;</td><td class="mdescRight">Core method. Calculates host individual fitness simply as the number of exposed pathogens.  <a href="#aeb8564ae54f8bbf968a62f13c61083df">More...</a><br /></td></tr>
<tr class="separator:aeb8564ae54f8bbf968a62f13c61083df"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a6079285b3b8bb5280d52500e6b7fc7ef"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_host.html#a6079285b3b8bb5280d52500e6b7fc7ef">calculateFitnessAccChromSize</a> ()</td></tr>
<tr class="memdesc:a6079285b3b8bb5280d52500e6b7fc7ef"><td class="mdescLeft">&#160;</td><td class="mdescRight">Core method. Calculates host individual fitness as the number of exposed pathogens divided by the genome size.  <a href="#a6079285b3b8bb5280d52500e6b7fc7ef">More...</a><br /></td></tr>
<tr class="separator:a6079285b3b8bb5280d52500e6b7fc7ef"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a614430aea49534f29b63232414cf4fcd"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_host.html#a614430aea49534f29b63232414cf4fcd">calculateFitnessForDrift</a> ()</td></tr>
<tr class="memdesc:a614430aea49534f29b63232414cf4fcd"><td class="mdescLeft">&#160;</td><td class="mdescRight">Core method. Sets fitness as 1, a fixed value to simulate genetic drift.  <a href="#a614430aea49534f29b63232414cf4fcd">More...</a><br /></td></tr>
<tr class="separator:a614430aea49534f29b63232414cf4fcd"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:aca12c9829c01995669e842876478c7fb"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_host.html#aca12c9829c01995669e842876478c7fb">calculateFitnessAlphaXSqr</a> (double alpha)</td></tr>
<tr class="memdesc:aca12c9829c01995669e842876478c7fb"><td class="mdescLeft">&#160;</td><td class="mdescRight">Core method. Calculates host individual fitness in proportion to one over the square of the number of genes scaled to by factor <img class="formulaInl" alt="$ \alpha $" src="form_12.png"/>:  <a href="#aca12c9829c01995669e842876478c7fb">More...</a><br /></td></tr>
<tr class="separator:aca12c9829c01995669e842876478c7fb"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a9134661842de046453acc0342ef74eb6"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_host.html#a9134661842de046453acc0342ef74eb6">calculateFitnessExpFunc</a> (double alpha)</td></tr>
<tr class="memdesc:a9134661842de046453acc0342ef74eb6"><td class="mdescLeft">&#160;</td><td class="mdescRight">Core method. Calculates host individual fitness in proportion to the Gaussian function of the number of genes scaled to by factor <img class="formulaInl" alt="$ \alpha $" src="form_12.png"/>:  <a href="#a9134661842de046453acc0342ef74eb6">More...</a><br /></td></tr>
<tr class="separator:a9134661842de046453acc0342ef74eb6"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ad185a5e02c8ed0930e4d400e9c0e6a9a"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_host.html#ad185a5e02c8ed0930e4d400e9c0e6a9a">calculateFitnessExpFuncUniqAlleles</a> (double alpha)</td></tr>
<tr class="memdesc:ad185a5e02c8ed0930e4d400e9c0e6a9a"><td class="mdescLeft">&#160;</td><td class="mdescRight">Core method. Calculates host individual fitness in proportion to the Gaussian function of the number of unique MHC alleles in the chromosome scaled to by factor <img class="formulaInl" alt="$ \alpha $" src="form_12.png"/>:  <a href="#ad185a5e02c8ed0930e4d400e9c0e6a9a">More...</a><br /></td></tr>
<tr class="separator:ad185a5e02c8ed0930e4d400e9c0e6a9a"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a87d357c93783916037446bf8096ed45d"><td class="memItemLeft" align="right" valign="top">double&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_host.html#a87d357c93783916037446bf8096ed45d">getFitness</a> ()</td></tr>
<tr class="memdesc:a87d357c93783916037446bf8096ed45d"><td class="mdescLeft">&#160;</td><td class="mdescRight">Core method. Returns current value of host's fitness.  <a href="#a87d357c93783916037446bf8096ed45d">More...</a><br /></td></tr>
<tr class="separator:a87d357c93783916037446bf8096ed45d"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a481373a4981aad450aa666962f0c8252"><td class="memItemLeft" align="right" valign="top">std::string&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_host.html#a481373a4981aad450aa666962f0c8252">stringChromosomes</a> ()</td></tr>
<tr class="memdesc:a481373a4981aad450aa666962f0c8252"><td class="mdescLeft">&#160;</td><td class="mdescRight">Data harvesting method. Gives a host's genome in a human-readable format. With all the gene specs.  <a href="#a481373a4981aad450aa666962f0c8252">More...</a><br /></td></tr>
<tr class="separator:a481373a4981aad450aa666962f0c8252"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a6d1d5bf85ae82cc62c7f693a575b2990"><td class="memItemLeft" align="right" valign="top">unsigned long int&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_host.html#a6d1d5bf85ae82cc62c7f693a575b2990">getOneGeneFromOne</a> (int indx)</td></tr>
<tr class="memdesc:a6d1d5bf85ae82cc62c7f693a575b2990"><td class="mdescLeft">&#160;</td><td class="mdescRight">Core method. Fetches a gene (as integer) with a given position on the Chromosome One.  <a href="#a6d1d5bf85ae82cc62c7f693a575b2990">More...</a><br /></td></tr>
<tr class="separator:a6d1d5bf85ae82cc62c7f693a575b2990"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a787c36f60a723e1567dd1a596ecec33e"><td class="memItemLeft" align="right" valign="top">unsigned long int&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_host.html#a787c36f60a723e1567dd1a596ecec33e">getOneGeneFromTwo</a> (int indx)</td></tr>
<tr class="memdesc:a787c36f60a723e1567dd1a596ecec33e"><td class="mdescLeft">&#160;</td><td class="mdescRight">Core method. Fetches a gene (as integer) with a given position on the Chromosome Two.  <a href="#a787c36f60a723e1567dd1a596ecec33e">More...</a><br /></td></tr>
<tr class="separator:a787c36f60a723e1567dd1a596ecec33e"><td class="memSeparator" colspan="2">&#160;</td></tr>
</table><table class="memberdecls">
<tr class="heading"><td colspan="2"><h2 class="groupheader"><a name="pub-attribs"></a>
Public Attributes</h2></td></tr>
<tr class="memitem:a0d79bd02d6147f8e94ae975dc42ae026"><td class="memItemLeft" align="right" valign="top">std::vector&lt; int &gt;&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_host.html#a0d79bd02d6147f8e94ae975dc42ae026">PathoSpecInfecting</a></td></tr>
<tr class="separator:a0d79bd02d6147f8e94ae975dc42ae026"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ac8b084352039979629a7896e2aaefb45"><td class="memItemLeft" align="right" valign="top">std::vector&lt; int &gt;&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_host.html#ac8b084352039979629a7896e2aaefb45">PathogesPresented</a></td></tr>
<tr class="separator:ac8b084352039979629a7896e2aaefb45"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ac173100d13da6dfeb93387f0d689c349"><td class="memItemLeft" align="right" valign="top">unsigned&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_host.html#ac173100d13da6dfeb93387f0d689c349">NumOfPathogesInfecting</a></td></tr>
<tr class="separator:ac173100d13da6dfeb93387f0d689c349"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ad0411c059e63ed72507b0c370880b5ab"><td class="memItemLeft" align="right" valign="top">unsigned&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_host.html#ad0411c059e63ed72507b0c370880b5ab">NumOfPathogesPresented</a></td></tr>
<tr class="separator:ad0411c059e63ed72507b0c370880b5ab"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ab56ce3fe28076749e087269cb100299a"><td class="memItemLeft" align="right" valign="top">unsigned&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_host.html#ab56ce3fe28076749e087269cb100299a">NumOfMhcAlleles</a></td></tr>
<tr class="separator:ab56ce3fe28076749e087269cb100299a"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a50bc219031501dd2c19a69d4e3f19533"><td class="memItemLeft" align="right" valign="top">int&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_host.html#a50bc219031501dd2c19a69d4e3f19533">SelectedForReproduction</a></td></tr>
<tr class="separator:a50bc219031501dd2c19a69d4e3f19533"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a4ac8f1f9b0c989efbade931db22e236d"><td class="memItemLeft" align="right" valign="top">int&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_host.html#a4ac8f1f9b0c989efbade931db22e236d">TimeOfRecombin</a></td></tr>
<tr class="separator:a4ac8f1f9b0c989efbade931db22e236d"><td class="memSeparator" colspan="2">&#160;</td></tr>
</table>
<a name="details" id="details"></a><h2 class="groupheader">Detailed Description</h2>
<div class="textblock"><p>Core class. Stores and handles a single host object. Each host has multiple instances of <a class="el" href="class_gene.html" title="Core class. Stores and handles a single gene object. Has methods to access and to mutate a gene...">Gene</a> class objects and stores them in 2 vectors called ChromosomeOne and ChromosomeTwo which simulate a diploid genome. </p>

<p>Definition at line <a class="el" href="_host_8h_source.html#l00040">40</a> of file <a class="el" href="_host_8h_source.html">Host.h</a>.</p>
</div><h2 class="groupheader">Constructor &amp; Destructor Documentation</h2>
<a class="anchor" id="a369c12e7425c030528bab0784da40e4b"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">Host::Host </td>
          <td>(</td>
          <td class="paramname"></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Definition at line <a class="el" href="_host_8cpp_source.html#l00038">38</a> of file <a class="el" href="_host_8cpp_source.html">Host.cpp</a>.</p>

</div>
</div>
<a class="anchor" id="a6c14f16c39ae6bc5db2672e0141adccd"></a>
<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">Host::~Host </td>
          <td>(</td>
          <td class="paramname"></td><td>)</td>
          <td></td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">virtual</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">

<p>Definition at line <a class="el" href="_host_8cpp_source.html#l00044">44</a> of file <a class="el" href="_host_8cpp_source.html">Host.cpp</a>.</p>

</div>
</div>
<h2 class="groupheader">Member Function Documentation</h2>
<a class="anchor" id="ac6ecd452f531bbcaf469d5d9d42b8028"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void Host::assignChromOne </td>
          <td>(</td>
          <td class="paramtype"><a class="el" href="_h2_pinteraction_8cpp.html#adcfab8f2ed9ed0c694cb95d35c91aa78">chromovector</a>&#160;</td>
          <td class="paramname"><em>One</em></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Core method. Assigns a new chromosome to host's Chromosome ONE. </p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">One</td><td>- a STL vector of genes </td></tr>
  </table>
  </dd>
</dl>

<p>Definition at line <a class="el" href="_host_8cpp_source.html#l00480">480</a> of file <a class="el" href="_host_8cpp_source.html">Host.cpp</a>.</p>

</div>
</div>
<a class="anchor" id="ab3e04fdf569863ddeee00d6d2f5c990d"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void Host::assignChromTwo </td>
          <td>(</td>
          <td class="paramtype"><a class="el" href="_h2_pinteraction_8cpp.html#adcfab8f2ed9ed0c694cb95d35c91aa78">chromovector</a>&#160;</td>
          <td class="paramname"><em>Two</em></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Core method. Assigns a new chromosome to host's Chromosome TWO. </p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">One</td><td>- a STL vector of genes </td></tr>
  </table>
  </dd>
</dl>

<p>Definition at line <a class="el" href="_host_8cpp_source.html#l00489">489</a> of file <a class="el" href="_host_8cpp_source.html">Host.cpp</a>.</p>

</div>
</div>
<a class="anchor" id="a6079285b3b8bb5280d52500e6b7fc7ef"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void Host::calculateFitnessAccChromSize </td>
          <td>(</td>
          <td class="paramname"></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Core method. Calculates host individual fitness as the number of exposed pathogens divided by the genome size. </p>

<p>Definition at line <a class="el" href="_host_8cpp_source.html#l00512">512</a> of file <a class="el" href="_host_8cpp_source.html">Host.cpp</a>.</p>

</div>
</div>
<a class="anchor" id="aca12c9829c01995669e842876478c7fb"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void Host::calculateFitnessAlphaXSqr </td>
          <td>(</td>
          <td class="paramtype">double&#160;</td>
          <td class="paramname"><em>alpha</em></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Core method. Calculates host individual fitness in proportion to one over the square of the number of genes scaled to by factor <img class="formulaInl" alt="$ \alpha $" src="form_12.png"/>: </p>
<p><img class="formulaInl" alt="$ F = \frac{P}{(\alpha \cdot N)^{2}} $" src="form_8.png"/></p>
<p>where <img class="formulaInl" alt="$0 < \alpha < 1 $" src="form_9.png"/>, <img class="formulaInl" alt="$ P $" src="form_10.png"/> is the number of pathogens exposed and <img class="formulaInl" alt="$ N $" src="form_11.png"/> in the sum of number of genes in both chromosomes.</p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">alpha</td><td>- scaling parameter for F() shape </td></tr>
  </table>
  </dd>
</dl>

<p>Definition at line <a class="el" href="_host_8cpp_source.html#l00549">549</a> of file <a class="el" href="_host_8cpp_source.html">Host.cpp</a>.</p>

</div>
</div>
<a class="anchor" id="a9134661842de046453acc0342ef74eb6"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void Host::calculateFitnessExpFunc </td>
          <td>(</td>
          <td class="paramtype">double&#160;</td>
          <td class="paramname"><em>alpha</em></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Core method. Calculates host individual fitness in proportion to the Gaussian function of the number of genes scaled to by factor <img class="formulaInl" alt="$ \alpha $" src="form_12.png"/>: </p>
<p><img class="formulaInl" alt="$ F = P \cdot \exp \left[-(\alpha N)^{2}\right] $" src="form_15.png"/></p>
<p>where <img class="formulaInl" alt="$0 < \alpha < 1 $" src="form_9.png"/> and scales the slope of the Gaussian function, <img class="formulaInl" alt="$ P $" src="form_10.png"/> is the number of pathogens exposed and <img class="formulaInl" alt="$ N $" src="form_11.png"/> in the sum of number of genes in both chromosomes.</p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">alpha</td><td>- scaling parameter for F() shape </td></tr>
  </table>
  </dd>
</dl>

<p>Definition at line <a class="el" href="_host_8cpp_source.html#l00570">570</a> of file <a class="el" href="_host_8cpp_source.html">Host.cpp</a>.</p>

</div>
</div>
<a class="anchor" id="ad185a5e02c8ed0930e4d400e9c0e6a9a"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void Host::calculateFitnessExpFuncUniqAlleles </td>
          <td>(</td>
          <td class="paramtype">double&#160;</td>
          <td class="paramname"><em>alpha</em></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Core method. Calculates host individual fitness in proportion to the Gaussian function of the number of unique MHC alleles in the chromosome scaled to by factor <img class="formulaInl" alt="$ \alpha $" src="form_12.png"/>: </p>
<p><img class="formulaInl" alt="$ F = P \cdot \exp \left[-(\alpha N)^{2}\right] $" src="form_15.png"/></p>
<p>where <img class="formulaInl" alt="$0 < \alpha < 1 $" src="form_9.png"/> and scales the slope of the Gaussian function, <img class="formulaInl" alt="$ P $" src="form_10.png"/> is the number of pathogens exposed and <img class="formulaInl" alt="$ N $" src="form_11.png"/> in the sum of number of unique MHC alleles in both chromosomes.</p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">alpha</td><td>- scaling parameter for F() shape </td></tr>
  </table>
  </dd>
</dl>

<p>Definition at line <a class="el" href="_host_8cpp_source.html#l00593">593</a> of file <a class="el" href="_host_8cpp_source.html">Host.cpp</a>.</p>

<p><div class="dynheader">
Here is the call graph for this function:</div>
<div class="dyncontent">
<div class="center"><img src="class_host_ad185a5e02c8ed0930e4d400e9c0e6a9a_cgraph.png" border="0" usemap="#class_host_ad185a5e02c8ed0930e4d400e9c0e6a9a_cgraph" alt=""/></div>
<map name="class_host_ad185a5e02c8ed0930e4d400e9c0e6a9a_cgraph" id="class_host_ad185a5e02c8ed0930e4d400e9c0e6a9a_cgraph">
<area shape="rect" id="node2" href="class_host.html#ad84a6fab9904782aa633deab0ad5e506" title="Returns the number of unique MHC alleles in the Chromosome One. " alt="" coords="230,5,442,32"/><area shape="rect" id="node3" href="class_host.html#ad05b417630dd1095b967de411bdb350a" title="Returns the number of unique MHC alleles in the Chromosome Two. " alt="" coords="229,56,443,83"/></map>
</div>
</p>

</div>
</div>
<a class="anchor" id="a614430aea49534f29b63232414cf4fcd"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void Host::calculateFitnessForDrift </td>
          <td>(</td>
          <td class="paramname"></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Core method. Sets fitness as 1, a fixed value to simulate genetic drift. </p>

<p>Definition at line <a class="el" href="_host_8cpp_source.html#l00534">534</a> of file <a class="el" href="_host_8cpp_source.html">Host.cpp</a>.</p>

</div>
</div>
<a class="anchor" id="aeb8564ae54f8bbf968a62f13c61083df"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void Host::calculateFitnessJustInfection </td>
          <td>(</td>
          <td class="paramname"></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Core method. Calculates host individual fitness simply as the number of exposed pathogens. </p>

<p>Definition at line <a class="el" href="_host_8cpp_source.html#l00526">526</a> of file <a class="el" href="_host_8cpp_source.html">Host.cpp</a>.</p>

</div>
</div>
<a class="anchor" id="a5911792258205cae25a2a4796061fe49"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void Host::chromoMutProcess </td>
          <td>(</td>
          <td class="paramtype">double&#160;</td>
          <td class="paramname"><em>mut_probabl</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">int&#160;</td>
          <td class="paramname"><em>timeStamp</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Core method. Decides (on a random basis) if there will be any mutations in the genome. </p>
<p>Iterates through the both chromosomes and calls gene mutation function. Genes mutate at random depending on the probability which was user-defined.</p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">mut_probabl</td><td>- mutation probability, a probability a gene will be replaced by a new one </td></tr>
    <tr><td class="paramname">timeStamp</td><td>- current time (number of the model iteration) </td></tr>
  </table>
  </dd>
</dl>

<p>Definition at line <a class="el" href="_host_8cpp_source.html#l00109">109</a> of file <a class="el" href="_host_8cpp_source.html">Host.cpp</a>.</p>

</div>
</div>
<a class="anchor" id="a48c01d4345818b873ceed7597588bd6d"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void Host::chromoMutProcessWithDelDupl </td>
          <td>(</td>
          <td class="paramtype">double&#160;</td>
          <td class="paramname"><em>mut_probabl</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">double&#160;</td>
          <td class="paramname"><em>del</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">double&#160;</td>
          <td class="paramname"><em>dupli</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">int&#160;</td>
          <td class="paramname"><em>maxGene</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">int&#160;</td>
          <td class="paramname"><em>timeStamp</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Core method. Decides (on a random basis) if there will be any mutations in the genome: deletion, duplication or change of a gene. </p>
<p>Iterates through the both chromosomes and calls gene mutation function. Genes mutate at random depending on the probability which was user-defined. Also at random a gene can be deleted or duplicated.</p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">mut_probabl</td><td>- mutation probability, a probability a gene will be replaced by a new one </td></tr>
    <tr><td class="paramname">del</td><td>- mutation probability, probability a gene will be deleted </td></tr>
    <tr><td class="paramname">dupli</td><td>- mutation probability, probability a gene will be duplicated (and added at the end of the Chromosome vector) </td></tr>
    <tr><td class="paramname">maxGene</td><td>- maximal allowed number of genes in a chromosome (user defined parameter). </td></tr>
    <tr><td class="paramname">timeStamp</td><td>- current time (current number of the model iteration) </td></tr>
  </table>
  </dd>
</dl>

<p>Definition at line <a class="el" href="_host_8cpp_source.html#l00137">137</a> of file <a class="el" href="_host_8cpp_source.html">Host.cpp</a>.</p>

<p><div class="dynheader">
Here is the call graph for this function:</div>
<div class="dyncontent">
<div class="center"><img src="class_host_a48c01d4345818b873ceed7597588bd6d_cgraph.png" border="0" usemap="#class_host_a48c01d4345818b873ceed7597588bd6d_cgraph" alt=""/></div>
<map name="class_host_a48c01d4345818b873ceed7597588bd6d_cgraph" id="class_host_a48c01d4345818b873ceed7597588bd6d_cgraph">
<area shape="rect" id="node2" href="class_random_numbs.html#a364c415e5105061274013a1e68576a64" title="Core method. Gets the instance of a singleton class. " alt="" coords="253,5,443,32"/><area shape="rect" id="node3" href="class_random_numbs.html#aaae77654579060738e55b46c8caba9a6" title="Core method. Generates one real random number from a given interval with an uniform distribution..." alt="" coords="260,56,436,83"/></map>
</div>
</p>

</div>
</div>
<a class="anchor" id="a303c0d61ab48aaa709fccfcdd14a156e"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void Host::chromoMutProcessWithDelDuplPointMuts </td>
          <td>(</td>
          <td class="paramtype">double&#160;</td>
          <td class="paramname"><em>pm_mut_probabl</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">double&#160;</td>
          <td class="paramname"><em>del</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">double&#160;</td>
          <td class="paramname"><em>dupli</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">int&#160;</td>
          <td class="paramname"><em>maxGene</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">int&#160;</td>
          <td class="paramname"><em>timeStamp</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Core method. Decides (on a random basis) if there will be any mutations in the genome. </p>
<p>Iterates through the both chromosomes and calls gene mutation function. Point sites in genes are mutated at random depending on the probability which was user-defined. Also at random a gene can be deleted or duplicated.</p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">pm_mut_probabl</td><td>- point mutation probability, a probability a gene will be replaced by a new one </td></tr>
    <tr><td class="paramname">del</td><td>- mutation probability, probability a gene will be deleted </td></tr>
    <tr><td class="paramname">dupli</td><td>- mutation probability, probability a gene will be duplicated (and added at the end of the Chromosome vector) </td></tr>
    <tr><td class="paramname">maxGene</td><td>- maximal allowed number of genes in a chromosome (user defined parameter). </td></tr>
    <tr><td class="paramname">timeStamp</td><td>- current time (current number of the model iteration) </td></tr>
  </table>
  </dd>
</dl>

<p>Definition at line <a class="el" href="_host_8cpp_source.html#l00183">183</a> of file <a class="el" href="_host_8cpp_source.html">Host.cpp</a>.</p>

<p><div class="dynheader">
Here is the call graph for this function:</div>
<div class="dyncontent">
<div class="center"><img src="class_host_a303c0d61ab48aaa709fccfcdd14a156e_cgraph.png" border="0" usemap="#class_host_a303c0d61ab48aaa709fccfcdd14a156e_cgraph" alt=""/></div>
<map name="class_host_a303c0d61ab48aaa709fccfcdd14a156e_cgraph" id="class_host_a303c0d61ab48aaa709fccfcdd14a156e_cgraph">
<area shape="rect" id="node2" href="class_random_numbs.html#a364c415e5105061274013a1e68576a64" title="Core method. Gets the instance of a singleton class. " alt="" coords="253,5,443,32"/><area shape="rect" id="node3" href="class_random_numbs.html#aaae77654579060738e55b46c8caba9a6" title="Core method. Generates one real random number from a given interval with an uniform distribution..." alt="" coords="260,56,436,83"/></map>
</div>
</p>

</div>
</div>
<a class="anchor" id="ada5564c9d57e18ce046e131a5968e08f"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void Host::chromoRecombination </td>
          <td>(</td>
          <td class="paramtype">double&#160;</td>
          <td class="paramname"><em>recomb_prob</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">int&#160;</td>
          <td class="paramname"><em>timeStamp</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Core method. Performs recombination between host's chromosomes by inserting a brake in randomly selected genes on Chromosome One and Two and swapping the resulting parts of the chromosomes. </p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">recomb_prob</td><td>- probability of the recombination event in host's genome </td></tr>
    <tr><td class="paramname">timeStamp</td><td>- current time (current number of the model iteration) </td></tr>
  </table>
  </dd>
</dl>

<p>Definition at line <a class="el" href="_host_8cpp_source.html#l00220">220</a> of file <a class="el" href="_host_8cpp_source.html">Host.cpp</a>.</p>

<p><div class="dynheader">
Here is the call graph for this function:</div>
<div class="dyncontent">
<div class="center"><img src="class_host_ada5564c9d57e18ce046e131a5968e08f_cgraph.png" border="0" usemap="#class_host_ada5564c9d57e18ce046e131a5968e08f_cgraph" alt=""/></div>
<map name="class_host_ada5564c9d57e18ce046e131a5968e08f_cgraph" id="class_host_ada5564c9d57e18ce046e131a5968e08f_cgraph">
<area shape="rect" id="node2" href="class_random_numbs.html#a364c415e5105061274013a1e68576a64" title="Core method. Gets the instance of a singleton class. " alt="" coords="243,5,432,32"/><area shape="rect" id="node3" href="class_random_numbs.html#aaae77654579060738e55b46c8caba9a6" title="Core method. Generates one real random number from a given interval with an uniform distribution..." alt="" coords="249,56,425,83"/></map>
</div>
</p>

</div>
</div>
<a class="anchor" id="a7c21e1938c61966dc2657a437aeec9b9"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void Host::clearInfections </td>
          <td>(</td>
          <td class="paramname"></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Core method. Clears data regarding infections and fitness. </p>

<p>Definition at line <a class="el" href="_host_8cpp_source.html#l00615">615</a> of file <a class="el" href="_host_8cpp_source.html">Host.cpp</a>.</p>

</div>
</div>
<a class="anchor" id="a505527db337d87670e5ab37eb9818fa3"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname"><a class="el" href="_h2_pinteraction_8cpp.html#adcfab8f2ed9ed0c694cb95d35c91aa78">chromovector</a> Host::doCrossAndMeiosis </td>
          <td>(</td>
          <td class="paramtype">double&#160;</td>
          <td class="paramname"><em>corssing_prob</em></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Core method. Performing crossing over and chromosome selection before mating. </p>
<p>It mixes up genes from two host chromosomes to form one child chromosome later pushed for mating. One can regulate which chromosome is preferred as a source of genes by changing crossing-over probability parameter.</p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">corssing_prob</td><td>-probability of a gene crossing over; p=0.5 means that each chromosome gives half of its genes to the resulting meiotic chromosome. p lesser then 0.5 means ChromosomeOne will give majority of genes; p larger then 0.5 will favour ChromosomeTwo as a donor.</td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>a Chromosome vector. </dd></dl>

<p>Definition at line <a class="el" href="_host_8cpp_source.html#l00242">242</a> of file <a class="el" href="_host_8cpp_source.html">Host.cpp</a>.</p>

<p><div class="dynheader">
Here is the call graph for this function:</div>
<div class="dyncontent">
<div class="center"><img src="class_host_a505527db337d87670e5ab37eb9818fa3_cgraph.png" border="0" usemap="#class_host_a505527db337d87670e5ab37eb9818fa3_cgraph" alt=""/></div>
<map name="class_host_a505527db337d87670e5ab37eb9818fa3_cgraph" id="class_host_a505527db337d87670e5ab37eb9818fa3_cgraph">
<area shape="rect" id="node2" href="class_random_numbs.html#a364c415e5105061274013a1e68576a64" title="Core method. Gets the instance of a singleton class. " alt="" coords="229,5,419,32"/><area shape="rect" id="node3" href="class_random_numbs.html#aaae77654579060738e55b46c8caba9a6" title="Core method. Generates one real random number from a given interval with an uniform distribution..." alt="" coords="236,56,412,83"/></map>
</div>
</p>

</div>
</div>
<a class="anchor" id="aec728ae16d7c1186739730510a2c7e2e"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">unsigned Host::getChromoOneSize </td>
          <td>(</td>
          <td class="paramname"></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Returns the number of genes in the Chromosome One. </p>
<dl class="section return"><dt>Returns</dt><dd>number of MHC genes in the Chromosome One. </dd></dl>

<p>Definition at line <a class="el" href="_host_8cpp_source.html#l00331">331</a> of file <a class="el" href="_host_8cpp_source.html">Host.cpp</a>.</p>

</div>
</div>
<a class="anchor" id="ad84a6fab9904782aa633deab0ad5e506"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">double Host::getChromoOneUniqAlleles </td>
          <td>(</td>
          <td class="paramname"></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Returns the number of unique MHC alleles in the Chromosome One. </p>
<dl class="section return"><dt>Returns</dt><dd>number of unique MHC alleles in the Chromosome One. </dd></dl>

<p>Definition at line <a class="el" href="_host_8cpp_source.html#l00351">351</a> of file <a class="el" href="_host_8cpp_source.html">Host.cpp</a>.</p>

<p><div class="dynheader">
Here is the caller graph for this function:</div>
<div class="dyncontent">
<div class="center"><img src="class_host_ad84a6fab9904782aa633deab0ad5e506_icgraph.png" border="0" usemap="#class_host_ad84a6fab9904782aa633deab0ad5e506_icgraph" alt=""/></div>
<map name="class_host_ad84a6fab9904782aa633deab0ad5e506_icgraph" id="class_host_ad84a6fab9904782aa633deab0ad5e506_icgraph">
<area shape="rect" id="node2" href="class_host.html#ad185a5e02c8ed0930e4d400e9c0e6a9a" title="Core method. Calculates host individual fitness in proportion to the Gaussian function of the number ..." alt="" coords="267,5,443,46"/></map>
</div>
</p>

</div>
</div>
<a class="anchor" id="ac387176ba203f6aa1f20610f5679385a"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname"><a class="el" href="_h2_pinteraction_8cpp.html#adcfab8f2ed9ed0c694cb95d35c91aa78">chromovector</a> Host::getChromosomeOne </td>
          <td>(</td>
          <td class="paramname"></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Core method. Returns the first chromosome. </p>
<dl class="section return"><dt>Returns</dt><dd>Chromosome One </dd></dl>

<p>Definition at line <a class="el" href="_host_8cpp_source.html#l00448">448</a> of file <a class="el" href="_host_8cpp_source.html">Host.cpp</a>.</p>

<p><div class="dynheader">
Here is the caller graph for this function:</div>
<div class="dyncontent">
<div class="center"><img src="class_host_ac387176ba203f6aa1f20610f5679385a_icgraph.png" border="0" usemap="#class_host_ac387176ba203f6aa1f20610f5679385a_icgraph" alt=""/></div>
<map name="class_host_ac387176ba203f6aa1f20610f5679385a_icgraph" id="class_host_ac387176ba203f6aa1f20610f5679385a_icgraph">
<area shape="rect" id="node2" href="class_h2_pinteraction.html#a46703648dcef05c6478d6249aeadd2c8" title="Core method. Checks if a host gets infected with a pathogen. Heterozygote has an advantage here over ..." alt="" coords="233,5,423,46"/><area shape="rect" id="node3" href="class_environment.html#aecd592b734359557235bb0a6980f77f9" title="Core method. Iterates through the host population and the parasite population to &quot;infect&quot; the hosts w..." alt="" coords="472,5,661,46"/></map>
</div>
</p>

</div>
</div>
<a class="anchor" id="aa5afa38408ebba644f56667364213f0d"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname"><a class="el" href="_h2_pinteraction_8cpp.html#adcfab8f2ed9ed0c694cb95d35c91aa78">chromovector</a> Host::getChromosomeTwo </td>
          <td>(</td>
          <td class="paramname"></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Core method. Returns the first chromosome. </p>
<dl class="section return"><dt>Returns</dt><dd>Chromosome Two </dd></dl>

<p>Definition at line <a class="el" href="_host_8cpp_source.html#l00457">457</a> of file <a class="el" href="_host_8cpp_source.html">Host.cpp</a>.</p>

<p><div class="dynheader">
Here is the caller graph for this function:</div>
<div class="dyncontent">
<div class="center"><img src="class_host_aa5afa38408ebba644f56667364213f0d_icgraph.png" border="0" usemap="#class_host_aa5afa38408ebba644f56667364213f0d_icgraph" alt=""/></div>
<map name="class_host_aa5afa38408ebba644f56667364213f0d_icgraph" id="class_host_aa5afa38408ebba644f56667364213f0d_icgraph">
<area shape="rect" id="node2" href="class_h2_pinteraction.html#a46703648dcef05c6478d6249aeadd2c8" title="Core method. Checks if a host gets infected with a pathogen. Heterozygote has an advantage here over ..." alt="" coords="233,5,423,46"/><area shape="rect" id="node3" href="class_environment.html#aecd592b734359557235bb0a6980f77f9" title="Core method. Iterates through the host population and the parasite population to &quot;infect&quot; the hosts w..." alt="" coords="472,5,661,46"/></map>
</div>
</p>

</div>
</div>
<a class="anchor" id="a03e175acb665dc22d96bb3a038571af7"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">unsigned Host::getChromoTwoSize </td>
          <td>(</td>
          <td class="paramname"></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Returns the number of genes in the Chromosome Two. </p>
<dl class="section return"><dt>Returns</dt><dd>number of MHC genes in the Chromosome Two. </dd></dl>

<p>Definition at line <a class="el" href="_host_8cpp_source.html#l00340">340</a> of file <a class="el" href="_host_8cpp_source.html">Host.cpp</a>.</p>

</div>
</div>
<a class="anchor" id="ad05b417630dd1095b967de411bdb350a"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">double Host::getChromoTwoUniqAlleles </td>
          <td>(</td>
          <td class="paramname"></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Returns the number of unique MHC alleles in the Chromosome Two. </p>
<dl class="section return"><dt>Returns</dt><dd>number of unique MHC alleles in the Chromosome Two. </dd></dl>

<p>Definition at line <a class="el" href="_host_8cpp_source.html#l00383">383</a> of file <a class="el" href="_host_8cpp_source.html">Host.cpp</a>.</p>

<p><div class="dynheader">
Here is the caller graph for this function:</div>
<div class="dyncontent">
<div class="center"><img src="class_host_ad05b417630dd1095b967de411bdb350a_icgraph.png" border="0" usemap="#class_host_ad05b417630dd1095b967de411bdb350a_icgraph" alt=""/></div>
<map name="class_host_ad05b417630dd1095b967de411bdb350a_icgraph" id="class_host_ad05b417630dd1095b967de411bdb350a_icgraph">
<area shape="rect" id="node2" href="class_host.html#ad185a5e02c8ed0930e4d400e9c0e6a9a" title="Core method. Calculates host individual fitness in proportion to the Gaussian function of the number ..." alt="" coords="267,5,443,46"/></map>
</div>
</p>

</div>
</div>
<a class="anchor" id="a87d357c93783916037446bf8096ed45d"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">double Host::getFitness </td>
          <td>(</td>
          <td class="paramname"></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Core method. Returns current value of host's fitness. </p>
<dl class="section return"><dt>Returns</dt><dd>current value of host's fitness </dd></dl>

<p>Definition at line <a class="el" href="_host_8cpp_source.html#l00608">608</a> of file <a class="el" href="_host_8cpp_source.html">Host.cpp</a>.</p>

</div>
</div>
<a class="anchor" id="a10ec4eb6494084cbc99a4c28164b0761"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">unsigned Host::getGenomeSize </td>
          <td>(</td>
          <td class="paramname"></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Returns the total number of genes in the genome. </p>
<dl class="section return"><dt>Returns</dt><dd>number of all MHC genes in the host </dd></dl>

<p>Definition at line <a class="el" href="_host_8cpp_source.html#l00322">322</a> of file <a class="el" href="_host_8cpp_source.html">Host.cpp</a>.</p>

</div>
</div>
<a class="anchor" id="a88a454efb7a483985eea815e2b285809"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">unsigned long Host::getHostIndvTag </td>
          <td>(</td>
          <td class="paramname"></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">

</div>
</div>
<a class="anchor" id="a44931bdccc9732456a67db776b5e69c4"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">unsigned long Host::getHostMotherTag </td>
          <td>(</td>
          <td class="paramname"></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">

</div>
</div>
<a class="anchor" id="a6d1d5bf85ae82cc62c7f693a575b2990"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">unsigned long int Host::getOneGeneFromOne </td>
          <td>(</td>
          <td class="paramtype">int&#160;</td>
          <td class="paramname"><em>indx</em></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Core method. Fetches a gene (as integer) with a given position on the Chromosome One. </p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">indx</td><td>- index of the gene you wanna get. </td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>an integer representation of a gene with a given index. </dd></dl>

<p>Definition at line <a class="el" href="_host_8cpp_source.html#l00276">276</a> of file <a class="el" href="_host_8cpp_source.html">Host.cpp</a>.</p>

</div>
</div>
<a class="anchor" id="a787c36f60a723e1567dd1a596ecec33e"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">unsigned long int Host::getOneGeneFromTwo </td>
          <td>(</td>
          <td class="paramtype">int&#160;</td>
          <td class="paramname"><em>indx</em></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Core method. Fetches a gene (as integer) with a given position on the Chromosome Two. </p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">indx</td><td>- index of the gene you wanna get. </td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>an integer representation of a gene with a given index. </dd></dl>

<p>Definition at line <a class="el" href="_host_8cpp_source.html#l00300">300</a> of file <a class="el" href="_host_8cpp_source.html">Host.cpp</a>.</p>

</div>
</div>
<a class="anchor" id="a8b8d1bb28ee49fcd337a0a5656595497"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname"><a class="el" href="_gene_8cpp.html#a108efa7670466d15a5ecafb60bbc9c37">genestring</a> Host::getSingleGeneFromOne </td>
          <td>(</td>
          <td class="paramtype">int&#160;</td>
          <td class="paramname"><em>indx</em></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Core method. Fetches a gene (bit string) with a given position on the chromosome One. </p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">indx</td><td>- index of the gene you wanna get. </td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>a bit-string representation of a gene with a given index. </dd></dl>

<p>Definition at line <a class="el" href="_host_8cpp_source.html#l00416">416</a> of file <a class="el" href="_host_8cpp_source.html">Host.cpp</a>.</p>

</div>
</div>
<a class="anchor" id="a8e57071a6138a8a975dba81be1d39e44"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname"><a class="el" href="_gene_8cpp.html#a108efa7670466d15a5ecafb60bbc9c37">genestring</a> Host::getSingleGeneFromTwo </td>
          <td>(</td>
          <td class="paramtype">int&#160;</td>
          <td class="paramname"><em>indx</em></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Core method. Fetches a gene (bit string) with a given position on the chromosome Two. </p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">indx</td><td>- index of the gene you wanna get. </td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>a bitstring representation of a gene with a given index. </dd></dl>

<p>Definition at line <a class="el" href="_host_8cpp_source.html#l00433">433</a> of file <a class="el" href="_host_8cpp_source.html">Host.cpp</a>.</p>

</div>
</div>
<a class="anchor" id="a55edfefc42f7306c85983474500ef10b"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname"><a class="el" href="_h2_pinteraction_8cpp.html#adcfab8f2ed9ed0c694cb95d35c91aa78">chromovector</a> Host::mergeChromosomes </td>
          <td>(</td>
          <td class="paramname"></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Data harvesting method. Merges both chromosomes to create an object easier to handle is some situations. </p>
<dl class="section return"><dt>Returns</dt><dd>Chromosome vector containing genes from both, Chromosome One and Two </dd></dl>

<p>Definition at line <a class="el" href="_host_8cpp_source.html#l00467">467</a> of file <a class="el" href="_host_8cpp_source.html">Host.cpp</a>.</p>

</div>
</div>
<a class="anchor" id="a9c86ac277ceed9299afccff49c954837"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void Host::setHostIndvTag </td>
          <td>(</td>
          <td class="paramtype">unsigned long&#160;</td>
          <td class="paramname"><em>theTag</em></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">

</div>
</div>
<a class="anchor" id="a8829231a3659a4b206734b4fa14341f0"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void Host::setHostMotherTag </td>
          <td>(</td>
          <td class="paramtype">unsigned long&#160;</td>
          <td class="paramname"><em>theTag</em></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">

</div>
</div>
<a class="anchor" id="acea27ac91fb04a2d711d5af25200c557"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void Host::setNewHomozygHost </td>
          <td>(</td>
          <td class="paramtype">int&#160;</td>
          <td class="paramname"><em>num_of_loci</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">int&#160;</td>
          <td class="paramname"><em>gene_size</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">int&#160;</td>
          <td class="paramname"><em>timeStamp</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Core method. Sets a new <a class="el" href="class_host.html" title="Core class. Stores and handles a single host object. Each host has multiple instances of Gene class o...">Host</a> object and assigns a user-defined number of genes of user-defined number of bits (as genes are bit-strings). </p>
<p>A host has two chromosome of equal length. Each corresponding pair of loci is homozygous.</p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">num_of_loci</td><td>- number of gene loci in a chromosome </td></tr>
    <tr><td class="paramname">gene_size</td><td>- the length of the bit-string representing a gene </td></tr>
    <tr><td class="paramname">timeStamp</td><td>- current time (number of the model iteration) </td></tr>
  </table>
  </dd>
</dl>

<p>Definition at line <a class="el" href="_host_8cpp_source.html#l00082">82</a> of file <a class="el" href="_host_8cpp_source.html">Host.cpp</a>.</p>

</div>
</div>
<a class="anchor" id="aafa1539a8942e3e99dc6b49110a618c1"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void Host::setNewHost </td>
          <td>(</td>
          <td class="paramtype">int&#160;</td>
          <td class="paramname"><em>num_of_loci</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">int&#160;</td>
          <td class="paramname"><em>gene_size</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">int&#160;</td>
          <td class="paramname"><em>timeStamp</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Core method. Sets a new <a class="el" href="class_host.html" title="Core class. Stores and handles a single host object. Each host has multiple instances of Gene class o...">Host</a> object and assigns a user-defined number of genes of user-defined number of bits (as genes are bit-strings). </p>
<p>A host has two chromosome of equal length. Each gene is set at random and loci are most likely to be heterozygous.</p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">num_of_loci</td><td>- number of gene loci in a chromosome </td></tr>
    <tr><td class="paramname">gene_size</td><td>- the length of the bit-string representing a gene </td></tr>
    <tr><td class="paramname">timeStamp</td><td>- current time (number of the model iteration) </td></tr>
  </table>
  </dd>
</dl>

<p>Definition at line <a class="el" href="_host_8cpp_source.html#l00058">58</a> of file <a class="el" href="_host_8cpp_source.html">Host.cpp</a>.</p>

</div>
</div>
<a class="anchor" id="a481373a4981aad450aa666962f0c8252"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">std::string Host::stringChromosomes </td>
          <td>(</td>
          <td class="paramname"></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Data harvesting method. Gives a host's genome in a human-readable format. With all the gene specs. </p>
<dl class="section return"><dt>Returns</dt><dd>a STL string containing the host's genome and its annotations in a human-readable format. </dd></dl>

<p>Definition at line <a class="el" href="_host_8cpp_source.html#l00631">631</a> of file <a class="el" href="_host_8cpp_source.html">Host.cpp</a>.</p>

</div>
</div>
<a class="anchor" id="ab30250ea6fdd47221d65914b3e6de4e0"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void Host::swapChromosomes </td>
          <td>(</td>
          <td class="paramname"></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Core method. Randomly swaps places of Chromosome One and Chromosome Two to avoid situation when they effectively become two separate populations. </p>

<p>Definition at line <a class="el" href="_host_8cpp_source.html#l00497">497</a> of file <a class="el" href="_host_8cpp_source.html">Host.cpp</a>.</p>

<p><div class="dynheader">
Here is the call graph for this function:</div>
<div class="dyncontent">
<div class="center"><img src="class_host_ab30250ea6fdd47221d65914b3e6de4e0_cgraph.png" border="0" usemap="#class_host_ab30250ea6fdd47221d65914b3e6de4e0_cgraph" alt=""/></div>
<map name="class_host_ab30250ea6fdd47221d65914b3e6de4e0_cgraph" id="class_host_ab30250ea6fdd47221d65914b3e6de4e0_cgraph">
<area shape="rect" id="node2" href="class_random_numbs.html#a364c415e5105061274013a1e68576a64" title="Core method. Gets the instance of a singleton class. " alt="" coords="229,5,419,32"/><area shape="rect" id="node3" href="class_random_numbs.html#aaae77654579060738e55b46c8caba9a6" title="Core method. Generates one real random number from a given interval with an uniform distribution..." alt="" coords="236,56,412,83"/></map>
</div>
</p>

</div>
</div>
<h2 class="groupheader">Member Data Documentation</h2>
<a class="anchor" id="ab56ce3fe28076749e087269cb100299a"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">unsigned Host::NumOfMhcAlleles</td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Definition at line <a class="el" href="_host_8h_source.html#l00050">50</a> of file <a class="el" href="_host_8h_source.html">Host.h</a>.</p>

</div>
</div>
<a class="anchor" id="ac173100d13da6dfeb93387f0d689c349"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">unsigned Host::NumOfPathogesInfecting</td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Definition at line <a class="el" href="_host_8h_source.html#l00048">48</a> of file <a class="el" href="_host_8h_source.html">Host.h</a>.</p>

</div>
</div>
<a class="anchor" id="ad0411c059e63ed72507b0c370880b5ab"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">unsigned Host::NumOfPathogesPresented</td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Definition at line <a class="el" href="_host_8h_source.html#l00049">49</a> of file <a class="el" href="_host_8h_source.html">Host.h</a>.</p>

</div>
</div>
<a class="anchor" id="ac8b084352039979629a7896e2aaefb45"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">std::vector&lt;int&gt; Host::PathogesPresented</td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Definition at line <a class="el" href="_host_8h_source.html#l00047">47</a> of file <a class="el" href="_host_8h_source.html">Host.h</a>.</p>

</div>
</div>
<a class="anchor" id="a0d79bd02d6147f8e94ae975dc42ae026"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">std::vector&lt;int&gt; Host::PathoSpecInfecting</td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Definition at line <a class="el" href="_host_8h_source.html#l00046">46</a> of file <a class="el" href="_host_8h_source.html">Host.h</a>.</p>

</div>
</div>
<a class="anchor" id="a50bc219031501dd2c19a69d4e3f19533"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">int Host::SelectedForReproduction</td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Definition at line <a class="el" href="_host_8h_source.html#l00051">51</a> of file <a class="el" href="_host_8h_source.html">Host.h</a>.</p>

</div>
</div>
<a class="anchor" id="a4ac8f1f9b0c989efbade931db22e236d"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">int Host::TimeOfRecombin</td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Definition at line <a class="el" href="_host_8h_source.html#l00052">52</a> of file <a class="el" href="_host_8h_source.html">Host.h</a>.</p>

</div>
</div>
<hr/>The documentation for this class was generated from the following files:<ul>
<li><a class="el" href="_host_8h_source.html">Host.h</a></li>
<li><a class="el" href="_host_8cpp_source.html">Host.cpp</a></li>
</ul>
</div><!-- contents -->
</div><!-- doc-content -->
<!-- start footer part -->
<div id="nav-path" class="navpath"><!-- id is needed for treeview function! -->
  <ul>
    <li class="navelem"><a class="el" href="class_host.html">Host</a></li>
    <li class="footer">Generated on Tue Aug 9 2016 16:16:54 for Modelling MHC Evolutionary Dynamics by
    <a href="http://www.doxygen.org/index.html">
    <img class="footer" src="doxygen.png" alt="doxygen"/></a> 1.8.7 </li>
  </ul>
</div>
</body>
</html>
