<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml">
<head>
<meta http-equiv="Content-Type" content="text/xhtml;charset=UTF-8"/>
<meta http-equiv="X-UA-Compatible" content="IE=9"/>
<meta name="generator" content="Doxygen 1.8.13"/>
<meta name="viewport" content="width=device-width, initial-scale=1"/>
<title>Modelling MHC Evolutionary Dynamics: &quot;Modelling MHC evolutionary dynamics&quot;</title>
<link href="tabs.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="jquery.js"></script>
<script type="text/javascript" src="dynsections.js"></script>
<link href="navtree.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="resize.js"></script>
<script type="text/javascript" src="navtreedata.js"></script>
<script type="text/javascript" src="navtree.js"></script>
<script type="text/javascript">
  $(document).ready(initResizable);
</script>
<link href="search/search.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="search/searchdata.js"></script>
<script type="text/javascript" src="search/search.js"></script>
<link href="doxygen.css" rel="stylesheet" type="text/css" />
</head>
<body>
<div id="top"><!-- do not remove this div, it is closed by doxygen! -->
<div id="titlearea">
<table cellspacing="0" cellpadding="0">
 <tbody>
 <tr style="height: 56px;">
  <td id="projectalign" style="padding-left: 0.5em;">
   <div id="projectname">Modelling MHC Evolutionary Dynamics
   &#160;<span id="projectnumber">0.9.3</span>
   </div>
  </td>
 </tr>
 </tbody>
</table>
</div>
<!-- end header part -->
<!-- Generated by Doxygen 1.8.13 -->
<script type="text/javascript">
var searchBox = new SearchBox("searchBox", "search",false,'Search');
</script>
<script type="text/javascript" src="menudata.js"></script>
<script type="text/javascript" src="menu.js"></script>
<script type="text/javascript">
$(function() {
  initMenu('',true,false,'search.php','Search');
  $(document).ready(function() { init_search(); });
});
</script>
<div id="main-nav"></div>
</div><!-- top -->
<div id="side-nav" class="ui-resizable side-nav-resizable">
  <div id="nav-tree">
    <div id="nav-tree-contents">
      <div id="nav-sync" class="sync"></div>
    </div>
  </div>
  <div id="splitbar" style="-moz-user-select:none;" 
       class="ui-resizable-handle">
  </div>
</div>
<script type="text/javascript">
$(document).ready(function(){initNavTree('index.html','');});
</script>
<div id="doc-content">
<!-- window showing the filter options -->
<div id="MSearchSelectWindow"
     onmouseover="return searchBox.OnSearchSelectShow()"
     onmouseout="return searchBox.OnSearchSelectHide()"
     onkeydown="return searchBox.OnSearchSelectKey(event)">
</div>

<!-- iframe showing the search results (closed by default) -->
<div id="MSearchResultsWindow">
<iframe src="javascript:void(0)" frameborder="0" 
        name="MSearchResults" id="MSearchResults">
</iframe>
</div>

<div class="header">
  <div class="headertitle">
<div class="title">"Modelling MHC evolutionary dynamics" </div>  </div>
</div><!--header-->
<div class="contents">
<div class="textblock"><dl class="section author"><dt>Authors</dt><dd>Coded by <b>Piotr Bentkowski</b>: <a href="#" onclick="location.href='mai'+'lto:'+'ben'+'tk'+'ows'+'ki'+'.pi'+'ot'+'r@g'+'ma'+'il.'+'co'+'m
?'+'su'+'bje'+'ct'+'=MH'+'C '+'evo'+'lu'+'tio'+'na'+'ry '+'dy'+'nam'+'ic'+'s -'+' m'+'ode'+'l '+'cod'+'e '+'v1.'+'0'; return false;">bentkowski.piotr@gmail.com</a> <br />
in collaboration with <b>Jacek Radwan</b>: <a href="#" onclick="location.href='mai'+'lto:'+'jra'+'dw'+'an@'+'am'+'u.e'+'du'+'.pl'+'
?'+'sub'+'je'+'ct='+'MH'+'C e'+'vo'+'lut'+'io'+'nar'+'y '+'dyn'+'am'+'ics'+' -'+' mo'+'de'+'l c'+'od'+'e v'+'1.'+'0'; return false;">jradwan@amu.edu.pl</a> <br />
<br />
<a href="https://sites.google.com/site/evobiolab/home">Evolutionary Biology Group</a> <br />
Faculty of Biology <br />
Adam Mickiewicz University <br />
ul. Umultowska 89 <br />
61-614 Poznań <br />
Poland <br />
</dd></dl>
<dl class="section version"><dt>Version</dt><dd>0.9.1</dd></dl>
<h1><a class="anchor" id="intro"></a>
Introduction</h1>
<p>This program is an implementation of the evolution dynamics of MHC described in a number of papers e.g.: <br />
</p><ol type="1">
<li>Borghans, J. a M., Beltman, J. B., &amp; De Boer, R. J. (2004). <em>MHC polymorphism under host-pathogen coevolution.</em> Immunogenetics, 55(11), 732–9. <a href="http://doi.org/10.1007/s00251-003-0630-5">http://doi.org/10.1007/s00251-003-0630-5</a> <br />
</li>
<li>Ejsmond, M., Babik, W., &amp; Radwan, J. (2010).<em> MHC allele frequency distributions under parasite-driven selection: A simulation model.</em> BMC Evolutionary Biology, 10, 332. <a href="http://doi.org/10.1186/1471-2148-10-332">http://doi.org/10.1186/1471-2148-10-332</a> <br />
</li>
<li>Ejsmond, M. J., &amp; Radwan, J. (2009). <em>MHC diversity in bottlenecked populations: a simulation model.</em> Conservation Genetics, 12(1), 129–137. <a href="http://doi.org/10.1007/s10592-009-9998-6">http://doi.org/10.1007/s10592-009-9998-6</a> <br />
</li>
<li>Ejsmond, M. J., Radwan, J., &amp; Wilson, A. B. (2014). <em>Sexual selection and the evolutionary dynamics of the major histocompatibility complex.</em> Proceedings of the Royal Society B: Biological Sciences, 281, 20141662. <a href="http://dx.doi.org/10.1098/rspb.2014.1662">http://dx.doi.org/10.1098/rspb.2014.1662</a> <br />
</li>
<li>Ejsmond, M. J., &amp; Radwan, J. (2015). <em>Red Queen Processes Drive Positive Selection on Major Histocompatibility Complex (MHC) Genes.</em> PLOS Comput. Biol. 11:e1004627. <a href="http://doi.org/10.1371/journal.pcbi.1004627">http://doi.org/10.1371/journal.pcbi.1004627</a> <br />
 <br />
<b>Read them before you start tweaking anything on your own.</b></li>
</ol>
<h1><a class="anchor" id="Compilation"></a>
Compilation</h1>
<p>The program was written in <a href="https://en.wikipedia.org/wiki/C%2B%2B14">C++14 standard</a> so if you are using GCC then version gcc 4.8 seems to be the minimum requirement (I have 5.4). This program has some serious dependencies on <a href="http://www.boost.org/">C++ Boost Libraries</a>. Should compile smoothly on most modern GNU/Linux distros with Boost Libs installed. Having <a href="http://www.scons.org/">Scons build tool</a> might be useful too. Basic compilation works fine on Ubuntu 16.04 LTS with mentioned packages installed by running the command: <br />
<br />
$<b> g++ -O3 -o SCBuild/MHC_model main.cpp nlohmann/json.hpp <a class="el" href="_gene_8cpp.html">Gene.cpp</a> <a class="el" href="_antigen_8cpp.html">Antigen.cpp</a> <a class="el" href="_host_8cpp.html">Host.cpp</a> <a class="el" href="_pathogen_8cpp.html">Pathogen.cpp</a> <a class="el" href="_h2_pinteraction_8cpp.html">H2Pinteraction.cpp</a> <a class="el" href="_random_8cpp.html">Random.cpp</a> <a class="el" href="_tagging__system_8cpp.html">Tagging_system.cpp</a> <a class="el" href="_environment_8cpp.html">Environment.cpp</a> <a class="el" href="_data_handler_8cpp.html">DataHandler.cpp</a> -fopenmp -std=c++14 </b><br />
 <br />
The code here can be also used like a toolbox for your research. You can stitch your own <em>main_yourown.cpp</em> file with your scenario and custom procedures (we did so for our research) and compile it using Scons script which is part of this code bundle. To do so run:<br />
<br />
$<b> scons -Q scenario="Scenarios/main_default.cpp" linking="dynamic"</b><br />
 <br />
replacing "Scenarios/main_default.cpp" with the path to your own custom <em>main_xyz.cpp</em>. Use our <em>main_core.cpp</em> file as the base to develop your own scenarios of MHC evolution dynamics. WARNING! Using our Scons script overwrites the <em>main.cpp</em> file in the main source directory - always store your scenarios somewhere else then the <em>main.cpp</em> file!<br />
<br />
Sometimes the HPC Cluster you have is lame and old and it has fairly outdated compiler (e.g. gcc &lt; 4.8). Then you can statically link the libraries on your fancy brand new PC running latest Linux distro and send the no-dependencies executable to cluster. Compile like this: <br />
 <br />
$<b> g++ -static -O3 -o MHC_model main.cpp <a class="el" href="_gene_8cpp.html">Gene.cpp</a> <a class="el" href="_antigen_8cpp.html">Antigen.cpp</a> <a class="el" href="_host_8cpp.html">Host.cpp</a> <a class="el" href="_pathogen_8cpp.html">Pathogen.cpp</a> <a class="el" href="_h2_pinteraction_8cpp.html">H2Pinteraction.cpp</a> RandomNumbs.cpp <a class="el" href="_tagging__system_8cpp.html">Tagging_system.cpp</a> <a class="el" href="_environment_8cpp.html">Environment.cpp</a> <a class="el" href="_data_handler_8cpp.html">DataHandler.cpp</a> -std=c++14 </b><br />
<br />
Or run the Scons script:<br />
<br />
$<b> scons -Q scenario="Scenarios/main_default.cpp" linking="static" </b><br />
 <br />
But it may throw some warnings e.g.: <br />
/usr/lib/gcc/x86_64-linux-gnu/5/libgomp.a(target.o): In function gomp_target_init : (.text+0xba): warning: Using 'dlopen' in statically linked applications requires at runtime the shared libraries from the glibc version used for linking <br />
<br />
 </p>
<h1><a class="anchor" id="Parameters"></a>
Parameters</h1>
<p><b>In the most advanced scenario the program takes exactly 17 parameters. These are:</b> <br />
<b> 00 </b>- Program's name <br />
<b> 01 </b>- Number of threads program will try to use on a multi-core CPU. Giving 0 will make the program use all CPU cores available. <br />
<b> 02 </b>- Number of bits in a gene. <br />
<b> 03 </b>- Number of bits in an antigen. <br />
<b> 04 </b>- <a class="el" href="class_host.html" title="Core class. Stores and handles a single host object. Each host has multiple instances of Gene class o...">Host</a> population size. <br />
<b> 05 </b>- <a class="el" href="class_pathogen.html" title="Core class. Stores and handles a single pathogen object. Each pathogen can have just one instance of ...">Pathogen</a> population size.<br />
<b> 06 </b>- Number of pathogen species.<br />
<b> 07 </b>- Number of genes in one host chromosome (they have two chromosomes) when the model is being initialised. <br />
<b> 08 </b>- Number of pathogen generations per one host generation. <br />
<b> 09 </b>- Number of host generations (effective length of model run). <br />
<b> 10 </b>- Probability of mutation in hosts ([0,1] range). <br />
<b> 11 </b>- Probability of mutation in pathogens ([0,1] range). <br />
<b> 12 </b>- The heterozygote advantage / lack of advantage mode. It has to be 10 for heterozygote advantage or 11 for lack of thereof. <br />
<b> 13 </b>- Probability of deleting a gene in the host ([0,1] range). <br />
<b> 14 </b>- Probability of duplicating a gene in the host ([0,1] range). <br />
<b> 15 </b>- Maximal number of genes permitted in one host chromosome. <br />
<b> 16 </b>- Number of sexual partners an individual checks out before selecting one for mating. <br />
<b> 17 </b>- Alpha factor for the host fitness function ([0,1] range).<br />
Historically we also had one more argument, that can be added if you want : <br />
 18- Fraction of the antigen's bits which are forbidden from changing (a.k.a. "No Mutation Bits"). <br />
<br />
This program can recognise simple errors in the argument list (a probability value out of [0,1] range, negative values when only positive are allowed etc.), but will not recognise when they don't make a 'biological' sense.<br />
<br />
Example run call: <br />
$ <b>./main_default 0 16 3200 1000 8000 8 1 10 3000 0.0001 5e-05 10 0.001 0.001 50 10 0.02</b><br />
<br />
 There are 8 pathogen species here, each has 1000 individuals, hence the total pathogen population size is 8000. <br />
<br />
 </p>
<h1><a class="anchor" id="The"></a>
output and data visualisation</h1>
<p>Program produces a number of text files containing desired data. The file <em>InputParameters.json</em> contains the values of parameters fed to the model. The file <em>HostsGeneDivers.csv</em> contains the main statistics of the model, including the time stamp used in the visualisation process of the other files with data. This two files are the basics to analyse the simulation output. Other files are optional and they can be switched off by commenting out some lines in the main function (<em>main.cpp</em> file) what should speed up the program runtime. <br />
<br />
 All the files are:<br />
<b><em>InputParameters.json </em></b> - contains run's parametrisation in JSON text format <br />
<b><em>HostsGeneDivers.csv</em></b> - contain basic statistics of the output for host genomes <br />
<b><em>HostGenomesFile.[t].csv</em></b> - contains all the genomes of all the cells for host population in time <em>t</em>. There can be more then one file like this for different time snapshots. <br />
<b><em>PathoGenomesFile.[t].csv</em></b> - contains all the genomes of all the cells for pathogen population in time <em>t</em>. There can be more then one file like this for different time snapshots. <br />
<b><em>HostGeneNumbTotal.csv </em></b> - total number of genes in each individual host in each time step (linked to <em>HostMHCsNumbUniq.csv </em>, that each column in <em>HostGeneNumbTotal.csv </em> is represents the same cells as in <em>HostMHCsNumbUniq.csv </em>). <br />
<b><em>HostMHCsNumbUniq.csv </em></b> - number of unique MHC alleles in each individual host in each time step (linked to <em>HostGeneNumbTotal.csv </em>, that each column in <em>HostGeneNumbTotal.csv </em> is represents the same cells as in <em>HostMHCsNumbUniq.csv </em>). <br />
<b><em> NoMutationInPathoList.csv </em></b> - list of conserved antigen sites. Each line contains one pathogen species, each number indicates the index of a "no-mutation" site in this species antigen. When empty it means there is no mutation restrictions. <br />
<b><em> NumberOfMhcAfterMating.csv </em></b> -number of the unique MHC types in each individual host in each time step after the mating procedure took place. <br />
<b><em> NumberOfMhcBeforeMating.csv </em></b> - number of the unique MHC types in each individual host in each time step before the mating procedure took place. <br />
<b><em> NumberOfMhcInMother.csv </em></b> - number of the unique MHC types in each individual host that is selecting a partner (a.k.a. "mother") in each time step during mating procedure. Each individual has a corresponding partner at the same index in the file <em>NumberOfMhcInFather.csv</em>. <br />
<b><em> NumberOfMhcInFather.csv </em></b> - number of the unique MHC types in each individual host that has been selected as a mating (a.k.a. "father") in each time step during mating procedure. Each individual has a corresponding partner at the same index in the file <em>NumberOfMhcInMother.csv</em>. <br />
<b><em> PresentedPathogenNumbers.csv </em></b> - number of presented pathogens by each individual in each time step. <br />
 <br />
Visualisation is done using Python 3.6 scripts containing a a lot of calls to Numpy, Matplotlib and other scietific Python libraries. You may wish to consider using the <a href="https://www.anaconda.com/download/">Python Anaconda project</a>. These scripts can be found in <em>PyScripts</em> directory. <br />
To load the environment for output analysis run: <br />
<br />
$ <b>conda env create -f mhcEvoEnv.yaml</b> <br />
</p>
</div></div><!-- contents -->
</div><!-- doc-content -->
<!-- start footer part -->
<div id="nav-path" class="navpath"><!-- id is needed for treeview function! -->
  <ul>
    <li class="footer">Generated on Sun Oct 6 2019 18:26:48 for Modelling MHC Evolutionary Dynamics by
    <a href="http://www.doxygen.org/index.html">
    <img class="footer" src="doxygen.png" alt="doxygen"/></a> 1.8.13 </li>
  </ul>
</div>
</body>
</html>
